---
title: "Transcriptomic variability in E. coli"
author: "Tsuru"
date: "`r format(Sys.time(), '%Y/%m/%d')`"
output: 
  html_document:
  code_folding: hide
---
# Outlook
This script is used to analyze the following datasets of transcriptional profiles mainly.
the Env dataset: RNA-seq compendium of E. coli K-12 MG1655 in different growth conditions
the Evo dataset: RNA-seq compendium of different natural isolates of E. coli grown in LB medium
the Mut dataset: Microarray compedium of different lineages of the mutation accumulation (MA) experiments using E. coli MG1655 delt-dnaQ::kan

Section 0: General settings
Section 1: Define name space of genes and set Transcriptional regulatory network (TRN), the KEGG pathways
Section 2: Load the Mut dataset and calculate transcriptional variability (DMmut)
Section 3: Load the Env dataset and calculate transcriptional variability (DMenv) 
Section 4: Load the Evo dataset and calculate transcriptional variability (DMevo)
Section 5: Comparison of the DM values between the three datasets
Section 6: GSEA using the KEGG pathways
Section 7: GSEA using TRN
Section 8: TRN statistics
Section 9: PCA
Section 10: Gene Ontology

Section 11: Analysis of the Noise_ch and Noise_pl dataset (Taniguchi, Science, 2010 /  Urchueguia, PLoS Biol., 2021)
Section 14: Analysis of the Prtn dataset (Schmidt, Nat. Biotechnol., 2016)
Section 15: Analysis of the Strs dataset (Maeda and Iwasawa, Nat. Comm., 2020)
Section 16: Analysis of the Strs_hp and Strs_he (Horinouchi, Sci. Rep., 2017)
Section 17: Bow-tie decomposition
Section 18:
Section 19:
Section 20: Output for supplementary tables


# 0. General Settings
## 0.1. Load workspace
```{r workspace_load,warning=FALSE, message=FALSE}
# load("./Variability_Ecoli_Main.RData")
```
## 0.2. Save workspace
```{r workspace_save,warning=FALSE, message=FALSE}
# save.image("./Variability_Ecoli_Main.RData")
```
## 0.3. Library loading
```{r library,warning=FALSE, message=FALSE}
library(viridis) # color palette
library(RColorBrewer) # color palette
library(ggsci) # color palette

library(ggplot2) # generic visualization
library(ggpubr) # generic visualization
library(cowplot) # generic visualization
library(ggrepel) # generic visualization
library(gtable) # generic visualization
library(gridExtra) # generic visualization
library(patchwork) # generic visualization
library(grid) # generic visualization
library(ggpmisc) # generic visualization
library(ggtext) # generic visualization
library(glue) # generic visualization
library(scales) # generic visualization
library(matrixStats) # calc statistics
library(rstatix) # visualizion of statistical test
library(dplyr) # handling dataframe
library(tidyverse) # handling dataframe
library(preprocessCore)
library(readr) # read_csv
library(readxl) # read_xls
library(openxlsx) # open xlsx files
library(corrplot) # correlation analysis and visualization
library(GGally) # correlation analysis and visualization
library(Hmisc)
library(magrittr) # pipe
library(dendextend) # dendrogram
library(FactoMineR) # PCA
library(factoextra) # PCA
library(proxy)
# library(plot3D)
library(clusterProfiler) # enrichment analysis
library(pathview) # enrichment analysis
library(org.EcK12.eg.db) # enrichment analysis
library(rvest) # enrichment analysis
library(corrr) # correlation analysis and visualization
library(GOSemSim) # enrichment analysis
library(rrvgo) # for heatmapPlot to visualize dfgoslim3
library(qdapTools)
library(AnnotationDbi) # enrichment analysis
library(ggVennDiagram) #venn diagram
#library(rJava)
library(venneuler) #venn diagram
library(igraph) # Network visualization
library(tidygraph) # Network visualization
# library(recoRdseq) # for Record-seq

library(KEGGREST) # enrichment analysis
# library(plot3D)
library(ggvenn) #venn diagram

#library(scales)
library(ape) # phylogenetic tree of 16 natural isolates of E. coli
library(ggtree) # phylogenetic tree of 16 natural isolates of E. coli
library(qvalue) # enrichment analysis
library(scatterpie) # Network visualization
#library(graphlayouts)

library(ggcorrplot) # correlation analysis and visualization
library(phangorn) # phylogenetic analysis
library(ggbeeswarm) # beeswarm plot
# library(ggcorrplot2)

Sys.time()
```
## 0.4. Define versatile user functions
The [qw](https://ja.coder.work/so/r/240447) function in Perl.  
The [createEmptyDf](https://htsuda.net/archives/2560) function.
The [g_legend](https://github.com/hadley/ggplot2/wiki/Share-a-legend-between-two-ggplot2-graphs) function to extract legend.
```{r functions}
qw <- function(...) {
  sapply(match.call()[-1], deparse)
}

createEmptyDf = function( nrow, ncol, colnames = c() ){
  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )
}

g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

pValue_sp_text=function(dtf){
  pcor_sp<-cor.test(x=dtf[,1],y=dtf[,2],method="spearman")
  rho_sp<-round(pcor_sp$estimate,digits = 2)
  pv_sp<-ifelse(pcor_sp$p.value<0.05," P<0.05",paste(" P==",round(pcor_sp$p.value,digits = 1),sep = ""))
  rhopvtxt_sp<-paste("list(italic(rho)==", rho_sp,", ",pv_sp,")")
  return(rhopvtxt_sp)
}

pValue_pr_text=function(dtf){
  pcor_pr<-cor.test(x=dtf[,1],y=dtf[,2],method="pearson")
  r_pr<-round(pcor_pr$estimate,digits = 2)
  pv_pr<-ifelse(pcor_pr$p.value<0.05," P<0.05",paste(" P==",round(pcor_pr$p.value,digits = 1),sep = ""))
  rpvtxt_pr<-paste("list(italic(r)==", r_pr,", ",pv_pr,")")
  return(rpvtxt_pr)
}

gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

Sys.time()
```
## 0.5. Define versatile color palletes
```{r color}
mycolpal<-get_palette("lancet",9)
print(c("#00468BFF","#ED0000FF","#42B540FF","#0099B4FF","#925E9FFF","#FDAF91FF","#AD002AFF","#ADB6B6FF","#1B1919FF"))
mycolpal.BuRd<-c(get_palette(palette = "Blues", 4)[2:4],get_palette(palette = "Reds", 4)[2:4])
print(c("#FEE5D9","#FCAE91","#FB6A4A","#CB181D"))
print(c("#EFF3FF","#BDD7E7","#6BAED6","#2171B5"))
Sys.time()
```
## 0.6. Generate a blank panel
```{r blank}
blank <- grid.rect(gp=gpar(col="white"))

Sys.time()
```
## 0.7. Directry
https://journals.asm.org/doi/10.1128/mBio.02096-17
```{r Correlation_substrate,warning=FALSE, message=FALSE}
# d<-getwd()
ddatraw<-paste(getwd(),"MUT_DataSet","Data_20211027.csv",sep="/") # transcriptome data of the MA experiment
ddatinfo<-paste(getwd(),"MUT_DataSet","SampleIinfo_20211027.csv",sep="/") # Sample information of the MA lineages
dgenename.w3110<-paste(getwd(),"W3110_genes","W3110BiotypeList.txt",sep="/") # Name Space based on W3110
ddess<-paste(getwd(),"BW25113_genes","BW25113ESSCDSList.txt",sep="/")  # Essential CDS in W3110 name space
dgenename.mg1655<-paste(getwd(),"MG1655_genes","MG1655_gnftr.csv",sep="/") # Gene features of MG1655 including CDS, ncRNA, tRNA, rRNA
dprobe2id<-paste(getwd(),"W3110_genes","Probe2IDs.csv",sep="/") # Name Space based on W3110
dgenename.mds42<-paste(getwd(),"MDS42_genes","MDS42_gnftr.csv",sep="/") # Gene features of MG1655 including CDS, ncRNA, tRNA, rRNA

Sys.time()
```



#
#
#
# 1. Name space of genes, TRN and KEGG pathways
## 1.1. Define namse space of genes
```{r Global_Bnb,warning=FALSE, message=FALSE,fig.height=1.4,fig.width=1.4,dpi=300}
# gene name of W3110
genename.w3110<-data.frame(read.delim(dgenename.w3110,stringsAsFactors=F))
# gene name of MG1655 including CDS, ncRNA, tRNA, rRNA
genename.mg1655<-as.data.frame(read_csv(dgenename.mg1655))
genename.mg1655<-genename.mg1655[,c(2:ncol(genename.mg1655))]
colnames(genename.mg1655)[colnames(genename.mg1655)=="locus_tag"]<-"Bnb"
colnames(genename.mg1655)[colnames(genename.mg1655)=="GeneID"]<-"EntrezID"
# essentiality
dess<-data.frame(read.delim(ddess,stringsAsFactors=F,header = F))
colnames(dess)<-qw(Name,ECKnb)
dess<-transform(dess,Essentiality="Essential")
dess<-dess[,2:3]
# dnds
ddnds<-read.csv(paste(getwd(),"dNdS","dat2_20210423.csv",sep="/"),stringsAsFactors = F)
ddnds<-ddnds[,which(colnames(ddnds) %in% qw(b_number,dNdS,dN,dS,dN.dS.inter.species,dN.inter.species,dS.inter.species))]
colnames(ddnds)<-qw(Bnb,dNdSwthn,dNwthn,dSwthn,dNdSbtwn,dNbtwn,dSbtwn)
# ko growth
dkogr<-read.delim(paste(getwd(),"KOGrowth","KOGrowthList.txt",sep="/"),stringsAsFactors = F)[,2:4]
colnames(dkogr)<-qw(GrowthRate,ODmax,ECKnb)

dBnb<-genename.mg1655
dBnb<-dBnb[!is.na(dBnb$Bnb),]
dBnb$presence_pr2<-NA # log_tpm_pl.pr2
# dBnb$presence_gut<-NA # rna2.tpm.qnorm
dBnb$presence_mut<-NA # datlog.norm
dBnb$presence_evo<-NA # dcol16.qnorm


# load PR2 dataset
temp<-as.data.frame(read.csv(paste(getwd(),"ENV_DataSet","data_files","log_tpm_qc.csv",sep="/"),stringsAsFactors = F,skip = 0))
colnames(temp)[1]<-"Bnb"
# Presence in PR2
temp<-temp[!is.na(temp$Bnb),]
rownames(temp)<-temp$Bnb
temp<-na.omit(temp[,which(colnames(temp) %nin% "Bnb")])
temp1<-which(rowVars(as.matrix(temp))<10^-20)
if(length(temp1)>0){
  temp<-temp[-which(rowVars(as.matrix(temp))<10^-20),]
}
dBnb[dBnb$Bnb %in% rownames(temp),]$presence_pr2<-"presence"


# load MUT dataset
datinfo<-read.csv(ddatinfo,stringsAsFactors=F)
datinfo<-transform(datinfo,label=ifelse(datinfo$Transfer==0,paste(datinfo$Lineage,paste("rep",datinfo$Rep,sep=""),datinfo$Transfer,sep="_"),paste(datinfo$Lineage,datinfo$Transfer,sep="_")))
datinfo$label<-as.character(datinfo$label)
temp<-read.csv(ddatraw,stringsAsFactors=F)
colnames(temp)<-c("ProbeName",datinfo$label)
lowlim_lin<-3
for(i in 2:ncol(temp)){
  temp<-temp[(temp[,i]>3),]
}


# probe name of microarray
probe2id<-as.data.frame(read_csv(dprobe2id))
probe2id<-probe2id[,c(2:ncol(probe2id))]
colnames(probe2id)[colnames(probe2id)=="ECKnbs"]<-"ECKnb"
colnames(probe2id)[colnames(probe2id)=="JWnbs"]<-"JWnb"
colnames(probe2id)[colnames(probe2id)=="Bnbs"]<-"Bnb"

temp<-left_join(temp,probe2id[(probe2id$IStag==F)&(probe2id$Count==1)&(!grepl("\\+",probe2id$ECKnb)),which(colnames(probe2id) %in% qw(ProbeName,ECKnb))],by="ProbeName")
temp<-left_join(temp,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(Bnb,ECKnb))],by="ECKnb")
# omit #Bnb=NA
temp<-temp[!is.na(temp$Bnb),]

# remove duplicated Bnbs
temp1<-data.frame(Bnb=unique(temp$Bnb),Count=1)
for(i in 1:nrow(temp1)){
  temp1$Count[i]<-nrow(temp[temp$Bnb==temp1$Bnb[i],])
}
temp1<-temp[temp$Bnb %in% temp1[temp1$Count>1,]$Bnb,]
temp<-temp[temp$ProbeName %nin% temp1[!grepl("[A-Z]$",temp1$ProbeName),]$ProbeName,]

# Presence in MA
rownames(temp)<-temp$Bnb
temp<-temp[,which(colnames(temp) %nin% qw(ECKnb,ProbeName,Bnb))]
temp<-na.omit(temp)
temp1<-which(rowVars(as.matrix(temp))<10^-20)
if(length(temp1)>0){
  temp<-temp[-which(rowVars(as.matrix(temp))<10^-20),]
}
dBnb[dBnb$Bnb %in% rownames(temp),]$presence_mut<-"presence"


# load EVO dataset
temp0<-data.frame(read_csv(paste(getwd(),"EVO_DataSet","TableS6.csv",sep="/")))
col16.strain<-colnames(temp0)[colnames(temp0)!="Gene"]
temp0<-left_join(temp0,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(gene,Bnb,ECKnb))],by=c("Gene"="gene"))

temp<-data.frame(Gene=temp0[is.na(temp0$Bnb),]$Gene,gene_synonym=NA)
temp1<-genename.mg1655[genename.mg1655$Bnb %nin% temp0[!is.na(temp0$Bnb),]$Bnb,]$gene_synonym
temp1<-temp1[grepl(",",temp1)]
for(i in 1:nrow(temp)){
  temp2<-temp1[grepl(temp$Gene[i],temp1)]
  if(length(temp2)>0){
    temp$gene_synonym[i]<-temp1[grepl(temp$Gene[i],temp1)]
  }
}
temp$ECKnb<-str_extract(temp$gene_synonym,"ECK[0-9]{4}")
temp<-temp[!is.na(temp$ECKnb),which(colnames(temp) %nin% "gene_synonym")]
temp<-left_join(temp,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(gene,ECKnb))],by="ECKnb")
temp1<-temp0[!is.na(temp0$Bnb),]
temp2<-temp0[is.na(temp0$Bnb),]
temp2<-temp2[temp2$Gene %in% temp$Gene,]
for(i in 1:nrow(temp2)){
  temp2$Gene[i]<-temp[temp$Gene==temp2$Gene[i],]$gene
}
temp2<-temp2[,which(colnames(temp2) %nin% qw(gene,Bnb,ECKnb))]
temp2<-left_join(temp2,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(gene,Bnb,ECKnb))],by=c("Gene"="gene"))

temp0<-bind_rows(temp1,temp2)
temp0<-temp0[!is.na(temp0$Bnb),]
rownames(temp0)<-temp0$Bnb
temp<-temp0[,which(colnames(temp0) %nin% qw(Gene,Bnb,ECKnb))]


# Presence in EVO
temp<-na.omit(temp)
temp1<-rownames(temp)
temp<-normalize.quantiles(as.matrix(temp))
rownames(temp)<-temp1


temp1<-which(rowVars(as.matrix(temp))<10^-20)
if(length(temp1)>0){
  temp<-temp[-which(rowVars(as.matrix(temp))<10^-20),]
}
dBnb[dBnb$Bnb %in% rownames(temp),]$presence_evo<-"presence"

# temp<-qw(presence_pr2,presence_gut,presence_mut,presence_evo)
temp<-qw(presence_pr2,presence_mut,presence_evo)
dBnb.comp<-dBnb
for(i in 1:length(temp)){
  dBnb.comp<-dBnb.comp[!is.na(dBnb.comp[,which(colnames(dBnb.comp)==temp[i])]),]
}

Sys.time()
```
## 1.2. Define TRN
TRN based on RegulonDB & Ecocyc
```{r ENV_DataSet_TRN,warning=FALSE, message=FALSE,fig.height=2,fig.width=2,dpi=300}
# Load TRN dataset from RegulonDB
TRN<-as.data.frame(read_csv(paste(getwd(),"RegulonDB","2022_ver_11","Output","trn.regulator_gene.csv",sep="/")))
# rename regulator
# CRP to Crp, Dan to TtdR, FNR to Fnr, HigB-A to HigBA, MatA to EcpR
TRN[TRN$regulator=="CRP",]$regulator<-"Crp"
TRN[TRN$regulator=="Dan",]$regulator<-"TtdR"
TRN[TRN$regulator=="FNR",]$regulator<-"Fnr"
TRN[TRN$regulator=="HigB-HigA",]$regulator<-"HigBA"
TRN[TRN$regulator=="MatA",]$regulator<-"EcpR"

# Load dataset of ECOCYC
ecocyc.all.genes<-as.data.frame(read_csv(paste(getwd(),"ECOCYC","2022_10_09_Ecocyc","All_Genes.csv",sep="/")))
ecocyc.sub.genes<-ecocyc.all.genes[!is.na(ecocyc.all.genes$Bnb),]
# length(unique(ecocyc.sub.genes$Bnb))
# load data set for TF
ecocyc.all.tfs<-as.data.frame(read_csv(paste(getwd(),"ECOCYC","2022_10_09_Ecocyc","All_transcription_factors.csv",sep="/")))
temp<-as.data.frame(read_csv(paste(getwd(),"ECOCYC","2022_10_09_Ecocyc","Cure01_transcription_factors.csv",sep="/")))
ecocyc.all.tfs<-bind_rows(ecocyc.all.tfs,temp)
# check name space of regulator
ecocyc.all.tfs[ecocyc.all.tfs$regulator %nin% unique(TRN$regulator),]$regulator
unique(TRN[TRN$regulator %nin% ecocyc.all.tfs$regulator,]$regulator)

mycolname<-qw(regulator,gene_name,gene_id,effect,evidence,source,category,regulator_label,Use,regulator_id)
for(i in 1:nrow(ecocyc.all.tfs)){
  temp0<-unique(str_split(ecocyc.all.tfs$Regulated_Bnb[i],",")[[1]])
  temp0<-temp0[grepl("^b",temp0)]
  temp1<-createEmptyDf(nrow=length(temp0),ncol=length(mycolname),colnames = mycolname)
  temp1$regulator<-ecocyc.all.tfs$regulator[i]
  temp1$regulator_id<-ecocyc.all.tfs$Bnb[i]
  temp1$category<-ecocyc.all.tfs$Category[i]
  temp1$gene_id<-temp0
  temp1.1<-c()
  if(!is.na(ecocyc.all.tfs$Inhibited_Bnb[i])){
    temp1.1<-unique(str_split(ecocyc.all.tfs$Inhibited_Bnb[i],",")[[1]])
    temp1.1<-temp1.1[grepl("^b",temp1.1)]
  }
  temp1.2<-c()
  if(!is.na(ecocyc.all.tfs$Activated_Bnb[i])){
    temp1.2<-unique(str_split(ecocyc.all.tfs$Activated_Bnb[i],",")[[1]])
    temp1.2<-temp1.2[grepl("^b",temp1.2)]
  }
  for(j in 1:nrow(temp1)){
    temp1$gene_name[j]<-ecocyc.sub.genes[ecocyc.sub.genes$Bnb==temp1$gene_id[j],]$Gene
    if(length(temp1.1[temp1.1==temp1$gene_id[j]])>0&length(temp1.2[temp1.2==temp1$gene_id[j]])>0){
      temp1$effect[j]<-"+,-"
    }else if(length(temp1.1[temp1.1==temp1$gene_id[j]])>0&length(temp1.2[temp1.2==temp1$gene_id[j]])==0){
      temp1$effect[j]<-"-"
    }else if(length(temp1.1[temp1.1==temp1$gene_id[j]])==0&length(temp1.2[temp1.2==temp1$gene_id[j]])>0){
      temp1$effect[j]<-"+"
    }
  }
  if(i==1){
    temp4.1<-temp1
  }else{
    temp4.1<-bind_rows(temp4.1,temp1)
  }
}
temp4.1$source<-"Ecocyc"

# load data set for Sigma
ecocyc.sigmulon<-as.data.frame(read_csv(paste(getwd(),"ECOCYC","2022_10_09_Ecocyc","Sigmulon.csv",sep="/")))
for(i in 1:nrow(ecocyc.sigmulon)){
  temp2.1<-unique(str_split(ecocyc.sigmulon$Activation[i],",")[[1]])
  temp2.1<-unique(ecocyc.sub.genes[ecocyc.sub.genes$Gene %in% temp2.1,]$Bnb)
  temp2.1<-temp2.1[grepl("^b",temp2.1)]
  temp2.2<-createEmptyDf(nrow=length(temp2.1),ncol=length(mycolname),colnames = mycolname)
  temp2.2$regulator<-ecocyc.sigmulon$Sigma[i]
  temp2.2$regulator_id<-ecocyc.sigmulon$Bnb[i]
  temp2.2$effect<-"+"
  temp2.2$category<-"sigma"
  temp2.2$gene_id<-temp2.1
  for(j in 1:nrow(temp2.2)){
    temp2.2$gene_name[j]<-ecocyc.sub.genes[ecocyc.sub.genes$Bnb==temp2.2$gene_id[j],]$Gene
  }
  if(i==1){
    temp4.2<-temp2.2
  }else{
    temp4.2<-bind_rows(temp4.2,temp2.2)
  }
}
temp4.2$source<-"Ecocyc"
temp4<-bind_rows(temp4.1,temp4.2)

# fix Bnb
temp5<-temp4 %>% distinct(regulator,.keep_all = T)
for(i in 1:nrow(temp5)){
  temp6<-TRN[TRN$regulator==temp5$regulator[i],]
  if(nrow(temp6)>0){
    temp5$regulator_id[i]<-unique(temp6$regulator_id)
  }
}
for(i in 1:nrow(temp5)){
  temp4[temp4$regulator==temp5$regulator[i],]$regulator_id<-temp5$regulator_id[i]
}



# add regulator_id (Bnb)
# TRN<-left_join(TRN,temp5[,which(colnames(temp5) %in% qw(regulator,regulator_id)),],by="regulator")

# update effect
for (i in 1:nrow(TRN)){
  if(TRN$effect[i] %nin% c("+","-","+,-")&!is.na(TRN$regulator_id[i])){
    temp6<-temp4[(temp4$regulator_id==TRN$regulator_id[i])&temp4$gene_id==TRN$gene_id[i],]
    if(nrow(temp6)==1){
      if(!is.na(temp6$effect)){
        TRN$effect[i]<-temp6$effect
      }
    }
  }
}


# Unique interactions in Ecocyc smart table
temp7<-temp4
temp7$prsn_in_TRN<-NA
for(i in 1:nrow(temp7)){
  if(nrow(TRN[(TRN$regulator==temp7$regulator[i])&(TRN$gene_id==temp7$gene_id[i]),])>0){
    temp7$prsn_in_TRN[i]<-"Yes"
  }
}
temp7<-temp7[is.na(temp7$prsn_in_TRN),]
temp7[is.na(temp7$effect),]$effect<-""
temp7<-temp7[temp7$gene_id %in% genename.mg1655$Bnb,]
temp8<-TRN %>% distinct(regulator,.keep_all = T)
for(i in 1:nrow(temp7)){
  temp9<-genename.mg1655[genename.mg1655$Bnb==temp7$gene_id[i],]$gene
  if(length(temp9)==1){
    temp7$gene_name[i]<-temp9
  }
  # temp10<-temp8[temp8$regulator==temp7$regulator[i],]$regulator_label
  # if(length(temp10)==1){
  #   temp7$regulator_label[i]<-temp10
  #   temp7$Use[i]<-temp8[temp8$regulator==temp7$regulator[i],]$Use
  # }else{
  #   temp7$regulator_label[i]<-temp7$regulator[i]
  #   temp7$Use[i]<-"Yes"
  # }
}
temp7$evidence<-NA
TRN<-bind_rows(TRN,temp7[,which(colnames(temp7) %nin% qw(prsn_in_TRN))])
TRN$prsn_in_gBnb<-NA
temp11<-TRN[(TRN$category %in% qw(TF,sigma))&!is.na(TRN$regulator_id),] %>% distinct(regulator,.keep_all = T)
for(i in 1:nrow(temp11)){
  if(str_count(temp11$regulator_id[i],",")==0){
    temp12<-dBnb.comp[dBnb.comp$Bnb==temp11$regulator_id[i],]
    if(nrow(temp12)==1){
      temp11$prsn_in_gBnb[i]<-"Yes"
    }
  }else{
    temp12<-dBnb.comp[dBnb.comp$Bnb %in% str_split(temp11$regulator_id,",")[[1]],]
    if(nrow(temp12)==2){
      temp11$prsn_in_gBnb[i]<-"Yes"
    }
  }
}
TRN[(TRN$gene_id %in% dBnb.comp$Bnb), ]$prsn_in_gBnb<-"Yes" # Yes if the target gene is present in three datasets

# Use gene name space in MG1655
TRN<-TRN[TRN$gene_id %in% genename.mg1655$Bnb,]
i<-1
for(i in 1:nrow(TRN)){
  if(grepl("^b",TRN$gene_id[i])){
    TRN$gene_name[i]<-genename.mg1655[genename.mg1655$Bnb==TRN$gene_id[i],]$gene
  }
}

# Check duplication of interactions
temp<-TRN %>% distinct(regulator,gene_id,.keep_all = T)
j<-1
k<-1
temp2<-createEmptyDf(nrow=0,ncol=1,colnames="x")
nrow(temp2)
for(i in 1:nrow(temp)){
  temp1<-TRN[(TRN$regulator==temp$regulator[i])&(TRN$gene_id==temp$gene_id[i]),]
  if(nrow(temp1)>1){
    if(j==1){
      temp2<-temp1
      j<-j+1
    }else{
      temp2<-rbind(temp2,temp1)
    }
  }else{
    if(k==1){
      temp3<-temp1
      k<-k+1
    }else{
      temp3<-rbind(temp3,temp1)
    }
  }
}
if(nrow(temp2)>0){
  temp2<-temp2[!is.na(temp2$evidence),]
  temp2<-temp2 %>% distinct(regulator,gene_id,.keep_all = T)
  TRN<-bind_rows(temp3,temp2)
}

# Action for STRONG EVIDENCE 
# TRN<-TRN[!is.na(TRN$evidence)&(TRN$evidence=="STRONG"),]

TRNsub<-TRN[!is.na(TRN$prsn_in_gBnb),]


dTRNsub_stat<-TRNsub[,which(colnames(TRNsub) %in% qw(regulator,regulator_id,category))]
dTRNsub_stat<-dTRNsub_stat %>% distinct(regulator,.keep_all = T)
dTRNsub_stat$size<-0
for(i in 1:nrow(dTRNsub_stat)){
  dTRNsub_stat$size[i]<-nrow(TRNsub[TRNsub$regulator==dTRNsub_stat$regulator[i],])
}

# create gene sets for GSEA
dtrn.each.gsea<-TRNsub[(TRNsub$category %in% qw(TF,sigma)),] #<------------------------------- use TRNsub
dtrn.each.gsea<-left_join(dtrn.each.gsea,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(Bnb,EntrezID))],by=c("gene_id"="Bnb"))
dtrn.each.gsea<-dtrn.each.gsea[dtrn.each.gsea$regulator %in% dTRNsub_stat[(dTRNsub_stat$size>4)&(dTRNsub_stat$size<1001),]$regulator,]
dtrn.each.gsea<-dtrn.each.gsea[,which(colnames(dtrn.each.gsea) %in% qw(regulator,EntrezID))]


# Create TRN comprising of TFs and Sigmas
temp<-TRN %>% distinct(regulator,.keep_all = T)
temp2<-unique(unlist(str_split(temp[str_count(temp$regulator_id,",")>0,]$regulator_id,",")))
temp1<-TRN[TRN$gene_id %in% unique(unlist(str_split(temp$regulator_id,","))),] # extract subset of TRN of which only TF and sigma consist
temp1$target<-NA
temp1.1<-temp1[temp1$gene_id %nin% temp2,] # TRN whose targets are not complex
for(i in 1:nrow(temp1.1)){
  temp1.1$target[i]<-temp[temp$regulator_id==temp1.1$gene_id[i],]$regulator
}
temp1.2<-temp1[temp1$gene_id %in% temp2,] # TRN whose targets are not complex
temp3<-temp[grepl(",",temp$regulator_id),]
for(i in 1:nrow(temp1.2)){
  # temp1.2$target[i]<-paste(temp3[grepl(temp1.2$gene_id[i],temp3$regulator_id),]$regulator,collapse=",")
  temp1.2$target[i]<-paste(temp[grepl(temp1.2$gene_id[i],temp$regulator_id),]$regulator,collapse=",")
}
temp3<-bind_rows(temp1.1,temp1.2)
temp3<-temp3[,which(colnames(temp3) %in% qw(regulator,effect,target))]
temp4<-createEmptyDf(nrow=1,ncol=ncol(temp3),colnames=colnames(temp3))
for(i in 1:nrow(temp3)){
  temp3.1<-str_split(temp3$target[i],",")[[1]]
  if(length(temp3.1)==1){
    temp4<-bind_rows(temp4,temp3[i,])
  }else{
    temp3.2<-createEmptyDf(nrow=length(temp3.1),ncol=ncol(temp3),colnames=colnames(temp3))
    temp3.2$regulator<-temp3$regulator[i]
    temp3.2$effect<-temp3$effect[i]
    temp3.2$target<-temp3.1
    temp4<-bind_rows(temp4,temp3.2)
  }
}
temp4<-temp4[2:nrow(temp4),]
temp4[temp4$effect %in% c("-","-,?","?,-"),]$effect<--1
temp4[temp4$effect %in% c("+","+,?","?,+"),]$effect<-1
temp4[temp4$effect %nin% c(-1,1),]$effect<-0
temp4$effect<-as.numeric(temp4$effect)
temp4<-temp4 %>% distinct(regulator,effect,target,.keep_all = T)
temp5<-temp4 %>% distinct(regulator,target,.keep_all = T)
for(i in 1:nrow(temp5)){
  temp4.1<-as.data.frame(temp4[temp4$regulator==temp5$regulator[i]&temp4$target==temp5$target[i],])
  if(nrow(temp4.1)>1){
    temp4.2<-sum(temp4.1$effect)
    if(temp4.2<0){
      temp5$effect[i]<--1
    }else if(temp4.2>0){
      temp5$effect[i]<-1
    }else{
      temp5$effect[i]<-0
    }
  }
}
TRN.ts<-temp5

# Solo or Multi with/without RpoD
temp<-data.frame(regulator=c(unique(TRN$regulator),"No"),solo.wo.rpod=NA,solo.wt.rpod=NA,multi.wt.rpod=NA,multi.wo.rpod=NA)
temp1<-unique(TRN[TRN$regulator=="RpoD",]$gene_id) # target genes regulated by RpoD with or without any TFs
temp2<-TRN[TRN$gene_id %nin% temp1,]$gene_id # target genes unregulated by RpoD
temp3<-dBnb[dBnb$Bnb %nin% TRN$gene_id,]$Bnb # non-target genes for any TRN and sigmas
temp4<-data.frame(gene_id=unique(TRN$gene_id),Flag=NA,regulators=NA) # target genes regulated by any TFs and sigmas
for(i in 1:nrow(temp4)){
  j<-NA
  temp5.1<-unique(TRN[(TRN$gene_id==temp4$gene_id[i]),]$regulator)
  temp4$regulators[i]<-paste(temp5.1,collapse = ",")
  j<-ifelse(length(temp5.1[temp5.1=="RpoD"])==0,0,1) # 0 if RpoD is present
  if(j==0){
    temp4$Flag[i]<-ifelse(length(temp5.1)==1,"solo.wo.rpod","multi.wo.rpod")
  }else{
    temp5.2<-temp5.1[temp5.1!="RpoD"]
    if(length(temp5.2)==0){
      temp4$Flag[i]<-"rpod.only"
    }else{
      temp4$Flag[i]<-ifelse(length(temp5.2)==1,"solo.wt.rpod","multi.wt.rpod")
    }
  }
}
temp6<-left_join(TRN,temp4,by="gene_id")

i<-1
for(i in 1:nrow(temp)){
  temp7<-temp6[temp6$regulator==temp$regulator[i],]
  temp8<-temp7[temp7$Flag=="solo.wo.rpod",]$gene_id
  temp$solo.wo.rpod[i]<-ifelse(length(temp8)>0,paste(temp8,collapse = ","),NA)
  temp8<-temp7[temp7$Flag=="solo.wt.rpod",]$gene_id
  temp$solo.wt.rpod[i]<-ifelse(length(temp8)>0,paste(temp8,collapse = ","),NA)
  temp8<-temp7[temp7$Flag=="multi.wo.rpod",]$gene_id
  temp$multi.wo.rpod[i]<-ifelse(length(temp8)>0,paste(temp8,collapse = ","),NA)
  temp8<-temp7[temp7$Flag=="multi.wt.rpod",]$gene_id
  temp$multi.wt.rpod[i]<-ifelse(length(temp8)>0,paste(temp8,collapse = ","),NA)
}
temp[temp$regulator=="RpoD",which(colnames(temp) %nin% "regulator")]<-NA
temp[temp$regulator=="RpoD",]$solo.wt.rpod<-paste(temp6[temp6$Flag=="rpod.only",]$gene_id,collapse=",")
temp[temp$regulator=="No",]$solo.wo.rpod<-paste(temp3,collapse=",")
temp<-temp %>% pivot_longer(cols=colnames(temp)[colnames(temp) %nin% "regulator"],names_to = "mode",values_to = "gene_id")
temp<-temp[!is.na(temp$gene_id),]
for(i in 1:nrow(temp)){
  temp9<-str_split(temp$gene_id[i],",")[[1]]
  temp10<-createEmptyDf(nrow=length(temp9),ncol=ncol(temp),colnames=colnames(temp))
  temp10$regulator<-temp$regulator[i]
  temp10$mode<-temp$mode[i]
  temp10$gene_id<-temp9
  if(i==1){
    temp11<-temp10
  }else{
    temp11<-bind_rows(temp11,temp10)
  }
}
dtrn_slmlt<-left_join(temp11,TRN,by=c("regulator","gene_id"))
dtrget_slmlt<-temp4



Sys.time()
```
## 1.3. Prepare KEGG pathways
```{r Prep_gseKEGG,warning=FALSE, message=FALSE,fig.height=4,fig.width=4,dpi=300}

dkegg.each  <- keggLink("pathway", "eco") %>%
  tibble(Pathway_ID = sub("path:", "", .), Bnb = sub("eco:", "", names(.))) %>%
  as.data.frame()
dkegg.each<-dkegg.each[,which(colnames(dkegg.each) %in% qw(Pathway_ID,Bnb))]
dkegg.each<-left_join(dkegg.each,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(Bnb,EntrezID))],by="Bnb")
dkegg.each$PrsAbs<-"Abs"
dkegg.each[dkegg.each$Bnb %in% dBnb.comp$Bnb,]$PrsAbs<-"Prs" 

dkegg <- keggList("pathway", "eco") %>% 
    tibble(Pathway_ID = sub("path:", "", names(.)), Pathway_name = .) %>%
  as.data.frame()


dkegg$Pathway_name<-str_remove_all(dkegg$Pathway_name," - Escherichia coli K-12 MG1655")
dkegg$Pathway_name<-str_extract(dkegg$Pathway_name,"([^;]+)|(.*)")
dkegg$size<-0
dkegg$Bnb<-NA
dkegg$EntrezID<-NA

# Use genes which are present in dBnb.comp
for(i in 1:nrow(dkegg)){
  dkegg$Bnb[i]<-paste(unique(dkegg.each[(dkegg.each$Pathway_ID==dkegg$Pathway_ID[i])&(dkegg.each$PrsAbs=="Prs"),]$Bnb),collapse = ",")
  dkegg$size[i]<-str_count(dkegg$Bnb[i],",")+1
  dkegg$EntrezID[i]<-paste(unique(dkegg.each[(dkegg.each$Pathway_ID==dkegg$Pathway_ID[i])&(dkegg.each$PrsAbs=="Prs"),]$EntrezID),collapse = ",")
}

# Filter gene sets comprising of less than five genes or more than 300 genes out
dkegg<-dkegg[dkegg$size>4&dkegg$size<301,]
dkegg$label<-seq(1,nrow(dkegg))

dkegg.each.gsea<-dkegg.each[(dkegg.each$PrsAbs=="Prs")&dkegg.each$Pathway_ID %in% dkegg$Pathway_ID,]
dkegg.each.gsea<-dkegg.each.gsea[,which(colnames(dkegg.each.gsea) %in% qw(Pathway_ID,EntrezID))]

Sys.time()
```


#
#
#
# 2. The Mut dataset
Load the Mut dataset and evaluate variability in response to mutations (DMmut)
How to obtain transcritome data of the MA lineages
The mutaion accumulation (MA) experiment was started from two ancestral strains of MG1655_delta_dnaQ, AncHa and AncHb, respectively.
6 MA lineages (3 lineages/Anc) were subjected to transcriptome analysis. We have two time points for each linages, R10 and R30.
Frozen stocks were constructed two ways. First, the colony suspensions used in the MA experiments were mixed with glycerol. The mixtures were then transferred to LB liquid medium and shaken at 32C overnight. The overnight cultures were finally mixed with glycerol. These two types of mixtures were stored at -80C. The latter stocks were used for analysis of transcriptome in LB and growth in different conditions. The glycerol stock was thawed completely at room temperature and diluted with LB/glycerol solution (10-fold dilution). This diluted solution was directly stored at -80C without cultivation. The dilution was sent to Osaka RIKEN, and grown in LB at 32C for about 6 hours. Total mRNA was extracted from the grown culture according to a custom protocol in Osaka RIKEN.
## 2.0. Action to submit the log2-transformed quantile-normalized transcriptome data to GEO
```{r Mut,warning=FALSE, message=FALSE,fig.height=4,fig.width=4,dpi=300}
# transcriptome data
my_dat<-read_csv(ddatraw)
my_col<-colnames(my_dat[,2:ncol(my_dat)])
# log2-transformation
my_dat[,2:ncol(my_dat)]<-log2(my_dat[,2:ncol(my_dat)])
# quantile-normalization
my_dat[,2:ncol(my_dat)]<-normalize.quantiles(as.matrix(my_dat[,2:ncol(my_dat)]))
# save
# write_csv(my_dat,file=paste(getwd(),"Submit2GEO","log2_quant_norm_allprobes.csv",sep="/"))
Sys.time()
```

## 2.1. Load transcriptome dataset
Raw data was plotted.
Red lines represent the lower limit of measurement.
```{r Mut,warning=FALSE, message=FALSE,fig.height=4,fig.width=4,dpi=300}
# transcriptome data
dat<-read.csv(ddatraw,stringsAsFactors=F)
colnames(dat)<-c("ProbeName",datinfo$label)
# probe name of microarray
probe2id<-as.data.frame(read_csv(dprobe2id))
probe2id<-probe2id[,c(2:ncol(probe2id))]
colnames(probe2id)[colnames(probe2id)=="ECKnbs"]<-"ECKnb"
colnames(probe2id)[colnames(probe2id)=="JWnbs"]<-"JWnb"
colnames(probe2id)[colnames(probe2id)=="Bnbs"]<-"Bnb"


lowlim_lin<-3
lowlim_lin_hc<-100
# Check
ggscatter(dat,x="AncHa_rep1_0",y="H03_30",alpha=0.1, size = 0.2,
          subtitle="Raw, AncHa vs H03 at R30",
          xlab="mRNA expression level of Anc [a.u.]", 
          ylab="mRNA expression level of Evo [a.u.]")+
  xscale("log10", .format = TRUE)+
  yscale("log10", .format = TRUE)+
  # annotate("text",x=10,y=Inf,label= "AncHa vs H03 at R30",hjust = 0,vjust = 2.5)+
  geom_hline(yintercept = lowlim_lin,color=mycolpal[2])+
  geom_vline(xintercept = lowlim_lin,color=mycolpal[2])+
  geom_abline(intercept = 0, slope = 1,colour=mycolpal[3])

Sys.time()
```
## 2.2. Log2-transformation
https://nanx.me/ggsci/reference/pal_lancet.html
The raw linear data was log10-transformed.
```{r Mut,warning=FALSE, message=FALSE,fig.height=8,fig.width=8,dpi=300}
datlin<-left_join(dat,probe2id[(probe2id$IStag==F)&(probe2id$Count==1)&(!grepl("\\+",probe2id$ECKnb)),which(colnames(probe2id) %in% qw(ProbeName,ECKnb))],by="ProbeName")
datlin<-left_join(datlin,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(Start,End,Strand,Bnb,EntrezID,gene_biotype,gene,product,ECKnb))],by="ECKnb")
datlin<-dplyr::left_join(datlin,dess,by="ECKnb")
datlin$Essentiality<-as.character(datlin$Essentiality)
datlin$Essentiality[is.na(datlin$Essentiality)]<-"Nonessential"
datlin$Essentiality[datlin$gene_biotype!="protein_coding"]<-"Others"
datlin$Essentiality<-factor(datlin$Essentiality,levels = qw(Nonessential,Essential,Others))
datlin<-dplyr::left_join(datlin,ddnds,by="Bnb")
datlin<-dplyr::left_join(datlin,dkogr,by="ECKnb")
datlin[datlin$Essentiality=="Essential",]$GrowthRate<-0
datlin[datlin$Essentiality=="Essential",]$ODmax<-0

# omit #Bnb=NA
datlin<-datlin[!is.na(datlin$Bnb),] 
datlin<-datlin[datlin$Bnb %in% dBnb.comp$Bnb,] #<----------- use global Bnb

#
temp<-data.frame(Bnb=unique(datlin$Bnb),Count=1)
for(i in 1:nrow(temp)){
  temp$Count[i]<-nrow(datlin[datlin$Bnb==temp$Bnb[i],])
}
temp<-datlin[datlin$Bnb %in% temp[temp$Count>1,]$Bnb,]
datlin<-datlin[datlin$ProbeName %nin% temp[!grepl("[A-Z]$",temp$ProbeName),]$ProbeName,]


# log2-transgormation
datlog<-datlin
datlog[,c(2:17)]<-log2(datlog[,c(2:17)])
lowlim<-log2(lowlim_lin)
lowlim_hc<-log2(lowlim_lin_hc)
datlog.tmpA<-datlog[,c(2:3,6:8,12:14)]
colnames(datlog.tmpA)[c(1,2)]<-qw(AncHa_rep1,AncHa_rep2)
datlog.tmpB<-datlog[,c(4:5,9:11,15:17)]
colnames(datlog.tmpB)[c(1,2)]<-qw(AncHb_rep1,AncHb_rep2)
lowerfun <- function(data,mapping){
  ggplot(data = data, mapping = mapping)+
    geom_point(alpha = 0.1,size=0.1)+
    geom_hline(yintercept=lowlim,colour= mycolpal[2])+
    geom_vline(xintercept=lowlim,colour= mycolpal[2])+
    geom_abline(intercept = 0, slope = 1,colour=mycolpal[3])
}
# ggpairs(datlog.tmpA, lower = list(continuous = wrap(lowerfun)),title = "Unfiltered data for AncHa and its descendants")+theme_pubr()
# ggpairs(datlog.tmpB, lower = list(continuous = wrap(lowerfun)),title = "Unfiltered data for AncHb and its descendants")+theme_pubr()

Sys.time()
```
## 2.3. Filtering, Quantile-Normalization, Calculation of Anc mean
The data below the lower limit (lowlim) was removed. and the subset was subject to quantile normalization.
Check point 1
```{r Mut,warning=FALSE, message=FALSE,fig.height=4,fig.width=4,dpi=300}
# filtering
datlog.norm<-datlog
for (i in 2:17){
  datlog.norm[,i]<-ifelse(datlog.norm[,i]>=lowlim,datlog.norm[,i],NA)
}
# Quantile-normalization
temp<-normalize.quantiles(as.matrix(datlog.norm[,2:17]))
for (i in 2:17){
  datlog.norm[,i]<-temp[,i-1]
}
# Calculation of mean for each Anc
datlog.norm<-transform(datlog.norm,AncHa.mean=rowMeans(datlog.norm[,c(2,3)],na.rm=T),AncHb.mean=rowMeans(datlog.norm[,c(4,5)],na.rm=T))

MAlin.sel<-c(datinfo$label[datinfo$label %nin% c("AncHa_rep1_0","AncHa_rep2_0","AncHb_rep1_0","AncHb_rep2_0")],"AncHa.mean","AncHb.mean")

# Check
ggscatter(datlog.norm,x="AncHa_rep1_0",y="H03_30",alpha=0.1,size = 0.2)+
  xlab(expression(log[2]~"mRNA expression level of Anc [a.u.]"))+
  ylab(expression(log[2]~"mRNA expression level of Evo [a.u.]"))+
  # annotate("text",x=-Inf,y=Inf,label= "AncHa vs H03 at R30",hjust = -0.2,vjust = 2.5)+
  geom_hline(yintercept=lowlim,colour= mycolpal[2])+
  geom_vline(xintercept=lowlim,colour= mycolpal[2])+
  geom_abline(intercept = 0, slope = 1,colour=mycolpal[3])

Sys.time()
```
## 2.4. Calculation of Mean, Variance and StdDev
### 2.4.1. Calculated Separately
We have two groups, descendants of AncHa and those of AncHb.
Variance was calculated for each group first, then mean of variance was calculated.
```{r Mut,warning=FALSE, message=FALSE,fig.height=5,fig.width=5,dpi=300}
datlog.norm<-transform(datlog.norm,
                       HaDesc_10_Av=rowMeans(as.matrix(datlog.norm[,6:8],na.rm=T)),
                       HbDesc_10_Av=rowMeans(as.matrix(datlog.norm[,9:11],na.rm=T)),
                       HaDesc_30_Av=rowMeans(as.matrix(datlog.norm[,12:14],na.rm=T)),
                       HbDesc_30_Av=rowMeans(as.matrix(datlog.norm[,15:17],na.rm=T)),
                       HaDesc_10_Var=rowVars(as.matrix(datlog.norm[,6:8],na.rm=T)),
                       HbDesc_10_Var=rowVars(as.matrix(datlog.norm[,9:11],na.rm=T)),
                       HaDesc_30_Var=rowVars(as.matrix(datlog.norm[,12:14],na.rm=T)),
                       HbDesc_30_Var=rowVars(as.matrix(datlog.norm[,15:17],na.rm=T)),
                       HaDesc_10_SD=rowSds(as.matrix(datlog.norm[,6:8],na.rm=T)),
                       HbDesc_10_SD=rowSds(as.matrix(datlog.norm[,9:11],na.rm=T)),
                       HaDesc_30_SD=rowSds(as.matrix(datlog.norm[,12:14],na.rm=T)),
                       HbDesc_30_SD=rowSds(as.matrix(datlog.norm[,15:17],na.rm=T)))
datlog.norm<-transform(datlog.norm,
                       HabDesc_10_Var.sep=rowMeans(as.matrix(cbind(datlog.norm$HaDesc_10_Var,datlog.norm$HbDesc_10_Var),na.rm=T)),
                       HabDesc_30_Var.sep=rowMeans(as.matrix(cbind(datlog.norm$HaDesc_30_Var,datlog.norm$HbDesc_30_Var),na.rm=T)))
datlog.norm<-transform(datlog.norm,
                       HabDesc_10_SD.sep=datlog.norm$HabDesc_10_Var.sep^0.5,
                       HabDesc_30_SD.sep=datlog.norm$HabDesc_30_Var.sep^0.5)
Sys.time()
```
### 2.4.2. Pooled
Two populations were pooled for each round
Variance was calculated for each round.
```{r Mut,warning=FALSE, message=FALSE,fig.height=5,fig.width=5,dpi=300}                       
datlog.norm<-transform(datlog.norm,
                       AncHab.mean=rowMeans(as.matrix(datlog.norm[,2:5],na.rm=T)),
                       HabDesc_10_Av.pool=rowMeans(as.matrix(datlog.norm[,6:11],na.rm=T)),
                       HabDesc_30_Av.pool=rowMeans(as.matrix(datlog.norm[,12:17],na.rm=T)),
                       HabDesc_10_Var.pool=rowVars(as.matrix(datlog.norm[,6:11],na.rm=T)),
                       HabDesc_30_Var.pool=rowVars(as.matrix(datlog.norm[,12:17],na.rm=T)),
                       HabDesc_10_SD.pool=rowSds(as.matrix(datlog.norm[,6:11],na.rm=T)),
                       HabDesc_30_SD.pool=rowSds(as.matrix(datlog.norm[,12:17],na.rm=T)))
Sys.time()
```
## 2.5. Reproducibility
```{r Mut,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
my_R30<-qw(H03_30,H04_30,H06_30,H08_30,H10_30,H11_30)
## AncHa and its descendants
ggscatter(datlog.norm,x="AncHa_rep1_0",y="AncHa_rep2_0",
          alpha=0.1,size=0.2,
          cor.coef = T,cor.method = "spearman",
          legend.title="")+
  geom_abline(slope = 1,intercept = 0,color=mycolpal[2])+
  xlab(expression(atop(log[2]~"mRNA expression level", paste("of replicate 1 [a.u.]"))))+
  ylab(expression(atop(log[2]~"mRNA expression level", paste("of replicate 2 [a.u.]"))))

ggscatter(datlog.norm,x="AncHa.mean",y="H03_30",
          alpha=0.1,size=0.2,
          cor.coef = T,cor.method = "spearman",
          legend.title="")+
  geom_abline(slope = 1,intercept = 0,color=mycolpal[2])+
  xlab(expression(atop(log[2]~"mRNA expression level", paste("of AncHa [a.u.]"))))+
  ylab(expression(atop(log[2]~"mRNA expression level", paste("of AncHa_30 [a.u.]"))))

## AncHb and its descendants
ggscatter(datlog.norm,x="AncHb_rep1_0",y="AncHb_rep2_0",
          alpha=0.1,size=0.2,
          cor.coef = T,cor.method = "spearman",
          legend.title="")+
  geom_abline(slope = 1,intercept = 0,color=mycolpal[2])+
  xlab(expression(atop(log[2]~"mRNA expression level", paste("of replicate 1 [a.u.]"))))+
  ylab(expression(atop(log[2]~"mRNA expression level", paste("of replicate 2 [a.u.]"))))


ggscatter(datlog.norm,x="AncHb.mean",y="H08_30",
          alpha=0.1,size=0.2,
          cor.coef = T,cor.method = "spearman",
          legend.title="")+
  geom_abline(slope = 1,intercept = 0,color=mycolpal[2])+
  xlab(expression(atop(log[2]~"mRNA expression level", paste("of AncHb [a.u.]"))))+
  ylab(expression(atop(log[2]~"mRNA expression level", paste("of AncHb_30 [a.u.]"))))

Sys.time()
```
## 2.6. Calculate Running Median, Smooth spline and Distance from Median (DMs)
```{r Mut,warning=FALSE, message=FALSE}
temp<-datlog.norm
Meanlist<-qw(HabDesc_10_Av.pool,HabDesc_30_Av.pool)
SDlist<-qw(HabDesc_10_SD.sep,HabDesc_30_SD.sep,HabDesc_10_SD.pool,HabDesc_30_SD.pool)
RMlist<-qw(RM.SD.sep.R10,RM.SD.sep.R30,RM.SD.pool.R10,RM.SD.pool.R30)
DMlist<-qw(DM.SD.sep.R10,DM.SD.sep.R30,DM.SD.pool.R10,DM.SD.pool.R30)
SSlist<-qw(SS.SD.sep.R10,SS.SD.sep.R30,SS.SD.pool.R10,SS.SD.pool.R30)

## Action for R10
SDlist.R10<-createEmptyDf(nrow = nrow(datlog.norm),ncol = length(SDlist)/2,colnames = SDlist[c(1,3)])
for(i in 1:ncol(SDlist.R10)){
  SDlist.R10[,i]<-temp[,c(which(colnames(temp)==colnames(SDlist.R10)[i]))]
}
SDlist.R10<-transform(SDlist.R10,Bnb=temp$Bnb,logx=temp$HabDesc_10_Av.pool)
SDlist.R10<-SDlist.R10[!is.na(SDlist.R10$logx),] # remove NA
SDlist.R10<-SDlist.R10[order(SDlist.R10$logx),] # sorting 
RMlist.R10<-createEmptyDf(nrow = nrow(SDlist.R10),ncol = length(RMlist)/2,colnames = RMlist[c(1,3)])
SSvislist.R10<-createEmptyDf(nrow=1000,ncol = length(SSlist)/2,colnames = SSlist[c(1,3)])
SSvislist.R10<-transform(SSvislist.R10,logx=seq(min(SDlist.R10$logx),max(SDlist.R10$logx),length=nrow(SSvislist.R10)))
SSlist.R10<-createEmptyDf(nrow = nrow(SDlist.R10),ncol = length(SSlist)/2,colnames = SSlist[c(1,3)])
DMlist.R10<-createEmptyDf(nrow = nrow(SDlist.R10),ncol = length(DMlist)/2,colnames = DMlist[c(1,3)])
SPlist.R10<-list()
for(i in 1:ncol(RMlist.R10)){
  RMlist.R10[,i]<-runmed(SDlist.R10[,i],41)
  SPlist.R10[[i]]<-smooth.spline(data.frame(x=SDlist.R10$logx,y=RMlist.R10[,i]),df=12)
  SSvislist.R10[,i]<-predict(SPlist.R10[[i]],SSvislist.R10$logx)[2]
  SSlist.R10[,i]<-predict(SPlist.R10[[i]],SDlist.R10$logx)[2]
  DMlist.R10[,i]<-SDlist.R10[,i]-SSlist.R10[,i]
}
RMlist.R10<-transform(RMlist.R10,Bnb=SDlist.R10$Bnb)
SSlist.R10<-transform(SSlist.R10,Bnb=SDlist.R10$Bnb)
DMlist.R10<-transform(DMlist.R10,Bnb=SDlist.R10$Bnb)

## Action for R30
SDlist.R30<-createEmptyDf(nrow = nrow(datlog.norm),ncol = length(SDlist)/2,colnames = SDlist[c(2,4)])
for(i in 1:ncol(SDlist.R30)){
  SDlist.R30[,i]<-temp[,c(which(colnames(temp)==colnames(SDlist.R30)[i]))]
}
SDlist.R30<-transform(SDlist.R30,Bnb=temp$Bnb,logx=temp$HabDesc_30_Av.pool)
SDlist.R30<-SDlist.R30[!is.na(SDlist.R30$logx),] # remove NA
SDlist.R30<-SDlist.R30[order(SDlist.R30$logx),] # sorting 
RMlist.R30<-createEmptyDf(nrow = nrow(SDlist.R30),ncol = length(RMlist)/2,colnames = RMlist[c(2,4)])
SSvislist.R30<-createEmptyDf(nrow=1000,ncol = length(SSlist)/2,colnames = SSlist[c(2,4)])
SSvislist.R30<-transform(SSvislist.R30,logx=seq(min(SDlist.R30$logx),max(SDlist.R30$logx),length=nrow(SSvislist.R30)))
SSlist.R30<-createEmptyDf(nrow = nrow(SDlist.R30),ncol = length(SSlist)/2,colnames = SSlist[c(2,4)])
DMlist.R30<-createEmptyDf(nrow = nrow(SDlist.R30),ncol = length(DMlist)/2,colnames = DMlist[c(2,4)])
SPlist.R30<-list()
for(i in 1:ncol(RMlist.R30)){
  RMlist.R30[,i]<-runmed(SDlist.R30[,i],41)
  SPlist.R30[[i]]<-smooth.spline(data.frame(x=SDlist.R30$logx,y=RMlist.R30[,i]),df=12)
  SSvislist.R30[,i]<-predict(SPlist.R30[[i]],SSvislist.R30$logx)[2]
  SSlist.R30[,i]<-predict(SPlist.R30[[i]],SDlist.R30$logx)[2]
  DMlist.R30[,i]<-SDlist.R30[,i]-SSlist.R30[,i]
}
RMlist.R30<-transform(RMlist.R30,Bnb=SDlist.R30$Bnb)
SSlist.R30<-transform(SSlist.R30,Bnb=SDlist.R30$Bnb)
DMlist.R30<-transform(DMlist.R30,Bnb=SDlist.R30$Bnb)

datlog.norm<-dplyr::left_join(datlog.norm,RMlist.R10,by="Bnb")
datlog.norm<-dplyr::left_join(datlog.norm,SSlist.R10,by="Bnb")
datlog.norm<-dplyr::left_join(datlog.norm,DMlist.R10,by="Bnb")
datlog.norm<-dplyr::left_join(datlog.norm,RMlist.R30,by="Bnb")
datlog.norm<-dplyr::left_join(datlog.norm,SSlist.R30,by="Bnb")
datlog.norm<-dplyr::left_join(datlog.norm,DMlist.R30,by="Bnb")
# head(datlog.norm)
Sys.time()
```
## 2.7. Labelig according to DM
```{r Mut,warning=FALSE, message=FALSE}
DMFlaglist<-paste(DMlist,".Flag",sep="")
DMthr.list<-list()
for (i in 1:length(DMFlaglist)){
  DMthr.list[[i]]<-quantile(datlog.norm[,c(which(colnames(datlog.norm)==DMlist[i]))], c(0.33, 0.66),na.rm = T) # L,M,H
}
temp<-createEmptyDf(nrow=nrow(datlog.norm),ncol=length(DMFlaglist),colnames = DMFlaglist)
for(i in 1:length(DMlist)){
  temp1<-datlog.norm[,c(which(colnames(datlog.norm)==DMlist[i]))]
  temp[,c(which(colnames(temp)==DMFlaglist[i]))]<-ifelse(temp1<DMthr.list[[i]][[1]],"L",                                          ifelse(temp1>=DMthr.list[[i]][[1]]&temp1<DMthr.list[[i]][[2]],"M",
                        ifelse(temp1>=DMthr.list[[i]][[2]],"H",NA)))
  temp[,c(which(colnames(temp)==DMFlaglist[i]))]<-factor(temp[,c(which(colnames(temp)==DMFlaglist[i]))],
                                                         levels = c("L", "M", "H"))
}
datlog.norm<-cbind(datlog.norm,temp)
# head(datlog.norm)
Sys.time()
```
## 2.8. Classification based on mean expression level
We would like to calculate mean/median of DMs for gated subpopulations of the cells within narrow windows of expression levels separated by thin vertical lines, where the window size is 0.5.
```{r Mut,warning=FALSE, message=FALSE}
mywidth<-0.5
datlog.norm<-transform(datlog.norm,exp.class.R10=cut_width(HabDesc_10_Av.pool,mywidth,labels=FALSE,boundary = mywidth))
datlog.norm<-transform(datlog.norm,exp.class.R30=cut_width(HabDesc_30_Av.pool,mywidth,labels=FALSE,boundary = mywidth))

Sys.time()
```
## 2.9. Define DMmut
DMmut was defined as DM.SD.sep.R30
```{r Mut,warning=FALSE, message=FALSE}
datlog.norm$DMmut<-datlog.norm$DM.SD.sep.R30
datlog.norm$SDmut<-datlog.norm$HabDesc_30_SD.sep
datlog.norm$MEANmut<-datlog.norm$HabDesc_30_Av.pool
datlog.norm$RMmut<-datlog.norm$RM.SD.sep.R30
datlog.norm$SSmut<-datlog.norm$SS.SD.sep.R30
Sys.time()
```
## 2.10. SDmut vs. MEANmut, DMmut vs. MEANmut
```{r Mut,warning=FALSE, message=FALSE,fig.height=3,fig.width=4,dpi=300}
my_fontsize<-16
p<-ggscatter(datlog.norm,x="MEANmut",y="SDmut", alpha = 0.3, size=1.5,fill="black",
              legend="",shape=21,stroke=0,cor.coef = T,cor.coeff.args = list(method = "spearman", label.x = 10.5, label.y=3, label.sep = "\n",size=5))+
              xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
              ylab(expression("Standard deviation [a.u.]"))+
  scale_x_continuous(expand = expansion(mult = c(0.1, 0.2)),breaks=get_breaks(from = 4,to = 16,by = 4))+
          geom_line(data=datlog.norm,aes(x=MEANmut,y=RMmut),colour="darkorange", size=0.7)+
          geom_line(data=SSvislist.R30,aes(x=logx,y=SSvislist.R30[,1]),colour="deepskyblue", size=1)
p<-ggpar(p,font.x=my_fontsize,font.y=my_fontsize,font.tickslab = my_fontsize)
p

p<-ggscatter(datlog.norm,x="MEANmut",y="DMmut", alpha = 0.3, size=1.5,fill="black",
          legend="",shape=21,stroke=0,
          cor.coef = T,
          cor.coeff.args = list(method = "spearman", label.x = 11, label.y=2.5, label.sep = "\n",size=5))+
    scale_x_continuous(expand = expansion(mult = c(0.1, 0.2)),breaks=get_breaks(from = 4,to = 16,by = 4))+
          xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
          ylab(expression(DM[mut]~"[a.u.]"))
p<-ggpar(p,font.x=my_fontsize,font.y=my_fontsize,font.tickslab = my_fontsize)
p
Sys.time()
```
## 2.11. Representative genes
```{r Mut,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
my_R30<-qw(H03_30,H04_30,H06_30,H08_30,H10_30,H11_30)
temp<-datlog.norm[datlog.norm$Bnb %in% qw(b3952,b3591,b4177,b3478,b0763,b0432),which(colnames(datlog.norm) %in% c("Bnb",my_R30))]
temp2<-temp %>% pivot_longer(cols=my_R30,names_to = "MAlineage",values_to ="log_ex_norm")
temp2<-left_join(temp2,datlog.norm[,which(colnames(datlog.norm) %in% qw(Bnb,gene))],by="Bnb")
temp2$gene<-factor(temp2$gene,levels=qw(pflC,nikC,selA,modA,purA,cyoA))
my_fontsize<-16
p<-ggerrorplot(temp2,x="gene",y="log_ex_norm",desc_stat = "mean_sd",
            add="jitter",add.params =list(size = 1, jitter = 0.3,alpha=0.4,color="black"),
            size=0.4,color=mycolpal[2],
            xlab = "")+
  # ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression(atop(log[2]~"mRNA", paste("expression level [a.u.]"))))+
  scale_y_continuous(breaks = get_breaks(by = 3, from = 0),limits =  c(2, 13))+
  font("x.text", face = "italic")+rotate_x_text()
p<-ggpar(p, xlab = FALSE, font.tickslab = list(size=my_fontsize),font.y = list(size=my_fontsize))
p
Sys.time()
```
## 2.12. Representative two genes
```{r Mut,warning=FALSE, message=FALSE,fig.height=3,fig.width=2,dpi=300}
temp<-datlog.norm[datlog.norm$Bnb %in% qw(b3591,b0763),which(colnames(datlog.norm) %in% c("Bnb",my_R30))]
temp2<-temp %>% pivot_longer(cols=my_R30,names_to = "MAlineage",values_to ="log_ex_norm")
temp2<-left_join(temp2,datlog.norm[,which(colnames(datlog.norm) %in% qw(Bnb,gene))],by="Bnb")
temp2$gene<-factor(temp2$gene,levels=qw(selA,modA))
p<-ggerrorplot(temp2,x="gene",y="log_ex_norm",desc_stat = "mean_sd",
            add="jitter",add.params =list(size = 1, jitter = 0.2,alpha=0.4,color="black"),
            size=0.4,color=mycolpal[2],
            xlab = "")+
  ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(6, 10))+
  font("x.text", face = "italic")+rotate_x_text()
p<-ggpar(p, xlab = FALSE)
p

temp2$Anc<-ifelse(temp2$MAlineage %in% qw(H03_30,H04_30,H06_30),"Anc_HA","Anc_HB")
p<-ggerrorplot(temp2,x="gene",y="log_ex_norm",desc_stat = "mean_sd",
            size=0.4,color="black")+
  ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
  geom_jitter(aes(color=Anc),alpha=0.5,width=0.2,size=0.9)+
  scale_color_manual(values = mycolpal[1:2])
ggpar(p,xlab=F,legend="",ylim=c(6,10))



Sys.time()
```
## 2.13. Output of dat.log.norm
```{r Mut,warning=FALSE, message=FALSE,fig.height=4,fig.width=3,dpi=300}
# write.csv(datlog.norm,"datlog_norm.csv",row.names=F)
Sys.time()
```



#
#
#
# 3. The Env dataset
Load the Env dataset (PRECISE2) and evaluate variability in response to environmental changes (DMenv)
https://imodulondb.org/dataset.html?organism=e_coli&dataset=precise2
https://doi.org/10.1101/2021.04.08.439047
https://github.com/SBRG/precise2/tree/master/data/regulation
## 3.1. Load Env dataset and TRN based on RegulonDB & Ecocyc
```{r Env,warning=FALSE, message=FALSE,fig.height=2,fig.width=2,dpi=300}
gene_info<-as.data.frame(read.csv(paste(getwd(),"ENV_DataSet","gene_files","gene_info.csv",sep="/"),stringsAsFactors = F,skip = 0))
colnames(gene_info)[1]<-"Bnb"
meta.pr2<-as.data.frame(read.csv(paste(getwd(),"ENV_DataSet","data_files","sample_table.csv",sep="/"),stringsAsFactors = F,skip = 0))
colnames(meta.pr2)[1]<-"sample_ID"
fileinfo_plasticity.pr2<-meta.pr2[(meta.pr2$Strain.Description=="Escherichia coli K-12 MG1655")&(meta.pr2$Evolved.Sample=="No"),]
log_tpm.pr2<-as.data.frame(read.csv(paste(getwd(),"ENV_DataSet","data_files","log_tpm_qc.csv",sep="/"),stringsAsFactors = F,skip = 0))
colnames(log_tpm.pr2)[1]<-"Bnb"
log_tpm.pr2<-log_tpm.pr2[log_tpm.pr2$Bnb %in% dBnb.comp$Bnb,] #<------------- Use global Bnb

Sys.time()
```
## 3.2 Quantile normalization and Calculate Plasticity (DMenv)
### 3.2.1. all conditions
```{r Env,warning=FALSE, message=FALSE,fig.height=1.5,fig.width=1.5,dpi=300}
# quantile normalization
temp<-normalize.quantiles(as.matrix(log_tpm.pr2[,2:ncol(log_tpm.pr2)]))
temp<-as.data.frame(temp)
colnames(temp)<-colnames(log_tpm.pr2)[2:ncol(log_tpm.pr2)]
log_tpm.pr2.norm<-transform(temp,Bnb=log_tpm.pr2$Bnb)
fileinfo_plasticity.pr2$sample_group=str_sub(fileinfo_plasticity.pr2$sample,start = 1,end = -4)
fileinfo_plasticity.pr2$condition_concat=NA
temp2<-c(which(colnames(fileinfo_plasticity.pr2)=="Base.Media"):which(colnames(fileinfo_plasticity.pr2)=="Culture.Type"))
temp2<-temp2[temp2!=which(colnames(fileinfo_plasticity.pr2) == "Antibiotic.for.selection")]
for (i in 1:nrow(fileinfo_plasticity.pr2)){
  fileinfo_plasticity.pr2$condition_concat[i]<-paste(fileinfo_plasticity.pr2[i,temp2],collapse="__")
}
print(unique(fileinfo_plasticity.pr2[fileinfo_plasticity.pr2$Carbon.Source..g.L.!="glucose?"&fileinfo_plasticity.pr2$Base.Media!="01xLB",]$condition_concat))
print(unique(fileinfo_plasticity.pr2[fileinfo_plasticity.pr2$Carbon.Source..g.L.!="glucose?"&fileinfo_plasticity.pr2$Base.Media!="01xLB",]$Supplement))

fileinfo_plasticity.pr2[grepl(pattern = "^[0-9]",fileinfo_plasticity.pr2$sample),]$sample<-paste("Temp",fileinfo_plasticity.pr2[grepl(pattern = "^[0-9]",fileinfo_plasticity.pr2$sample),]$sample,sep="_")
fileinfo_plasticity.pr2[grepl(pattern = "^[0-9]",fileinfo_plasticity.pr2$sample_group),]$sample_group<-paste("Temp",fileinfo_plasticity.pr2[grepl(pattern = "^[0-9]",fileinfo_plasticity.pr2$sample_group),]$sample_group,sep="_")

fileinfo_plasticity.pr2$sample<-str_replace_all(fileinfo_plasticity.pr2$sample,pattern = "\\-",replacement = "_")
fileinfo_plasticity.pr2$sample_group<-str_replace_all(fileinfo_plasticity.pr2$sample_group,pattern = "\\-",replacement = "_")

# write.csv(fileinfo_plasticity.pr2,"fileinfo_plasticity_pr2.csv",row.names=F)

## Activate if summing up replicates
# temp<-data.frame(sample_group=unique(fileinfo_plasticity.pr2$sample_group),reps=0,files=NA)
# for(i in 1:nrow(temp)){
#   temp1<-fileinfo_plasticity.pr2[fileinfo_plasticity.pr2$sample_group==temp$sample_group[i],]
#   temp$reps[i]<-nrow(temp1)
#   temp$files[i]=paste(temp1$sample_ID,collapse = ",")
# }
# log_tpm_pl.pr2<-createEmptyDf(nrow=nrow(log_tpm.pr2.norm),ncol=nrow(temp)+1,colnames=c("Bnb",temp$sample_group))
# log_tpm_pl.pr2$Bnb<-log_tpm.pr2.norm$Bnb
# i<-1
# for(i in 1:nrow(temp)){
#   temp1<-log_tpm.pr2.norm[,which(colnames(log_tpm.pr2.norm) %in% c(str_split(temp$files[i],",")[[1]]))]
#   temp1$mean<-rowMeans(as.matrix(temp1),na.rm=T)
#   log_tpm_pl.pr2[,i+1]<-temp1$mean
# }

log_tpm_pl.pr2<-createEmptyDf(nrow=nrow(log_tpm.pr2.norm),ncol=nrow(fileinfo_plasticity.pr2)+1,colnames=c("Bnb",fileinfo_plasticity.pr2$sample))
log_tpm_pl.pr2$Bnb<-log_tpm.pr2.norm$Bnb
for(i in 1:nrow(fileinfo_plasticity.pr2)){
  log_tpm_pl.pr2[,i+1]<-log_tpm.pr2.norm[,which(colnames(log_tpm.pr2.norm)==fileinfo_plasticity.pr2$sample_ID[i])]
}

prtemp2.pr2<-createEmptyDf(nrow=nrow(log_tpm_pl.pr2),ncol=4,colnames = qw(Bnb,PRMean,PRVar,PRSD))
prtemp2.pr2$Bnb<-log_tpm_pl.pr2$Bnb
temp<-log_tpm_pl.pr2[,!(colnames(log_tpm_pl.pr2) %in% "Bnb")]
for(i in 1:nrow(prtemp2.pr2)){
  prtemp2.pr2$PRMean[i]<-rowMeans(as.matrix(temp[i,]))
  prtemp2.pr2$PRVar[i]<-rowVars(as.matrix(temp[i,]))
  prtemp2.pr2$PRSD[i]<-rowSds(as.matrix(temp[i,]))
}

temp<-datlog.norm[,which(colnames(datlog.norm) %in% qw(gene,gene_biotype,EntrezID,Strand,product,Bnb,ECKnb,Essentiality))]
prtemp2.pr2<-dplyr::left_join(prtemp2.pr2,temp,by="Bnb")
prtemp2.pr2<-prtemp2.pr2[!is.na(prtemp2.pr2$gene),]

prtemp2.pr2<-prtemp2.pr2[!is.na(prtemp2.pr2$PRMean),]
prtemp2.pr2<-prtemp2.pr2[order(prtemp2.pr2$PRMean),]
prtemp2.pr2<-transform(prtemp2.pr2,PRRunMed=NA,PRSmoothSp=NA,PRDM=NA)
prtemp2.pr2$PRRunMed<-runmed(prtemp2.pr2$PRSD,41) #41
PRSSvislist<-createEmptyDf(nrow=1000,ncol=2, colnames = qw(logx,SS))
PRSSvislist$logx=seq(min(prtemp2.pr2$PRMean),max(prtemp2.pr2$PRMean),length=nrow(PRSSvislist))
misplist<-list()
misplist[[1]]<-smooth.spline(data.frame(x=prtemp2.pr2$PRMean,y=prtemp2.pr2$PRRunMed),df=12)
PRSSvislist$SS<-predict(misplist[[1]],PRSSvislist$logx)[[2]]
prtemp2.pr2$PRSmoothSp<-predict(misplist[[1]],prtemp2.pr2$PRMean)[[2]]
prtemp2.pr2$PRDM<-prtemp2.pr2$PRSD-prtemp2.pr2$PRSmoothSp  # <------------------------------ DMenv
temp<-datlog.norm[,which(colnames(datlog.norm) %in% c("Bnb","HabDesc_30_SD.sep","HabDesc_30_SD.pool","DM.SD.sep.R30","DM.SD.pool.R30"))]
prtemp2.pr2<-dplyr::left_join(prtemp2.pr2,temp,by="Bnb")

Pthr<-quantile(prtemp2.pr2$PRDM, c(0.33, 0.66),na.rm=T)
prtemp2.pr2<-transform(prtemp2.pr2,PFlag=ifelse(prtemp2.pr2$PRDM<Pthr[[1]],"L",
                                    ifelse(prtemp2.pr2$PRDM>=Pthr[[1]]&prtemp2.pr2$PRDM<Pthr[[2]],"M",
                                    ifelse(prtemp2.pr2$PRDM>=Pthr[[2]],"H","NA"))))
prtemp2.pr2$PFlag<-factor(prtemp2.pr2$PFlag, levels = c("L", "M", "H"))
prtemp2.pr2_wo_abx<-

Sys.time()
```
### 3.2.2. w/o the abx project
```{r Env,warning=FALSE, message=FALSE,fig.height=1.5,fig.width=1.5,dpi=300}
# remove the abx project
fileinfo_plasticity.pr2_wo_abx<-fileinfo_plasticity.pr2[fileinfo_plasticity.pr2$project %nin% "abx_media",]
# quantile normalization
my_tmp<-log_tpm.pr2[,colnames(log_tpm.pr2) %in% fileinfo_plasticity.pr2_wo_abx$sample_ID]
my_cln<-colnames(my_tmp)
my_tmp<-normalize.quantiles(as.matrix(my_tmp)) %>% as.data.frame()
colnames(my_tmp)<-my_cln
my_tmp$Bnb<-log_tpm.pr2$Bnb
log_tpm.pr2.norm_wo_abx<-my_tmp %>% select(Bnb,everything())
my_tmp<-data.frame(sample_ID=colnames(log_tpm.pr2.norm_wo_abx)[2:ncol(log_tpm.pr2.norm_wo_abx)])
my_tmp<-left_join(my_tmp,fileinfo_plasticity.pr2_wo_abx)
log_tpm_pl.pr2_wo_abx<-log_tpm.pr2.norm_wo_abx
colnames(log_tpm_pl.pr2_wo_abx)<-c("Bnb",my_tmp$sample)
# calc DMenv_wo_abx



prtemp2.pr2_wo_abx<-createEmptyDf(nrow=nrow(log_tpm_pl.pr2_wo_abx),ncol=4,colnames = qw(Bnb,PRMean,PRVar,PRSD))
prtemp2.pr2_wo_abx$Bnb<-log_tpm_pl.pr2_wo_abx$Bnb
temp<-log_tpm_pl.pr2_wo_abx[,!(colnames(log_tpm_pl.pr2_wo_abx) %in% "Bnb")]
for(i in 1:nrow(prtemp2.pr2_wo_abx)){
  prtemp2.pr2_wo_abx$PRMean[i]<-rowMeans(as.matrix(temp[i,]))
  prtemp2.pr2_wo_abx$PRVar[i]<-rowVars(as.matrix(temp[i,]))
  prtemp2.pr2_wo_abx$PRSD[i]<-rowSds(as.matrix(temp[i,]))
}

temp<-datlog.norm[,which(colnames(datlog.norm) %in% qw(gene,gene_biotype,EntrezID,Strand,product,Bnb,ECKnb,Essentiality))]
prtemp2.pr2_wo_abx<-dplyr::left_join(prtemp2.pr2_wo_abx,temp,by="Bnb")
prtemp2.pr2_wo_abx<-prtemp2.pr2_wo_abx[!is.na(prtemp2.pr2_wo_abx$gene),]

prtemp2.pr2_wo_abx<-prtemp2.pr2_wo_abx[!is.na(prtemp2.pr2_wo_abx$PRMean),]
prtemp2.pr2_wo_abx<-prtemp2.pr2_wo_abx[order(prtemp2.pr2_wo_abx$PRMean),]
prtemp2.pr2_wo_abx<-transform(prtemp2.pr2_wo_abx,PRRunMed=NA,PRSmoothSp=NA,PRDM=NA)
prtemp2.pr2_wo_abx$PRRunMed<-runmed(prtemp2.pr2_wo_abx$PRSD,41) #41
PRSSvislist<-createEmptyDf(nrow=1000,ncol=2, colnames = qw(logx,SS))
PRSSvislist$logx=seq(min(prtemp2.pr2_wo_abx$PRMean),max(prtemp2.pr2_wo_abx$PRMean),length=nrow(PRSSvislist))
misplist<-list()
misplist[[1]]<-smooth.spline(data.frame(x=prtemp2.pr2_wo_abx$PRMean,y=prtemp2.pr2_wo_abx$PRRunMed),df=12)
PRSSvislist$SS<-predict(misplist[[1]],PRSSvislist$logx)[[2]]
prtemp2.pr2_wo_abx$PRSmoothSp<-predict(misplist[[1]],prtemp2.pr2_wo_abx$PRMean)[[2]]
prtemp2.pr2_wo_abx$PRDM<-prtemp2.pr2_wo_abx$PRSD-prtemp2.pr2_wo_abx$PRSmoothSp  # <------------------------------ DMenv
temp<-datlog.norm[,which(colnames(datlog.norm) %in% c("Bnb","HabDesc_30_SD.sep","HabDesc_30_SD.pool","DM.SD.sep.R30","DM.SD.pool.R30"))]
prtemp2.pr2_wo_abx<-dplyr::left_join(prtemp2.pr2_wo_abx,temp,by="Bnb")

Pthr<-quantile(prtemp2.pr2_wo_abx$PRDM, c(0.33, 0.66),na.rm=T)
prtemp2.pr2_wo_abx<-transform(prtemp2.pr2_wo_abx,PFlag=ifelse(prtemp2.pr2_wo_abx$PRDM<Pthr[[1]],"L",
                                    ifelse(prtemp2.pr2_wo_abx$PRDM>=Pthr[[1]]&prtemp2.pr2_wo_abx$PRDM<Pthr[[2]],"M",
                                    ifelse(prtemp2.pr2_wo_abx$PRDM>=Pthr[[2]],"H","NA"))))
prtemp2.pr2_wo_abx$PFlag<-factor(prtemp2.pr2_wo_abx$PFlag, levels = c("L", "M", "H"))

Sys.time()
```
## 3.3 Define DMenv
```{r Env,warning=FALSE, message=FALSE}
prtemp2.pr2$DMenv=prtemp2.pr2$PRDM
prtemp2.pr2$SDenv=prtemp2.pr2$PRSD
prtemp2.pr2$MEANenv=prtemp2.pr2$PRMean
prtemp2.pr2$RMenv=prtemp2.pr2$PRRunMed
prtemp2.pr2$SSenv=prtemp2.pr2$PRSmoothSp

prtemp2.pr2_wo_abx$DMenv_wo_abx=prtemp2.pr2_wo_abx$PRDM
prtemp2.pr2_wo_abx$SDenv_wo_abx=prtemp2.pr2_wo_abx$PRSD
prtemp2.pr2_wo_abx$MEANenv_wo_abx=prtemp2.pr2_wo_abx$PRMean
prtemp2.pr2_wo_abx$RMenv_wo_abx=prtemp2.pr2_wo_abx$PRRunMed
prtemp2.pr2_wo_abx$SSenv_wo_abx=prtemp2.pr2_wo_abx$PRSmoothSp
Sys.time()
```
## 3.4 SDenv vs. MEANenv, DMenv vs. MEANenv
```{r Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=4,dpi=300}
my_fontsize<-16
# Mean vs SD in PR2
p<-ggscatter(prtemp2.pr2,x="MEANenv",y="SDenv", alpha = 0.3,size=1.5,
          legend="",shape=21,stroke=0,
          fill="black",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman", label.x = 9, label.y=4, label.sep = "\n",size=5))+
            xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
            ylab(expression("Standard deviation [a.u.]"))+
  scale_x_continuous(expand = expansion(mult = c(0.1, 0.23)),breaks=get_breaks(from = 0,to = 16,by = 4))+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
          geom_line(data=prtemp2.pr2,aes(x=MEANenv,y=RMenv),colour="darkorange", size=0.7)+
          geom_line(data=PRSSvislist,aes(x=logx,y=SS),colour="deepskyblue", size=1)
p<-ggpar(p,font.x=my_fontsize,font.y=my_fontsize,font.tickslab = my_fontsize)
p

# Mean vs DM in PR2
p<-ggscatter(prtemp2.pr2,x="MEANenv",y="DMenv", alpha = 0.3,size=1.5,
          legend="",shape=21,stroke=0,
          fill="black",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman", label.x = 8.5, label.y=3, label.sep = "\n",size=5))+
    scale_x_continuous(expand = expansion(mult = c(0.1, 0.23)),breaks=get_breaks(from = 0,to = 16,by = 4))+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
          xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
          ylab(expression(DM[env]~"[a.u.]"))
p<-ggpar(p,font.x=my_fontsize,font.y=my_fontsize,font.tickslab = my_fontsize)
p

Sys.time()
```
## 3.5 Representative genes
```{r Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp<-log_tpm_pl.pr2[log_tpm_pl.pr2$Bnb %in% qw(b3952,b3591,b4177,b3478,b0763,b0432),]
temp2<-temp %>% pivot_longer(cols=colnames(temp)[2:ncol(temp)],names_to = qw(Condition),values_to =qw(log_tpm_norm))
temp2<-left_join(temp2,dBnb[,which(colnames(dBnb) %in% qw(Bnb,gene))],by="Bnb")
temp2$gene<-factor(temp2$gene,levels=qw(pflC,nikC,selA,modA,purA,cyoA))
my_fontsize<-16
p<-ggerrorplot(temp2,x="gene",y="log_tpm_norm",desc_stat = "mean_sd",
            add="jitter",add.params =list(size = 0.4, jitter = 0.3,alpha=0.4,color="black"),
            size=0.4,color=mycolpal[2])+
  # ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
    ylab(expression(atop(log[2]~"mRNA", paste("expression level [a.u.]"))))+
font("x.text", face = "italic")+rotate_x_text()
p<-ggpar(p, xlab = FALSE, font.tickslab = list(size=my_fontsize),font.y = list(size=my_fontsize))
p
Sys.time()
```
## 3.6 Representative two genes
```{r Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=2,dpi=300}
temp<-log_tpm_pl.pr2[log_tpm_pl.pr2$Bnb %in% qw(b3591,b0763),]
temp2<-temp %>% pivot_longer(cols=colnames(temp)[2:ncol(temp)],names_to = qw(Condition),values_to =qw(log_tpm_norm))
temp2<-left_join(temp2,dBnb[,which(colnames(dBnb) %in% qw(Bnb,gene))],by="Bnb")
temp2$gene<-factor(temp2$gene,levels=qw(selA,modA))

p<-ggerrorplot(temp2,x="gene",y="log_tpm_norm",desc_stat = "mean_sd",
            add="jitter",add.params =list(size = 0.4, jitter = 0.2,alpha=0.4,color="black"),
            size=0.4,color=mycolpal[2])+
  ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
font("x.text", face = "italic")+rotate_x_text()
p<-ggpar(p, xlab = FALSE)
p

p<-ggerrorplot(temp2,x="gene",y="log_tpm_norm",desc_stat = "mean_sd",
            size=0.4,color="black")+
  ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
  geom_jitter(aes(color=Condition),alpha=0.3,width=0.2,size=0.3)
ggpar(p,xlab=F,legend="")



Sys.time()
```


#
#
#
# 4. The Evo dataset
Load the Evo dataset and evaluate variability in evolution (DMevo)
Transcriptome data of 16 E. coli species in LB
https://www.nature.com/articles/s41564-020-00839-y#MOESM4
## 4.1. Load Table S6 RNA-seq, 41-1Ti9 was renamed to 41_1Ti9
```{r Evo,warning=FALSE, message=FALSE,fig.height=1.5,fig.width=2.2,dpi=300}
ddgloess<-paste(getwd(),"EVO_DataSet","TableS5.csv",sep="/")
dgloess<-data.frame(read_csv(ddgloess))
colnames(dgloess)<-qw(alos_LB,all_LB,alos_M9,all_M9,alos_GMM,all_GMM)
ddcol16<-paste(getwd(),"EVO_DataSet","TableS6.csv",sep="/")
dcol16<-data.frame(read_csv(ddcol16))
col16.strain<-colnames(dcol16)[colnames(dcol16)!="Gene"]
dcol16<-left_join(dcol16,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(gene,Bnb,ECKnb,gene_biotype,EntrezID))],by=c("Gene"="gene"))

temp<-data.frame(Gene=dcol16[is.na(dcol16$Bnb),]$Gene,gene_synonym=NA)
temp1<-genename.mg1655[genename.mg1655$Bnb %nin% dcol16[!is.na(dcol16$Bnb),]$Bnb,]$gene_synonym
temp1<-temp1[grepl(",",temp1)]
for(i in 1:nrow(temp)){
  rm(temp2)
  temp2<-temp1[grepl(temp$Gene[i],temp1)]
  if(length(temp2)>0){
    temp$gene_synonym[i]<-temp1[grepl(temp$Gene[i],temp1)]
  }
}
temp$ECKnb<-str_extract(temp$gene_synonym,"ECK[0-9]{4}")
temp<-temp[!is.na(temp$ECKnb),which(colnames(temp) %nin% "gene_synonym")]
temp<-left_join(temp,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(gene,ECKnb))],by="ECKnb")
temp1<-dcol16[!is.na(dcol16$Bnb),]
temp2<-dcol16[is.na(dcol16$Bnb),]
temp2<-temp2[temp2$Gene %in% temp$Gene,]
for(i in 1:nrow(temp2)){
  temp2$Gene[i]<-temp[temp$Gene==temp2$Gene[i],]$gene
}
temp2<-temp2[,which(colnames(temp2) %nin% qw(gene,Bnb,ECKnb,gene_biotype,EntrezID))]
temp2<-left_join(temp2,genename.mg1655[,which(colnames(genename.mg1655) %in% qw(gene,Bnb,ECKnb,gene_biotype,EntrezID))],by=c("Gene"="gene"))
dcol16<-bind_rows(temp1,temp2)
dcol16<-left_join(dcol16,dess,by="ECKnb")
dcol16[is.na(dcol16$Essentiality),]$Essentiality<-"Nonessential"
dcol16$GlobalEssentiality<-"Nonessential"
dcol16[dcol16$Gene %in% dgloess[!is.na(dgloess$all_LB),]$all_LB,]$GlobalEssentiality<-"Essential"
dcol16<-dcol16[!is.na(dcol16$Bnb),]
dcol16<-dcol16[dcol16$Bnb %in% dBnb.comp$Bnb,] #<------------- Use global Bnbs
temp<-dcol16[,which(colnames(dcol16) %nin% qw(Gene,Bnb,ECKnb,gene_biotype,EntrezID,Essentiality,GlobalEssentiality))]
temp1<-createEmptyDf(nrow=nrow(dcol16),ncol = 3,colnames = qw(mean,var,sd))
temp1$mean<-rowMeans(as.matrix(temp),na.rm=T)
temp1$var<-rowVars(as.matrix(temp),na.rm=T)
temp1$sd<-rowSds(as.matrix(temp),na.rm=T)
dcol16<-cbind(dcol16,temp1)

Sys.time()
```
## 4.2. Quantile normalization for log rpkm (reads-per-kilobase-per-million)
```{r Evo,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=3,dpi=300}
temp<-dcol16[,which(colnames(dcol16) %in% col16.strain)]
temp<-normalize.quantiles(as.matrix(temp))
colnames(temp)<-col16.strain
temp1<-createEmptyDf(nrow=nrow(dcol16),ncol = 3,colnames = qw(mean,var,sd))
temp1$mean<-rowMeans(as.matrix(temp),na.rm=T)
temp1$var<-rowVars(as.matrix(temp),na.rm=T)
temp1$sd<-rowSds(as.matrix(temp),na.rm=T)
dcol16.qnorm<-cbind(dcol16[,which(colnames(dcol16) %in% qw(Gene,Bnb,ECKnb,gene_biotype,EntrezID,Essentiality,GlobalEssentiality))],temp,temp1)

dcol16.qnorm<-dcol16.qnorm[order(dcol16.qnorm$mean),] # sorting 
dcol16.qnorm<-transform(dcol16.qnorm,RunmedVar=runmed(dcol16.qnorm$sd,41),lmg=dcol16.qnorm$mean)
lmgn<-data.frame(x=dcol16.qnorm$lmg,y=dcol16.qnorm$RunmedVar)
sp<-smooth.spline(lmgn,df=12)
x<-seq(min(lmgn$x),max(lmgn$x),length=1000)
pred<-as.data.frame(x)
pred<-transform(pred,smsp=predict(sp,x)[2])
colnames(pred)<-c("x","smoothsp")
pred2<-dcol16.qnorm$mean
pred2<-transform(pred2,y=predict(sp,pred2)[2])
colnames(pred2)<-c("x","smoothsp")
pred_evo<-pred2
dcol16.qnorm<-transform(dcol16.qnorm,SmoothSpline=pred2$smoothsp)
dcol16.qnorm<-transform(dcol16.qnorm,DMevo=dcol16.qnorm$sd-dcol16.qnorm$SmoothSpline) # calculate distance from smooth spline to SD 

temp<-datlog.norm[,which(colnames(datlog.norm) %in% c("Bnb","HabDesc_30_SD.sep","HabDesc_30_SD.pool","DM.SD.sep.R30","DM.SD.pool.R30"))]
dcol16.qnorm<-left_join(dcol16.qnorm,temp,by="Bnb")
dcol16.qnorm<-left_join(dcol16.qnorm,prtemp2.pr2[,which(colnames(prtemp2.pr2) %in% qw(Bnb,PRDM))],by="Bnb")

Ethr<-quantile(dcol16.qnorm$DMevo, c(0.33, 0.66),na.rm=T)
dcol16.qnorm<-transform(dcol16.qnorm,EFlag=ifelse(dcol16.qnorm$DMevo<Ethr[[1]],"L",
                                    ifelse(dcol16.qnorm$DMevo>=Ethr[[1]]&dcol16.qnorm$DMevo<Ethr[[2]],"M",
                                    ifelse(dcol16.qnorm$DMevo>=Ethr[[2]],"H","NA"))))
dcol16.qnorm$EFlag<-factor(dcol16.qnorm$EFlag, levels = c("L", "M", "H"))

# ggscatter(dcol16.qnorm,x="PRDM",y="DMevo",
#           # color="GlobalEssentiality",palette = rev(mycolpal),
#           cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
#           cor.coeff.args = list(method = "spearman", label.x = -0.5, label.y=2.2, label.sep = "\n"),
#           alpha = 0.2, size=0.3,legend="")+
#   labs(x="DMenv [a.u]",y="DMevo [a.u.]")
# 
# ggscatter(dcol16.qnorm,x="DM.SD.sep.R30",y="DMevo",
#           # color="GlobalEssentiality",palette = rev(mycolpal),
#           cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
#           cor.coeff.args = list(method = "spearman", label.x = 2.2, label.y=2.2, label.sep = "\n"),
#           alpha = 0.2, size=0.3,legend="")+
#   # scale_x_continuous(breaks = get_breaks(by = 0.5, from = 0),limits =  c(-0.15, 1))+
#   labs(x="DMmut [a.u]",y="DMevo [a.u.]")

Sys.time()
```
## 4.3. Define DMevo
```{r Evo,warning=FALSE, message=FALSE}
dcol16.qnorm$DMevo=dcol16.qnorm$DMevo
dcol16.qnorm$SDevo=dcol16.qnorm$sd
dcol16.qnorm$MEANevo=dcol16.qnorm$mean
dcol16.qnorm$RMevo=dcol16.qnorm$RunmedVar
dcol16.qnorm$SSevo=dcol16.qnorm$SmoothSpline
Sys.time()
```
## 4.4. SDevo vs. MEANevo, DMevo vs. MEANevo
```{r Evo,warning=FALSE, message=FALSE,fig.height=3,fig.width=4,dpi=300}
my_fontsize<-16
p<-ggscatter(dcol16.qnorm,x="MEANevo",y="SDevo", alpha = 0.3, size=1.5,fill="black",
              legend="",shape=21,stroke=0,
              cor.coef = T,
              cor.coeff.args = list(method = "spearman", label.x = 9, label.y=3, label.sep = "\n",size=5))+
    scale_x_continuous(expand = expansion(mult = c(0.1, 0.2)),breaks=get_breaks(from = 0,to = 16,by = 4))+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
              xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
              ylab(expression("Standard deviation [a.u.]"))+
          geom_line(data=dcol16.qnorm,aes(x=MEANevo,y=RMevo),colour="darkorange", size=0.7)+
          geom_line(data=pred_evo,aes(x=x,y=smoothsp), colour="deepskyblue", size=1) # "deepskyblue"
p<-ggpar(p,font.x=my_fontsize,font.y=my_fontsize,font.tickslab = my_fontsize)
p

p<-ggscatter(dcol16.qnorm,x="MEANevo",y="DMevo", alpha = 0.3, size=1.5,fill="black",
          legend="",shape=21,stroke=0,
          cor.coef = T,
          cor.coeff.args = list(method = "spearman", label.x = 10, label.y=3, label.sep = "\n",size=5))+
    scale_x_continuous(expand = expansion(mult = c(0.1, 0.2)),breaks=get_breaks(from = 0,to = 16,by = 4))+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
          xlab(expression(log[2]~"mRNA expression level"))+
          ylab(expression(DM[evo]~"[a.u.]"))
p<-ggpar(p,font.x=my_fontsize,font.y=my_fontsize,font.tickslab = my_fontsize)
p

Sys.time()
```
## 4.5. Representative genes in Eco16
```{r Evo,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
my_fontsize<-16
temp<-dcol16.qnorm[dcol16.qnorm$Bnb %in% qw(b3952,b3591,b4177,b3478,b0763,b0432),]
temp2<-temp %>% pivot_longer(cols=colnames(temp)[8:23],names_to = qw(Condition),values_to =qw(log_rpkm_norm))
temp2$Gene<-factor(temp2$Gene,levels=qw(pflC,nikC,selA,modA,purA,cyoA))
p<-ggerrorplot(temp2,x="Gene",y="log_rpkm_norm",desc_stat = "mean_sd",
            add="jitter",add.params =list(size = 1, jitter = 0.3,alpha=0.4,color="black"),
            size=0.4,color=mycolpal[2])+
  scale_y_continuous(breaks = get_breaks(by = 3, from = 0),limits =  c(0, 13))+
  # ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression(atop(log[2]~"mRNA", paste("expression level [a.u.]"))))+
font("x.text", face = "italic")+rotate_x_text()
p<-ggpar(p, xlab = FALSE, font.tickslab = list(size=my_fontsize),font.y = list(size=my_fontsize))
p

Sys.time()
```


## 4.6. Phylogenetic tree
```{r Evo,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
strain.tbl<-data.frame(read_csv(paste(getwd(),"EVO_DataSet","StrainTable.csv",sep="/")))
myTree <- ape::read.tree(text='(MG1655:0.06630,E1114:0.05239,(HS:0.05242,((H120:0.04415,E1167:0.04161)1.000:0.05385,(((TA280:0.12521,(M114:0.12137,(TA249:0.07872,41-1Ti9:0.07266)1.000:0.05107)1.000:0.02434)1.000:0.05772,(ROAR8:0.14647,(JJ1886:0.07611,(CFT073:0.05805,(UTI89:0.00117,(APECO1:0.00212,S88:0.01318)1.000:0.00164)1.000:0.05856)1.000:0.03941)1.000:0.16718)1.000:0.05304)1.000:0.09489,(TA447:0.09574,(TA054:0.08624,E101:0.08764)1.000:0.02826)1.000:0.03405)1.000:0.04809)1.000:0.03715)1.000:0.02048);')
to_drop <- strain.tbl[is.na(strain.tbl$Transcriptome),]$Strain.name
myTree<-ape::drop.tip(myTree, to_drop)
strain.tbl<-strain.tbl[!is.na(strain.tbl$Transcriptome),]
p<-ggtree(myTree)+ geom_treescale(fontsize=3, linesize=1, offset=0.5,x=0.5,y=1)
p<-p %<+% strain.tbl + 
  geom_tiplab(aes(color = factor(Phylogroup)),offset=.01)+
  xlim(0, 0.72) 
p<-ggpar(p,legend.title = "")
p


temp<-ape::cophenetic.phylo(myTree)
myTree2<-phangorn::upgma(temp)
p2<-ggtree(myTree2, branch.length='none')
p2<-p2 %<+% strain.tbl + 
  geom_tiplab(aes(color = factor(Phylogroup)),offset=.01)+
  xlim(0, 10) 
p2<-ggpar(p2,legend="")
p2

Sys.time()
```
## 4.5.b Representative two genes in Eco16
```{r Evo,warning=FALSE, message=FALSE,fig.height=3,fig.width=2,dpi=300}
temp<-dcol16.qnorm[dcol16.qnorm$Bnb %in% qw(b3591,b0763),]
temp2<-temp %>% pivot_longer(cols=colnames(temp)[8:23],names_to = qw(Condition),values_to =qw(log_rpkm_norm))
temp2$Gene<-factor(temp2$Gene,levels=qw(selA,modA))
my_strains<-strain.tbl %>% select(Strain.name,Phylogroup)
colnames(my_strains)[1]<-"Condition"
my_strains[my_strains$Condition=="41-1Ti9",]$Condition<-"X41_1Ti9"
temp2<-left_join(temp2,my_strains,by="Condition")

p<-ggerrorplot(temp2,x="Gene",y="log_rpkm_norm",desc_stat = "mean_sd",
            add="jitter",add.params =list(size = 1, jitter = 0.2,alpha=0.4,color="black"),
            size=0.4,color=mycolpal[2])+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(4, 9))+
  ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
font("x.text", face = "italic")+rotate_x_text()
p<-ggpar(p, xlab = FALSE)
p

p<-ggerrorplot(temp2,x="Gene",y="log_rpkm_norm",desc_stat = "mean_sd",
            size=0.4,color="black")+
  ylab(expression(log[2]~"mRNA expression level [a.u.]"))+
  geom_jitter(aes(color=factor(Phylogroup)),alpha=0.6,width=0.2,size=0.5)
ggpar(p,xlab=F,legend="")

Sys.time()
```
#
#
#
# 5. Comparison of the DM values between the three datasets
## 5.1. Comparison between DMenv, DMevo and DMmut at the gene levels
```{r All_DM,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=2.5,dpi=300}
df_DMall<-dBnb.comp
df_DMall<-left_join(df_DMall,datlog.norm[,which(colnames(datlog.norm) %in% qw(Bnb,DMmut))])
df_DMall<-left_join(df_DMall,prtemp2.pr2[,which(colnames(prtemp2.pr2) %in% qw(Bnb,DMenv))])
df_DMall<-left_join(df_DMall,dcol16.qnorm[,which(colnames(dcol16.qnorm) %in% qw(Bnb,DMevo))])
df_DMall<-left_join(df_DMall,prtemp2.pr2_wo_abx[,which(colnames(prtemp2.pr2_wo_abx) %in% qw(Bnb,DMenv_wo_abx))])

my_fontsize<-16
# DMenv vs. DMmut
p<-ggscatter(df_DMall,x="DMenv",y="DMmut",
          size=0.2,alpha=0.2,
          color="gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=5,label.x=0.75,label.y=2.8))+
          xlab(expression(DM[env]~"[a.u.]"))+
          ylab(expression(DM[mut]~"[a.u.]"))
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.tickslab = my_fontsize)
p

# DMenv vs. DMevo
p<-ggscatter(df_DMall,x="DMenv",y="DMevo",
          size=0.2,alpha=0.2,
          color="gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=5,label.x=-0.5,label.y=3))+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
          xlab(expression(DM[env]~"[a.u.]"))+
          ylab(expression(DM[evo]~"[a.u.]"))

p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.tickslab = my_fontsize)
p

# DMmut vs. DMevo
p<-ggscatter(df_DMall,x="DMmut",y="DMevo",
          size=0.2,alpha=0.2,
          color="gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=5,label.x=1,label.y=2.9))+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
          xlab(expression(DM[mut]~"[a.u.]"))+
          ylab(expression(DM[evo]~"[a.u.]"))
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.tickslab = my_fontsize)
p

Sys.time()
```
## 5.2. Comparison between DMenv, DMevo and DMmut at the functional category levels
```{r All_DM,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
df_DMall_cat<-dkegg
df_DMall_cat$DMenv<-NA
df_DMall_cat$DMevo<-NA
df_DMall_cat$DMmut<-NA
i<-1
for(i in 1:nrow(df_DMall_cat)){
  temp_genes<-str_split(df_DMall_cat$Bnb[i],",")[[1]]
  temp_df<-df_DMall[df_DMall$Bnb %in% temp_genes,]
  df_DMall_cat$DMenv[i]<-mean(temp_df$DMenv)
  df_DMall_cat$DMevo[i]<-mean(temp_df$DMevo)
  df_DMall_cat$DMmut[i]<-mean(temp_df$DMmut)
}
df_DMall_cat$label.sel<-""
# for display
mypathway<-qw(eco03010,eco00910,eco02030,eco00020,eco00290,eco00920,eco00052,eco02026,eco00240,eco00071,eco02040)
df_DMall_cat[df_DMall_cat$Pathway_ID %in% mypathway,]$label.sel<-df_DMall_cat[df_DMall_cat$Pathway_ID %in% mypathway,]$label

my_tmp<-df_DMall_cat %>% select(DMenv,DMevo,DMmut)
my_tmp<-as.matrix(my_tmp)
df_DMall_cat$DM_mean<-rowMeans(my_tmp)
df_DMall_cat$label.sel<-""
my_tmp<-quantile(df_DMall_cat$DM_mean)
df_DMall_cat[df_DMall_cat$DM_mean %in% my_tmp,]$label.sel<-df_DMall_cat[df_DMall_cat$DM_mean %in% my_tmp,]$label
my_fontsize<-16
# DMenv vs. DMmut
p<-ggscatter(df_DMall_cat,x="DMenv",y="DMmut",
          size=2,alpha=0.5,
          color="gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=0,label.y=1))+
          xlab(expression("Mean"~DM[env]~"[a.u.]"))+
          ylab(expression("Mean"~DM[mut]~"[a.u.]"))+
  geom_text_repel(data=df_DMall_cat,
                  aes(label=label.sel),segment.size=0.3,segment.alpha = 0.5,nudge_x = 0.6, lineheight = 0.1,size=3)
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p

# DMenv vs. DMevo
p<-ggscatter(df_DMall_cat,x="DMenv",y="DMevo",
          size=2,alpha=0.5,
          color="gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=0,label.y=1))+
          xlab(expression("Mean"~DM[env]~"[a.u.]"))+
          ylab(expression("Mean"~DM[evo]~"[a.u.]"))+
  geom_text_repel(data=df_DMall_cat,
                  aes(label=label.sel),segment.size=0.3,segment.alpha = 0.5,nudge_x = 0.6, lineheight = 0.1,size=3)
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p

# DMmut vs. DMevo
p<-ggscatter(df_DMall_cat,x="DMmut",y="DMevo",
          size=2,alpha=0.5,
          color="gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1))+
          xlab(expression("Mean"~DM[mut]~"[a.u.]"))+
          ylab(expression("Mean"~DM[evo]~"[a.u.]"))+
  geom_text_repel(data=df_DMall_cat,
                  aes(label=label.sel),segment.size=0.3,segment.alpha = 0.5,nudge_x = 0.6, lineheight = 0.1,size=3)
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p

Sys.time()
```
## 5.3. table of labeled KEGG pathways
```{r All_DM,warning=FALSE, message=FALSE,fig.height=2,fig.width=5,dpi=300}
my_tmp<-df_DMall_cat%>% filter(label.sel!="") %>% select(label.sel,Pathway_ID,Pathway_name)
colnames(my_tmp)<-c("Label","KEGG ID","Pathway name")
ggtexttable(my_tmp,rows=NULL)
Sys.time()
```
#
#
#
# 6. GSEA using KEGG pathways
## 6.1. GSEA in KEGG for the Env dataset
### 6.1.1. perform GSEA
```{r All_GSEA_KEGG,warning=FALSE, message=FALSE,fig.height=8,fig.width=6,dpi=300}
my_DM<-"DMenv"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$EntrezID
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)

## KEGG Pathway analysis
set.seed(123)
res_gsea <- GSEA(geneList     = gene_list,
              TERM2GENE = dkegg.each.gsea,
              seed = TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 300, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)
# pAdjusted and qvalues
# https://support.bioconductor.org/p/88811/
# https://www.biostars.org/p/128931/
# https://support.bioconductor.org/p/49864/
# calculation of pAjusted and qvalues in gsea 
# https://rdrr.io/github/GuangchuangYu/DOSE/src/R/gsea.R
# https://rdrr.io/github/GuangchuangYu/DOSE/src/R/utilities.R

gene_count<-res_gsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(res_gsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
dot_df<- left_join(dot_df,dkegg[,which(colnames(dkegg) %in% qw(Pathway_ID,label,Pathway_name))],by=c("ID"="Pathway_ID"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))

dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]

# identify leading edge subset
dot_df$core_enrichment_Bnb<-NA
for (i in 1:nrow(dot_df)){
  temp<-str_split(dot_df$core_enrichment[i],"/")[[1]]
  dot_df$core_enrichment_Bnb[i]<-paste(dkegg.each[dkegg.each$EntrezID %in% temp,]$Bnb,collapse = "/")
}
dot_df$log_qvalue<-log10(dot_df$qvalue)
dgsea_kegg_env<-dot_df
Sys.time()
```
### 6.1.2. visualization
```{r All_GSEA_KEGG,warning=FALSE, message=FALSE,fig.height=8,fig.width=6,dpi=300}
figtemp1<-dgsea_kegg_env[dgsea_kegg_env$qvalue<0.05,]
figtemp1$Path_lab<-paste(figtemp1$Pathway_name, " (",figtemp1$label,")",sep="")
p<-ggdotchart(figtemp1,x="Path_lab",y="NES", # "enrichmentScore"
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="top",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
  scale_size(range = c(1, 4)) +
  scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "bottom",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 90),ticks.linewidth = 1,order=1))+
  guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
p
Sys.time()
```
## 6.2. GSEA in KEGG for the Evo dataset
### 6.2.1. perform GSEA
```{r All_GSEA_KEGG,warning=FALSE, message=FALSE,fig.height=5,fig.width=4,dpi=300}
my_DM<-"DMevo"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$EntrezID
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)

## KEGG Pathway analysis
set.seed(123)
res_gsea <- GSEA(geneList     = gene_list,
              TERM2GENE = dkegg.each.gsea,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 300, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)

gene_count<-res_gsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(res_gsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
dot_df<- left_join(dot_df,dkegg[,which(colnames(dkegg) %in% qw(Pathway_ID,label,Pathway_name))],by=c("ID"="Pathway_ID"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]

# identify leading edge subset
dot_df$core_enrichment_Bnb<-NA
for (i in 1:nrow(dot_df)){
  temp<-str_split(dot_df$core_enrichment[i],"/")[[1]]
  dot_df$core_enrichment_Bnb[i]<-paste(dkegg.each[dkegg.each$EntrezID %in% temp,]$Bnb,collapse = "/")
}
dot_df$log_qvalue<-log10(dot_df$qvalue)
dgsea_kegg_evo<-dot_df
Sys.time()
```
### 6.2.2. visualization
```{r All_GSEA_KEGG,warning=FALSE, message=FALSE,fig.height=5,fig.width=4,dpi=300}
figtemp1<-dgsea_kegg_evo[dgsea_kegg_evo$qvalue<0.05,]
figtemp1$Path_lab<-paste(figtemp1$Pathway_name, " (",figtemp1$label,")",sep="")
p<-ggdotchart(figtemp1,x="Path_lab",y="NES", # "enrichmentScore"
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="top",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
  scale_size(range = c(1, 4)) +
  scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "bottom",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 90),ticks.linewidth = 1,order=1))+
  guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
p
Sys.time()
```
## 6.3. GSEA in KEGG for the Mut dataset
### 6.3.1. perform GSEA
```{r All_GSEA_KEGG,warning=FALSE, message=FALSE,fig.height=6,fig.width=5,dpi=300}
my_DM<-"DMmut"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$EntrezID
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)

## KEGG Pathway analysis
set.seed(123)
res_gsea <- GSEA(geneList     = gene_list,
              TERM2GENE = dkegg.each.gsea,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 300, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)

gene_count<-res_gsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(res_gsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
dot_df<- left_join(dot_df,dkegg[,which(colnames(dkegg) %in% qw(Pathway_ID,label,Pathway_name))],by=c("ID"="Pathway_ID"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]

# identify leading edge subset
dot_df$core_enrichment_Bnb<-NA
for (i in 1:nrow(dot_df)){
  temp<-str_split(dot_df$core_enrichment[i],"/")[[1]]
  dot_df$core_enrichment_Bnb[i]<-paste(dkegg.each[dkegg.each$EntrezID %in% temp,]$Bnb,collapse = "/")
}
dot_df$log_qvalue<-log10(dot_df$qvalue)
dgsea_kegg_mut<-dot_df
Sys.time()
```
### 6.3.2. visualization
```{r All_GSEA_KEGG,warning=FALSE, message=FALSE,fig.height=6,fig.width=5,dpi=300}
figtemp1<-dgsea_kegg_mut[dgsea_kegg_mut$qvalue<0.05,]
figtemp1$Path_lab<-paste(figtemp1$Pathway_name, " (",figtemp1$label,")",sep="")
p<-ggdotchart(figtemp1,x="Path_lab",y="NES", # "enrichmentScore"
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="top",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
  scale_size(range = c(1, 4)) +
  scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "bottom",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 90),ticks.linewidth = 1,order=1))+
  guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
p
Sys.time()
```
## 6.4. Visualize all datasets
```{r All_GSEA_KEGG,warning=FALSE, message=FALSE,fig.height=7.5,fig.width=7,dpi=300}
figtemp1<-dgsea_kegg_env[dgsea_kegg_env$qvalue<0.05,]
figtemp1$Dataset<-"Env"
figtemp2<-dgsea_kegg_evo[dgsea_kegg_evo$qvalue<0.05,]
figtemp2$Dataset<-"Evo"
figtemp3<-dgsea_kegg_mut[dgsea_kegg_mut$qvalue<0.05,]
figtemp3$Dataset<-"Mut"
figtemp<-bind_rows(figtemp1,figtemp2,figtemp3)
my_set12<-intersect(figtemp1$ID,figtemp2$ID)
my_set23<-intersect(figtemp2$ID,figtemp3$ID)
my_set31<-intersect(figtemp3$ID,figtemp1$ID)
myset123<-intersect(intersect(my_set12,my_set23),my_set31)
myset12_23_31<-unique(c(my_set12,my_set23,my_set31))
myset12_23_31<-myset12_23_31[myset12_23_31 %nin% myset123]
figtemp$col<-ifelse(figtemp$ID %in% myset123,"#AD002AFF",ifelse(figtemp$ID %in% myset12_23_31,"#ED0000FF","black"))

my_tmp<-data.frame(Dataset=qw(Env,Evo,Mut),Count=0,Width=0,BlankWidth=0,Length=0,Height=0,BlankHeight=0)
for(i in 1:nrow(my_tmp)){
  my_tmp$Count[i]<-nrow(figtemp[figtemp$Dataset==my_tmp$Dataset[i],])
  my_tmp$Length[i]<-max(str_length(figtemp[figtemp$Dataset==my_tmp$Dataset[i],]$Pathway_name))
}
my_tmp$Width<-0.3+0.7*my_tmp$Count/max(my_tmp$Count)
my_tmp$BlankWidth<-1-my_tmp$Width
my_tmp$Height<- 0.43+0.57*(my_tmp$Length/max(my_tmp$Length))
my_tmp$BlankHeight<-1-my_tmp$Height

# my_lo<-rbind(c(1,1),c(2,3),c(4,4))
my_lo<-rbind(c(1,2),c(3,3))
for(i in 1:nrow(my_tmp)){
  figtemp1<-figtemp[figtemp$Dataset==my_tmp$Dataset[i],]
  figtemp1<-figtemp1[order(figtemp1$NES,decreasing = T),]
  p<-ggdotchart(figtemp1,x="Pathway_name",y="NES", # "enrichmentScore"
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="right",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
              scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
              scale_size(range = c(1, 4),breaks = c(0.3,0.6,0.9),limits = c(0,1)) +
              scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                                    guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "right",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 0),ticks.linewidth = 1,order=1))+
              guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
  p<-ggpar(p,font.xtickslab = list(color=figtemp1$col))
  my_w<-c(my_tmp$Width[i],my_tmp$BlankWidth[i])
  my_h<-c(my_tmp$Height[i],my_tmp$BlankHeight[i])
  grid.arrange(p,blank,blank,
           # nrow=2,
           layout_matrix=my_lo,
           heights=c(my_h),
           widths=c(my_w))
}

Sys.time()
```
#
#
#
# 7. GSEA using TRN
## 7.1. GSEA in TRN for the Env dataset
### 7.1.1. perform GSEA
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3.5,fig.width=6,dpi=300}
my_DM<-"DMenv"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$EntrezID
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)

## KEGG Pathway analysis
set.seed(123)
res_gsea <- GSEA(geneList     = gene_list,
              TERM2GENE = dtrn.each.gsea,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 1000, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)
gene_count<-res_gsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(res_gsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
dot_df<- left_join(dot_df,dkegg[,which(colnames(dkegg) %in% qw(Pathway_ID,label,Pathway_name))],by=c("ID"="Pathway_ID"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]

# identify leading edge subset
dot_df$core_enrichment_Bnb<-NA
for (i in 1:nrow(dot_df)){
  temp<-str_split(dot_df$core_enrichment[i],"/")[[1]]
  dot_df$core_enrichment_Bnb[i]<-paste(dBnb.comp[dBnb.comp$EntrezID %in% temp,]$Bnb,collapse = "/")
}
dot_df$log_qvalue<-log10(dot_df$qvalue)
dgsea_trn_env<-dot_df
Sys.time()
```
### 7.1.2. visualization
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3.5,fig.width=6,dpi=300}
figtemp1<-dgsea_trn_env[dgsea_trn_env$qvalue<0.05,]
p<-ggdotchart(figtemp1,x="Description",y="NES",
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="top",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
  scale_size(range = c(1, 4)) +
  scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "bottom",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 90),ticks.linewidth = 1,order=1))+
  guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))

p
q<-ggpar(p,legend = "top",legend.title = list("Gene ratio","q-value"),font.ytickslab= c(10))
leg <- ggpubr::get_legend(q)
as_ggplot(leg)

# # of enriched TRs
print(paste("Positive NES: ", nrow(figtemp1[figtemp1$NES>0,])))
print(paste("Negative NES: ", nrow(figtemp1[figtemp1$NES<0,])))

Sys.time()
```
## 7.2. GSEA in TRN for the Evo dataset
### 7.2.1. perform GSEA
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3.5,fig.width=6,dpi=300}
my_DM<-"DMevo"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$EntrezID
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)

## KEGG Pathway analysis
set.seed(123)
res_gsea <- GSEA(geneList     = gene_list,
              TERM2GENE = dtrn.each.gsea,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 1000, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)
gene_count<-res_gsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(res_gsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
dot_df<- left_join(dot_df,dkegg[,which(colnames(dkegg) %in% qw(Pathway_ID,label,Pathway_name))],by=c("ID"="Pathway_ID"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]

# identify leading edge subset
dot_df$core_enrichment_Bnb<-NA
for (i in 1:nrow(dot_df)){
  temp<-str_split(dot_df$core_enrichment[i],"/")[[1]]
  dot_df$core_enrichment_Bnb[i]<-paste(dBnb.comp[dBnb.comp$EntrezID %in% temp,]$Bnb,collapse = "/")
}
dot_df$log_qvalue<-log10(dot_df$qvalue)

dgsea_trn_evo<-dot_df
Sys.time()
```
### 7.2.2. visualization
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3.5,fig.width=6,dpi=300}
figtemp1<-dgsea_trn_evo[dgsea_trn_evo$qvalue<0.05,]
p<-ggdotchart(figtemp1,x="Description",y="NES",
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="top",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
  scale_size(range = c(1, 4)) +
  scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "bottom",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 90),ticks.linewidth = 1,order=1))+
  guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))

p
q<-ggpar(p,legend = "top",legend.title = list("Gene ratio","q-value"),font.ytickslab= c(10))
leg <- ggpubr::get_legend(q)
as_ggplot(leg)

# # of enriched TRs
print(paste("Positive NES: ", nrow(figtemp1[figtemp1$NES>0,])))
print(paste("Negative NES: ", nrow(figtemp1[figtemp1$NES<0,])))

Sys.time()
```
## 7.3. GSEA in TRN for the Mut dataset
### 7.3.1. perform GSEA
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3.5,fig.width=6,dpi=300}
my_DM<-"DMmut"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$EntrezID
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)

## KEGG Pathway analysis
set.seed(123)
res_gsea <- GSEA(geneList     = gene_list,
              TERM2GENE = dtrn.each.gsea,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 1000, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)
gene_count<-res_gsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(res_gsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
dot_df<- left_join(dot_df,dkegg[,which(colnames(dkegg) %in% qw(Pathway_ID,label,Pathway_name))],by=c("ID"="Pathway_ID"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]

# identify leading edge subset
dot_df$core_enrichment_Bnb<-NA
for (i in 1:nrow(dot_df)){
  temp<-str_split(dot_df$core_enrichment[i],"/")[[1]]
  dot_df$core_enrichment_Bnb[i]<-paste(dBnb.comp[dBnb.comp$EntrezID %in% temp,]$Bnb,collapse = "/")
}
dot_df$log_qvalue<-log10(dot_df$qvalue)

dgsea_trn_mut<-dot_df
Sys.time()
```
### 7.3.2. visualization
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3.5,fig.width=6,dpi=300}
figtemp1<-dgsea_trn_mut[dgsea_trn_mut$qvalue<0.05,]
p<-ggdotchart(figtemp1,x="Description",y="NES",
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="top",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
  scale_size(range = c(1, 4)) +
  scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "bottom",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 90),ticks.linewidth = 1,order=1))+
  guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))

p
q<-ggpar(p,legend = "top",legend.title = list("Gene ratio","Adjusted p-value"),font.ytickslab= c(10))
leg <- ggpubr::get_legend(q)
as_ggplot(leg)

# # of enriched TRs
print(paste("Positive NES: ", nrow(figtemp1[figtemp1$NES>0,])))
print(paste("Negative NES: ", nrow(figtemp1[figtemp1$NES<0,])))

Sys.time()
```
## 7.4. Venn diagram of enriched TRs in higher DM values in three datasets and define TRcom
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
venn_all<-list(
  Evo=dgsea_trn_evo[(dgsea_trn_evo$NES>0)&(dgsea_trn_evo$qvalue<0.05),]$ID,
  Env=dgsea_trn_env[(dgsea_trn_env$NES>0)&(dgsea_trn_env$qvalue<0.05),]$ID,
  Mut=dgsea_trn_mut[(dgsea_trn_mut$NES>0)&(dgsea_trn_mut$qvalue<0.05),]$ID
)
venn <- process_data(Venn(venn_all))
venn$regionData$name
venn$regionData$count
v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))

TRcom<-venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]]

v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)
  # text(x=v$centers[1,1]-0.2, y=v$centers[1,2]-0.24, labels='Evo')+ # Evo label (reg 1)
  # text(x=v$centers[1,1]+0.3, y=v$centers[1,2]+0.3, labels='Env')+ # Env label (reg 2)
  # text(x=v$centers[1,1]-0.38, y=v$centers[1,2]+0.3, labels='Mut') # Mut label (reg 3)

for(i in 1:length(venn$regionData$item)){
  print(paste(venn$regionData$name[[i]],paste(venn$regionData$item[[i]],collapse=", "),sep=": "))
}

Sys.time()
```
## 7.5. Venn diagram of enriched TRs in lower DM values in three datasets
```{r All_GSEA_TRN,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
venn_all<-list(
  Evo=dgsea_trn_evo[(dgsea_trn_evo$NES<0)&(dgsea_trn_evo$qvalue<0.05),]$ID,
  Env=dgsea_trn_env[(dgsea_trn_env$NES<0)&(dgsea_trn_env$qvalue<0.05),]$ID,
  Mut=dgsea_trn_mut[(dgsea_trn_mut$NES<0)&(dgsea_trn_mut$qvalue<0.05),]$ID
)

venn <- process_data(Venn(venn_all))
venn$regionData$name
venn$regionData$count
v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))

v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)
  # text(x=v$centers[1,1]-0.2, y=v$centers[1,2]-0.24, labels='Evo')+ # Evo label (reg 1)
  # text(x=v$centers[1,1]+0.3, y=v$centers[1,2]+0.3, labels='Env')+ # Env label (reg 2)
  # text(x=v$centers[1,1]-0.38, y=v$centers[1,2]+0.3, labels='Mut') # Mut label (reg 3)

for(i in 1:length(venn$regionData$item)){
  print(paste(venn$regionData$name[[i]],paste(venn$regionData$item[[i]],collapse=", "),sep=": "))
}

Sys.time()
```
#
#
#
# 8. TRN statics
## 8.1. Ranking of # of target genes
```{r TRN,warning=FALSE, message=FALSE,fig.height=4,fig.width=2.5,dpi=300}
temp<-TRN %>% distinct(regulator,.keep_all = T)
temp<-temp[,which(colnames(temp) %in% qw(regulator,regulator_id,category))]
temp2<-unique(unlist(str_split(temp[str_count(temp$regulator_id,",")>0,]$regulator_id,",")))
temp1.1<-TRN[TRN$gene_id %in% unique(unlist(str_split(temp$regulator_id,","))),] # extract subset of TRN comprising TF and sigma genes as a target
temp1.2<-TRN[TRN$gene_id %nin% unique(unlist(str_split(temp$regulator_id,","))),] # extract subset of TRN comprising non-TF genes as a target
temp$ts_target<-0
temp$nts_target<-0
for(i in 1:nrow(temp)){
  temp$ts_target[i]<-nrow(temp1.1[temp1.1$regulator==temp$regulator[i],])
  temp$nts_target[i]<-nrow(temp1.2[temp1.2$regulator==temp$regulator[i],])
}
temp$all_target<-temp$ts_target+temp$nts_target
temp<-temp[order(temp$all_target,decreasing = T),]
temp$rank<-seq(1:nrow(temp))
temp$labelcol<-"Black"
temp[temp$regulator %in% TRcom,]$labelcol<-mycolpal[2]
temp$TRcomflag<-"Others"
temp[temp$regulator %in% TRcom,]$TRcomflag<-"TRcom"
temp$label<-""
temp[temp$regulator %in% TRcom,]$label<-temp[temp$regulator %in% TRcom,]$regulator
temp<- temp %>% pivot_longer(cols=qw(ts_target,nts_target),names_to = "target",values_to = "Nb_target")
trn.stat.table<-temp

mythr<-22
# p<-ggbarplot(trn.stat.table[trn.stat.table$rank<mythr,], x="regulator", "Nb_target",
# fill = "target",color = "target", 
# ylim=c(0,600), # ylim=c(0,160),
# # palette = ,
# ylab="Number of target genes",xlab="",legend="right")+
#   rotate_x_text()
# p<-ggpar(p,font.xtickslab = list(color=trn.stat.table[(trn.stat.table$rank<mythr)&(seq(1:nrow(trn.stat.table)) %% 2==0),]$labelcol))
# p


# Target genes: TFs and Sigmas
tempcol<- c(mycolpal[2],"gray70","black")
temp1<-trn.stat.table[trn.stat.table$target=="ts_target",]
temp1<-temp1[order(temp1$Nb_target,decreasing = T),]
temp1<-temp1[temp1$Nb_target>0,]
temp1$xpos<-seq(1:nrow(temp1))
temp1$TRcomflag2<-as.character(temp1$TRcomflag)
temp1[temp1$TRcomflag2=="TRcom",]$TRcomflag2<-"TRcom"
temp1$TRcomflag2<-factor(temp1$TRcomflag2,levels=qw(TRcom,Others))
temp1$label2<-temp1$label
temp1[temp1$label2!="",]$label2[c(1:c(nrow(temp1[temp1$label2!="",])-5))]<-""
p1<-ggbarplot(temp1, x="regulator", "Nb_target",fill = "TRcomflag2",color = "TRcomflag2", size=0.8,palette = tempcol,ylab="Number of target genes",xlab="",legend="")+
  rotate_x_text()
p1<-ggpar(p1,font.xtickslab = list(color=temp1$labelcol,size=0),font.ytickslab=c(18))+
  scale_y_log10(labels = trans_format("log10", math_format(10^.x)),breaks = 10^(0:2),limits = c(1, 200))+
  # yscale("log10", .format = TRUE,breaks=)+
  # scale_y_continuous(trans='log10',
  #                    breaks=trans_breaks('log10', function(x) 10^x),
  #                    labels=trans_format('log10', math_format(10^.x)))+
  geom_text_repel(aes(x = xpos,y = Nb_target,label = label2), 
                  ylim = c(0, NA), xlim = c(0, NA),  ## EDIT
                  segment.size=0.3,segment.alpha = 0.5,segment.color=mycolpal[2],
                  size = 5,nudge_y = 1, nudge_x = 2, lineheight = 0.1)+
  theme(axis.ticks.x = element_blank(),axis.text.x= element_blank(),axis.title.x=element_blank(),axis.title.y=element_blank())
# p1

# Target genes: non-TFs and non-Sigmas
tempcol<- c(mycolpal[2],"gray70","black")
temp2<-trn.stat.table[trn.stat.table$target=="nts_target",]
temp2<-temp2[order(temp2$Nb_target,decreasing = T),]
temp2<-temp2[temp2$Nb_target>0,]
temp2$xpos<-seq(1:nrow(temp2))
temp2$TRcomflag2<-as.character(temp2$TRcomflag)
temp2[temp2$TRcomflag2=="TRcom",]$TRcomflag2<-"TRcom"
temp2$TRcomflag2<-factor(temp2$TRcomflag2,levels=qw(TRcom,Others))
temp2$label2<-temp2$label
temp2[temp2$label2!="",]$label2[c(1:c(nrow(temp2[temp2$label2!="",])-5))]<-""

p2<-ggbarplot(temp2, x="regulator", "Nb_target",fill = "TRcomflag2",color = "TRcomflag2", size=0.8,palette = tempcol,ylab="Number of target genes",xlab="",legend="")+
  rotate_x_text()
p2<-ggpar(p2,font.xtickslab = list(color=temp2$labelcol,size=0),font.ytickslab=c(18))+
  scale_y_log10(labels = trans_format("log10", math_format(10^.x)),breaks = 10^(c(0,2,4)),limits = c(1, 10000))+
  geom_text_repel(aes(x = xpos,y = Nb_target,label = label2), 
                  ylim = c(0, NA), xlim = c(0, NA),  ## EDIT
                  segment.size=0.3,segment.alpha = 0.5,segment.color=mycolpal[2],
                  size = 5,nudge_y = 1, nudge_x = 100, lineheight = 0.1)+
  theme(axis.ticks.x = element_blank(),axis.text.x= element_blank(),axis.title.x=element_blank(),axis.title.y=element_blank())

# p2
# q<-ggpar(p2,legend = "right",legend.title = "",font.ytickslab= c(10))
# leg <- ggpubr::get_legend(q)
# as_ggplot(leg)


lay<-t(rbind(seq(1,3)))
blank <- grid.rect(gp=gpar(col="white"))
grid.arrange(p1,blank,p2,
             layout_matrix=lay,
             heights=c(1,0.1,1),
             bottom=textGrob(expression("Regulators"),gp=gpar(fontsize=18)),
             left=textGrob("Number of target genes", rot=90,gp=gpar(fontsize=18)))
Sys.time()
```
## 8.2. TRN comprising all TRs
https://qiita.com/saltcooky/items/4e55d97c5e86dfb208cd 
```{r TRN,warning=FALSE, message=FALSE,fig.height=6,fig.width=7,dpi=300}
temp<-data.frame(who=TRN.ts$regulator,whom=TRN.ts$target,effect=TRN.ts$effect)
temp0<-temp # used for an initial input for g
# create node infomation
temp<-data.frame(Name=unique(c(TRN.ts$regulator,TRN.ts$target)),enrichmentScore=0)
temp3<-temp
temp3[temp3$Name %in% TRcom,]$enrichmentScore<-1
temp3$enrichmentScore<-factor(temp3$enrichmentScore,levels=c(1,0))
temp3$mycol<-ifelse(temp3$enrichmentScore==0,"gray40",ifelse(temp3$enrichmentScore==1,"red","blue"))
temp3$mygroup<-ifelse(temp3$enrichmentScore==1,"TRcom","Others")
temp3<-temp3[order(temp3$enrichmentScore,decreasing = T),]
temp3$TRcomlabs<-""
temp3[temp3$mygroup=="TRcom",]$TRcomlabs<-temp3[temp3$mygroup=="TRcom",]$regulator
temp3$TRcom_refsigma_lab<-temp3$TRcomlabs
temp3[(temp3$TRcom_refsigma_lab=="")&(temp3$Name %in% qw(RpoD,RpoE,RpoN,RpoH,RpoS,FliA,FecI)),]$TRcom_refsigma_lab<-temp3[(temp3$TRcom_refsigma_lab=="")&(temp3$Name %in% qw(RpoD,RpoE,RpoN,RpoH,RpoS,FliA,FecI)),]$Name

temp3.1<-data.frame(Name=c(temp3[temp3$mygroup=="TRcom",]$Name,temp3[(temp3$Name %in% qw(RpoD,RpoE,RpoN,RpoH,RpoS,FliA,FecI))&(temp3$mygroup!="TRcom"),]$Name),rank=NA)

temp3.2<-data.frame(Name=temp3[temp3$Name %nin% temp3.1$Name,]$Name,rank=NA)
set.seed(111)
temp3.2$rand<-runif(nrow(temp3.2))
set.seed(NULL)
temp3.2<-temp3.2[order(temp3.2$rand),]
temp3.2<-temp3.2[,which(colnames(temp3.2) %nin% qw(rand))]

ii<-nrow(temp3.2)%/%nrow(temp3.1)
jj<-nrow(temp3.2)%%nrow(temp3.1)
temp3.1$rank<-seq(1,nrow(temp3.1)*c(ii+1),by=c(ii+1))[1:nrow(temp3.1)]
jj<-seq(1:nrow(temp3))
temp3.2$rank<-jj[jj %nin% temp3.1$rank]
temp3.1<-rbind(temp3.1,temp3.2)
temp3.1<-temp3.1[order(temp3.1$rank,decreasing = F),]
rownames(temp3.1)<-temp3.1$rank

temp3<-left_join(temp3,temp3.1)
temp3<-temp3[order(temp3$rank,decreasing = F),]
temp3$id<-seq(1:nrow(temp3))
temp3$angle <- 0 + 360 * temp3$id / nrow(temp3)
temp3$hjust <- ifelse((temp3$angle > 90)&(temp3$angle<270), 1, 0)
temp3$angle <- ifelse((temp3$angle > 90)&(temp3$angle<270), temp3$angle+180, temp3$angle)
temp3$labcol<-"black"
temp3[temp3$mygroup=="TRcom",]$labcol<-mycolpal[2]

temp0$edgecol<-"black"
temp0[(temp0$who %in% temp3[temp3$mygroup=="TRcom",]$Name)&(temp0$whom %in% temp3[temp3$mygroup=="TRcom",]$Name),]$edgecol<-"red"
temp0$edgewidth<-0.1
temp0[(temp0$who %in% temp3[temp3$mygroup=="TRcom",]$Name)&(temp0$whom %in% temp3[temp3$mygroup=="TRcom",]$Name),]$edgewidth<-0.8
# temp3$mygroup<-factor(temp3$mygroup,levels=qw(TFenv,Others))
temp3$TRcomflag<-ifelse(temp3$TRcomlabs!="",2,1) # TRcom: 2, Others: 1
temp3$TRcomflag<-factor(temp3$TRcomflag,levels=c(1,2))


dnet_edgeinfo_ts.global<-temp0
dnet_nodeinfo_ts.global<-temp3

g<-as_tbl_graph(dnet_edgeinfo_ts.global,directed = T)
# g<-g %>%
#   mutate(centrality=centrality_betweenness(),community = as.factor(group_edge_betweenness()))
g<-g %>%
  mutate(centrality=centrality_degree())
g<-g %>%
    activate(nodes) %>%
    left_join(dnet_nodeinfo_ts.global, by = c("name" = "Name"))

# Modularity of TRcom including sigma
Modularity.TRcom.global<-g %>%
  activate(nodes) %>% 
  # mutate(modularity = graph_modularity(group=eTFflag)) %>% 
  mutate(modularity = graph_modularity(group=TRcomflag)) %>% 
  pull(modularity) %>% 
  head(1)




# order the nodes
g <- permute(g, match(V(g)$name, dnet_nodeinfo_ts.global$Name))

g %>% 
  ggraph::ggraph(layout="circle")+
  # ggraph::geom_edge_link(alpha=0.5,width=0.1,arrow = arrow(length = unit(1, 'mm')),aes(color = as.factor(edgecol)))+
  ggraph::geom_edge_link(alpha=0.5,arrow = arrow(length = unit(1, 'mm')),aes(color = as.factor(edgecol),width=edgewidth))+
  ggraph::geom_node_point(aes(size=centrality,color=mygroup,alpha=1,x=x*1.05,y=y*1.05))+
  ggraph::geom_node_text(aes(x = x*1.15, y=y*1.15,angle=angle, label=TRcom_refsigma_lab, hjust=hjust,colour=mygroup), size=7) +
  # scale_color_manual(values = c(mycolpal[2],get_palette(palette = "Reds",3)[2],"gray40"))+
  scale_color_manual(values = c("gray40",mycolpal[2]))+
  ggraph::theme_graph(background="white",)+
  theme(plot.margin=unit(c(0,0,0,0),"cm"))+
  # expand_limits(x = c(-1.2, 1.2), y = c(-1.2, 1.2))+
  expand_limits(x = c(-2, 2), y = c(-2, 2))+
  scale_size(range = c(2, 10))
Sys.time()
```
## 8.3. TRN comprising TRcom
https://qiita.com/saltcooky/items/4e55d97c5e86dfb208cd 
```{r TRN,warning=FALSE, message=FALSE,fig.height=3.5,fig.width=4.5,dpi=300}
temp<-data.frame(who=TRN.ts$regulator,whom=TRN.ts$target,effect=TRN.ts$effect)
temp<-temp[(temp$who %in% TRcom)&(temp$whom %in% TRcom),]
temp0<-temp # used for an initial input for g
# create node infomation
temp<-data.frame(Name=unique(c(TRN.ts$regulator,TRN.ts$target)),enrichmentScore=0)
temp<-temp[temp$Name %in% TRcom,]
temp3<-temp
temp3[temp3$Name %in% TRcom,]$enrichmentScore<-1
temp3$enrichmentScore<-factor(temp3$enrichmentScore,levels=c(1,0))
temp3$mycol<-ifelse(temp3$enrichmentScore==0,"gray40",ifelse(temp3$enrichmentScore==1,"red","blue"))
# temp3$mygroup<-ifelse(temp3$enrichmentScore==1,"eTFs","Others")
temp3<-temp3[order(temp3$enrichmentScore,decreasing = T),]
temp3$eTFlabs<-temp3$Name
# temp3[temp3$mygroup=="eTFs",]$eTFlabs<-temp3[temp3$mygroup=="eTFs",]$regulator
temp3$id<-seq(1:nrow(temp3))-1

temp3$angle <- 0 + 360 * temp3$id / nrow(temp3)
temp3$hjust <- ifelse((temp3$angle > 90)&(temp3$angle<270), 1, 0)
temp3$angle <- ifelse((temp3$angle > 90)&(temp3$angle<270), temp3$angle+180, temp3$angle)
temp3$labcol<-mycolpal[2]

temp0$edgecol<-"gray40"
temp0[temp0$effect==-1,]$edgecol<-mycolpal[1]
temp0[temp0$effect==1,]$edgecol<-mycolpal[2]
temp0$edgewidth<-0.8

dnet_edgeinfo_ts.global.sbst<-temp0
dnet_nodeinfo_ts.global.sbst<-temp3

g<-as_tbl_graph(dnet_edgeinfo_ts.global.sbst,directed = T)
# g<-g %>%
#   mutate(centrality=centrality_betweenness(),community = as.factor(group_edge_betweenness()))
g<-g %>%
  mutate(centrality=centrality_degree())


temp4.1<-as.data.frame(igraph::degree(g,mode="in"))
colnames(temp4.1)<-"In"
temp4.1$Name<-rownames(temp4.1)
temp4.2<-as.data.frame(igraph::degree(g,mode="out"))
colnames(temp4.2)<-"Out"
temp4.2$Name<-rownames(temp4.2)
temp4<-left_join(temp4.1,temp4.2)
temp4.3<-as.data.frame(ends(g,es = E(g))[which_loop(g,E(g)),])
temp4$Loop<-0
temp4[temp4$Name %in% temp4.3$V1,]$Loop<-1
temp4$In<-temp4$In-temp4$Loop
temp4$Out<-temp4$Out-temp4$Loop

g<-g %>%
    activate(nodes) %>%
    left_join(dnet_nodeinfo_ts.global.sbst, by = c("name" = "Name")) %>%
  left_join(temp4,by = c("name" = "Name"))

# order the nodes
g <- permute(g, match(V(g)$name, dnet_nodeinfo_ts.global.sbst$Name))

g2<-g %>% 
  ggraph::ggraph(layout="circle")+
  # ggraph::geom_edge_link(alpha=0.5,width=0.1,arrow = arrow(length = unit(1, 'mm')),aes(color = as.factor(edgecol)))+
  ggraph::geom_edge_fan(alpha=0.5,arrow = arrow(length = unit(3, 'mm')),aes(color = as.factor(edgecol),width=edgewidth))+
  ggraph::geom_edge_loop(alpha=0.5,arrow = arrow(length = unit(3, 'mm')),aes(color = as.factor(edgecol),width=edgewidth),angle_calc = "along")+
  ggraph::geom_node_point(aes(size=centrality,alpha=1,x=x*1.05,y=y*1.05))+
  ggraph::geom_node_text(aes(x = x*1.2, y=y*1.2,angle=angle, label=eTFlabs, hjust=hjust,colour=mygroup), size=7) +
  # geom_scatterpie(cols=qw(In,Out,Loop),data = igraph::as_data_frame(g, "vertices"),pie_scale = 2)+
  # scale_color_manual(values = c(mycolpal[2],get_palette(palette = "Reds",3)[2],"gray40"))+
  scale_color_manual(values = c(mycolpal[2],"gray40"))+
  ggraph::theme_graph(background="white",)+
  theme(plot.margin=unit(c(0,0,0,0),"cm"))+
  # expand_limits(x = c(-1.2, 1.2), y = c(-1.2, 1.2))+
  expand_limits(x = c(-2, 2), y = c(-2, 2))+
  scale_size(range = c(5, 10))

#https://www.r-bloggers.com/2020/03/ggraph-tricks-for-common-problems/

ggraph::ggraph(g, "manual", x = g2$data$x, y = g2$data$y)+
  ggraph::geom_edge_fan(alpha=0.5,arrow = arrow(length = unit(3, 'mm')),aes(color = as.factor(edgecol),width=edgewidth))+
  # ggraph::geom_edge_loop(alpha=0.5,arrow = arrow(length = unit(3, 'mm')),aes(color = as.factor(edgecol),width=edgewidth),angle_calc = "along")+
  # ggraph::geom_node_point(aes(size=centrality,color=mygroup,alpha=1,x=x*1.05,y=y*1.05))+
  ggraph::geom_node_text(aes(x = x*1.2, y=y*1.2,angle=angle, label=eTFlabs, hjust=hjust), size=7) +
  geom_scatterpie(cols=qw(In,Out,Loop),data = as.data.frame(g2$data),pie_scale = 4)+
  scale_color_manual(values = c(mycolpal[2],"gray40"))+
  scale_fill_manual(values = c("#FDE725FF","#22A884FF", "#440154FF"))+
  ggraph::theme_graph(background="white",)+
  theme(plot.margin=unit(c(0,0,0,0),"cm"))+
  # expand_limits(x = c(-1.2, 1.2), y = c(-1.2, 1.2))+
  expand_limits(x = c(-2, 2), y = c(-2, 2))+
  scale_size(range = c(5, 10))
# https://kateto.net/network-visualization
# plot(g,edge.curved=.3,vertex.frame.color="#ffffff",edge.arrow.size=.4,edge.color=E(g)$edgecol,vertex.size=30,
     # layout=layout_in_circle)

Sys.time()
```
## 8.4. TRN comprising Non-TRcom
```{r TRN,warning=FALSE, message=FALSE,fig.height=6,fig.width=7,dpi=300}
dnet_edgeinfo_ts.global.others<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$who %nin% TRcom & dnet_edgeinfo_ts.global$whom %nin% TRcom,]
dnet_nodeinfo_ts.global.others<-dnet_nodeinfo_ts.global[dnet_nodeinfo_ts.global$Name %in% unique(c(dnet_edgeinfo_ts.global.others$who,dnet_edgeinfo_ts.global.others$whom)), ]
dnet_edgeinfo_ts.global.others$edgecol<-"gray40"
dnet_edgeinfo_ts.global.others[dnet_edgeinfo_ts.global.others$effect==-1,]$edgecol<-mycolpal[1]
dnet_edgeinfo_ts.global.others[dnet_edgeinfo_ts.global.others$effect==1,]$edgecol<-mycolpal[2]
dnet_edgeinfo_ts.global.others$edgewidth<-0.2

g<-as_tbl_graph(dnet_edgeinfo_ts.global.others,directed = T)
# g<-g %>%
#   mutate(centrality=centrality_betweenness(),community = as.factor(group_edge_betweenness()))
# g<-g %>%
#   mutate(centrality=centrality_degree())
g<-g %>%
    activate(nodes) %>%
    left_join(dnet_nodeinfo_ts.global.others, by = c("name" = "Name"))

# order the nodes
g <- permute(g, match(V(g)$name, dnet_nodeinfo_ts.global.others$Name))

g %>% 
  ggraph::ggraph(layout="circle")+
  # ggraph::geom_edge_link(alpha=0.5,width=0.1,arrow = arrow(length = unit(1, 'mm')),aes(color = as.factor(edgecol)))+
  ggraph::geom_edge_link(alpha=0.5,arrow = arrow(length = unit(1, 'mm')),aes(color = as.factor(edgecol),width=edgewidth))+
  ggraph::geom_node_point(aes(size=1,color=mygroup,alpha=1,x=x*1.05,y=y*1.05))+
  # ggraph::geom_node_text(aes(x = x*1.15, y=y*1.15,angle=angle, label=eTF_refsigma_lab, hjust=hjust,colour=mygroup), size=7) +
  # scale_color_manual(values = c(mycolpal[2],get_palette(palette = "Reds",3)[2],"gray40"))+
  scale_color_manual(values = c("gray40"))+
  ggraph::theme_graph(background="white",)+
  theme(plot.margin=unit(c(0,0,0,0),"cm"))+
  # expand_limits(x = c(-1.2, 1.2), y = c(-1.2, 1.2))+
  expand_limits(x = c(-2, 2), y = c(-2, 2))+
  scale_size(range = c(1, 1))

Sys.time()
```
## 8.5. Calculate edge density of randomly sampled subnetwork
About 13 regulators are sampled.
Self-loops are omitted from the calculation.
```{r TRN,warning=FALSE, message=FALSE,fig.height=3.1,fig.width=3,dpi=300}
temp0<-trn.stat.table[trn.stat.table$all_target>4,] %>% distinct(regulator,.keep_all = T)
trial<-10000
temp<-createEmptyDf(nrow=trial,ncol = 4,colnames = qw(test,regulators,ratio.TRcom,test.btw))
for(i in 1:nrow(temp)){
  temp1<-temp0[sample(seq(1,nrow(temp0)),size = length(TRcom),replace = F),]$regulator
  temp$regulators[i]<-paste(temp1,collapse = ",")
  temp$ratio.TRcom[i]<-length(temp1[temp1 %in% TRcom])/length(TRcom)
  temp2<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$who %in% temp1 & dnet_edgeinfo_ts.global$whom %in% temp1,]
  temp2<-temp2[temp2$who!=temp2$whom,] # remove self-loop
  g<-as_tbl_graph(temp2,directed = T)
  # temp$test[i]<-ecount(g)/((length(TRcom)*(length(TRcom)-1))+length(TRcom)) # consider self-loop
  temp$test[i]<-ecount(g)/((length(TRcom)*(length(TRcom)-1))) # not consider self-loop
  temp1.2<-temp0[temp0$regulator %nin% temp1,] # the rest TFs
  temp2.2<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$who %in% temp1 & dnet_edgeinfo_ts.global$whom %in% temp1.2$regulator,]
  temp2.3<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$whom %in% temp1 & dnet_edgeinfo_ts.global$who %in% temp1.2$regulator,]
  temp2.4<-bind_rows(temp2.2,temp2.3)
  temp$test.btw[i]<-nrow(temp2.4)/(2*length(TRcom)*nrow(temp1.2))
}
temp3<-dnet_edgeinfo_ts.global.sbst
temp3<-temp3[temp3$who!=temp3$whom,]
g<-as_tbl_graph(temp3,directed = T)
# temp3<-ecount(g)/((length(TRcom)*(length(TRcom)-1))+length(TRcom)) # consider self-loop
temp3<-ecount(g)/((length(TRcom)*(length(TRcom)-1))) # not consider self-loop
temp4<-nrow(as.data.frame(temp[temp$test>temp3,]))/trial

# Edge density within randomly sampled 13 regulators
df.rndmpick.edgedens<-temp
p<-gghistogram(df.rndmpick.edgedens,x="test",y="..density..",bins=30,
            fill="black",
            color="darkgray",alpha=1,size=0,
            # add="mean",add.params = list(color="black",linetype=2),
            xlab="Edge density",ylab="Relative frequency [a.u.]")+
  geom_vline(xintercept = temp3,linetype=1,color=mycolpal[2])+
  scale_x_continuous(breaks = get_breaks(by = 0.1, from = 0),limits =  c(0, 0.2))
p<-ggpar(p,font.x=c(17),font.y = c(17),font.xtickslab = c(17),font.ytickslab = c(17))
if(nrow(as.data.frame(df.rndmpick.edgedens[df.rndmpick.edgedens$test>temp3,]))==0){
  p<-p+annotate("text",x=temp3,y=2,label=paste("p< ",1/trial ,sep=""),hjust = 1.1,vjust = -3, parse=F,size=6)
}else{
  p<-p+annotate("text",x=temp3,y=2,label=paste("p= ",temp4 ,sep=""),hjust = 1.1,vjust = -3, parse=F,size=6)
}
p  


# ggscatter(temp,x="ratio.etfs",y="test",size=0.1)

Sys.time()
```
## 8.6. Plot edge density within communities
Self-loops are omitted.
```{r TRN,warning=FALSE, message=FALSE,fig.height=3,fig.width=2.5,dpi=300}

# Edge density of the TRN including sigma
### Edge density of subnetwork
# https://michaelgastner.com/DAVisR2021/essentials-of-network-analysis.html
# for a directed network,the # of possible connections between nodes (N) is N*(N-1)
df.edge.dens<-createEmptyDf(nrow=3,ncol=4,colnames=qw(Class,EdgeDensity,Low_sd,High_sd))
df.edge.dens$Class<-qw(TRcom,Others,All)
myTRcomflag=c(2,1)
for(i in 1:nrow(df.edge.dens)){
  if(i!=3){
    temp1<-dnet_nodeinfo_ts.global[dnet_nodeinfo_ts.global$TRcomflag==myTRcomflag[i],]$Name
    temp2<-dnet_edgeinfo_ts.global[(dnet_edgeinfo_ts.global$who %in% temp1)&(dnet_edgeinfo_ts.global$whom %in% temp1),]
    temp2<-temp2[temp2$who!=temp2$whom,]
    # df.edge.dens$EdgeDensity[i]<-nrow(temp2)/((length(temp1)*(length(temp1)-1))+length(temp1))# consider self-loop
    df.edge.dens$EdgeDensity[i]<-nrow(temp2)/((length(temp1)*(length(temp1)-1)))# not consider self-loop
  }else{
    temp2<-dnet_edgeinfo_ts.global
    temp2<-temp2[temp2$who!=temp2$whom,]
    # df.edge.dens$EdgeDensity[i]<-nrow(dnet_edgeinfo_ts.global)/((nrow(dnet_edgeinfo_ts.global)*(nrow(dnet_edgeinfo_ts.global)-1))+nrow(dnet_edgeinfo_ts.global))
    df.edge.dens$EdgeDensity[i]<-nrow(temp2)/((nrow(dnet_nodeinfo_ts.global)*(nrow(dnet_nodeinfo_ts.global)-1)))# not consider self-loop
  }
}
df.edge.dens<-rbind(df.edge.dens,c("Random",mean(df.rndmpick.edgedens$test),mean(df.rndmpick.edgedens$test)-sd(df.rndmpick.edgedens$test),mean(df.rndmpick.edgedens$test)+sd(df.rndmpick.edgedens$test)))
df.edge.dens$EdgeDensity<-as.numeric(df.edge.dens$EdgeDensity)
df.edge.dens$Low_sd<-as.numeric(df.edge.dens$Low_sd)
df.edge.dens[df.edge.dens$Class=="Random",]$Low_sd<-ifelse(df.edge.dens[df.edge.dens$Class=="Random",]$Low_sd<0,0,df.edge.dens[df.edge.dens$Class=="Random",]$Low_sd) # for visualization
df.edge.dens$High_sd<-as.numeric(df.edge.dens$High_sd)


temp<- c(mycolpal[2],"gray70","black","white")
df.edge.dens$Class2<-df.edge.dens$Class
df.edge.dens$Class2[df.edge.dens$Class2=="TRcom"]<-"TRcom"
df.edge.dens$Class2<-factor(df.edge.dens$Class2,levels=qw(TRcom,Others,All,Random))
df.edge.dens$Class3<-c(glue("TR[com]"),"Others","All","Random")
my_labs<-c(bquote(TR[com]),bquote(Non-TR[com]),bquote(All),bquote(Random))
names(my_labs)<-df.edge.dens$Class2
p<-ggbarplot(df.edge.dens,x="Class2",y="EdgeDensity",fill="Class2",palette = temp,xlab="",ylab="Edge density",legend.title="",legend="")+
  geom_linerange(aes(ymax=High_sd,ymin=Low_sd),width=0.5)+
  scale_y_continuous(breaks = get_breaks(by = 0.1, from = 0),limits =  c(0, 0.22))+rotate_x_text(90)+
  scale_x_discrete(labels = my_labs)
p<-ggpar(p,font.y=c(17),font.xtickslab = c(17),font.ytickslab = c(17),xlab=FALSE)
p
Sys.time()
```
## 8.7. Calculate edge density between TRcoms and Others
The number of sample size was equal to that of TRcoms
```{r TRN,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-trn.stat.table[trn.stat.table$all_target>4,] %>% distinct(regulator,.keep_all = T)
temp1<-temp0[temp0$regulator %nin% TRcom,]
temp2.1<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$who %in% TRcom & dnet_edgeinfo_ts.global$whom %in% temp1$regulator,]
temp2.2<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$whom %in% TRcom & dnet_edgeinfo_ts.global$who %in% temp1$regulator,]
temp2<-bind_rows(temp2.1,temp2.2)
egedens.btwn.TRcom_Others<-nrow(temp2)/(2*length(TRcom)*nrow(temp1))
print(paste("Edge density between TRcoms and Others:",egedens.btwn.TRcom_Others))

# Edge density between randomly sampled 13 regulators and the rest regulators
p2<-gghistogram(df.rndmpick.edgedens,x="test.btw",y="..density..",bins=30,
            fill="black",
            color="darkgray",alpha=1,size=0,
            # add="mean",add.params = list(color="black",linetype=2),
            xlab="Edge density",ylab="Relative frequency [a.u.]")+
  geom_vline(xintercept = egedens.btwn.TRcom_Others,linetype=1,color=mycolpal[2])+
  scale_x_continuous(breaks = get_breaks(by = 0.03, from = 0),limits =  c(0, 0.06))

temp3<-nrow(as.data.frame(df.rndmpick.edgedens[df.rndmpick.edgedens$test.btw>egedens.btwn.TRcom_Others,]))
if(temp3==0){
  p2<-p2+annotate("text",x=egedens.btwn.TRcom_Others,y=2,label=paste("p< ",1/nrow(df.rndmpick.edgedens) ,sep=""),hjust = 1.1,vjust = -3, parse=F)
}else{
  p2<-p2+annotate("text",x=egedens.btwn.TRcom_Others,y=2,label=paste("p= ",temp3/nrow(df.rndmpick.edgedens) ,sep=""),hjust = 1.1,vjust = -3, parse=F)
}
p2


Sys.time()
```
## 8.8. Plot edge density between communities
```{r TRN,warning=FALSE, message=FALSE,fig.height=3,fig.width=2,dpi=300}

# Edge density of the TRN including sigma
### Edge density of subnetwork
# https://michaelgastner.com/DAVisR2021/essentials-of-network-analysis.html
# for a directed network,the # of possible connections between nodes (N) is N*(N-1)
df.edge.dens_btwn<-createEmptyDf(nrow=2,ncol=4,colnames=qw(Class,EdgeDensity,Low_sd,High_sd))
df.edge.dens_btwn$Class<-qw(TRcom,Random)
df.edge.dens_btwn[df.edge.dens_btwn$Class=="TRcom",]$EdgeDensity<-egedens.btwn.TRcom_Others
df.edge.dens_btwn[df.edge.dens_btwn$Class=="Random",]$EdgeDensity<-mean(df.rndmpick.edgedens$test.btw)
df.edge.dens_btwn[df.edge.dens_btwn$Class=="Random",]$Low_sd<-mean(df.rndmpick.edgedens$test.btw)-sd(df.rndmpick.edgedens$test.btw)
df.edge.dens_btwn[df.edge.dens_btwn$Class=="Random",]$High_sd<-mean(df.rndmpick.edgedens$test.btw)+sd(df.rndmpick.edgedens$test.btw)
df.edge.dens_btwn[df.edge.dens_btwn$Class=="Random",]$Low_sd<-ifelse(df.edge.dens_btwn[df.edge.dens_btwn$Class=="Random",]$Low_sd<0,0,df.edge.dens_btwn[df.edge.dens_btwn$Class=="Random",]$Low_sd) # for visualization

temp<- c(mycolpal[2],"white")
df.edge.dens_btwn$Class2<-df.edge.dens_btwn$Class
df.edge.dens_btwn$Class2[df.edge.dens_btwn$Class2=="TRcom"]<-"TRcom"
df.edge.dens_btwn$Class2<-factor(df.edge.dens_btwn$Class2,levels=qw(TRcom,Random))
df.edge.dens_btwn$Class3<-c(glue("TR[com]"),"Random")
my_labs<-c(bquote(TR[com]),bquote(Random))
names(my_labs)<-df.edge.dens_btwn$Class2
p<-ggbarplot(df.edge.dens_btwn,x="Class2",y="EdgeDensity",fill="Class2",palette = temp,xlab="",ylab="Edge density",legend.title="",legend="")+
  geom_linerange(aes(ymax=High_sd,ymin=Low_sd),width=0.5)+
  scale_y_continuous(breaks = get_breaks(by = 0.1, from = 0),limits =  c(0, 0.22))+rotate_x_text(90)+
  scale_x_discrete(labels = my_labs)
p<-ggpar(p,font.y=c(17),font.xtickslab = c(17),font.ytickslab = c(17),xlab=FALSE)
p
Sys.time()
```
## 8.9. Number of in-degree vs. DM values
```{r TRN,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
dnet_degree<-TRN %>% distinct(regulator,.keep_all = T)
dnet_degree<-dnet_degree[dnet_degree$regulator %in% unique(c(TRN.ts$regulator,TRN.ts$target)), ]
rownames(dnet_degree)<-seq(1,nrow(dnet_degree))
dnet_degree<-dnet_degree[,which(colnames(dnet_degree) %in% qw(regulator,category,regulator_id))]
g<-as_tbl_graph(dnet_edgeinfo_ts.global,directed = T)
dnet_degree<-transform(dnet_degree,
                       in_degree=igraph::degree(graph = g,v = dnet_degree$regulator,mode = "in",loops = F),
                       out_degree=igraph::degree(graph = g,v = dnet_degree$regulator,mode = "out",loops = F),
                       all_degree=igraph::degree(graph = g,v = dnet_degree$regulator,mode = "all",loops = F))
dnet_degree$DMenv<-NA
dnet_degree$DMevo<-NA
dnet_degree$DMmut<-NA

for(i in 1:nrow(dnet_degree)){
  temp_id<-str_split(dnet_degree$regulator_id[i],",")[[1]]
  df_temp<-df_DMall[df_DMall$Bnb %in% temp_id,]
  dnet_degree$DMenv[i]<-mean(df_temp$DMenv, na.rm=T)
  dnet_degree$DMevo[i]<-mean(df_temp$DMevo, na.rm=T)
  dnet_degree$DMmut[i]<-mean(df_temp$DMmut, na.rm=T)
}

ggscatter(dnet_degree,x="in_degree",y="DMenv",alpha=0.3)+
  stat_cor(method = "spearman", label.x = 1, label.y = 2)
ggscatter(dnet_degree,x="in_degree",y="DMevo",alpha=0.3)+
  stat_cor(method = "spearman", label.x = 1, label.y = 2)
ggscatter(dnet_degree,x="in_degree",y="DMmut",alpha=0.3)+
  stat_cor(method = "spearman", label.x = 1, label.y = 2)

Sys.time()
```
## 8.10. Count the number of regulations (TRcom-containing or TRcom-less) for each target genes
```{r TRN,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=2,dpi=300}
# temp<-dtrn_slmlt
# temp<-left_join(temp,df_DMall[,which(colnames(df_DMall) %in% qw(Bnb,DMenv,DMevo,DMmut))],by=c("gene_id"="Bnb"))
# temp<-temp[temp$regulator %in% TRcom,]
# temp<-temp %>% pivot_longer(cols = qw(DMenv,DMevo,DMmut),names_to="Dataset", values_to ="DM")

temp2<-dtrget_slmlt
temp2$Count.allregs<-str_count(temp2$regulators,",")+1
temp2$Count.TRcom<-0
temp2$Count.TRcom.wosigma<-0
temp2$Count.sigma.of.TRcom<-0
temp2$Count.rpod<-0
temp2$Count.sigmas<-0

mysigmas<-qw(RpoD,RpoE,RpoN,RpoH,RpoS,FliA,FecI)
for(i in 1:nrow(temp2)){
  temp_reg<-str_split(temp2$regulators[i],",")[[1]]
  temp2$Count.TRcom[i]<-length(temp_reg[temp_reg %in% TRcom])
  temp2$Count.TRcom.wosigma[i]<-length(temp_reg[temp_reg %in% TRcom[TRcom %nin% mysigmas]])
  temp2$Count.sigma.of.TRcom[i]<-length(temp_reg[temp_reg %in% mysigmas[mysigmas %in% TRcom]])
  temp2$Count.rpod[i]<-length(temp_reg[temp_reg %in% "RpoD"])
  temp2$Count.sigmas[i]<-length(temp_reg[temp_reg %in% mysigmas])
}
temp2$Percent.TRcom<-temp2$Count.TRcom/temp2$Count.allregs

dtrget_stat<-temp2
dtrget_stat<-left_join(dtrget_stat,df_DMall[,which(colnames(df_DMall) %in% qw(Bnb,DMenv,DMevo,DMmut))],by=c("gene_id"="Bnb"))
dtrget_stat<-dtrget_stat[dtrget_stat$gene_id %in% dBnb.comp$Bnb,]
dtrget_stat$TRcomflag<-ifelse(dtrget_stat$Count.TRcom>0,"TRcom","Others")
dtrget_stat$Use<-NA
myflag<-qw(TRcom,Others)
for(j in myflag){
  l<-0
  for(i in 1: max(dtrget_stat$Count.allregs)){
    if(nrow(dtrget_stat[(dtrget_stat$Count.allregs==i)&(dtrget_stat$TRcomflag==j),])<5){
      l<-c(l,i)
    }
  }
  dtrget_stat[dtrget_stat$TRcomflag==j&dtrget_stat$Count.allregs %nin% l,]$Use<-"Use"
}

# add genes who have no regulators
temp3<-createEmptyDf(nrow=nrow(dtrn_slmlt[dtrn_slmlt$regulator=="No",]),ncol=ncol(dtrget_stat[,which(colnames(dtrget_stat) %nin% qw(DMenv,DMevo,DMmut))]),colnames =colnames(dtrget_stat[,which(colnames(dtrget_stat) %nin% qw(DMenv,DMevo,DMmut))]))
temp3$gene_id<-dtrn_slmlt[dtrn_slmlt$regulator=="No",]$gene_id
temp3$Flag<-"no.regs"
temp3$Count.allregs<-0
temp3$Count.TRcom<-0
temp3$Count.TRcom.wosigma<-0
temp3$Count.sigma.of.TRcom<-0
temp3$Count.rpod<-0
temp3$Count.sigmas<-0
temp3$Percent.TRcom<-0
temp3$TRcomflag<-"Null"
temp3$Use<-"Use"
temp3<-temp3[temp3$gene_id %in% dBnb.comp$Bnb,]
temp3<-left_join(temp3,df_DMall[,which(colnames(df_DMall) %in% qw(Bnb,DMenv,DMevo,DMmut))],by=c("gene_id"="Bnb"))
dtrget_stat<-bind_rows(dtrget_stat,temp3)
dtrget_stat$TRcomflag2<-ifelse(dtrget_stat$TRcomflag=="TRcom","TRcom-containing",ifelse(dtrget_stat$TRcomflag=="Others","TRcom-less","Null"))
dtrget_stat$TRcomflag<-factor(dtrget_stat$TRcomflag,levels=qw(TRcom,Others,Null))
dtrget_stat$TRcomflag2<-factor(dtrget_stat$TRcomflag2,levels=c("TRcom-containing","TRcom-less","Null"))
df.regcount<-dtrget_stat
Sys.time()
```
## 8.11.1. The relation between the number of regulations (TRcom-containing or TRcom-less) and DM values
All TRcomm including sigma factors are considered
```{r TRN,warning=FALSE, message=FALSE,fig.height=6,fig.width=4.5,dpi=300}
figtemp2<-df.regcount

# First, perform statistical tests
# Wilcox test between DMenv of TRcom-containing (TRcom) and that of TRcom-less (Others) 
stat.test1 <- dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="Null"&dtrget_stat$Count.allregs<6,] %>%
  df_group_by(Count.allregs) %>%
  wilcox_test(DMenv ~ TRcomflag) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance("p.adj")%>%
  add_xy_position(x = "Count.allregs", dodge = 0.8)%>%
  mutate(x = seq(2,max(x)+1,by=1),xmin=x-0.2,xmax=x+0.2,y.position=2)

# Wilcox test between DMevo of TRcom-containing (TRcom) and that of TRcom-less (Others) 
stat.test2 <- dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="Null"&dtrget_stat$Count.allregs<6,] %>%
  df_group_by(Count.allregs) %>%
  wilcox_test(DMevo ~ TRcomflag) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance("p.adj")%>%
  add_xy_position(x = "Count.allregs", dodge = 0.8)%>%
  mutate(x = seq(2,max(x)+1,by=1),xmin=x-0.2,xmax=x+0.2,y.position=1.2)

# Wilcox test between DMmut of TRcom-containing (TRcom) and that of TRcom-less (Others) 
stat.test3 <- dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="Null"&dtrget_stat$Count.allregs<6,] %>%
  df_group_by(Count.allregs) %>%
  wilcox_test(DMmut ~ TRcomflag) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance("p.adj")%>%
  add_xy_position(x = "Count.allregs", dodge = 0.8)%>%
  mutate(x = seq(2,max(x)+1,by=1),xmin=x-0.2,xmax=x+0.2,y.position=1.2)

### Count # of target genes of each in-degree class
figtemp1.1<-dtrget_stat[!is.na(dtrget_stat$Use),]
figtemp1.2<-data.frame(Count.allregs=rep(seq(min(figtemp1.1$Count.allregs),max(figtemp1.1$Count.allregs)),each=2))
figtemp1.2$TRcomflag2<-rep(c("TRcom-containing","TRcom-less"),nrow(figtemp1.2)/2)
figtemp1.2[figtemp1.2$Count.allregs==0,]$TRcomflag2<-"Null"
figtemp1.2$Count<-0
for(i in 1:nrow(figtemp1.2)){
  figtemp1.2$Count[i]<-nrow(figtemp1.1[figtemp1.1$TRcomflag2==figtemp1.2$TRcomflag2[i]&figtemp1.1$Count.allregs==figtemp1.2$Count.allregs[i],])
}
figtemp1.2$label<-figtemp1.2$Count
figtemp1.2[figtemp1.2$Count==0,]$label<-""
figtemp1.2$TRcomflag2<-factor(figtemp1.2$TRcomflag2,levels=c("TRcom-containing","TRcom-less","Null"))
figtemp1.2$DMenv=-0.5
figtemp1.2[figtemp1.2$Count==0,]$DMenv<-NA
figtemp1.2<-figtemp1.2[figtemp1.2$Count!=0,]
figtemp1.2<-figtemp1.2 %>% distinct(Count.allregs,TRcomflag2,.keep_all = T)
p0<-ggbarplot(figtemp1.2,x="Count.allregs",y="DMenv",fill="TRcomflag2",width=0.5,size=0,
              ylim=c(0.4,0.5),
              palette = c(mycolpal[2],"gray50",mycolpal[1]),
              position = position_dodge(width=0.7))+
   ylab(expression(DM[env]~"[a.u.]"))
p0<-ggpar(p0,legend="")+
  geom_text(aes(label=label,x=Count.allregs,y=0.5,fill=TRcomflag2,color=TRcomflag2),position=position_dodge(width=1),angle=90,size=3)+
  scale_y_continuous(limits=c(0.4,0.6))+
  clean_theme()
p0


# Error plots
figlist<-list()
##### Env
p<-ggerrorplot(dtrget_stat[!is.na(dtrget_stat$Use),],x="Count.allregs",y="DMenv",
          color="TRcomflag2",palette = c(mycolpal[2],"gray50",mycolpal[1]),
          desc_stat = "mean_sd",
          ylim=c(-0.7,2.8),legend="",
          xlab="Number of regulation")+
  ylab(expression(DM[env]~"[a.u.]"))+
  stat_pvalue_manual(stat.test1, label = "p.adj.signif", tip.length = 0,hide.ns=F,size = 4) +
  stat_cor(data=dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="Others"&dtrget_stat$Count.allregs<6,],method = "spearman", label.x = 6.7, label.y = 2.6,color=mycolpal[2],size=5)+ # Corr for 
  stat_cor(data=dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="TRcom"&dtrget_stat$Count.allregs<6,],method = "spearman", label.x = 6.7, label.y = 2,color="gray10",size=5)+
  scale_y_continuous(breaks = get_breaks(by = 1, from = -0.5),limits =  c(-0.7, 2.7))
p<-p+theme(axis.text.x= element_blank(),axis.title.x=element_blank())
p<-ggpar(p,font.y = c(17))
figlist[[1]]<-p
##### Evo
p<-ggerrorplot(dtrget_stat[!is.na(dtrget_stat$Use),],x="Count.allregs",y="DMevo",
          color="TRcomflag2",palette = c(mycolpal[2],"gray50",mycolpal[1]),
          desc_stat = "mean_sd",
          ylim=c(-0.6,2.1),legend="",
          xlab="Number of regulation")+
  ylab(expression(DM[evo]~"[a.u.]"))+
  stat_pvalue_manual(stat.test2, label = "p.adj.signif", tip.length = 0,hide.ns=F,size = 4) +
  stat_cor(data=dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="Others"&dtrget_stat$Count.allregs<6,],method = "spearman", label.x = 6.7, label.y = 1.9,color=mycolpal[2],size=5)+
  stat_cor(data=dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="TRcom"&dtrget_stat$Count.allregs<6,],method = "spearman", label.x = 6.7, label.y = 1.4,color="gray10", size=5)+
  scale_y_continuous(breaks = get_breaks(by = 1, from = -0.5),limits =  c(-0.7, 2))
p<-p+theme(axis.text.x= element_blank(),axis.title.x=element_blank())
p<-ggpar(p,font.y = c(17))
figlist[[2]]<-p
##### Mut
p<-ggerrorplot(dtrget_stat[!is.na(dtrget_stat$Use),],x="Count.allregs",y="DMmut",
          color="TRcomflag2",palette = c(mycolpal[2],"gray50",mycolpal[1]),
          desc_stat = "mean_sd",
          ylim=c(-0.6,1.7),legend="",
          xlab="Number of unique regulators of target genes")+
  ylab(expression(DM[mut]~"[a.u.]"))+
   stat_pvalue_manual(stat.test3, label = "p.adj.signif", tip.length = 0,hide.ns=F,size = 4) +
    stat_cor(data=dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="Others"&dtrget_stat$Count.allregs<6,],method = "spearman", label.x = 6.7, label.y = 1.5,color=mycolpal[2],size=5)+
  stat_cor(data=dtrget_stat[!is.na(dtrget_stat$Use)&dtrget_stat$TRcomflag!="TRcom"&dtrget_stat$Count.allregs<6,],method = "spearman", label.x = 6.7, label.y = 1.1,color="gray10",size=5)+
  scale_y_continuous(breaks = get_breaks(by = 1, from = -0.5),limits =  c(-0.7, 2))
p<-ggpar(p,font.y = c(17),xlab = FALSE)
figlist[[3]]<-p

my_labs<-c(bquote(TR[com]-containing),bquote(TR[com]-less),bquote(Null))
names(my_labs)<-c("TRcom-containing","TRcom-less","Null")
q<-ggpar(p,legend = "top",legend.title = "",font.ytickslab= c(10),font.legend = c(15))+
  # scale_color_discrete(labels = my_labs)+
  scale_color_manual(values = c(mycolpal[2],"gray50",mycolpal[1]),labels=my_labs)
leg <- ggpubr::get_legend(q)

lay<-t(rbind(seq(1,5)))
blank <- grid.rect(gp=gpar(col="white"))
grid.arrange(as_ggplot(leg),blank,figlist[[1]],figlist[[2]],figlist[[3]],
             layout_matrix=lay,
             bottom=textGrob("Number of unique regulators",gp=gpar(fontsize=16)),
             heights=c(0.2,0.1,rep(1.1,3)))
Sys.time()
```
## 8.11.2.
```{r TRN,warning=FALSE, message=FALSE,fig.height=6,fig.width=4.5,dpi=300}
# Error plots
figlist<-list()
p<-ggerrorplot(dtrget_stat[!is.na(dtrget_stat$Use),],x="Count.allregs",y="DMenv",
          # color="TRcomflag2",palette = c(mycolpal[2],"gray50",mycolpal[1]),
          desc_stat = "mean_sd",
          ylim=c(-0.7,2),legend="",
          xlab="Number of regulation")+
  stat_cor(data = dtrget_stat[!is.na(dtrget_stat$Use),],method="spearman",label.x = 1,label.y = 1.8,size=6)+
  ylab(expression(DM[env]~"[a.u.]"))
p<-p+theme(axis.text.x= element_blank(),axis.title.x=element_blank())
p<-ggpar(p,font.y = c(17),xlab = FALSE)
figlist[[1]]<-p

p<-ggerrorplot(dtrget_stat[!is.na(dtrget_stat$Use),],x="Count.allregs",y="DMevo",
          # color="TRcomflag2",palette = c(mycolpal[2],"gray50",mycolpal[1]),
          desc_stat = "mean_sd",
          ylim=c(-0.3,1.1),legend="",
          xlab="Number of regulation")+
  stat_cor(data = dtrget_stat[!is.na(dtrget_stat$Use),],method="spearman",label.x = 1,label.y = 1,size=6)+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 1.1))+
  ylab(expression(DM[evo]~"[a.u.]"))
p<-p+theme(axis.text.x= element_blank(),axis.title.x=element_blank())
p<-ggpar(p,font.y = c(17),xlab = FALSE)
figlist[[2]]<-p


p<-ggerrorplot(dtrget_stat[!is.na(dtrget_stat$Use),],x="Count.allregs",y="DMmut",
          # color="TRcomflag2",palette = c(mycolpal[2],"gray50",mycolpal[1]),
          desc_stat = "mean_sd",
          ylim=c(-0.3,1.2),legend="",
          xlab="Number of regulation")+
   stat_cor(data = dtrget_stat[!is.na(dtrget_stat$Use),],method="spearman",label.x = 1,label.y = 1.1,size=6)+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 1.2))+
  ylab(expression(DM[mut]~"[a.u.]"))
p<-p+theme(axis.title.x=element_blank())
p<-ggpar(p,font.y = c(17),xlab = FALSE)
figlist[[3]]<-p

lay<-t(rbind(seq(1,3)))
grid.arrange(figlist[[1]],figlist[[2]],figlist[[3]],
             layout_matrix=lay,
             bottom=textGrob("Number of unique regulators",gp=gpar(fontsize=16)),
             heights=c(1,1,1.1))
Sys.time()
```
## 8.12. Example of correlation between DMenv and the Number of TRcoms for target genes which has 5 unique regulators
### 8.12.1. Include all sigma factors
```{r TRN,warning=FALSE, message=FALSE,fig.height=5,fig.width=4,dpi=300}
# Example of correlation between DMenv and the Number of TRcoms for target genes which has 5 unique regulators
i<-5
figtemp3<-dtrget_stat[dtrget_stat$Count.allregs==i,]
## Error plot
p4<-ggerrorplot(figtemp3,x="Count.TRcom",y="DMenv",
                desc_stat = "mean_sd",
                legend="",
          color = "TRcomflag",palette=c(mycolpal[2],"gray50"),
          subtitle="Number of unique regulators: 5")+
  xlab(expression(Number~of~unique~paste(TR[com])))+
  ylab(expression(DM[env]~"[a.u.]"))+
  stat_cor(data=figtemp3,method="spearman",label.x=1.5,label.y = 1.5,size=5)+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))
p4<-ggpar(p4,font.x=c(15),font.y=c(15),font.subtitle = c(15))

## DMenv
p4.1<-ggerrorplot(figtemp3,x="Count.TRcom",y="DMenv",
                desc_stat = "mean_sd",
                legend="",
          color = "TRcomflag",palette=c(mycolpal[2],"gray50"),
          subtitle="Number of regulation: 5")+
  xlab(expression(Number~of~regulation~by~paste(TR[com])))+
  ylab(expression(DM[env]~"[a.u.]"))+
  stat_cor(data=figtemp3,method="spearman",label.x=0.5,label.y = 1)
p4.1<-p4.1+theme(axis.text.x= element_blank(),axis.title.x=element_blank())

## DMevo
p4.2<-ggerrorplot(figtemp3,x="Count.TRcom",y="DMevo",
                desc_stat = "mean_sd",
                legend="",
          color = "TRcomflag",palette=c(mycolpal[2],"gray50"))+
    xlab(expression(Number~of~regulation~by~paste(TR[com])))+
  ylab(expression(DM[evo]~"[a.u.]"))+
  stat_cor(data=figtemp3,method="spearman",label.x=0.5,label.y = 0.7)
p4.2<-p4.2+theme(axis.text.x= element_blank(),axis.title.x=element_blank())

## DMmut
p4.3<-ggerrorplot(figtemp3,x="Count.TRcom",y="DMmut",
                desc_stat = "mean_sd",
                legend="",
          color = "TRcomflag",palette=c(mycolpal[2],"gray50"))+
    xlab(expression(Number~of~regulation~by~paste(TR[com])))+
  ylab(expression(DM[mut]))+
  stat_cor(data=figtemp3,method="spearman",label.x=0.5,label.y = 1)

grid.arrange(p4.1,p4.2,p4.3,ncol=1,
             heights=c(1.1,1,1.3))

# Correlation was calculatedfor all TRcoms including sigma factors
figtemp4<-createEmptyDf(nrow=4*3,ncol=4,colnames=qw(Count.allregs,DM,Correlation,Pval))
figtemp4$Count.allregs<-rep(seq(2,5),3)
figtemp4$DM<-rep(qw(DMenv,DMevo,DMmut),each=4)

i<-1
for(i in 1:nrow(figtemp4)){
  temp14<-dtrget_stat[dtrget_stat$Count.allregs==figtemp4$Count.allregs[i],which(colnames(dtrget_stat) %in% c("Count.TRcom",figtemp4$DM[i]))]
  # test<-cor.test(x=temp14[,1],y=temp14[,2],method="spearman")
  figtemp4$Correlation[i]<-cor_test(data=temp14,method="spearman")$cor
  figtemp4$Pval[i]<-cor_test(data=temp14,method="spearman")$p
}
figtemp4$label<-ifelse(figtemp4$Pval>0.05,"ns",ifelse(figtemp4$Pval>0.01,"*",ifelse(figtemp4$Pval>0.001,"**",ifelse(figtemp4$Pval>0.0001,"***","****"))))
figtemp4$Count.allregs<-factor(figtemp4$Count.allregs,levels=seq(2,5))
figtemp4<-transform(figtemp4,Consider="All TRcom")
figtemp4$facet_lab<-ifelse(figtemp4$DM=="DMenv","DM[env]",ifelse(figtemp4$DM=="DMevo","DM[evo]","DM[mut]"))
levels(figtemp4$facet_lab)<-c("DM[env]","DM[evo]","DM[mut]")

p5<-ggbarplot(figtemp4,x="Count.allregs",y="Correlation",size=0,fill="black",label=figtemp4$label,
               xlab="Number of unique regulators",ylab="Spearman's correlation",ylim=c(0,0.38))
# p5<-p5+ facet_grid(.~facet_lab)
p5<-p5+facet_grid(.~facet_lab, labeller = label_parsed)+theme(strip.text = element_text(size = 15))
p5<-ggpar(p5,font.x = c(15),font.y=c(15))

  
lay<-t(rbind(seq(1,2)))
grid.arrange(p4,p5,
             layout_matrix=lay,
             heights=c(0.7,1))
Sys.time()
```
### 8.12.2. Remove all sigma factors
```{r TRN,warning=FALSE, message=FALSE,fig.height=5,fig.width=4,dpi=300}
i<-5
figtemp5<-dtrget_stat[(dtrget_stat$Count.allregs==i),]
figtemp5<-figtemp5[figtemp5$Count.sigma.of.TRcom==0,]

## Error plot
## DMenv
p6<-ggerrorplot(figtemp5,x="Count.TRcom.wosigma",y="DMenv",
                desc_stat = "mean_sd",
                legend="",
          color = "TRcomflag",palette=c(mycolpal[2],"gray50"),
          subtitle="Number of regulation: 5")+
    xlab("Number of regulation by non-sigma TRcom")+
  ylab(expression(DM[env]))+
  stat_cor(data=figtemp5,method="spearman",label.x=3,label.y = 2)

figtemp6<-createEmptyDf(nrow=4*3,ncol=4,colnames=qw(Count.allregs,DM,Correlation,Pval))
figtemp6$Count.allregs<-rep(seq(2,5),3)
figtemp6$DM<-rep(qw(DMenv,DMevo,DMmut),each=4)
i<-1
for(i in 1:nrow(figtemp6)){
  temp15<-dtrget_stat[dtrget_stat$Count.allregs==figtemp4$Count.allregs[i],which(colnames(dtrget_stat) %in% c("Count.TRcom","Count.TRcom.wosigma","Count.sigma.of.TRcom",figtemp6$DM[i]))]
  temp15<-temp15[temp15$Count.sigma.of.TRcom==0,]
  # test<-cor.test(x=temp14[,1],y=temp14[,2],method="spearman")
  figtemp6$Correlation[i]<-cor_test(data=temp15[,which(colnames(temp15) %in% c("Count.TRcom.wosigma",figtemp6$DM[i]))],method="spearman")$cor
  figtemp6$Pval[i]<-cor_test(data=temp15[,which(colnames(temp15) %in% c("Count.TRcom.wosigma",figtemp6$DM[i]))],method="spearman")$p
}
figtemp6$label<-ifelse(figtemp6$Pval>0.05,"ns",ifelse(figtemp6$Pval>0.01,"*",ifelse(figtemp6$Pval>0.001,"**",ifelse(figtemp6$Pval>0.0001,"***","****"))))
figtemp6$Count.allregs<-factor(figtemp6$Count.allregs,levels=seq(2,5))
figtemp6<-transform(figtemp6,Consider="Non-sigma TRcom")

p7<-ggbarplot(figtemp6,x="Count.allregs",y="Correlation",facet="DM",size=0,fill="black",label=figtemp6$label,
              xlab="Number of unique regulators of target genes",ylab="Spearman's correlation",ylim=c(0,0.45))

lay<-t(rbind(seq(1,2)))
grid.arrange(p6,p7,
             layout_matrix=lay,
             heights=c(1,1))

Sys.time()
```
### 8.12.3.
```{r TRN,warning=FALSE, message=FALSE,fig.height=5,fig.width=4,dpi=300}
figtemp7<-bind_rows(figtemp4,figtemp6)
p8<-ggplot(figtemp7,aes(x=Count.allregs,y=Correlation,fill=Consider))+ #enrichmentScore
      geom_bar(stat="identity",aes(fill=Consider),width=0.5,position=position_dodge(width=0.7))+
      geom_text(aes(label=label,x=Count.allregs,y=Correlation+0.02),position=position_dodge(width=0.7))+
   scale_fill_manual(values =  get_palette(palette = "Reds",3)[c(3,2)])
p8<-ggpar(p8,legend="top",legend.title = "",xlab="Number of unique regulators of target genes",ylab="Spearman's correlation")+theme_pubr()
p8<-facet(p8,facet.by = "DM")

lay<-t(rbind(seq(1,2)))
grid.arrange(p4,p8,
             layout_matrix=lay,
             heights=c(1,1.3))

Sys.time()
```
## 8.13. Relationship between the number of unique regulators, including indirect regulations, and DM values
```{r TRN,warning=FALSE, message=FALSE,fig.height=3,fig.width=4,dpi=300}
my_g<-as_tbl_graph(dnet_edgeinfo_ts.global,directed = T)
TRN.ts.subcomp<-data.frame(Regulator=V(my_g)$name,Subcomponent=NA,Subcomp_size=0)
for(i in 1:length(V(my_g))){
  my_subcomp<-igraph::subcomponent(graph = my_g,v = V(my_g)[i],mode = "in")
  TRN.ts.subcomp$Subcomponent[i]<-paste(my_subcomp$name,collapse = ",")
  TRN.ts.subcomp$Subcomp_size[i]<-length(my_subcomp)
}
df.regcount.subcomp<-df.regcount
df.regcount.subcomp$Subcomponent<-NA
df.regcount.subcomp$Subcomp_size<-0
i<-1
for(i in 1:nrow(df.regcount.subcomp)){
  if(!is.na(df.regcount.subcomp$regulators[i])){
    my_regs<-df.regcount.subcomp$regulators[i]
    my_df<-TRN.ts.subcomp[TRN.ts.subcomp$Regulator %in% str_split(my_regs,",")[[1]],]
    my_regs<-unique(str_split(paste(my_df$Subcomponent,collapse = ","),",")[[1]])
    df.regcount.subcomp$Subcomponent[i]<-paste(my_regs,collapse = ",")
    df.regcount.subcomp$Subcomp_size[i]<-length(my_regs)
  }
}
ggscatter(df.regcount.subcomp,y="DMenv",x="Subcomp_size",size=0.1,alpha=0.1)+
  stat_cor(method="spearman",label.x = 3,label.y = 0)
Sys.time()
```
#
#
#
# 9. PCA
https://personal.utdallas.edu/~herve/abdi-awPCA2010.pdf # detailed manual of PCA
https://stats.stackexchange.com/questions/495342/pca-and-variable-contributions-to-first-n-dimensions
http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/112-pca-principal-component-analysis-essentials/
## 9.1. PCA for the Env dataset
### 9.1.1 Detect outliers in PC1~PC3
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-log_tpm_pl.pr2[log_tpm_pl.pr2$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
figtemp<-as.data.frame(t(temp0[,which(colnames(temp0) %nin% "Bnb")]))# transpose the matrix so genes are as columns
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
env_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    env_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(env_outlier[!is.na(env_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
### 9.1.2 Remove outliers if it is necessary and examine PCA
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
mypcs<-paste("PC",seq(1:3),sep="")
temp<-log_tpm_pl.pr2[log_tpm_pl.pr2$Bnb %in% dBnb.comp$Bnb,]
if(!is.na(env_outlier)){
  temp<-temp[,colnames(temp) %nin% env_outlier]
}
temp<-temp[order(temp$Bnb,decreasing = F),]
figtemp<-as.data.frame(t(temp[,which(colnames(temp) %nin% "Bnb")]))
colnames(figtemp)<-temp$Bnb
figtemp<-figtemp %>% select_if(negate(anyNA))
res.pca.global <- prcomp(figtemp, scale. = T)
res.pca.global_env<-res.pca.global
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_env <- res.pca.global$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_env$label<-rownames(pc_scores_env)

pc_eigenvalues_env <- res.pca.global_env$sdev^2
pc_eigenvalues_env <- tibble(PC = factor(1:length(pc_eigenvalues_env)), 
                         variance = pc_eigenvalues_env) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained
# pc_eigenvalues_env %>% 
#   ggplot(aes(x = PC)) +
#   geom_col(aes(y = pct)) +
#   geom_line(aes(y = pct_cum, group = 1)) + 
#   geom_point(aes(y = pct_cum)) +
#   labs(x = "Principal component", y = "Fraction variance explained")

# Scree plot
figtemp1<-pc_eigenvalues_env
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree.env<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,color="#619CFF",plot_type = "l",
       xlab="Components",ylab="Fraction variance explained (%)")+
  scale_x_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,80))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# PC1 vs. PC2
ind.p<-ggscatter(data=pc_scores_env,x="PC1",y="PC2",alpha=0.2,
                 # color = "Condition",label="Condition",repel=T,font.label=4,
                 legend="")+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
ind.p<-ggpar(ind.p,
      xlab = paste("PC1 (", round(pc_eigenvalues_env$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_env$pct[2],digits = 0),"%)",sep=""))
ind.p
# q<-ggpar(ind.p,legend = "right",legend.title="",font.legend = c(4))
# leg <- ggpubr::get_legend(q)
# as_ggplot(leg)


# reconstruct original dataset using only PCx
# https://sinhrks.hatenablog.com/entry/2014/10/19/000929
# k<-3
# reconst_pck<-res.pca.global_env$x[,1:k] %*% t(res.pca.global_env$rotation[,1:k])
# reconst_pck <- scale(reconst_pck, center =F, scale=1/res.pca.global_env$scale)
# reconst_pck<- scale(reconst_pck, center =-res.pca.global_env$center, scale=F)

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_env$pct_cum[k] ,"% variance",sep=""))

pca.glob.var_env<-get_pca_var(res.pca.global_env)

Sys.time()
```
### 9.1.3. 2D plot (PC1~PC3)
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
my_tmp<-fileinfo_plasticity.pr2 %>% select(sample,sample_group,Growth.Rate..1.hr.)
colnames(my_tmp)[3]<-"Growth_rate"
figtemp<-left_join(pc_scores_env,my_tmp,by=c("Condition"="sample"))
figtemp<-figtemp %>% select(Condition,sample_group,PC1,PC2,PC3)
p1<-ggscatter(data=figtemp,x="PC1",y="PC2",
                 alpha=0.5,
                 color = "sample_group",legend.title="",legend="",
                 )+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,
      xlab = paste("PC1 (", round(pc_eigenvalues_env$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_env$pct[2],digits = 0),"%)",sep=""))

p2<-ggscatter(data=figtemp,x="PC2",y="PC3",
                 alpha=0.5,
                 color = "sample_group",legend.title="",legend="",
                 )+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,
      xlab = paste("PC2 (", round(pc_eigenvalues_env$pct[2],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_env$pct[3],digits = 0),"%)",sep=""))

p3<-ggscatter(data=figtemp,x="PC1",y="PC3",
                 alpha=0.5,
                 color = "sample_group",legend.title="",legend="",
                 )+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,
      xlab = paste("PC1 (", round(pc_eigenvalues_env$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_env$pct[3],digits = 0),"%)",sep=""))

# q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")
# leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,widths=c(1,1,1),nrow=1)
Sys.time()
```
### 9.1.4. Correlation between PCs and th growth rate
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
temp<-fileinfo_plasticity.pr2[,which(colnames(fileinfo_plasticity.pr2) %in% c("sample","Growth.Rate..1.hr.") )]
colnames(temp)[colnames(temp)=="Growth.Rate..1.hr."]<-"Growth_Rate"
figtemp<-left_join(pc_scores_env,temp,by=c("Condition"="sample"))
p1<-ggscatter(data=figtemp,x="PC1",y="Growth_Rate",alpha=0.5,legend="",
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=25,label.y=1))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,xlab = "PC1",ylab = "Growth rate [1/h]")

p2<-ggscatter(data=figtemp,x="PC2",y="Growth_Rate",alpha=0.5,legend="",
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=25,label.y=1))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,xlab = "PC2",ylab = "Growth rate [1/h]")

p3<-ggscatter(data=figtemp,x="PC3",y="Growth_Rate",alpha=0.5,legend="",
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=25,label.y=1))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,xlab = "PC3",ylab = "Growth rate [1/h]")

grid.arrange(p1,p2,p3,widths=c(1,1,1),nrow=1)
Sys.time()
```
### 9.1.5. Mapping the Evo dataset onto the PC space defined by the Env dataset
https://stats.stackexchange.com/questions/8630/principal-component-analysis-backwards-how-much-variance-of-the-data-is-expla
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-dcol16.qnorm[dcol16.qnorm$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% col16.strain)])
dscale<-data.frame(Bnb=rownames(temp0),center=rowMeans(temp0),scale=rowSds(as.matrix(temp0)))
temp0<-t(temp0) # transpose the matrix so genes are as columns
temp<-scale(temp0) # apply scaling to each column of the matrix (genes)

print(sum(pc_eigenvalues_env$variance)) # this should be the number of genes in the dataset (i.e. identical to nrow(dBnb.com))
# Calc % of variance of the Env explaind py PCs 
figtemp1<-pc_eigenvalues_env
figtemp1$PC<-as.numeric(figtemp1$PC)
figtemp1$variance.evo<-NA
figtemp1$pct.evo<-NA
figtemp1$pct_cum.evo<-NA
i<-1
for(i in 1:nrow(figtemp1)){
  temp2<-temp %*% res.pca.global_env$rotation[,i]
  figtemp1$variance.evo[i]<-var(temp2)
}
figtemp1$pct.evo<-figtemp1$variance.evo/sum(pc_eigenvalues_env$variance)*100
figtemp1$pct_cum.evo<-cumsum(figtemp1$pct.evo)
dscree.evo<-figtemp1

figtemp2<-figtemp1 %>% pivot_longer(cols=qw(pct_cum,pct_cum.evo),names_to = qw(Source),values_to =qw(Fraction_pct))
figtemp2[figtemp2$Source=="pct_cum",]$Source<-"Env"
figtemp2[figtemp2$Source=="pct_cum.evo",]$Source<-"Evo"
figtemp2$Source<-factor(figtemp2$Source,levels=qw(Env,Evo))
ggline(figtemp2,x="PC",y="Fraction_pct",numeric.x.axis = T,color="Source",palette = c("#619CFF","#F8766D"),
       legend.title="",plot_type ="l",
       xlab="Components",ylab="Fraction variance explained (%)")+
  scale_x_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,80))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

Sys.time()
```
### 9.1.6. Mapping the Mut dataset (Anc, R10, and R30) onto the PC plane defined by the Env dataset
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-datlog.norm[datlog.norm$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% MAlin.sel)])
dscale<-data.frame(Bnb=rownames(temp0),center=rowMeans(temp0),scale=rowSds(as.matrix(temp0)))
temp0<-t(temp0) # transpose the matrix so genes are as columns
temp<-scale(temp0) # apply scaling to each column of the matrix (genes)

print(sum(pc_eigenvalues_env$variance)) # this should be the number of genes in the dataset (i.e. identical to nrow(dBnb.com))
# Calc % of variance of the Env explaind py PCs 
figtemp1<-pc_eigenvalues_env
figtemp1$PC<-as.numeric(figtemp1$PC)
figtemp1$variance.mut<-NA
figtemp1$pct.mut<-NA
figtemp1$pct_cum.mut<-NA
i<-1
for(i in 1:nrow(figtemp1)){
  temp2<-temp %*% res.pca.global_env$rotation[,i]
  figtemp1$variance.mut[i]<-var(temp2)
}
figtemp1$pct.mut<-figtemp1$variance.mut/sum(pc_eigenvalues_env$variance)*100
figtemp1$pct_cum.mut<-cumsum(figtemp1$pct.mut)
dscree.mut<-figtemp1

figtemp2<-figtemp1 %>% pivot_longer(cols=qw(pct_cum,pct_cum.mut),names_to = qw(Source),values_to =qw(Fraction_pct))
figtemp2[figtemp2$Source=="pct_cum",]$Source<-"Env"
figtemp2[figtemp2$Source=="pct_cum.mut",]$Source<-"Mut"
figtemp2$Source<-factor(figtemp2$Source,levels=qw(Env,Mut))
ggline(figtemp2,x="PC",y="Fraction_pct",numeric.x.axis = T,color="Source",palette = c("#619CFF","#00BA38"),
       legend.title="",plot_type ="l",
       xlab="Components",ylab="Fraction variance explained (%)")+
  scale_x_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,80))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

Sys.time()
```
### 9.1.7. Mapping the Mut dataset (R30, R10, separately) onto the PC plane defined by the Env dataset
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# R30
MAlin.sel.R30<-MAlin.sel[grepl("_30",MAlin.sel)]
temp0<-datlog.norm[datlog.norm$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% MAlin.sel.R30)])
dscale.R30<-data.frame(Bnb=rownames(temp0),center=rowMeans(temp0),scale=rowSds(as.matrix(temp0)))
temp0<-t(temp0) # transpose the matrix so genes are as columns
temp<-scale(temp0) # apply scaling to each column of the matrix (genes)

# print(sum(pc_eigenvalues_env$variance)) # this should be the number of genes in the dataset (i.e. identical to nrow(dBnb.com))
# Calc % of variance of the Env explaind py PCs 
figtemp1<-pc_eigenvalues_env
figtemp1$PC<-as.numeric(figtemp1$PC)
figtemp1$variance.mut.R30<-NA
figtemp1$pct.mut.R30<-NA
figtemp1$pct_cum.mut.R30<-NA
i<-1
for(i in 1:nrow(figtemp1)){
  temp2<-temp %*% res.pca.global_env$rotation[,i]
  figtemp1$variance.mut.R30[i]<-var(temp2)
}
figtemp1$pct.mut.R30<-figtemp1$variance.mut.R30/sum(pc_eigenvalues_env$variance)*100
figtemp1$pct_cum.mut.R30<-cumsum(figtemp1$pct.mut.R30)
dscree.mut.R30<-figtemp1


# R10
MAlin.sel.R10<-MAlin.sel[grepl("_10",MAlin.sel)]
temp0<-datlog.norm[datlog.norm$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% MAlin.sel.R10)])
dscale.R10<-data.frame(Bnb=rownames(temp0),center=rowMeans(temp0),scale=rowSds(as.matrix(temp0)))
temp0<-t(temp0) # transpose the matrix so genes are as columns
temp<-scale(temp0) # apply scaling to each column of the matrix (genes)

# print(sum(pc_eigenvalues_env$variance)) # this should be the number of genes in the dataset (i.e. identical to nrow(dBnb.com))
# Calc % of variance of the Env explaind py PCs 
figtemp1<-pc_eigenvalues_env
figtemp1$PC<-as.numeric(figtemp1$PC)
figtemp1$variance.mut.R10<-NA
figtemp1$pct.mut.R10<-NA
figtemp1$pct_cum.mut.R10<-NA
i<-1
for(i in 1:nrow(figtemp1)){
  temp2<-temp %*% res.pca.global_env$rotation[,i]
  figtemp1$variance.mut.R10[i]<-var(temp2)
}
figtemp1$pct.mut.R10<-figtemp1$variance.mut.R10/sum(pc_eigenvalues_env$variance)*100
figtemp1$pct_cum.mut.R10<-cumsum(figtemp1$pct.mut.R10)
dscree.mut.R10<-figtemp1

figtemp2<-left_join(dscree.mut.R10,dscree.mut.R30[,which(colnames(dscree.mut.R30) %in% qw(PC,pct_cum.mut.R30))])
figtemp2<-figtemp2 %>% pivot_longer(cols=qw(pct_cum,pct_cum.mut.R30,pct_cum.mut.R10),names_to = qw(Source),values_to =qw(Fraction_pct))
figtemp2[figtemp2$Source=="pct_cum",]$Source<-"Env"
figtemp2[figtemp2$Source=="pct_cum.mut.R10",]$Source<-"MutR10"
figtemp2[figtemp2$Source=="pct_cum.mut.R30",]$Source<-"MutR30"
ggline(figtemp2,x="PC",y="Fraction_pct",numeric.x.axis = T,color="Source",palette = c("#619CFF",rep("#00BA38",2)),
       legend.title="",linetype = "Source",plot_type = "l",
       xlab="Components",ylab="Fraction variance explained (%)")+
  scale_x_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,80))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

Sys.time()
```
### 9.1.8. Scree plot for Env, Evo, and  Mut
```{r PCA_Env,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=2.5,dpi=300}
figtemp3<-dscree.env
figtemp3<-left_join(figtemp3,dscree.mut[,which(colnames(dscree.mut) %in% qw(PC,pct_cum.mut))])
figtemp3<-left_join(figtemp3,dscree.evo[,which(colnames(dscree.evo) %in% qw(PC,pct_cum.evo))])
figtemp3<-left_join(figtemp3,dscree.mut.R10[,which(colnames(dscree.mut.R10) %in% qw(PC,pct_cum.mut.R10))])
figtemp3<-left_join(figtemp3,dscree.mut.R30[,which(colnames(dscree.mut.R30) %in% qw(PC,pct_cum.mut.R30))])


figtemp3<-figtemp3 %>% pivot_longer(cols=qw(pct_cum,pct_cum.mut,pct_cum.evo,pct_cum.mut.R10,pct_cum.mut.R30),names_to = qw(Source),values_to =qw(Fraction_pct))
# figtemp3<-figtemp3 %>% pivot_longer(cols=qw(pct_cum,pct_cum.mut,pct_cum.evo),names_to = qw(Source),values_to =qw(Fraction_pct))
figtemp3[figtemp3$Source=="pct_cum",]$Source<-"Env"
figtemp3[figtemp3$Source=="pct_cum.mut",]$Source<-"Mut"
figtemp3[figtemp3$Source=="pct_cum.evo",]$Source<-"Evo"
figtemp3[figtemp3$Source=="pct_cum.mut.R10",]$Source<-"Mut(R10)"
figtemp3[figtemp3$Source=="pct_cum.mut.R30",]$Source<-"Mut(R30)"

# figtemp3$Source<-factor(figtemp3$Source,levels=qw(Env,Evo,Mut))
# temp<-c("#619CFF","#F8766D","#00BA38")
figtemp3$Source<-factor(figtemp3$Source,levels=c(qw(Env,Evo,Mut),"Mut(R10)","Mut(R30)"))
temp<-c("#619CFF","#F8766D","#00BA38",rep("#00BA38",3))

p<-ggline(figtemp3,x="PC",y="Fraction_pct",numeric.x.axis = T,color="Source",plot_type = "l",
       legend.title="",legend="",palette = temp,linetype = "Source",
       xlab="Components",ylab="Variance explained (%)")+
  scale_linetype_manual(values = c(rep("solid",3),"dotted","dashed"))+
  geom_vline(xintercept = c(3,30),color="gray70")+
  scale_x_continuous(breaks = get_breaks(by = 40, from = 0),limits = c(0,160))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))
p
q<-ggpar(p,legend = "right",legend.title="",font.legend = c(8))
leg <- ggpubr::get_legend(q)
as_ggplot(leg)

p1<-ggline(figtemp3[figtemp3$Source %in% qw(Env,Evo,Mut),],x="PC",y="Fraction_pct",numeric.x.axis = T,color="Source",plot_type = "l",
       legend.title="",legend="",palette = temp,linetype = "Source",
       xlab="Components of Env",ylab="Variance explained (%)")+
  scale_linetype_manual(values = c(rep("solid",3),"dotted","dashed"))+
  # geom_vline(xintercept = c(3,30),color="gray70")+
  scale_x_continuous(breaks = get_breaks(by = 40, from = 0),limits = c(0,160))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))
p1

print(paste("The Env dataset explains",figtemp3[(figtemp3$PC==max(figtemp3$PC))&figtemp3$Source=="Evo",]$Fraction_pct, "% of variance in the Evo dataset"))
print(paste("The Env dataset explains",figtemp3[(figtemp3$PC==max(figtemp3$PC))&figtemp3$Source=="Mut",]$Fraction_pct, "% of variance in the Mut dataset"))

Sys.time()
```
## 9.2. PCA for the Evo dataset
### 9.2.1 Detect outliers in PC1~PC3
```{r PCA_Evo,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-dcol16.qnorm[dcol16.qnorm$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% col16.strain)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
evo_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    evo_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(evo_outlier[!is.na(evo_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
### 9.2.2 Remove outliers if it is necessary and examine PCA
```{r PCA_Evo,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-dcol16.qnorm[dcol16.qnorm$Bnb %in% dBnb.comp$Bnb,]
if(!is.na(evo_outlier)){
  temp0<-temp0[,colnames(temp0) %nin% evo_outlier]
}

temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% col16.strain)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
res.pca.global_evo <- prcomp(figtemp, scale. = T)
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_evo <- res.pca.global_evo$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_evo$label<-rownames(pc_scores_evo)

pc_eigenvalues_evo <- res.pca.global_evo$sdev^2
pc_eigenvalues_evo <- tibble(PC = factor(1:length(pc_eigenvalues_evo)), 
                         variance = pc_eigenvalues_evo) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained

# Scree plot
figtemp1<-pc_eigenvalues_evo
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree_evo.evo<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,plot_type = "l",
        color="#F8766D",
       xlab="Components",ylab="Fraction of variance explained (%)")+
  scale_x_continuous(breaks = get_breaks(by = 4, from = 0),limits = c(0,16))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# PC1 vs. PC2
figtemp2<-pc_scores_evo
figtemp2[figtemp2$Condition=="X41_1Ti9",]$Condition<-"41-1Ti9"
figtemp2<-left_join(figtemp2,strain.tbl[,which(colnames(strain.tbl) %in% c("Strain.name","Phylogroup"))],by=c("Condition"="Strain.name"))
figtemp2$Phylogroup<-factor(figtemp2$Phylogroup,levels=qw(A,B1,B2,D,E,F))
ind.p<-ggscatter(data=figtemp2,x="PC1",y="PC2",
                 # alpha=0.2,
                 legend="",
                 color = "Phylogroup",label="Condition",repel=T,font.label=10
                 )+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
ind.p<-ggpar(ind.p,
      xlab = paste("PC1 (", round(pc_eigenvalues_evo$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_evo$pct[2],digits = 0),"%)",sep=""))
ind.p
my_fontsize<-12
q<-ggpar(ind.p,legend = "right",legend.title="",font.legend = c(8))+
  guides(colour = guide_legend(title= "Phylogroups",title.position = "left",label.theme = element_text(size=my_fontsize)))+
  theme(legend.title = element_text(angle = 90, hjust = 0.5, vjust = 0.5,size=my_fontsize))
leg <- ggpubr::get_legend(q)
as_ggplot(leg)

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_evo$pct_cum[k] ,"% variance",sep=""))

pca.glob.var_evo<-get_pca_var(res.pca.global_evo)

Sys.time()
```
### 9.2.3. 2D plot (PC1~PC3)
```{r PCA_Evo,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
figtemp<-pc_scores_evo
figtemp[figtemp$Condition=="X41_1Ti9",]$Condition<-"41-1Ti9"
figtemp<-left_join(figtemp,strain.tbl[,which(colnames(strain.tbl) %in% c("Strain.name","Phylogroup"))],by=c("Condition"="Strain.name"))
figtemp$Phylogroup<-factor(figtemp$Phylogroup,levels=qw(A,B1,B2,D,E,F))
p1<-ggscatter(data=figtemp,x="PC1",y="PC2",alpha=1,legend="",color = "Phylogroup")+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,
      xlab = paste("PC1 (", round(pc_eigenvalues_evo$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_evo$pct[2],digits = 0),"%)",sep=""))

p2<-ggscatter(data=figtemp,x="PC2",y="PC3",alpha=1,legend="",color = "Phylogroup")+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,
      xlab = paste("PC2 (", round(pc_eigenvalues_evo$pct[2],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_evo$pct[3],digits = 0),"%)",sep=""))

p3<-ggscatter(data=figtemp,x="PC1",y="PC3",alpha=1,legend="",color = "Phylogroup")+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,
      xlab = paste("PC1 (", round(pc_eigenvalues_evo$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_evo$pct[3],digits = 0),"%)",sep=""))

my_fontsize<-12
q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")+
    guides(colour = guide_legend(title= "Phylogroups",title.position = "left",label.theme = element_text(size=my_fontsize)))+
  theme(legend.title = element_text(angle = 90, hjust = 0.5, vjust = 0.5,size=my_fontsize))
leg <- ggpubr::get_legend(q)


grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,0.3),nrow=1)
Sys.time()
```
## 9.3. PCA for the Mut dataset
### 9.3.1. Detect outliers in PC1~PC3
```{r PCA_Mut,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-datlog.norm[datlog.norm$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% MAlin.sel)])
dscale<-data.frame(Bnb=rownames(temp0),center=rowMeans(temp0),scale=rowSds(as.matrix(temp0)))
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
mut_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    mut_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(mut_outlier[!is.na(mut_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
### 9.3.2. Remove outliers if it is necessary and examine PCA
```{r PCA_Mut,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-datlog.norm[datlog.norm$Bnb %in% dBnb.comp$Bnb,]
if(!is.na(mut_outlier)){
  temp0<-temp0[,colnames(temp0) %nin% mut_outlier]
}
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% MAlin.sel)])
dscale<-data.frame(Bnb=rownames(temp0),center=rowMeans(temp0),scale=rowSds(as.matrix(temp0)))
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
res.pca.global_mut <- prcomp(figtemp, scale. = T)
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_mut <- res.pca.global_mut$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_mut$label<-rownames(pc_scores_mut)

pc_eigenvalues_mut <- res.pca.global_mut$sdev^2
pc_eigenvalues_mut <- tibble(PC = factor(1:length(pc_eigenvalues_mut)), 
                         variance = pc_eigenvalues_mut) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained

# Scree plot
figtemp1<-pc_eigenvalues_mut
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree_mut.mut<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,plot_type = "l",
        color="#00BA38",
       xlab="Components",ylab="Fraction of variance explained (%)")+
  scale_x_continuous(breaks = get_breaks(by = 4, from = 0),limits = c(0,16))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# PC1 vs. PC2
figtemp2<-pc_scores_mut
figtemp2<-left_join(figtemp2, datinfo[,which(colnames(datinfo) %in% qw(Transfer,Lineage,Anc,label))],by=c("Condition"="label"))
figtemp2[figtemp2$Condition=="AncHa.mean",]$Transfer<-0
figtemp2[figtemp2$Condition=="AncHa.mean",]$Anc<-"AncHa"
figtemp2[figtemp2$Condition=="AncHb.mean",]$Transfer<-0
figtemp2[figtemp2$Condition=="AncHb.mean",]$Anc<-"AncHb"
figtemp2$col<-NA
figtemp2$label.2<-paste(figtemp2$Anc,figtemp2$Transfer,sep="_")
figtemp2$label.2<-factor(figtemp2$label.2,levels=paste(rep(c("AncHa","AncHb"),each=3),rep(c(0,10,30),2),sep="_"))


ind.p<-ggscatter(data=figtemp2,x="PC1",y="PC2",
                 # alpha=0.2,
                 color="label.2",palette = mycolpal.BuRd,
                 legend="")+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
ind.p<-ggpar(ind.p,
      xlab = paste("PC1 (", round(pc_eigenvalues_mut$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_mut$pct[2],digits = 0),"%)",sep=""))
ind.p
q<-ggpar(ind.p,legend = "right",legend.title="",font.legend = c(8))
leg <- ggpubr::get_legend(q)
as_ggplot(leg)

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_mut$pct_cum[k],"% variance",sep=""))

pca.glob.var_mut<-get_pca_var(res.pca.global_mut)

Sys.time()
```
### 9.3.3. 2D plot (PC1~PC3)
```{r PCA_Mut,warning=FALSE, message=FALSE,fig.height=3,fig.width=10,dpi=300}
figtemp<-pc_scores_mut
figtemp<-left_join(figtemp, datinfo[,which(colnames(datinfo) %in% qw(Transfer,Lineage,Anc,label))],by=c("Condition"="label"))
figtemp[figtemp$Condition=="AncHa.mean",]$Transfer<-0
figtemp[figtemp$Condition=="AncHa.mean",]$Anc<-"AncHa"
figtemp[figtemp$Condition=="AncHb.mean",]$Transfer<-0
figtemp[figtemp$Condition=="AncHb.mean",]$Anc<-"AncHb"
figtemp$col<-NA
figtemp$label.2<-paste(figtemp$Anc,figtemp$Transfer,sep="_")
figtemp$label.2<-factor(figtemp$label.2,levels=paste(rep(c("AncHa","AncHb"),each=3),rep(c(0,10,30),2),sep="_"))

p1<-ggscatter(data=figtemp,x="PC1",y="PC2",alpha=1,legend="",color="label.2",palette = mycolpal.BuRd)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,
      xlab = paste("PC1 (", round(pc_eigenvalues_mut$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_mut$pct[2],digits = 0),"%)",sep=""))

p2<-ggscatter(data=figtemp,x="PC2",y="PC3",alpha=1,legend="",color="label.2",palette = mycolpal.BuRd)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,
      xlab = paste("PC2 (", round(pc_eigenvalues_mut$pct[2],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_mut$pct[3],digits = 0),"%)",sep=""))

p3<-ggscatter(data=figtemp,x="PC1",y="PC3",alpha=1,legend="",color="label.2",palette = mycolpal.BuRd)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,
      xlab = paste("PC1 (", round(pc_eigenvalues_mut$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_mut$pct[3],digits = 0),"%)",sep=""))

q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")
leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,0.4),nrow=1)
Sys.time()
```
### 9.3.4. Correlation between PCs and th growth rate
```{r PCA_Mut,warning=FALSE, message=FALSE,fig.height=3,fig.width=10,dpi=300}
figtemp<-pc_scores_mut
figtemp<-left_join(figtemp, datinfo[,which(colnames(datinfo) %in% qw(Transfer,Lineage,Anc,label,Growth_Rate))],by=c("Condition"="label"))
figtemp[figtemp$Condition=="AncHa.mean",]$Transfer<-0
figtemp[figtemp$Condition=="AncHa.mean",]$Anc<-"AncHa"
figtemp[figtemp$Condition=="AncHb.mean",]$Transfer<-0
figtemp[figtemp$Condition=="AncHb.mean",]$Anc<-"AncHb"
figtemp$col<-NA
figtemp$label.2<-paste(figtemp$Anc,figtemp$Transfer,sep="_")
figtemp$label.2<-factor(figtemp$label.2,levels=paste(rep(c("AncHa","AncHb"),each=3),rep(c(0,10,30),2),sep="_"))
p1<-ggscatter(data=figtemp,x="PC1",y="Growth_Rate",legend="",
           color="label.2",palette = mycolpal.BuRd,   
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=20,label.y=1))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,xlab = "PC1",ylab = "Growth rate [1/h]")

p2<-ggscatter(data=figtemp,x="PC2",y="Growth_Rate",legend="",
              color="label.2",palette = mycolpal.BuRd,   
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=-20,label.y=1))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,xlab = "PC2",ylab = "Growth rate [1/h]")

p3<-ggscatter(data=figtemp,x="PC3",y="Growth_Rate",legend="",
              color="label.2",palette = mycolpal.BuRd,   
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=10,label.y=1))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,xlab = "PC3",ylab = "Growth rate [1/h]")


q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")
leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,0.4),nrow=1)
Sys.time()
```
## 9.4. Correlation between rotations
### 9.4.1. Correlation between rotation (eigenvector) for PC1~3 based on Env and rotation (eigenvector) for PC1~3 based on Evo or Mut
```{r PCA_all,warning=FALSE, message=FALSE,fig.height=4.5,fig.width=11,dpi=300}
#
my_lim<-c(0,0.65)

# Env vs. Evo
temp5<-data.frame(Env_PC=rep(seq(1,3),each=3),Evo_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp5)){
  temp1<-as.data.frame(res.pca.global$rotation[,temp5$Env_PC[i]])
  colnames(temp1)<-"rot.env"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_evo$rotation[,temp5$Evo_PC[i]])
  colnames(temp2)<-"rot.evo"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp5$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.evo, method = "spearman")$estimate
  temp5$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.evo, method = "spearman")$p.value
}

temp5.1<-matrix(rep(0,9),nrow=3)
rownames(temp5.1)<-mypcs
colnames(temp5.1)<-mypcs
temp5.2<-temp5.1
for(i in 1:nrow(temp5.1)){
  for(j in 1:ncol(temp5.1)){
    temp5.1[i,j]<-temp5[temp5$Env_PC==i&temp5$Evo_PC==j,]$Cor
    temp5.2[i,j]<-temp5[temp5$Env_PC==i&temp5$Evo_PC==j,]$p.value
  }
}
temp5.3<-abs(temp5.1)

# Env vs. Mut
temp6<-data.frame(Env_PC=rep(seq(1,3),each=3),Mut_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp6)){
  temp1<-as.data.frame(res.pca.global$rotation[,temp6$Env_PC[i]])
  colnames(temp1)<-"rot.env"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_mut$rotation[,temp6$Mut_PC[i]])
  colnames(temp2)<-"rot.mut"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp6$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.mut, method = "spearman")$estimate
  temp6$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.mut, method = "spearman")$p.value
}

temp6.1<-matrix(rep(0,9),nrow=3)
rownames(temp6.1)<-mypcs
colnames(temp6.1)<-mypcs
temp6.2<-temp6.1
for(i in 1:nrow(temp6.1)){
  for(j in 1:ncol(temp6.1)){
    temp6.1[i,j]<-temp6[temp6$Env_PC==i&temp6$Mut_PC==j,]$Cor
    temp6.2[i,j]<-temp6[temp6$Env_PC==i&temp6$Mut_PC==j,]$p.value
  }
}
temp6.3<-abs(temp6.1)

# Env vs. Mut
temp7<-data.frame(Mut_PC=rep(seq(1,3),each=3),Evo_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp7)){
  temp1<-as.data.frame(res.pca.global_mut$rotation[,temp7$Mut_PC[i]])
  colnames(temp1)<-"rot.mut"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_evo$rotation[,temp7$Evo_PC[i]])
  colnames(temp2)<-"rot.evo"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp7$Cor[i]<-cor.test(temp4$rot.mut,temp4$rot.evo, method = "spearman")$estimate
  temp7$p.value[i]<-cor.test(temp4$rot.mut,temp4$rot.evo, method = "spearman")$p.value
}

temp7.1<-matrix(rep(0,9),nrow=3)
rownames(temp7.1)<-mypcs
colnames(temp7.1)<-mypcs
temp7.2<-temp7.1
for(i in 1:nrow(temp7.1)){
  for(j in 1:ncol(temp7.1)){
    temp7.1[i,j]<-temp7[temp7$Mut_PC==i&temp7$Evo_PC==j,]$Cor
    temp7.2[i,j]<-temp7[temp7$Mut_PC==i&temp7$Evo_PC==j,]$p.value
  }
}
temp7.3<-abs(temp7.1)

my_lab_fontsize<-6

p4<-ggcorrplot::ggcorrplot(t(temp5.3),
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Evo", y = "Env")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p4<-ggpar(p4,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p5<-ggcorrplot::ggcorrplot(t(temp6.3),
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Env")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p5<-ggpar(p5,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p6<-ggcorrplot::ggcorrplot(temp7.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Evo")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p6<-ggpar(p6,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

# legend
my_fontsize<-16
my_tmp<-data.frame(x=seq(min(my_lim),max(my_lim),by=0.01),y=seq(min(my_lim),max(my_lim),by=0.01),z=seq(min(my_lim),max(my_lim),by=0.01))
p_led<-ggscatter(my_tmp,x="x",y="y",fill="z",legend.title = "|R|",legend="right")+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
q2<-ggpar(p_led,font.legend = my_fontsize)
leg2 <- ggpubr::get_legend(q2)

grid.arrange(p4,p5,p6,as_ggplot(leg2),widths=c(1,1,1,0.4),nrow=1)

Sys.time()
```
### 9.4.2. Correlation between rotation (eigenvector) of all PCs in Env and that in Evo or Mut
```{r PCA_all,warning=FALSE, message=FALSE,fig.height=54,fig.width=18,dpi=300}
# # Env vs. Evo
# temp5<-data.frame(Env_PC=rep(seq(1,ncol(res.pca.global_env$rotation)),each=ncol(res.pca.global_evo$rotation)),
#                   Evo_PC=rep(seq(1,ncol(res.pca.global_evo$rotation)),ncol(res.pca.global_env$rotation)),Cor=NA,p.value=NA)
# for(i in 1:nrow(temp5)){
#   temp1<-as.data.frame(res.pca.global_env$rotation[,temp5$Env_PC[i]])
#   colnames(temp1)<-"rot.env"
#   temp1$Bnb<-rownames(temp1)
#   temp2<-as.data.frame(res.pca.global_evo$rotation[,temp5$Evo_PC[i]])
#   colnames(temp2)<-"rot.evo"
#   temp2$Bnb<-rownames(temp2)
#   temp4<-left_join(temp1,temp2)
#   temp5$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.evo, method = "spearman")$estimate
#   temp5$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.evo, method = "spearman")$p.value
# }
# 
# temp5.1<-matrix(rep(0,nrow(temp5)),nrow=ncol(res.pca.global_env$rotation))
# rownames(temp5.1)<-paste("PC",seq(1:ncol(res.pca.global_env$rotation)),sep="")
# colnames(temp5.1)<-paste("PC",seq(1:ncol(res.pca.global_evo$rotation)),sep="")
# temp5.2<-temp5.1
# for(i in 1:nrow(temp5.1)){
#   for(j in 1:ncol(temp5.1)){
#     temp5.1[i,j]<-temp5[temp5$Env_PC==i&temp5$Evo_PC==j,]$Cor
#     temp5.2[i,j]<-temp5[temp5$Env_PC==i&temp5$Evo_PC==j,]$p.value
#   }
# }
# temp5.3<-abs(temp5.1)
# 
# 
# # Env vs. Mut
# temp6<-data.frame(Env_PC=rep(seq(1,ncol(res.pca.global_env$rotation)),each=ncol(res.pca.global_mut$rotation)),
#                   Mut_PC=rep(seq(1,ncol(res.pca.global_mut$rotation)),ncol(res.pca.global_env$rotation)),Cor=NA,p.value=NA)
# for(i in 1:nrow(temp6)){
#   temp1<-as.data.frame(res.pca.global_env$rotation[,temp6$Env_PC[i]])
#   colnames(temp1)<-"rot.env"
#   temp1$Bnb<-rownames(temp1)
#   temp2<-as.data.frame(res.pca.global_mut$rotation[,temp6$Mut_PC[i]])
#   colnames(temp2)<-"rot.mut"
#   temp2$Bnb<-rownames(temp2)
#   temp4<-left_join(temp1,temp2)
#   temp6$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.mut, method = "spearman")$estimate
#   temp6$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.mut, method = "spearman")$p.value
# }
# 
# temp6.1<-matrix(rep(0,nrow(temp6)),nrow=ncol(res.pca.global_env$rotation))
# rownames(temp6.1)<-paste("PC",seq(1:ncol(res.pca.global_env$rotation)),sep="")
# colnames(temp6.1)<-paste("PC",seq(1:ncol(res.pca.global_mut$rotation)),sep="")
# temp6.2<-temp6.1
# for(i in 1:nrow(temp6.1)){
#   for(j in 1:ncol(temp6.1)){
#     temp6.1[i,j]<-temp6[temp6$Env_PC==i&temp6$Mut_PC==j,]$Cor
#     temp6.2[i,j]<-temp6[temp6$Env_PC==i&temp6$Mut_PC==j,]$p.value
#   }
# }
# temp6.3<-abs(temp6.1)
# 
# 
# 
# # Env vs. Mut
# temp7<-data.frame(Mut_PC=rep(seq(1,ncol(res.pca.global_mut$rotation)),each=ncol(res.pca.global_evo$rotation)),
#                   Evo_PC=rep(seq(1,ncol(res.pca.global_evo$rotation)),ncol(res.pca.global_mut$rotation)),Cor=NA,p.value=NA)
# for(i in 1:nrow(temp7)){
#   temp1<-as.data.frame(res.pca.global_mut$rotation[,temp7$Mut_PC[i]])
#   colnames(temp1)<-"rot.mut"
#   temp1$Bnb<-rownames(temp1)
#   temp2<-as.data.frame(res.pca.global_evo$rotation[,temp7$Evo_PC[i]])
#   colnames(temp2)<-"rot.evo"
#   temp2$Bnb<-rownames(temp2)
#   temp4<-left_join(temp1,temp2)
#   temp7$Cor[i]<-cor.test(temp4$rot.mut,temp4$rot.evo, method = "spearman")$estimate
#   temp7$p.value[i]<-cor.test(temp4$rot.mut,temp4$rot.evo, method = "spearman")$p.value
# }
# 
# temp7.1<-matrix(rep(0,nrow(temp7)),nrow=ncol(res.pca.global_mut$rotation))
# rownames(temp7.1)<-paste("PC",seq(1:ncol(res.pca.global_mut$rotation)),sep="")
# colnames(temp7.1)<-paste("PC",seq(1:ncol(res.pca.global_evo$rotation)),sep="")
# temp7.2<-temp7.1
# for(i in 1:nrow(temp7.1)){
#   for(j in 1:ncol(temp7.1)){
#     temp7.1[i,j]<-temp7[temp7$Mut_PC==i&temp7$Evo_PC==j,]$Cor
#     temp7.2[i,j]<-temp7[temp7$Mut_PC==i&temp7$Evo_PC==j,]$p.value
#   }
# }
# temp7.3<-abs(temp7.1)
# 
# my_lab_fontsize<-6
# 
# p4<-ggcorrplot::ggcorrplot(t(temp5.3),
#                            legend.title = "Spearman's R",
#   lab = FALSE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
#   scale_fill_gradient(limit = c(0,0.5), low = "white", high =  "#E46726")+
#   labs(x = "Evo", y = "Env")+
#   theme(
#     axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
#     axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
#   )
# p4<-ggpar(p4,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())
# 
# p5<-ggcorrplot::ggcorrplot(t(temp6.3),
#                            legend.title = "Spearman's R",
#   lab = FALSE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
#   scale_fill_gradient(limit = c(0,0.5), low = "white", high =  "#E46726")+
#   labs(x = "Mut", y = "Env")+
#   theme(
#     axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
#     axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
#   )
# p5<-ggpar(p5,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())
# 
# p6<-ggcorrplot::ggcorrplot(temp7.3,
#                            legend.title = "Spearman's R",
#   lab = FALSE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
#   scale_fill_gradient(limit = c(0,0.5), low = "white", high =  "#E46726")+
#   labs(x = "Mut", y = "Evo")+
#   theme(
#     axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
#     axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
#   )
# p6<-ggpar(p6,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())
# 
# my_fontsize<-16
# q2<-ggpar(p6,legend = "right",legend.title="",font.legend = my_fontsize)+
#   scale_fill_continuous(limits = c(0,0.5), breaks = seq(0,0.5,0.1),low = "white", high =  "#E46726",
#                         guide=guide_colourbar(title= "Abs. of Spearman's R",title.position = "left",ticks.linewidth = 1))+
#   theme(legend.title = element_text(angle = 90, hjust = 10, vjust = 0.5,size=18),legend.title.align=10)
# leg2 <- ggpubr::get_legend(q2)
# # as_ggplot(leg2)
# 
# grid.arrange(p4,p5,p6,as_ggplot(leg2),widths=c(1.108,1,1,0.4),nrow=1)
# 
# Sys.time()
```
## 9.5. Alignment of major axes of the data set
https://academic.oup.com/evolut/article/69/11/2927/6852156
https://www.pnas.org/doi/10.1073/pnas.1821066116#supplementary-materials
```{r Alignment,warning=FALSE, message=FALSE,fig.height=4.5,fig.width=11,dpi=300}
vec_env<-res.pca.global_env$rotation[,1:3] # top three eigenvectors, PC1~PC3, of variance-covariance matrix in Env 
vec_evo<-res.pca.global_evo$rotation[,1:3] # top three eigenvectors, PC1~PC3, of variance-covariance matrix in Evo
vec_mut<-res.pca.global_mut$rotation[,1:3] # top three eigenvectors, PC1~PC3, of variance-covariance matrix in Mut

calc_theta<-function(x,y){
  x<-c(x)
  y<-c(y)
  theta_e<-((180/pi)*acos(x %*% y))
  theta_e<-ifelse(theta_e>90,180-theta_e,theta_e)
  return(c(theta_e))　# Difference in angle between x and y (0~90 degree)
}

print(paste("Env_PC1 vs. Evo_PC1: ",calc_theta(vec_env[,1],vec_evo[,1]),"degree"))
print(paste("Env_PC1 vs. Mut_PC1: ",calc_theta(vec_env[,1],vec_mut[,1]),"degree"))
print(paste("Evo_PC1 vs. Mut_PC1: ",calc_theta(vec_evo[,1],vec_mut[,1]),"degree"))
print(paste("Env_PC1 vs. Evo_PC2: ",calc_theta(vec_env[,1],vec_evo[,2]),"degree"))
print(paste("Env_PC1 vs. Mut_PC2: ",calc_theta(vec_env[,1],vec_mut[,2]),"degree"))

Sys.time()
```
## 9.6. GSEA using loadings of PC1~3 (investigate enrichment of only TRcoms)
Two regulatory modes (activation or inhibition) were distinguished.
### 9.6.1. perform GSEA
```{r GSEA_TRcom,warning=FALSE, message=FALSE,fig.height=6,fig.width=6,dpi=300}
TRN_TRcom<-TRN
TRN_TRcom<-TRN_TRcom[(TRN_TRcom$regulator %in% TRcom)&(TRN_TRcom$effect %in% c("+","-")),]
TRN_TRcom$reg_eff<-NA
TRN_TRcom[TRN_TRcom$effect=="+",]$reg_eff<-paste(TRN_TRcom[TRN_TRcom$effect=="+",]$regulator,"act",sep="_")
TRN_TRcom[TRN_TRcom$effect=="-",]$reg_eff<-paste(TRN_TRcom[TRN_TRcom$effect=="-",]$regulator,"inh",sep="_")
TRN_TRcom.sub<-TRN_TRcom[!is.na(TRN_TRcom$prsn_in_gBnb),]
TRN_TRcom.sub<-left_join(TRN_TRcom.sub,dBnb.comp[,which(colnames(dBnb.comp) %in% qw(Bnb,EntrezID))],by=c("gene_id"="Bnb"))
dtrn_TRcom<-TRN_TRcom.sub[,which(colnames(TRN_TRcom.sub) %in% qw(reg_eff,EntrezID))]


# For Env
# temp<-t(res.pca.global$sdev*t(res.pca.global$rotation)) : "loadings: correlation between PCx and original variables" identical to pca.glob.var$coord
# Principal component loadings: correlation between PCx and original variables
# see http://www.eeso.ges.kyoto-u.ac.jp/emm/materials/rmemo/pcrcom
my_datasets<-qw(Env,Evo,Mut)
my_pca.coord<-list()
my_pca.coord[[1]]<-get_pca_var(res.pca.global_env)
my_pca.coord[[2]]<-get_pca_var(res.pca.global_evo)
my_pca.coord[[3]]<-get_pca_var(res.pca.global_mut)
names(my_pca.coord)<-my_datasets

for (j in 1:length(my_datasets)){
  pca.var<-my_pca.coord[[j]]
  figtemp1<-as.data.frame(pca.var$coord[,1:ncol(pca.var$coord)])
  colnames(figtemp1)<-str_replace(colnames(figtemp1),"Dim.","PC")
  mydims<-colnames(figtemp1)
  figtemp1$Bnb<-rownames(figtemp1)
  figtemp1<-left_join(figtemp1,dBnb.comp[,colnames(dBnb) %in% qw(Bnb,EntrezID)])
  for(i in mypcs){
    test<-figtemp1[order(figtemp1[,which(colnames(figtemp1)==i)],decreasing = T),]
    original_gene_list <- test[,which(colnames(test)==i)]
    # name the vector
    names(original_gene_list) <- test$EntrezID
    # omit any NA values 
    gene_list<-na.omit(original_gene_list)
    # sort the list in decreasing order (required for clusterProfiler)
    gene_list = sort(gene_list, decreasing = TRUE)
    set.seed(123)
    res_tfgsea <- GSEA(geneList     = gene_list,
                  TERM2GENE = dtrn_TRcom,
                  seed=TRUE,
                   nPerm        = 10000,
                   minGSSize    = 5, # 5
                   maxGSSize    = 1000, # 100
                   pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
                   pAdjustMethod = "BH") #fdr
    set.seed(NULL)
    gene_count<-res_tfgsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
    dot_df<- left_join(res_tfgsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
    # dot_df<-dot_df[dot_df$count>4,]
    dot_df$type="Sensitive"
    dot_df$type[dot_df$NES < 0]="Robust"
    dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
    dot_df$PC<-i
    if(i=="PC1"){
      figtemp2<-dot_df
    }else{
      figtemp2<-bind_rows(figtemp2,dot_df)
    }
  }
  figtemp2$qval.fin<-qvalue(figtemp2$pvalue)$qvalues
  figtemp2$log_qval.fin<-log10(figtemp2$qval.fin)
  unique(figtemp2[figtemp2$qval.fin<0.05,]$ID)
  figtemp2<-figtemp2[figtemp2$qval.fin<0.05,]
  # figtemp2<-figtemp2[figtemp2$qvalue<0.05,] # <-------------------    Use standard qvalue
  figtemp2$regulator<-NA
  figtemp2$effect<-"-"
  for(ii in 1:nrow(figtemp2)){
    temp<-str_split(figtemp2$ID[ii],"_")[[1]]
    figtemp2$regulator[ii]<-temp[1]
    if(temp[2]=="act"){
      figtemp2$effect[ii]<-"+"
    }
  }
  figtemp2$col<-mycolpal[1]
  figtemp2[figtemp2$effect=="+",]$col<-mycolpal[2]
  figtemp2$dataset<-my_datasets[j]
  if(j==1){
    dTRcom_pca_gsea<-figtemp2
  }else{
    dTRcom_pca_gsea<-bind_rows(dTRcom_pca_gsea,figtemp2)
  }
}
Sys.time()
```
### 9.6.2. visualization
```{r GSEA_TRcom,warning=FALSE, message=FALSE,fig.height=6,fig.width=6,dpi=300}
my_ds<-qw(Env,Evo,Mut)
my_fontsize<-16
my_pc<-qw(PC1,PC2,PC3)

for (j in 1:length(my_ds)){
  figtemp2<-dTRcom_pca_gsea[dTRcom_pca_gsea$dataset==my_ds[j],]
  figlist4<-list()
  for(ii in 1:length(my_pc)){
    k<-my_pc[ii]
    temp<-figtemp2[figtemp2$PC==k,]
    temp<-temp[order(temp$enrichmentScore,decreasing = T),]
    p<-ggdotchart(figtemp2[figtemp2$PC==k,],x="Description",y="NES",
               color="log_qval.fin", #qval.fin
               sorting = "descending", 
               add = "segments", 
               add.params = list(color = "lightgray", size = 2),
               dot.size = "GeneRatio",
               legend="",legend.title="",
               xlab="",ylab="Normalized enrichment score")+
      scale_x_discrete(labels=temp$regulator)+
      scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
      scale_size_continuous(limits = c(0,1),breaks=c(0.2,0.4,0.6,0.8))+
      annotate("text",x=Inf,y=Inf,label=paste(my_pc[ii],"(",my_ds[j],")"),hjust = 1,vjust = 1,size=5)+
       # scale_size_area(max_size = 4,breaks=c(0.3,0.4,0.5,0.6,0.7,0.8)) +
      guides(size = guide_legend(title= "Gene ratio",label.theme = element_text(angle = 0,size=my_fontsize),title.position = "left"))+
      scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "left",label.vjust = 0,label.hjust = 0,label.theme = element_text(angle = 0,size=my_fontsize),ticks.linewidth = 1,order=1))
    p<-ggpar(p,font.xtickslab = list(size=my_fontsize,color=temp$col),font.ytickslab = my_fontsize)+
      theme(axis.title.x=element_blank(),axis.title.y = element_blank(),
        legend.title = element_text(size = my_fontsize, angle = 90),
        legend.title.align = 0.5,
        legend.direction = "vertical")
    figlist4[[ii]]<-p
    q<-ggpar(p,legend = "right",legend.title = list("Gene ratio","q-value (log)"),font.ytickslab= c(10))
    leg <- ggpubr::get_legend(q)
    as_ggplot(leg)
  }
  # lay<-rbind(c(1,1,1,1,4),c(2,2,5,5,4),c(3,3,3,3,4))
  lay<-rbind(c(1,1,1,1,4),c(2,2,2,2,4),c(3,3,3,3,4))
  grid.arrange(figlist4[[1]],figlist4[[2]],figlist4[[3]],as_ggplot(leg), #blank,
               widths=c(rep(1,4),1),
               left=textGrob("Normalized enrichment score", rot=90,gp=gpar(fontsize=20)),
               layout_matrix=lay)
}
Sys.time()
```
### 9.6.3. Venn diagram of enriched regulation in PC1~PC3
PC1~PC3
```{r GSEA_TRcom,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
venn_all<-list(
  Evo=dTRcom_pca_gsea[dTRcom_pca_gsea$dataset=="Evo",]$ID,
  Env=dTRcom_pca_gsea[dTRcom_pca_gsea$dataset=="Env",]$ID,
  Mut=dTRcom_pca_gsea[dTRcom_pca_gsea$dataset=="Mut",]$ID
)
# ggVennDiagram(figtemp)
venn <- process_data(Venn(venn_all))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))

v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)
  # text(x=v$centers[1,1]-0.2, y=v$centers[1,2]-0.24, labels='Evo')+ # Evo label (reg 1)
  # text(x=v$centers[1,1]+0.3, y=v$centers[1,2]+0.3, labels='Env')+ # Env label (reg 2)
  # text(x=v$centers[1,1]-0.38, y=v$centers[1,2]+0.3, labels='Mut') # Mut label (reg 3)

for(i in 1:length(venn$regionData$item)){
  print(paste(venn$regionData$name[[i]],paste(venn$regionData$item[[i]],collapse=", "),sep=": "))
}

Sys.time()
```
### 9.6.4. Venn diagram of enriched regulation in PC1
PC1 only
```{r GSEA_TRcom,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
venn_all<-list(
  Evo=dTRcom_pca_gsea[(dTRcom_pca_gsea$dataset=="Evo")&(dTRcom_pca_gsea$PC=="PC1"),]$ID,
  Env=dTRcom_pca_gsea[(dTRcom_pca_gsea$dataset=="Env")&(dTRcom_pca_gsea$PC=="PC1"),]$ID,
  Mut=dTRcom_pca_gsea[(dTRcom_pca_gsea$dataset=="Mut")&(dTRcom_pca_gsea$PC=="PC1"),]$ID
)
# ggVennDiagram(figtemp)
venn <- process_data(Venn(venn_all))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))
v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)
  # text(x=v$centers[1,1]-0.2, y=v$centers[1,2]-0.24, labels='Evo')+ # Evo label (reg 1)
  # text(x=v$centers[1,1]+0.3, y=v$centers[1,2]+0.3, labels='Env')+ # Env label (reg 2)
  # text(x=v$centers[1,1]-0.38, y=v$centers[1,2]+0.3, labels='Mut') # Mut label (reg 3)

for(i in 1:length(venn$regionData$item)){
  print(paste(venn$regionData$name[[i]],paste(venn$regionData$item[[i]],collapse=", "),sep=": "))
}

Sys.time()
```
## 9.7. GSEA using loadings of PC1~3 (investigate enrichment of all TRs)
Two regulatory modes (activation or inhibition) were distinguished.
### 9.7.1. perform GSEA
```{r GSEA_allTR,warning=FALSE, message=FALSE,fig.height=7,fig.width=14,dpi=300}
TRN_TRall<-TRN
TRN_TRall<-TRN_TRall[TRN_TRall$effect %in% c("+","-"),]
TRN_TRall$reg_eff<-NA
TRN_TRall[TRN_TRall$effect=="+",]$reg_eff<-paste(TRN_TRall[TRN_TRall$effect=="+",]$regulator,"act",sep="_")
TRN_TRall[TRN_TRall$effect=="-",]$reg_eff<-paste(TRN_TRall[TRN_TRall$effect=="-",]$regulator,"inh",sep="_")
TRN_TRall.sub<-TRN_TRall[!is.na(TRN_TRall$prsn_in_gBnb),]
TRN_TRall.sub<-left_join(TRN_TRall.sub,dBnb.comp[,which(colnames(dBnb.comp) %in% qw(Bnb,EntrezID))],by=c("gene_id"="Bnb"))
dTRN_TRall<-TRN_TRall.sub[,which(colnames(TRN_TRall.sub) %in% qw(reg_eff,EntrezID))]


# For Env
# temp<-t(res.pca.global$sdev*t(res.pca.global$rotation)) : "loadings: correlation between PCx and original variables" identical to pca.glob.var$coord
# Principal component loadings: correlation between PCx and original variables
# see http://www.eeso.ges.kyoto-u.ac.jp/emm/materials/rmemo/pcrcom
my_datasets<-qw(Env,Evo,Mut)
my_pca.coord<-list()
my_pca.coord[[1]]<-get_pca_var(res.pca.global_env)
my_pca.coord[[2]]<-get_pca_var(res.pca.global_evo)
my_pca.coord[[3]]<-get_pca_var(res.pca.global_mut)
names(my_pca.coord)<-my_datasets

for (j in 1:length(my_datasets)){
  pca.var<-my_pca.coord[[j]]
  figtemp1<-as.data.frame(pca.var$coord[,1:ncol(pca.var$coord)])
  colnames(figtemp1)<-str_replace(colnames(figtemp1),"Dim.","PC")
  mydims<-colnames(figtemp1)
  figtemp1$Bnb<-rownames(figtemp1)
  figtemp1<-left_join(figtemp1,dBnb.comp[,colnames(dBnb) %in% qw(Bnb,EntrezID)])
  for(i in mypcs){
    test<-figtemp1[order(figtemp1[,which(colnames(figtemp1)==i)],decreasing = T),]
    original_gene_list <- test[,which(colnames(test)==i)]
    # name the vector
    names(original_gene_list) <- test$EntrezID
    # omit any NA values 
    gene_list<-na.omit(original_gene_list)
    # sort the list in decreasing order (required for clusterProfiler)
    gene_list = sort(gene_list, decreasing = TRUE)
    set.seed(123)
    res_tfgsea <- GSEA(geneList     = gene_list,
                  TERM2GENE = dTRN_TRall,
                  seed=TRUE,
                   nPerm        = 10000,
                   minGSSize    = 5, # 5
                   maxGSSize    = 1000, # 100
                   pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
                   pAdjustMethod = "BH") #fdr
    set.seed(NULL)
    gene_count<-res_tfgsea@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
    dot_df<- left_join(res_tfgsea@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
    # dot_df<-dot_df[dot_df$count>4,]
    dot_df$type="Sensitive"
    dot_df$type[dot_df$NES < 0]="Robust"
    dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
    dot_df$PC<-i
    if(i=="PC1"){
      figtemp2<-dot_df
    }else{
      figtemp2<-bind_rows(figtemp2,dot_df)
    }
  }
  figtemp2$qval.fin<-qvalue(figtemp2$pvalue)$qvalues
  figtemp2$log_qval.fin<-log10(figtemp2$qval.fin)
  unique(figtemp2[figtemp2$qval.fin<0.05,]$ID)
  figtemp2<-figtemp2[figtemp2$qval.fin<0.05,]
  # figtemp2<-figtemp2[figtemp2$qvalue<0.05,] # <-------------------    Use standard qvalue
  figtemp2$regulator<-NA
  figtemp2$effect<-"-"
  for(ii in 1:nrow(figtemp2)){
    temp<-str_split(figtemp2$ID[ii],"_")[[1]]
    figtemp2$regulator[ii]<-temp[1]
    if(temp[2]=="act"){
      figtemp2$effect[ii]<-"+"
    }
  }
  figtemp2$col<-mycolpal[1]
  figtemp2[figtemp2$effect=="+",]$col<-mycolpal[2]
  figtemp2$dataset<-my_datasets[j]
  figtemp2$label<-figtemp2$regulator
  figtemp2[figtemp2$regulator %in% TRcom,]$label<-paste(figtemp2[figtemp2$regulator %in% TRcom,]$regulator,"*",sep="")
  if(j==1){
    dTRall_pca_gsea<-figtemp2
  }else{
    dTRall_pca_gsea<-bind_rows(dTRall_pca_gsea,figtemp2)
  }
}
Sys.time()
```
### 9.7.2. visualization
```{r GSEA_allTR,warning=FALSE, message=FALSE,fig.height=7,fig.width=14,dpi=300}
my_ds<-qw(Env,Evo,Mut)
my_fontsize<-20
my_pc<-qw(PC1,PC2,PC3)

for (j in 1:length(my_ds)){
  figtemp2<-dTRall_pca_gsea[dTRall_pca_gsea$dataset==my_ds[j],]
  figlist4<-list()
  for(ii in 1:length(my_pc)){
    k<-my_pc[ii]
    temp<-figtemp2[figtemp2$PC==k,]
    temp<-temp[order(temp$enrichmentScore,decreasing = T),]
    p<-ggdotchart(figtemp2[figtemp2$PC==k,],x="Description",y="NES",
               color="log_qval.fin", #qval.fin
               sorting = "descending", 
               add = "segments", 
               add.params = list(color = "lightgray", size = 2),
               dot.size = "GeneRatio",
               legend="",legend.title="",
               xlab="",ylab="Normalized enrichment score")+
      scale_x_discrete(labels=temp$regulator)+
      scale_x_discrete(labels=temp$label)+
      scale_y_continuous(breaks = get_breaks(by = 2, from = -2),expand = expansion(mult = c(0.1, 0.1)))+
      scale_size_continuous(limits = c(0,1),breaks=c(0.2,0.4,0.6,0.8))+
      annotate("text",x=Inf,y=Inf,label=paste(my_pc[ii],"(",my_ds[j],")",sep=""),hjust = 1,vjust = 1,size=8)+
       # scale_size_area(max_size = 4,breaks=c(0.3,0.4,0.5,0.6,0.7,0.8)) +
      guides(size = guide_legend(title= "Gene ratio",label.theme = element_text(angle = 0,size=my_fontsize),title.position = "left"))+
      scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                        guide=guide_colourbar(title="q-value (log)",title.position = "left",label.vjust = 0,label.hjust = 0,label.theme = element_text(angle = 0,size=my_fontsize),ticks.linewidth = 1,order=1))
    p<-ggpar(p,font.xtickslab = list(size=my_fontsize,color=temp$col),font.ytickslab = my_fontsize)+
      theme(axis.title.x=element_blank(),axis.title.y = element_blank(),
        legend.title = element_text(size = 24, angle = 90),
        legend.title.align = 0.5,
        legend.direction = "vertical")
    figlist4[[ii]]<-p
    q<-ggpar(p,legend = "right",legend.title = list("Gene ratio","q-value (log)"),font.ytickslab= c(10))
    leg <- ggpubr::get_legend(q)
    as_ggplot(leg)
  }
  # lay<-rbind(c(1,1,1,1,4),c(2,2,5,5,4),c(3,3,3,3,4))
  lay<-rbind(c(1,1,1,1,4),c(2,2,2,2,4),c(3,3,3,3,4))
  grid.arrange(figlist4[[1]],figlist4[[2]],figlist4[[3]],as_ggplot(leg), #blank,
               widths=c(rep(1,4),0.5),
               left=textGrob("Normalized enrichment score", rot=90,gp=gpar(fontsize=24)),
               layout_matrix=lay)
}

for(i in my_ds){
  for(j in my_pc){
    k<-length(unique(dTRall_pca_gsea[(dTRall_pca_gsea$dataset==i)&(dTRall_pca_gsea$PC==j)&(dTRall_pca_gsea$NES>0),]$Description))
    l<-length(unique(dTRall_pca_gsea[(dTRall_pca_gsea$dataset==i)&(dTRall_pca_gsea$PC==j)&(dTRall_pca_gsea$NES<0),]$Description))
    print(paste("# of TRs enriched in pos. of",j,"in",i,":",k))
    print(paste("# of TRs enriched in neg. of",j,"in",i,":",l))
  }
}
Sys.time()
```
### 9.7.3. Venn diagram of enriched regulation in PC1~PC3
PC1~PC3
```{r GSEA_allTR,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
venn_all<-list(
  Evo=dTRall_pca_gsea[dTRall_pca_gsea$dataset=="Evo",]$ID,
  Env=dTRall_pca_gsea[dTRall_pca_gsea$dataset=="Env",]$ID,
  Mut=dTRall_pca_gsea[dTRall_pca_gsea$dataset=="Mut",]$ID
)
# ggVennDiagram(figtemp)
venn <- process_data(Venn(venn_all))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))

v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)
  # text(x=v$centers[1,1]-0.2, y=v$centers[1,2]-0.24, labels='Evo')+ # Evo label (reg 1)
  # text(x=v$centers[1,1]+0.3, y=v$centers[1,2]+0.3, labels='Env')+ # Env label (reg 2)
  # text(x=v$centers[1,1]-0.38, y=v$centers[1,2]+0.3, labels='Mut') # Mut label (reg 3)

for(i in 1:length(venn$regionData$item)){
  print(paste(venn$regionData$name[[i]],paste(venn$regionData$item[[i]],collapse=", "),sep=": "))
}

Sys.time()
```
### 9.7.4. Venn diagram of enriched regulation in PC1
PC1 only
```{r GSEA_allTR,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
venn_all<-list(
  Evo=dTRall_pca_gsea[(dTRall_pca_gsea$dataset=="Evo")&(dTRall_pca_gsea$PC=="PC1"),]$ID,
  Env=dTRall_pca_gsea[(dTRall_pca_gsea$dataset=="Env")&(dTRall_pca_gsea$PC=="PC1"),]$ID,
  Mut=dTRall_pca_gsea[(dTRall_pca_gsea$dataset=="Mut")&(dTRall_pca_gsea$PC=="PC1"),]$ID
)
# ggVennDiagram(figtemp)
venn <- process_data(Venn(venn_all))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))
v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)
  # text(x=v$centers[1,1]-0.2, y=v$centers[1,2]-0.24, labels='Evo')+ # Evo label (reg 1)
  # text(x=v$centers[1,1]+0.3, y=v$centers[1,2]+0.3, labels='Env')+ # Env label (reg 2)
  # text(x=v$centers[1,1]-0.38, y=v$centers[1,2]+0.3, labels='Mut') # Mut label (reg 3)

for(i in 1:length(venn$regionData$item)){
  print(paste(venn$regionData$name[[i]],paste(venn$regionData$item[[i]],collapse=", "),sep=": "))
}

Sys.time()
```
## 9.8. PCA for the Env_wo_abx dataset
### 9.8.1. Detect outliers in PC1~PC3
```{r PCA_Env_wo_abx,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-log_tpm_pl.pr2_wo_abx[log_tpm_pl.pr2_wo_abx$Bnb %in% dBnb.comp$Bnb,]
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
figtemp<-as.data.frame(t(temp0[,which(colnames(temp0) %nin% "Bnb")]))# transpose the matrix so genes are as columns
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
env_wo_abx_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    env_wo_abx_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(env_wo_abx_outlier[!is.na(env_wo_abx_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
### 9.8.2. Remove outliers if it is necessary and examine PCA
```{r PCA_Env_wo_abx,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
mypcs<-paste("PC",seq(1:3),sep="")
temp<-log_tpm_pl.pr2_wo_abx[log_tpm_pl.pr2_wo_abx$Bnb %in% dBnb.comp$Bnb,]
if(!is.na(env_wo_abx_outlier)){
  temp<-temp[,colnames(temp) %nin% env_wo_abx_outlier]
}
temp<-temp[order(temp$Bnb,decreasing = F),]
figtemp<-as.data.frame(t(temp[,which(colnames(temp) %nin% "Bnb")]))
colnames(figtemp)<-temp$Bnb
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)
res.pca.global_env_wo_abx<-my_res
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_env_wo_abx <- my_res$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_env_wo_abx$label<-rownames(pc_scores_env_wo_abx)

pc_eigenvalues_env_wo_abx <- res.pca.global_env_wo_abx$sdev^2
pc_eigenvalues_env_wo_abx <- tibble(PC = factor(1:length(pc_eigenvalues_env_wo_abx)), 
                         variance = pc_eigenvalues_env_wo_abx) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained
# pc_eigenvalues_env %>% 
#   ggplot(aes(x = PC)) +
#   geom_col(aes(y = pct)) +
#   geom_line(aes(y = pct_cum, group = 1)) + 
#   geom_point(aes(y = pct_cum)) +
#   labs(x = "Principal component", y = "Fraction variance explained")

# Scree plot
figtemp1<-pc_eigenvalues_env_wo_abx
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree.env<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,color="#619CFF",plot_type = "l",
       xlab="Components",ylab="Fraction variance explained (%)")+
  scale_x_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,80))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# PC1 vs. PC2
ind.p<-ggscatter(data=pc_scores_env_wo_abx,x="PC1",y="PC2",alpha=0.2,
                 # color = "Condition",label="Condition",repel=T,font.label=4,
                 legend="")+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
ind.p<-ggpar(ind.p,
      xlab = paste("PC1 (", round(pc_eigenvalues_env_wo_abx$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_env_wo_abx$pct[2],digits = 0),"%)",sep=""))
ind.p
# q<-ggpar(ind.p,legend = "right",legend.title="",font.legend = c(4))
# leg <- ggpubr::get_legend(q)
# as_ggplot(leg)


# reconstruct original dataset using only PCx
# https://sinhrks.hatenablog.com/entry/2014/10/19/000929
# k<-3
# reconst_pck<-res.pca.global_env$x[,1:k] %*% t(res.pca.global_env$rotation[,1:k])
# reconst_pck <- scale(reconst_pck, center =F, scale=1/res.pca.global_env$scale)
# reconst_pck<- scale(reconst_pck, center =-res.pca.global_env$center, scale=F)

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_env$pct_cum[k] ,"% variance",sep=""))

pca.glob.var_env_wo_abx<-get_pca_var(res.pca.global_env_wo_abx)

Sys.time()
```
### 9.8.3. Correlation between rotations in PC1~3 of Env_wo_abx and those of Evo or Mut
```{r PCA_Env_wo_abx,warning=FALSE, message=FALSE,fig.height=4.5,fig.width=11,dpi=300}
# Env_wo_abx vs. Evo
temp5<-data.frame(Env_wo_abx_PC=rep(seq(1,3),each=3),Evo_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp5)){
  temp1<-as.data.frame(res.pca.global_env_wo_abx$rotation[,temp5$Env_wo_abx_PC[i]])
  colnames(temp1)<-"rot.env_wo_abx"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_evo$rotation[,temp5$Evo_PC[i]])
  colnames(temp2)<-"rot.evo"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp5$Cor[i]<-cor.test(temp4$rot.env_wo_abx,temp4$rot.evo, method = "spearman")$estimate
  temp5$p.value[i]<-cor.test(temp4$rot.env_wo_abx,temp4$rot.evo, method = "spearman")$p.value
}

temp5.1<-matrix(rep(0,9),nrow=3)
rownames(temp5.1)<-mypcs
colnames(temp5.1)<-mypcs
temp5.2<-temp5.1
for(i in 1:nrow(temp5.1)){
  for(j in 1:ncol(temp5.1)){
    temp5.1[i,j]<-temp5[temp5$Env_wo_abx_PC==i&temp5$Evo_PC==j,]$Cor
    temp5.2[i,j]<-temp5[temp5$Env_wo_abx_PC==i&temp5$Evo_PC==j,]$p.value
  }
}
temp5.3<-abs(temp5.1)

# Env vs. Mut
temp6<-data.frame(Env_wo_abx_PC=rep(seq(1,3),each=3),Mut_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp6)){
  temp1<-as.data.frame(res.pca.global_env_wo_abx$rotation[,temp6$Env_wo_abx_PC[i]])
  colnames(temp1)<-"rot.env_wo_abx"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_mut$rotation[,temp6$Mut_PC[i]])
  colnames(temp2)<-"rot.mut"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp6$Cor[i]<-cor.test(temp4$rot.env_wo_abx,temp4$rot.mut, method = "spearman")$estimate
  temp6$p.value[i]<-cor.test(temp4$rot.env_wo_abx,temp4$rot.mut, method = "spearman")$p.value
}

temp6.1<-matrix(rep(0,9),nrow=3)
rownames(temp6.1)<-mypcs
colnames(temp6.1)<-mypcs
temp6.2<-temp6.1
for(i in 1:nrow(temp6.1)){
  for(j in 1:ncol(temp6.1)){
    temp6.1[i,j]<-temp6[temp6$Env_wo_abx_PC==i&temp6$Mut_PC==j,]$Cor
    temp6.2[i,j]<-temp6[temp6$Env_wo_abx_PC==i&temp6$Mut_PC==j,]$p.value
  }
}
temp6.3<-abs(temp6.1)

# Env vs. Mut
temp7<-data.frame(Mut_PC=rep(seq(1,3),each=3),Evo_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp7)){
  temp1<-as.data.frame(res.pca.global_mut$rotation[,temp7$Mut_PC[i]])
  colnames(temp1)<-"rot.mut"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_evo$rotation[,temp7$Evo_PC[i]])
  colnames(temp2)<-"rot.evo"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp7$Cor[i]<-cor.test(temp4$rot.mut,temp4$rot.evo, method = "spearman")$estimate
  temp7$p.value[i]<-cor.test(temp4$rot.mut,temp4$rot.evo, method = "spearman")$p.value
}

temp7.1<-matrix(rep(0,9),nrow=3)
rownames(temp7.1)<-mypcs
colnames(temp7.1)<-mypcs
temp7.2<-temp7.1
for(i in 1:nrow(temp7.1)){
  for(j in 1:ncol(temp7.1)){
    temp7.1[i,j]<-temp7[temp7$Mut_PC==i&temp7$Evo_PC==j,]$Cor
    temp7.2[i,j]<-temp7[temp7$Mut_PC==i&temp7$Evo_PC==j,]$p.value
  }
}
temp7.3<-abs(temp7.1)

my_lab_fontsize<-6

p4<-ggcorrplot::ggcorrplot(t(temp5.3),
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = c(0,0.5), low = "white", high =  "#E46726")+
  labs(x = "Evo", y = "Env_wo_abx")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p4<-ggpar(p4,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p5<-ggcorrplot::ggcorrplot(t(temp6.3),
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = c(0,0.5), low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Env_wo_abx")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p5<-ggpar(p5,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p6<-ggcorrplot::ggcorrplot(temp7.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = c(0,0.5), low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Evo")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p6<-ggpar(p6,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

my_fontsize<-16
q2<-ggpar(p6,legend = "right",legend.title="",font.legend = my_fontsize)+
  scale_fill_continuous(limits = c(0,0.5), breaks = seq(0,0.5,0.1),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "Abs. of Spearman's R",title.position = "left",ticks.linewidth = 1))+
  theme(legend.title = element_text(angle = 90, hjust = 10, vjust = 0.5,size=18),legend.title.align=10)
leg2 <- ggpubr::get_legend(q2)
# as_ggplot(leg2)

grid.arrange(p4,p5,p6,as_ggplot(leg2),widths=c(1,1,1,0.4),nrow=1)

Sys.time()
```
#
#
#
# 10. Gene Ontology
## 10.1. Create a custom GO slim (a subset of GO, dfgoslim2, dfgoslim23)
method=Wang, threshold=0.7. score= size of GO
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=6,fig.width=8,dpi=300}
gocategory<-qw(BP,CC,MF)
k<-0
for (j in 1:length(gocategory)){
  for(i in 2:15){
    if((j<3)|(i<14)){
      ggo <- groupGO(gene=as.character(df_DMall$EntrezID),OrgDb= org.EcK12.eg.db,ont=gocategory[j],level= i,readable = F)
      temp<-ggo@result
      temp<-temp[(temp$Count>29)&(temp$Count<1001),]
      if(nrow(temp)>0){
        temp<-transform(temp,Category=gocategory[j])
        if(k==0){
          dfggo<-temp
          k<-k+1
        }else{
          dfggo<-rbind(dfggo,temp)
        }
      }
    }
  }
}
dfggo<-dfggo  %>%  dplyr::distinct(ID,.keep_all = T)
getGoSize <- function(terms, orgdb, keytype) {
  go <- suppressMessages(
          AnnotationDbi::select(orgdb,
                                keytype=keytype,
                                columns=c("GO", "ONTOLOGY"),
                                keys=AnnotationDbi::keys(orgdb, keytype=keytype)))
  go <- go[!is.na(go$GO), ]
  go <- go[go$GO %in% terms, ]
  # count
  counts   <- table(go$GO)
  go <- go[go$GO %in% terms, ]
  empty    <- terms[!(terms %in% names(counts))]
  nocounts <- setNames(rep(0, length(empty)), empty)
  c(counts, nocounts)
}

# Reduce redundant GOs based on semantic similarity
for (i in 1:length(gocategory)){
  temp2<-dfggo[dfggo$Category==gocategory[i],]
  # temp2<-temp2  %>%  dplyr::distinct(ID,.keep_all = T)
  myscores<-getGoSize(terms = rownames(temp2), orgdb = org.EcK12.eg.db,keytype = "ENTREZID")
  ecGO <-GOSemSim::godata(OrgDb = org.EcK12.eg.db,ont = gocategory[i])
  simMatrix<-GOSemSim::mgoSim(GO1 = temp2$ID,GO2 = temp2$ID,semData = ecGO,combine = NULL)
  reducedTerms<-rrvgo::reduceSimMatrix(simMatrix = simMatrix,threshold = 0.7,orgdb = org.EcK12.eg.db,scores = myscores)
  if(i==1){
    dfgoslim<-reducedTerms
  }else{
    dfgoslim<-rbind(dfgoslim,reducedTerms)
  }
  #heatmapPlot(simMatrix = simMatrix,reducedTerms = reducedTerms,annotateParent=TRUE,annotationLabel="parentTerm",fontsize=6)
}

dfgoslim<-dplyr::left_join(dfgoslim,dfggo,by=c("go"="ID"))
# Filtering and cerate dfgoslim3
dfgoslim2<-dfgoslim[dfgoslim$termDispensability==0,]
# dfgoslim2<-dfgoslim[dfgoslim$parentSimScore==1,] # old package

dfgoslim2<-transform(dfgoslim2,Bnb=NA)
for(i in 1:nrow(dfgoslim2)){
  temp<-str_split(dfgoslim2$geneID[i],"/")[[1]]
  # temp<-dBnbtoEntrezID[which(dBnbtoEntrezID$GeneID %in% temp),]$Locus.tag
  temp<-datlog.norm[datlog.norm$EntrezID %in% temp,]$Bnb
  dfgoslim2$Bnb[i]<-paste(temp,collapse = "_")
  temp<-data.frame(Bnb=temp,go=dfgoslim2$go[i],term=dfgoslim2$term[i],Count=dfgoslim2$Count[i],Category=dfgoslim2$Category[i])
  if(i==1){
    dfgoslim3<-temp
  }else{
    dfgoslim3<-rbind(dfgoslim3,temp)
  }
}
Sys.time()
```
## 10.2. Comparison between DMenv, DMevo and DMmut at the functional category levels based on GO
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=2.5,dpi=300}
df_DMall_gocat<-dfgoslim2
df_DMall_gocat$Custom_ID<-seq(1,nrow(df_DMall_gocat))
df_DMall_gocat$DMenv<-NA
df_DMall_gocat$DMevo<-NA
df_DMall_gocat$DMmut<-NA
i<-1
for(i in 1:nrow(df_DMall_gocat)){
  temp_genes<-str_split(df_DMall_gocat$Bnb[i],"_")[[1]]
  temp_df<-df_DMall[df_DMall$Bnb %in% temp_genes,]
  df_DMall_gocat$DMenv[i]<-mean(temp_df$DMenv)
  df_DMall_gocat$DMevo[i]<-mean(temp_df$DMevo)
  df_DMall_gocat$DMmut[i]<-mean(temp_df$DMmut)
}

df_DMall_gocat$Category<-factor(df_DMall_gocat$Category,levels=qw(BP,CC,MF))
colnames(df_DMall_gocat)[colnames(df_DMall_gocat)=="go"]<-"ID"
colnames(df_DMall_gocat)[colnames(df_DMall_gocat)=="term"]<-"Term Name"
df_DMall_gocat$Domain<-ifelse(df_DMall_gocat$Category=="BP","Biological Process",ifelse(df_DMall_gocat$Category=="CC","Cellular Component","Molecular Function"))
df_DMall_gocat$Domain<-factor(df_DMall_gocat$Domain,levels=c("Biological Process","Cellular Component","Molecular Function"))

my_tmp<-df_DMall_gocat %>% select(DMenv,DMevo,DMmut)
my_tmp<-as.matrix(my_tmp)
df_DMall_gocat$DM_mean<-rowMeans(my_tmp)
df_DMall_gocat$label.sel<-""
my_d<-qw(BP,CC,MF)
i<-1
for(i in 1:length(my_d)){
  my_min<-min(df_DMall_gocat[df_DMall_gocat$Category==my_d[i],]$DM_mean)
  df_DMall_gocat[(df_DMall_gocat$Category==my_d[i])&(df_DMall_gocat$DM_mean==my_min),]$label.sel<-df_DMall_gocat[(df_DMall_gocat$Category==my_d[i])&(df_DMall_gocat$DM_mean==my_min),]$Custom_ID
  my_max<-max(df_DMall_gocat[df_DMall_gocat$Category==my_d[i],]$DM_mean)
  df_DMall_gocat[(df_DMall_gocat$Category==my_d[i])&(df_DMall_gocat$DM_mean==my_max),]$label.sel<-df_DMall_gocat[(df_DMall_gocat$Category==my_d[i])&(df_DMall_gocat$DM_mean==my_max),]$Custom_ID
}


# for display
# my_id<-8*seq(1,10)
# 
# df_DMall_gocat[df_DMall_gocat$Custom_ID %in% my_id,]$label.sel<-df_DMall_gocat[df_DMall_gocat$Custom_ID %in% my_id,]$Custom_ID




my_fontsize<-16
# DMenv vs. DMmut
p<-ggscatter(df_DMall_gocat,x="DMenv",y="DMmut",
          size=2,alpha=0.5,
          label = "label.sel",repel = T,
          color="Domain",palette = mycolpal[1:3],
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=5,label.x=-0.1,label.y=0.6),legend="")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
  xlab(expression("Mean"~DM[env]~"[a.u.]"))+
  ylab(expression("Mean"~DM[mut]~"[a.u.]"))
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p

# DMenv vs. DMevo
p<-ggscatter(df_DMall_gocat,x="DMenv",y="DMevo",
          size=2,alpha=0.5,
          label = "label.sel",repel = T,
          color="Domain",palette = mycolpal[1:3],
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=5,label.x=-0.1,label.y=0.8),legend="")+
  scale_y_continuous(breaks = get_breaks(by = 0.2, from = 0),limits =  c(-0.16, 1.1))+
  xlab(expression("Mean"~DM[env]~"[a.u.]"))+
  ylab(expression("Mean"~DM[evo]~"[a.u.]"))
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p

# DMmut vs. DMevo
p<-ggscatter(df_DMall_gocat,x="DMmut",y="DMevo",
          size=2,alpha=0.5,
          label = "label.sel",repel = T,
          color="Domain",palette = mycolpal[1:3],
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=5,label.x=0.16,label.y=0.65),legend="")+
    # geom_text_repel(data=df_DMall_gocat %>% filter(label.sel!=""),
    #               aes(label=label.sel,x=DMmut,y=DMevo),segment.size=0.3,segment.alpha = 0.5,nudge_x = 0.1, lineheight = 0.1,size=3)+
    scale_x_continuous(breaks = get_breaks(by = 0.2, from = 0),limits =  c(-0.16, 0.7))+
    scale_y_continuous(breaks = get_breaks(by = 0.2, from = 0),limits =  c(-0.16, 1.1))+
          xlab(expression("Mean"~DM[mut]~"[a.u.]"))+
          ylab(expression("Mean"~DM[evo]~"[a.u.]"))
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p

# for legend
p<-ggscatter(df_DMall_gocat,x="DMenv",y="DMmut",
          size=2,alpha=0.5,
          # label = "label.sel",repel = T,
          color="Category",palette = mycolpal[1:3],
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=5,label.x=-0.1,label.y=0.6),legend="")+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.3)))+
  xlab(expression("Mean"~DM[env]~"[a.u.]"))+
  ylab(expression("Mean"~DM[mut]~"[a.u.]"))
p<-ggpar(p,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
q<-ggpar(p,legend = "right",legend.title = "Domain",font.ytickslab= c(10))
leg <- ggpubr::get_legend(q)
as_ggplot(leg)


Sys.time()
```
## 10.3. table of custom subset of GO terms
### 10.3.1.only labeled GO terms
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=10,dpi=300}
my_tmp<-df_DMall_gocat%>% filter(label.sel!="") %>% select(label.sel,Category,ID,`Term Name`)
colnames(my_tmp)[c(1,2)]<-qw(Label,Domain)
my_tmp$`Term Name`<-str_replace_all(string = my_tmp$`Term Name`,pattern = "ATP-binding cassette \\(ABC\\) ",replacement = "ABC ")

ggtexttable(my_tmp,rows=NULL)
Sys.time()
```
### 10.3.2. all 
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=24,fig.width=10,dpi=300}
ggtexttable(df_DMall_gocat[,which(colnames(df_DMall_gocat) %in% c("ID","Domain","Term Name","Count"))],rows=NULL)
Sys.time()
```
## 10.4. Enrichment analysis for DMs using GO slim (using all three categories)
### 10.4.1. perform GSEA for the Env dataset
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=5,dpi=300}
my_DM<-"DMenv"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$Bnb
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)
my_t2g<-dfgoslim3 %>% select(go,Bnb)
# my_t2g<-dfgoslim3[dfgoslim3$Category=="BP",] %>% select(go,Bnb)

## GO analysis
set.seed(123)
my_res <- GSEA(geneList     = gene_list,
              TERM2GENE = my_t2g,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 300, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)
gene_count<-my_res@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(my_res@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
my_tmp<-dfgoslim2 %>% select(go,term,Category)
dot_df<- left_join(dot_df,my_tmp,by=c("ID"="go"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))

dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]
dot_df$log_qvalue<-log10(dot_df$qvalue)
dgsea_go_env_all<-dot_df
dgsea_go_env_all$term_cat<-paste(dgsea_go_env_all$term,"(",dgsea_go_env_all$Category,")",sep="")

Sys.time()
```
### 10.4.2. perform GSEA for the Evo dataset
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=5,dpi=300}
my_DM<-"DMevo"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$Bnb
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)
my_t2g<-dfgoslim3 %>% select(go,Bnb)
# my_t2g<-dfgoslim3[dfgoslim3$Category=="BP",] %>% select(go,Bnb)

## GO analysis
set.seed(123)
my_res <- GSEA(geneList     = gene_list,
              TERM2GENE = my_t2g,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 300, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)
gene_count<-my_res@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(my_res@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
my_tmp<-dfgoslim2 %>% select(go,term,Category)
dot_df<- left_join(dot_df,my_tmp,by=c("ID"="go"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))

dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]
dot_df$log_qvalue<-log10(dot_df$qvalue)
dgsea_go_evo_all<-dot_df
dgsea_go_evo_all$term_cat<-paste(dgsea_go_evo_all$term,"(",dgsea_go_evo_all$Category,")",sep="")
Sys.time()
```
### 10.4.3. perform GSEA for the Mut dataset
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=5,dpi=300}
my_DM<-"DMmut"
gene_list <- df_DMall[,colnames(df_DMall)==my_DM] # Create a vector of the gene universe
names(gene_list)<-df_DMall$Bnb
gene_list = sort(gene_list, decreasing = TRUE) # sort the list in decreasing order (required for clusterProfiler)
my_t2g<-dfgoslim3 %>% select(go,Bnb)
# my_t2g<-dfgoslim3[dfgoslim3$Category=="BP",] %>% select(go,Bnb)

## GO analysis
set.seed(123)
my_res <- GSEA(geneList     = gene_list,
              TERM2GENE = my_t2g,
              seed=TRUE,
               nPerm        = 10000,
               minGSSize    = 5, # 5
               maxGSSize    = 300, # 300
               pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
               pAdjustMethod = "BH") #fdr
set.seed(NULL)
gene_count<-my_res@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
dot_df<- left_join(my_res@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
my_tmp<-dfgoslim2 %>% select(go,term,Category)
dot_df<- left_join(dot_df,my_tmp,by=c("ID"="go"))
dot_df$type="Sensitive"
dot_df$type[dot_df$NES < 0]="Robust"
dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))

dot_df<-dot_df[order(dot_df$enrichmentScore,decreasing = F),]
dot_df$log_qvalue<-log10(dot_df$qvalue)
dgsea_go_mut_all<-dot_df
dgsea_go_mut_all$term_cat<-paste(dgsea_go_mut_all$term,"(",dgsea_go_mut_all$Category,")",sep="")
Sys.time()
```
### 10.4.4. Visualization
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=7,dpi=300}
figtemp1<-dgsea_go_env_all[dgsea_go_env_all$qvalue<0.05,]
figtemp1$Dataset<-"Env"
figtemp2<-dgsea_go_evo_all[dgsea_go_evo_all$qvalue<0.05,]
figtemp2$Dataset<-"Evo"
figtemp3<-dgsea_go_mut_all[dgsea_go_mut_all$qvalue<0.05,]
figtemp3$Dataset<-"Mut"
figtemp<-bind_rows(figtemp1,figtemp2,figtemp3)
my_set12<-intersect(figtemp1$ID,figtemp2$ID)
my_set23<-intersect(figtemp2$ID,figtemp3$ID)
my_set31<-intersect(figtemp3$ID,figtemp1$ID)
myset123<-intersect(intersect(my_set12,my_set23),my_set31)
myset12_23_31<-unique(c(my_set12,my_set23,my_set31))
myset12_23_31<-myset12_23_31[myset12_23_31 %nin% myset123]
figtemp$col<-ifelse(figtemp$ID %in% myset123,"#AD002AFF",ifelse(figtemp$ID %in% myset12_23_31,"#ED0000FF","black"))

my_max=max(str_length(figtemp$term_cat))
figtemp$label<-""
for(i in 1:nrow(figtemp)){
  my_n<-my_max-str_length(figtemp$term_cat[i])
  figtemp$label[i]<-paste(paste(rep(" ",my_n),collapse =""),figtemp$term_cat[i],sep ="")
}

my_tmp<-data.frame(Dataset=qw(Env,Evo,Mut),Count=0,Width=0,BlankWidth=0,Length=0,Height=0,BlankHeight=0)
for(i in 1:nrow(my_tmp)){
  my_tmp$Count[i]<-nrow(figtemp[figtemp$Dataset==my_tmp$Dataset[i],])
  my_tmp$Length[i]<-max(str_length(figtemp[figtemp$Dataset==my_tmp$Dataset[i],]$term_cat))
}
my_tmp$Width<-0.2+0.8*my_tmp$Count/max(my_tmp$Count)
my_tmp$BlankWidth<-1-my_tmp$Width
my_tmp$Height<- 0.87+0.13*(my_tmp$Length/max(my_tmp$Length))
my_tmp$BlankHeight<-1-my_tmp$Height

my_lo<-rbind(c(1,2),c(3,3))
for(i in 1:nrow(my_tmp)){
  figtemp1<-figtemp[figtemp$Dataset==my_tmp$Dataset[i],]
  figtemp1<-figtemp1[order(figtemp1$NES,decreasing = T),]
  p<-ggdotchart(figtemp1,x="label",y="NES", # "enrichmentScore"
           color="log_qvalue",
           sorting = "descending", 
           add = "segments", 
           add.params = list(color = "lightgray", size = 2),
           dot.size = "GeneRatio",
           legend="right",legend.title="",
           xlab="",ylab="Normalized enrichment score")+
              scale_y_continuous(limits = c(-3,3))+
              scale_size(range = c(1, 4),breaks = c(0.3,0.6,0.9),limits = c(0,1)) +
              scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                                    guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "right",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 0),ticks.linewidth = 1,order=1))+
              guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
  p<-ggpar(p,font.xtickslab = list(color=figtemp1$col))
  my_w<-c(my_tmp$Width[i],my_tmp$BlankWidth[i])
  my_h<-c(my_tmp$Height[i],my_tmp$BlankHeight[i])
  grid.arrange(p,blank,blank,
           # nrow=2,
           layout_matrix=my_lo,
           heights=c(my_h),
           widths=c(my_w))
}

Sys.time()
```
### 10.4.5. Venn diagram
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=7,dpi=300}
# Higher DM values
my_venn<-list(
  Evo=dgsea_go_evo_all[(dgsea_go_evo_all$NES>0)&(dgsea_go_evo_all$qvalue<0.05),]$ID,
  Env=dgsea_go_env_all[(dgsea_go_env_all$NES>0)&(dgsea_go_env_all$qvalue<0.05),]$ID,
  Mut=dgsea_go_mut_all[(dgsea_go_mut_all$NES>0)&(dgsea_go_mut_all$qvalue<0.05),]$ID
)
venn <- process_data(Venn(my_venn))
venn$regionData$name
venn$regionData$count
my_df<-venn$regionData
my_tmp<-unlist(my_df[my_df$name=="Evo/Env/Mut","item"])
dfgoslim2[dfgoslim2$go %in% my_tmp,]$term


v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))

v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)

# Lower DM values
my_venn<-list(
  Evo=dgsea_go_evo_all[(dgsea_go_evo_all$NES<0)&(dgsea_go_evo_all$qvalue<0.05),]$ID,
  Env=dgsea_go_env_all[(dgsea_go_env_all$NES<0)&(dgsea_go_env_all$qvalue<0.05),]$ID,
  Mut=dgsea_go_mut_all[(dgsea_go_mut_all$NES<0)&(dgsea_go_mut_all$qvalue<0.05),]$ID
)
venn <- process_data(Venn(my_venn))
venn$regionData$name
venn$regionData$count
my_df<-venn$regionData
my_tmp<-unlist(my_df[my_df$name=="Evo/Env/Mut","item"])
dfgoslim2[dfgoslim2$go %in% my_tmp,]$term

v<-venneuler(c(Evo=length(venn$regionData$item[[which(venn$regionData$name=="Evo")]]),
               Env=length(venn$regionData$item[[which(venn$regionData$name=="Env")]]),
               Mut=length(venn$regionData$item[[which(venn$regionData$name=="Mut")]]),
               "Evo&Env"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env")]]),
             "Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Env/Mut")]]),
               "Mut&Evo"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Mut")]]),
             "Evo&Env&Mut"=length(venn$regionData$item[[which(venn$regionData$name=="Evo/Env/Mut")]])))

v$labels<-c("","","")

par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = c("#F8766D", "#619CFF", "#00BA38"), quantities = TRUE) # Evo(red), Env(blue), Mut(green)

Sys.time()
```
## 10.5. Enrichment analysis for loadings in PC1-3 using GO slim
### 10.5.1. perform GSEA for the Env dataset
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=10,dpi=300}
my_t2g<-dfgoslim3 %>% select(go,Bnb)
# my_t2g<-dfgoslim3[dfgoslim3$Category=="BP",] %>% select(go,Bnb)

# For Env
# temp<-t(res.pca.global$sdev*t(res.pca.global$rotation)) : "loadings: correlation between PCx and original variables" identical to pca.glob.var$coord
# Principal component loadings: correlation between PCx and original variables
# see http://www.eeso.ges.kyoto-u.ac.jp/emm/materials/rmemo/pcrcom


my_fontsize<-20

my_datasets<-qw(Env)
pca.var<-get_pca_var(res.pca.global_env)
figtemp1<-as.data.frame(pca.var$coord[,1:ncol(pca.var$coord)])
colnames(figtemp1)<-str_replace(colnames(figtemp1),"Dim.","PC")
figtemp1<-figtemp1 %>% select(mypcs)
my_dims<-colnames(figtemp1)
figtemp1$Bnb<-rownames(figtemp1)
for(i in mypcs){
  test<-figtemp1[order(figtemp1[,which(colnames(figtemp1)==i)],decreasing = T),]
  original_gene_list <- test[,which(colnames(test)==i)]
  # name the vector
  names(original_gene_list) <- test$Bnb
  # omit any NA values 
  gene_list<-na.omit(original_gene_list)
  # sort the list in decreasing order (required for clusterProfiler)
  gene_list = sort(gene_list, decreasing = TRUE)
  set.seed(123)
  my_res<- GSEA(geneList     = gene_list,
                TERM2GENE = my_t2g,
                seed=TRUE,
                 nPerm        = 10000,
                 minGSSize    = 5, # 5
                 maxGSSize    = 1000, # 100
                 pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
                 pAdjustMethod = "BH") #fdr
  set.seed(NULL)
  gene_count<-my_res@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
  dot_df<- left_join(my_res@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
  # dot_df<-dot_df[dot_df$count>4,]
  dot_df$type="Sensitive"
  dot_df$type[dot_df$NES < 0]="Robust"
  dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
  my_tmp<-dfgoslim2 %>% select(go,term,Category)
  dot_df<- left_join(dot_df,my_tmp,by=c("ID"="go"))
  dot_df$PC<-i
  if(i=="PC1"){
    figtemp2<-dot_df
  }else{
    figtemp2<-bind_rows(figtemp2,dot_df)
  }
}
figtemp2$qval.fin<-qvalue(figtemp2$pvalue)$qvalues
figtemp2$log_qval.fin<-log10(figtemp2$qval.fin)
figtemp2<-figtemp2[figtemp2$qval.fin<0.05,]
# figtemp2<-figtemp2[figtemp2$qvalue<0.05,] # <-------------------    Use standard qvalue
figtemp2$term_cat<-paste(figtemp2$term,"(",figtemp2$Category,")",sep="")
my_tmp<-figtemp2[(figtemp2$NES>0)&(figtemp2$PC=="PC1"),]$ID
my_tmp2<-figtemp2[(figtemp2$NES<0)&(figtemp2$PC=="PC1"),]$ID
figtemp2$col<-ifelse(figtemp2$ID %in% my_tmp,"#ED0000FF",ifelse(figtemp2$ID %in% my_tmp2,"#00468BFF","black"))

df_pc123_gsea_env<-figtemp2

my_tmp<-data.frame(PC=mypcs,Count=0,Width=0,BlankWidth=0,Length=0,Height=0,BlankHeight=0)
for(i in 1:nrow(my_tmp)){
  my_tmp$Count[i]<-nrow(figtemp2[figtemp2$PC==my_tmp$PC[i],])
  my_tmp$Length[i]<-max(str_length(figtemp2[figtemp2$PC==my_tmp$PC[i],]$term))
}
my_tmp$Width<-0.3+0.7*my_tmp$Count/max(my_tmp$Count)
my_tmp$BlankWidth<-1-my_tmp$Width
my_tmp$Height<- 0.45+0.55*(my_tmp$Length/max(my_tmp$Length))
my_tmp$BlankHeight<-1-my_tmp$Height

my_lo<-rbind(c(1,2),c(3,3))

for(ii in 1:length(mypcs)){
  k<-mypcs[ii]
  temp<-figtemp2[figtemp2$PC==k,]
  temp<-temp[order(temp$NES,decreasing = T),]
  p<-ggdotchart(figtemp2[figtemp2$PC==k,],x="term_cat",y="NES",
                color="log_qval.fin", #qval.fin
                sorting = "descending", 
                add = "segments", 
                add.params = list(color = "lightgray", size = 2),
                dot.size = "GeneRatio",
                legend="right",legend.title="",
                xlab="",ylab="Normalized enrichment score")+
              scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
              scale_size(range = c(1, 4),breaks = c(0.3,0.6,0.9),limits = c(0,1)) +
              scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                                    guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "right",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 0),ticks.linewidth = 1,order=1))+
              guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
  p<-ggpar(p,font.xtickslab = list(color=temp$col))
  my_w<-c(my_tmp$Width[ii],my_tmp$BlankWidth[ii])
  my_h<-c(my_tmp$Height[ii],my_tmp$BlankHeight[ii])
  grid.arrange(p,blank,blank,
           # nrow=2,
           layout_matrix=my_lo,
           heights=c(my_h),
           widths=c(my_w))
}
Sys.time()
```
### 10.5.2. perform GSEA for the Evo dataset
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=10,dpi=300}
my_t2g<-dfgoslim3 %>% select(go,Bnb)
my_fontsize<-20
my_datasets<-qw(Evo)
pca.var<-get_pca_var(res.pca.global_evo)
figtemp1<-as.data.frame(pca.var$coord[,1:ncol(pca.var$coord)])
colnames(figtemp1)<-str_replace(colnames(figtemp1),"Dim.","PC")
figtemp1<-figtemp1 %>% select(mypcs)
my_dims<-colnames(figtemp1)
figtemp1$Bnb<-rownames(figtemp1)
for(i in mypcs){
  test<-figtemp1[order(figtemp1[,which(colnames(figtemp1)==i)],decreasing = T),]
  original_gene_list <- test[,which(colnames(test)==i)]
  # name the vector
  names(original_gene_list) <- test$Bnb
  # omit any NA values 
  gene_list<-na.omit(original_gene_list)
  # sort the list in decreasing order (required for clusterProfiler)
  gene_list = sort(gene_list, decreasing = TRUE)
  set.seed(123)
  my_res<- GSEA(geneList     = gene_list,
                TERM2GENE = my_t2g,
                seed=TRUE,
                 nPerm        = 10000,
                 minGSSize    = 5, # 5
                 maxGSSize    = 1000, # 100
                 pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
                 pAdjustMethod = "BH") #fdr
  set.seed(NULL)
  gene_count<-my_res@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
  dot_df<- left_join(my_res@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
  # dot_df<-dot_df[dot_df$count>4,]
  dot_df$type="Sensitive"
  dot_df$type[dot_df$NES < 0]="Robust"
  dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
  my_tmp<-dfgoslim2 %>% select(go,term,Category)
  dot_df<- left_join(dot_df,my_tmp,by=c("ID"="go"))
  dot_df$PC<-i
  if(i=="PC1"){
    figtemp2<-dot_df
  }else{
    figtemp2<-bind_rows(figtemp2,dot_df)
  }
}
figtemp2$qval.fin<-qvalue(figtemp2$pvalue)$qvalues
figtemp2$log_qval.fin<-log10(figtemp2$qval.fin)
figtemp2<-figtemp2[figtemp2$qval.fin<0.05,]
# figtemp2<-figtemp2[figtemp2$qvalue<0.05,] # <-------------------    Use standard qvalue
figtemp2$term_cat<-paste(figtemp2$term,"(",figtemp2$Category,")",sep="")

my_max=max(str_length(figtemp2$term_cat))
figtemp2$label<-""
for(i in 1:nrow(figtemp2)){
  my_n<-my_max-str_length(figtemp2$term_cat[i])
  figtemp2$label[i]<-paste(paste(rep(" ",my_n),collapse =""),figtemp2$term_cat[i],sep ="")
}

figtemp2$col<-ifelse(figtemp2$ID %in% df_pc123_gsea_env[(df_pc123_gsea_env$PC=="PC1")&(df_pc123_gsea_env$NES>0),]$ID,"#ED0000FF",
                     ifelse(figtemp2$ID %in% df_pc123_gsea_env[(df_pc123_gsea_env$PC=="PC1")&(df_pc123_gsea_env$NES<0),]$ID,"#00468BFF","black"))

df_pc123_gsea_evo<-figtemp2

my_tmp<-data.frame(PC=mypcs,Count=0,Width=0,BlankWidth=0,Length=0,Height=0,BlankHeight=0)
for(i in 1:nrow(my_tmp)){
  my_tmp$Count[i]<-nrow(figtemp2[figtemp2$PC==my_tmp$PC[i],])
  my_tmp$Length[i]<-max(str_length(figtemp2[figtemp2$PC==my_tmp$PC[i],]$term_cat))
}
my_tmp$Width<-0.3+0.7*my_tmp$Count/max(my_tmp$Count)
my_tmp$BlankWidth<-1-my_tmp$Width
my_tmp$Height<- 0.36+0.64*(my_tmp$Length/max(my_tmp$Length))
my_tmp$BlankHeight<-1-my_tmp$Height

my_lo<-rbind(c(1,2),c(3,3))

for(ii in 1:length(mypcs)){
  k<-mypcs[ii]
  temp<-figtemp2[figtemp2$PC==k,]
  temp<-temp[order(temp$enrichmentScore,decreasing = T),]
  p<-ggdotchart(figtemp2[figtemp2$PC==k,],x="term_cat",y="NES",
                color="log_qval.fin", #qval.fin
                sorting = "descending", 
                add = "segments", 
                add.params = list(color = "lightgray", size = 2),
                dot.size = "GeneRatio",
                legend="right",legend.title="",
                xlab="",ylab="Normalized enrichment score")+
              scale_y_continuous(breaks = get_breaks(by = 2, from = -2),limit=c(-3.5,3.5))+
              scale_size(range = c(1, 4),breaks = c(0.3,0.6,0.9),limits = c(0,1)) +
              scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                                    guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "right",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 0),ticks.linewidth = 1,order=1))+
              guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
  p<-ggpar(p,font.xtickslab = list(color=temp$col))
  my_w<-c(my_tmp$Width[ii],my_tmp$BlankWidth[ii])
  my_h<-c(my_tmp$Height[ii],my_tmp$BlankHeight[ii])
  grid.arrange(p,blank,blank,
           # nrow=2,
           layout_matrix=my_lo,
           heights=c(my_h),
           widths=c(my_w))
}
Sys.time()
```
### 10.5.3. perform GSEA for the Mut dataset
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=10,fig.width=10,dpi=300}
my_t2g<-dfgoslim3 %>% select(go,Bnb)
my_fontsize<-20
my_datasets<-qw(Mut)
pca.var<-get_pca_var(res.pca.global_mut)
figtemp1<-as.data.frame(pca.var$coord[,1:ncol(pca.var$coord)])
colnames(figtemp1)<-str_replace(colnames(figtemp1),"Dim.","PC")
figtemp1<-figtemp1 %>% select(mypcs)
my_dims<-colnames(figtemp1)
figtemp1$Bnb<-rownames(figtemp1)
for(i in mypcs){
  test<-figtemp1[order(figtemp1[,which(colnames(figtemp1)==i)],decreasing = T),]
  original_gene_list <- test[,which(colnames(test)==i)]
  # name the vector
  names(original_gene_list) <- test$Bnb
  # omit any NA values 
  gene_list<-na.omit(original_gene_list)
  # sort the list in decreasing order (required for clusterProfiler)
  gene_list = sort(gene_list, decreasing = TRUE)
  set.seed(123)
  my_res<- GSEA(geneList     = gene_list,
                TERM2GENE = my_t2g,
                seed=TRUE,
                 nPerm        = 10000,
                 minGSSize    = 5, # 5
                 maxGSSize    = 1000, # 100
                 pvalueCutoff = 1, # <------------ set 1 to obtain enrichment scores for all gene sets 
                 pAdjustMethod = "BH") #fdr
  set.seed(NULL)
  gene_count<-my_res@result %>% group_by(ID) %>% summarise(count = sum(str_count(core_enrichment, "/")+1))
  dot_df<- left_join(my_res@result, gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)
  # dot_df<-dot_df[dot_df$count>4,]
  dot_df$type="Sensitive"
  dot_df$type[dot_df$NES < 0]="Robust"
  dot_df$type<-factor(dot_df$type,levels=c("Sensitive","Robust"))
  my_tmp<-dfgoslim2 %>% select(go,term,Category)
  dot_df<- left_join(dot_df,my_tmp,by=c("ID"="go"))
  dot_df$PC<-i
  if(i=="PC1"){
    figtemp2<-dot_df
  }else{
    figtemp2<-bind_rows(figtemp2,dot_df)
  }
}
figtemp2$qval.fin<-qvalue(figtemp2$pvalue)$qvalues
figtemp2$log_qval.fin<-log10(figtemp2$qval.fin)
figtemp2<-figtemp2[figtemp2$qval.fin<0.05,]
# figtemp2<-figtemp2[figtemp2$qvalue<0.05,] # <-------------------    Use standard qvalue
figtemp2$term_cat<-paste(figtemp2$term,"(",figtemp2$Category,")",sep="")

figtemp2$col<-ifelse(figtemp2$ID %in% df_pc123_gsea_env[(df_pc123_gsea_env$PC=="PC1")&(df_pc123_gsea_env$NES>0),]$ID,"#ED0000FF",
                     ifelse(figtemp2$ID %in% df_pc123_gsea_env[(df_pc123_gsea_env$PC=="PC1")&(df_pc123_gsea_env$NES<0),]$ID,"#00468BFF","black"))

df_pc123_gsea_mut<-figtemp2

my_tmp<-data.frame(PC=mypcs,Count=0,Width=0,BlankWidth=0,Length=0,Height=0,BlankHeight=0)
for(i in 1:nrow(my_tmp)){
  my_tmp$Count[i]<-nrow(figtemp2[figtemp2$PC==my_tmp$PC[i],])
  my_tmp$Length[i]<-max(str_length(figtemp2[figtemp2$PC==my_tmp$PC[i],]$term_cat))
}
my_tmp$Width<-0.3+0.7*my_tmp$Count/max(my_tmp$Count)
my_tmp$BlankWidth<-1-my_tmp$Width
my_tmp$Height<- 0.45+0.55*(my_tmp$Length/max(my_tmp$Length))
my_tmp$BlankHeight<-1-my_tmp$Height

my_lo<-rbind(c(1,2),c(3,3))

for(ii in 1:length(mypcs)){
  k<-mypcs[ii]
  temp<-figtemp2[figtemp2$PC==k,]
  temp<-temp[order(temp$NES,decreasing = T),]
  p<-ggdotchart(figtemp2[figtemp2$PC==k,],x="term_cat",y="NES",
                color="log_qval.fin", #qval.fin
                sorting = "descending", 
                add = "segments", 
                add.params = list(color = "lightgray", size = 2),
                dot.size = "GeneRatio",
                legend="right",legend.title="",
                xlab="",ylab="Normalized enrichment score")+
              scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
              scale_size(range = c(1, 4),breaks = c(0.3,0.6,0.9),limits = c(0,1)) +
              scale_colour_continuous(limits = c(-4.1,log10(0.05)), breaks = c(-4,-3,-2,-1),
                                    guide=guide_colourbar(title="q-value (log)",title.position = "top",label.position = "right",label.vjust = 0.4,label.hjust = 0.2,label.theme = element_text(angle = 0),ticks.linewidth = 1,order=1))+
              guides(size = guide_legend(title= "Gene ratio",title.position = "top",order=2))
  p<-ggpar(p,font.xtickslab = list(color=temp$col))
  my_w<-c(my_tmp$Width[ii],my_tmp$BlankWidth[ii])
  my_h<-c(my_tmp$Height[ii],my_tmp$BlankHeight[ii])
  grid.arrange(p,blank,blank,
           # nrow=2,
           layout_matrix=my_lo,
           heights=c(my_h),
           widths=c(my_w))
}
Sys.time()
```
### 10.5.4. Visualization
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=6,fig.width=4.5,dpi=300}
my_df<-data.frame(name=paste(rep(qw(Env,Evo,Mut),each=6),rep(rep(qw(PC1,PC2,PC3),each=2),3),rep(qw(Pos,Neg),9),sep="_"),
                  dataset=NA,pc=NA,sign=NA,growth=0,starvation=0,others=0,total=0)
for(i in 1:nrow(my_df)){
  my_tmp<-str_split(my_df$name[i],"_")[[1]]
  my_df$dataset[i]<-my_tmp[1]
  my_df$pc[i]<-my_tmp[2]
  my_df$sign[i]<-my_tmp[3]
}

my_term1<-df_pc123_gsea_env[df_pc123_gsea_env$col=="#ED0000FF",]$ID
my_term2<-df_pc123_gsea_env[df_pc123_gsea_env$col=="#00468BFF",]$ID
for(i in 1:nrow(my_df)){
  if(my_df$dataset[i]=="Env"){
    my_tmp<-df_pc123_gsea_env[df_pc123_gsea_env$PC==my_df$pc[i],]
    if(my_df$sign[i]=="Pos"){
      my_tmp1<-my_tmp[my_tmp$NES>0,]
      my_df$growth[i]<-nrow(my_tmp1[my_tmp1$col=="#ED0000FF",])
      my_df$starvation[i]<-nrow(my_tmp1[my_tmp1$col=="#00468BFF",])
      my_df$others[i]<-nrow(my_tmp1[my_tmp1$col=="black",])
    }else{
      my_tmp1<-my_tmp[my_tmp$NES<0,]
      my_df$growth[i]<-nrow(my_tmp1[my_tmp1$col=="#ED0000FF",])
      my_df$starvation[i]<-nrow(my_tmp1[my_tmp1$col=="#00468BFF",])
      my_df$others[i]<-nrow(my_tmp1[my_tmp1$col=="black",])
    }
  }
  if(my_df$dataset[i]=="Evo"){
    my_tmp<-df_pc123_gsea_evo[df_pc123_gsea_evo$PC==my_df$pc[i],]
    if(my_df$sign[i]=="Pos"){
      my_tmp1<-my_tmp[my_tmp$NES>0,]
      my_df$growth[i]<-nrow(my_tmp1[my_tmp1$col=="#ED0000FF",])
      my_df$starvation[i]<-nrow(my_tmp1[my_tmp1$col=="#00468BFF",])
      my_df$others[i]<-nrow(my_tmp1[my_tmp1$col=="black",])
    }else{
      my_tmp1<-my_tmp[my_tmp$NES<0,]
      my_df$growth[i]<-nrow(my_tmp1[my_tmp1$col=="#ED0000FF",])
      my_df$starvation[i]<-nrow(my_tmp1[my_tmp1$col=="#00468BFF",])
      my_df$others[i]<-nrow(my_tmp1[my_tmp1$col=="black",])
    }
  }
  if(my_df$dataset[i]=="Mut"){
    my_tmp<-df_pc123_gsea_mut[df_pc123_gsea_mut$PC==my_df$pc[i],]
    if(my_df$sign[i]=="Pos"){
      my_tmp1<-my_tmp[my_tmp$NES>0,]
      my_df$growth[i]<-nrow(my_tmp1[my_tmp1$col=="#ED0000FF",])
      my_df$starvation[i]<-nrow(my_tmp1[my_tmp1$col=="#00468BFF",])
      my_df$others[i]<-nrow(my_tmp1[my_tmp1$col=="black",])
    }else{
      my_tmp1<-my_tmp[my_tmp$NES<0,]
      my_df$growth[i]<-nrow(my_tmp1[my_tmp1$col=="#ED0000FF",])
      my_df$starvation[i]<-nrow(my_tmp1[my_tmp1$col=="#00468BFF",])
      my_df$others[i]<-nrow(my_tmp1[my_tmp1$col=="black",])
    }
  }
}
my_df$total<-my_df$growth+my_df$starvation+my_df$others
my_df$growth_frac<-my_df$growth/my_df$total
my_df$starvation_frac<-my_df$starvation/my_df$total
my_df$others_frac<-my_df$others/my_df$total

# count
my_df2<-my_df %>% pivot_longer(cols = c(growth,starvation,others),names_to = "family",values_to = "count")
my_df2$family<-factor(my_df2$family,levels=qw(growth,starvation,others))
my_df2$pc_sign<-paste(my_df2$pc,my_df2$sign,sep="_")

my_fontsize<-16
p1<-ggbarplot(my_df2,x="pc_sign",y="count",fill="family",palette = c("#ED0000FF","#2171B5","gray70"),legend="",size=0,
          # position = position_dodge(),
          xlab = "",ylab="Number of GO terms")+
  rotate_x_text(angle = 90)
p1<-ggpar(p1,font.y = my_fontsize)
p1<-facet(p1, facet.by = "dataset",ncol = 1,panel.labs.font = list(size=my_fontsize*0.8))

leg<-ggpar(p1,font.legend = my_fontsize,legend="top")
leg <- ggpubr::get_legend(leg)

# percent
my_df3<-my_df %>% pivot_longer(cols = c(growth_frac,starvation_frac,others_frac),names_to = "family",values_to = "fraction")
my_df3$family<-factor(my_df3$family,levels=qw(growth_frac,starvation_frac,others_frac))
my_df3$pc_sign<-paste(my_df3$pc,my_df3$sign,sep="_")

my_fontsize<-16
p2<-ggbarplot(my_df3,x="pc_sign",y="fraction",fill="family",palette = c("#ED0000FF","#2171B5","gray70"),legend="",size=0,
          # position = position_dodge(),
          xlab = "",ylab="Fraction of GO terms")+
  scale_y_continuous(limit=c(0,1),breaks=seq(0,1,0.5))+
  rotate_x_text(angle = 90)
p2<-ggpar(p2,font.y = my_fontsize)
p2<-facet(p2, facet.by = "dataset",ncol = 1,panel.labs.font = list(size=my_fontsize*0.8))

my_lo<-rbind(c(1,1),c(2,3))
grid.arrange(as_ggplot(leg),p1,p2,widths=c(0.6,0.65),heights=c(0.1,0.9),layout_matrix=my_lo)
Sys.time()
```
### 10.5.5. growth and starvation families
```{r GO_slim,warning=FALSE, message=FALSE,fig.height=4,fig.width=10,dpi=300}
my_df1<-df_pc123_gsea_env[(df_pc123_gsea_env$NES>0)&(df_pc123_gsea_env$PC=="PC1"),]
my_df1<-my_df1[order(my_df1$NES,decreasing = T),]
my_df1<-head(my_df1,5)
my_df1<-my_df1 %>% select(Category,ID,term)
colnames(my_df1)<-c("Domain","ID","Term Name")
my_df1$family<-"growth"

my_df2<-df_pc123_gsea_env[(df_pc123_gsea_env$NES<0)&(df_pc123_gsea_env$PC=="PC1"),]
my_df2<-my_df2[order(my_df2$NES,decreasing = F),]
my_df2<-head(my_df2,5)
my_df2<-my_df2 %>% select(Category,ID,term)
colnames(my_df2)<-c("Domain","ID","Term Name")
my_df2$family<-"starvation"

my_df<-bind_rows(my_df1,my_df2)
my_df$`Term Name`<-str_replace_all(string = my_df$`Term Name`,pattern = "ATP-binding cassette \\(ABC\\) ",replacement = "ABC ")

ggtexttable(my_df,rows=NULL)
Sys.time()
```
#
#
#
# 11. Noise-Plasticity Coupling
## 11.1. The Noise_ch dataset
Taniguchi 2010, Science
https://doi.org/10.1126/science.1188308
Strain: YFP Fusion Library
### 11.1.1. Load the Noise_ch dataset
```{r Noise_ch,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=3.2,dpi=300}
ddtani<-paste(getwd(),"NOISE_CH_DataSet","TableS6.xlsx",sep = "/")
dtani<-data.frame(read.xlsx(ddtani))

df_tani<-dtani[,which(colnames(dtani) %in% qw(B.Number,Mean_Protein,Noise_Protein))]
colnames(df_tani)<-qw(Bnb,TNMean,TNNoise)
df_tani$TNMean<-ifelse(is.na(df_tani$TNMean),NA,log10(df_tani$TNMean))
df_tani$TNNoise<-ifelse(is.na(df_tani$TNNoise),NA,log10(df_tani$TNNoise))
# df_tani<-dplyr::left_join(df_DMall,df_tani,by="Bnb")
df_tani<-dplyr::left_join(df_tani,df_DMall,by="Bnb")
df_tani<-df_tani[!is.na(df_tani$TNMean),]
df_tani<-df_tani[order(df_tani$TNMean),]
df_tani<-transform(df_tani,TNRunMed=NA,TNSmoothSp=NA,TNDM=NA)
df_tani$TNRunMed<-runmed(df_tani$TNNoise,41)
TNSSvislist<-createEmptyDf(nrow=1000,ncol=2, colnames = qw(logx,SS))
TNSSvislist$logx=seq(min(df_tani$TNMean),max(df_tani$TNMean),length=nrow(TNSSvislist))
misplist<-list()
misplist[[1]]<-smooth.spline(data.frame(x=df_tani$TNMean,y=df_tani$TNRunMed),df=12)
TNSSvislist$SS<-predict(misplist[[1]],TNSSvislist$logx)[[2]]
df_tani$TNSmoothSp<-predict(misplist[[1]],df_tani$TNMean)[[2]]
df_tani$TNDM<-df_tani$TNNoise-df_tani$TNSmoothSp
# TN_thr<-median(df_tani$TNMean)
TN_thr<-1 # from Taniguchi et al
df_tani$HL_label<-ifelse(df_tani$TNMean>=TN_thr,"High","Low")
df_tani$HL_label<-factor(df_tani$HL_label,levels = c("Low","High"))
my_tmp<-df.regcount %>% select(gene_id,Count.allregs,Count.TRcom,TRcomflag2)
df_tani<-left_join(df_tani,my_tmp,by=c("Bnb"="gene_id"))


ggscatter(df_tani,x="TNMean",y="TNNoise",alpha=0.5,size=0.3,color = "HL_label",palette = c("green4","green3"),legend="",
          cor.coef = TRUE,cor.coeff.args = list(method = "spearman", label.x = 1.5, label.y=0.8, label.sep = "\n",size=4))+
            xlab(expression(log[10]~"protein expression level [a.u.]"))+
            ylab(expression(log[10]~"protein noise [a.u.]"))+
  scale_x_continuous(breaks = get_breaks(by = 2, from = -2),limits =  c(-2, 4))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = -2),limits =  c(-2, 2))+
  geom_vline(xintercept = TN_thr,size=0.5,colour="darkgray")+
  geom_line(data=df_tani,aes(x=TNMean,y=TNRunMed),colour="darkorange", size=0.7)+
  geom_line(data=TNSSvislist,aes(x=logx,y=SS),colour="deepskyblue", size=1)+
  geom_hline(yintercept = -1,linetype="dashed", color=mycolpal[1])+
  geom_abline(slope = -1,intercept = 0,linetype="dashed",color=mycolpal[2])

ggscatter(df_tani,x="TNMean",y="TNDM",
          cor.coef = TRUE,cor.coeff.args = list(method = "spearman", label.x = -1.3, label.y=1.2, label.sep = "\n",size=4),alpha=0.5,size=0.3,color = "HL_label",palette = c("green4","green3"),legend="",)+
  scale_x_continuous(breaks = get_breaks(by = 2, from = -2),limits =  c(-2, 4))+
  # scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2))+
  xlab(expression(log[10]~"protein expression level [a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))

Sys.time()
```
### 11.1.2. Basic statistics of the dataset
```{r Noise,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
my_venn<-list(
  Three=dBnb.comp$Bnb,
  Noise_ch=df_tani$Bnb
)

# ggVennDiagram(figtemp)
venn <- process_data(Venn(my_venn))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Three=length(venn$regionData$item[[which(venn$regionData$name=="Three")]]),
               Noise_ch=length(venn$regionData$item[[which(venn$regionData$name=="Noise_ch")]]),
               "Three&Noise_ch"=length(venn$regionData$item[[which(venn$regionData$name=="Three/Noise_ch")]])))
my_tmp<-v$colors
my_tmp[names(my_tmp)=="Three"][[1]]<-"gray50"
my_tmp[names(my_tmp)!="Three"][[1]]<-"yellow"
v$colors<-my_tmp
v$labels<-c("","")
par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = v$colors, quantities = TRUE) # Three(gray), 

Sys.time()
```
### 11.1.3. Comprison of DMs
```{r Noise_ch,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
my_fontsize<-16

# Lowly expressed proteins
# DMenv vs. DMnoise
p1<-ggscatter(df_tani[df_tani$HL_label=="Low",],x="DMenv",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green4", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
  scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p1<-ggpar(p1,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMevo vs. DMnoise
p2<-ggscatter(df_tani[df_tani$HL_label=="Low",],x="DMevo",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green4", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
  scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p2<-ggpar(p2,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMmut vs. DMnoise
p3<-ggscatter(df_tani[df_tani$HL_label=="Low",],x="DMmut",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green4", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
  scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p3<-ggpar(p3,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)

p1+p2+p3

# Highly expressed proteins
# DMenv vs. DMnoise
p4<-ggscatter(df_tani[df_tani$HL_label=="High",],x="DMenv",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
  scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p4<-ggpar(p4,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMevo vs. DMnoise
p5<-ggscatter(df_tani[df_tani$HL_label=="High",],x="DMevo",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
  scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p5<-ggpar(p5,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMmut vs. DMnoise
p6<-ggscatter(df_tani[df_tani$HL_label=="High",],x="DMmut",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
  scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p6<-ggpar(p6,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)

p4+p5+p6

# Classifiation based on transcriptional regulation
# DMenv vs. DMnoise
## TRcom-containing
p7<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="TRcom-containing"),],x="DMenv",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p7<-ggpar(p7,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
## TRcom-less
p8<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="TRcom-less"),],x="DMenv",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p8<-ggpar(p8,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
## Null regulation
p9<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="Null"),],x="DMenv",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p9<-ggpar(p9,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p7+p8+p9

# DMevo vs. DMnoise
## TRcom-containing
p7<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="TRcom-containing"),],x="DMevo",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p7<-ggpar(p7,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
## TRcom-less
p8<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="TRcom-less"),],x="DMevo",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p8<-ggpar(p8,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
## Null regulation
p9<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="Null"),],x="DMevo",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p9<-ggpar(p9,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p7+p8+p9

# DMmut vs. DMnoise
## TRcom-containing
p7<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="TRcom-containing"),],x="DMmut",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p7<-ggpar(p7,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
## TRcom-less
p8<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="TRcom-less"),],x="DMmut",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p8<-ggpar(p8,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
## Null regulation
p9<-ggscatter(df_tani[(df_tani$HL_label=="High")&(df_tani$TRcomflag2=="Null"),],x="DMmut",y="TNDM",
          size=0.2,alpha=0.5,
          color = "green3", # "gray20",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.2))+
    scale_y_continuous(breaks = get_breaks(by = 0.5, from = -0.5),limits =  c(-0.6, 1.8))+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[noise_ch]~"[a.u.]"))
p9<-ggpar(p9,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p7+p8+p9


Sys.time()
```
### 11.1.4. 
```{r Noise,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
my_tmp<-df_tani[!is.na(df_tani$TRcomflag2),]
my_comparisons <- list( c("TRcom-containing", "TRcom-less"), c("TRcom-containing", "Null"), c("TRcom-less", "Null") )
ggboxplot(my_tmp[my_tmp$HL_label=="High",],x="TRcomflag2",y="TNDM",xlab="",ylab="DMnoise_ch")+
  stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
  scale_x_discrete(labels = c("TRcom-\ncontaining","TRcom-\nless","Null"))
ggscatter(my_tmp[my_tmp$HL_label=="High",],x="Count.allregs",y="TNDM",size=0.5,alpha=0.5,cor.coef = T)
Sys.time()
```
#
#
#
# 14. The Prtn dataset
Schmidt 2016 Nature biotechnology
https://www.nature.com/articles/nbt.3418#Sec24
Strain: BW25113
Supplementary Table S6: Final table with combined global absolute abundance estimations from both datasets including functional annotations using cluster of orthologues groups (COG) 
### 14.1. Load data and log2-transformation
```{r Prtn,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
Sch_prtn_meta<-as.data.frame(read_csv(paste(getwd(),"PRTN_DataSet","Table_S6_meta.csv",sep="/")))
Sch_prtn<-as.data.frame(read_xlsx(path = paste(getwd(),"PRTN_DataSet","41587_2016_BFnbt3418_MOESM18_ESM.xlsx",sep="/"),sheet = "Table S6",
                      skip=2)) # all data
Sch_prtn<-Sch_prtn[,Sch_prtn_meta[!is.na(Sch_prtn_meta$New_Col_name),]$Col_nb]
colnames(Sch_prtn)<-Sch_prtn_meta[!is.na(Sch_prtn_meta$New_Col_name),]$New_Col_name
Sch_prtn<-na.omit(Sch_prtn)
for(i in 1:c(ncol(Sch_prtn)-1)){
  Sch_prtn[Sch_prtn[,i]=="NA",i]<-NA
  Sch_prtn[,i]<-as.numeric(Sch_prtn[,i])
}
Sch_prtn<-na.omit(Sch_prtn)


# remove duplicated Bnbs
temp1<-data.frame(Bnb=unique(Sch_prtn$Bnb),Count=1)
for(i in 1:nrow(temp1)){
  temp1$Count[i]<-nrow(Sch_prtn[Sch_prtn$Bnb==temp1$Bnb[i],])
}
Sch_prtn<-Sch_prtn[Sch_prtn$Bnb %nin% temp1[temp1$Count>1,]$Bnb,]

# log2-transform
for(i in 1:c(ncol(Sch_prtn)-1)){
  Sch_prtn[,i]<-log(Sch_prtn[,i],2)
}
ggscatter(Sch_prtn,x="Glucose",y="LB",size=0.5,alpha=0.2)
Sys.time()
```
### 14.2. Basic statistics of the dataset
```{r Prtn,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
my_venn<-list(
  Three=dBnb.comp$Bnb,
  Prtn=Sch_prtn$Bnb
)

# ggVennDiagram(figtemp)
venn <- process_data(Venn(my_venn))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Three=length(venn$regionData$item[[which(venn$regionData$name=="Three")]]),
               Prtn=length(venn$regionData$item[[which(venn$regionData$name=="Prtn")]]),
               "Three&Prtn"=length(venn$regionData$item[[which(venn$regionData$name=="Three/Prtn")]])))
my_tmp<-v$colors
my_tmp[names(my_tmp)=="Three"][[1]]<-"gray50"
my_tmp[names(my_tmp)!="Three"][[1]]<-"yellow"
v$colors<-my_tmp
v$labels<-c("","")
par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = v$colors, quantities = TRUE) # Three(gray), 

Sys.time()
```
### 14.3. DM
#### 14.3.1. Calc DM
```{r Prtn,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=3,dpi=300}
temp3<-Sch_prtn[,which(colnames(Sch_prtn) %nin% "Bnb")]
temp5<-createEmptyDf(nrow=nrow(temp3),ncol = 3,colnames = qw(mean,var,sd))
temp5$mean<-rowMeans(as.matrix(temp3),na.rm=T)
temp5$var<-rowVars(as.matrix(temp3),na.rm=T)
temp5$sd<-rowSds(as.matrix(temp3),na.rm=T)
Sch_prtn<-cbind(Sch_prtn,temp5)

Sch_prtn<-Sch_prtn[order(Sch_prtn$mean),] # sorting 
Sch_prtn<-transform(Sch_prtn,RunmedVar=runmed(Sch_prtn$sd,41),lmg=Sch_prtn$mean)
lmgn<-data.frame(x=Sch_prtn$lmg,y=Sch_prtn$RunmedVar)
sp<-smooth.spline(lmgn,df=12)
x<-seq(min(lmgn$x),max(lmgn$x),length=1000)
pred<-as.data.frame(x)
pred<-transform(pred,smsp=predict(sp,x)[2])
colnames(pred)<-c("x","smoothsp")
pred2<-Sch_prtn$mean
pred2<-transform(pred2,y=predict(sp,pred2)[2])
colnames(pred2)<-c("x","smoothsp")
Sch_prtn<-transform(Sch_prtn,SmoothSpline=pred2$smoothsp)
Sch_prtn<-transform(Sch_prtn,DMprtn=Sch_prtn$sd-Sch_prtn$SmoothSpline) # calculate distance from smooth spline to SD 

df_prtn_sch<-left_join(df_DMall,Sch_prtn,by="Bnb")

## Running median
# mycolpal<-get_palette("lancet",2)
ggscatter(df_prtn_sch,x="mean",y="sd",
          cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
          cor.coeff.args = list(method = "spearman", label.x = 5, label.y=4, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"protein expression level [a.u.]"))+
  ylab(expression("Standard deviation [a.u.]"))+
  geom_line(data=df_prtn_sch,aes(x=mean,y=RunmedVar), colour="darkorange", size=0.5)+
  geom_line(data=pred,aes(x=x,y=smoothsp), colour="deepskyblue", size=0.5) # "deepskyblue"

ggscatter(df_prtn_sch,x="mean",y="DMprtn",
          cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
          cor.coeff.args = list(method = "spearman", label.x = 5, label.y=3, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"protein expression level [a.u.]"))+
  ylab(expression(DM[prtn]~"[a.u.]"))

Sys.time()
```
#### 14.3.2. Comprison of DMs
```{r Prtn,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
# DMenv vs. DMprtn
p1<-ggscatter(df_prtn_sch,x="DMenv",y="DMprtn",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=2,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.7, 3.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
    xlab(expression(DM[env]~"[a.u.]"))+
    ylab(expression(DM[prtn]~"[a.u.]"))
p1<-ggpar(p1,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMevo vs. DMprtn
p2<-ggscatter(df_prtn_sch,x="DMevo",y="DMprtn",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
    xlab(expression(DM[evo]~"[a.u.]"))+
    ylab(expression(DM[prtn]~"[a.u.]"))
p2<-ggpar(p2,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMmut vs. DMprtn
p3<-ggscatter(df_prtn_sch,x="DMmut",y="DMprtn",
          cor.coef = T,
          cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  # scale_x_continuous(breaks = get_breaks(by = 0.5, from = 0),limits =  c(-0.5, 2))+
    xlab(expression(DM[mut]~"[a.u.]"))+
    ylab(expression(DM[prtn]~"[a.u.]"))
p3<-ggpar(p3,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p1+p2+p3

Sys.time()
```
### 14.4. PCA
#### 14.4.1. Detect outliers in PC1~PC3
```{r Prtn,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-Sch_prtn
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% Sch_prtn_meta[(!is.na(Sch_prtn_meta$New_Col_name))&(Sch_prtn_meta$New_Col_name!="Bnb"),]$New_Col_name)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
pr_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    pr_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(pr_outlier[!is.na(pr_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
#### 14.4.2. Remove outliers if it is necessary and examine PCA
```{r Prtn,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-Sch_prtn
if(!is.na(pr_outlier)){
  temp0<-temp0[,colnames(temp0) %nin% pr_outlier]
}
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% Sch_prtn_meta[(!is.na(Sch_prtn_meta$New_Col_name))&(Sch_prtn_meta$New_Col_name!="Bnb"),]$New_Col_name)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
res.pca.global_prtn <- prcomp(figtemp, scale. = T)
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_prtn <- res.pca.global_prtn$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_prtn$label<-rownames(pc_scores_prtn)

pc_eigenvalues_prtn <- res.pca.global_prtn$sdev^2
pc_eigenvalues_prtn <- tibble(PC = factor(1:length(pc_eigenvalues_prtn)), 
                         variance = pc_eigenvalues_prtn) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained

# Scree plot
figtemp1<-pc_eigenvalues_prtn
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree_prtn<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,plot_type = "l",
        color="gray20",
       xlab="Components",ylab="Fraction of variance explained (%)")+
  # scale_x_continuous(breaks = get_breaks(by = 4, from = 0),limits = c(0,16))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_prtn$pct_cum[k] ,"% variance",sep=""))

Sys.time()
```
#### 14.4.3. 2D plot (PC1~PC3)
```{r Prtn,warning=FALSE, message=FALSE,fig.height=3,fig.width=12,dpi=300}
figtemp<-left_join(pc_scores_prtn,Sch_prtn_meta[!is.na(Sch_prtn_meta$New_Col_name),which(colnames(Sch_prtn_meta) %in% qw(New_Col_name,Media))],by=c("Condition"="New_Col_name"))
figtemp$Media<-factor(figtemp$Media,levels = unique(figtemp$Media))
p1<-ggscatter(data=figtemp,x="PC1",y="PC2",
alpha=0.5,
                 legend="",
                 color = "Media",
                 )+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,
      xlab = paste("PC1 (", round(pc_eigenvalues_prtn$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_prtn$pct[2],digits = 0),"%)",sep=""))

p2<-ggscatter(data=figtemp,x="PC2",y="PC3",
                 alpha=0.5,
                 legend="",
                 color = "Media",
                 )+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,
      xlab = paste("PC2 (", round(pc_eigenvalues_prtn$pct[2],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_prtn$pct[3],digits = 0),"%)",sep=""))

p3<-ggscatter(data=figtemp,x="PC1",y="PC3",
                 alpha=0.5,
                 legend="",
                 color = "Media",
                 )+
  # annotate("text",x=-Inf,y=Inf,label="PR2",hjust = -0.1,vjust = 1)+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,
      xlab = paste("PC1 (", round(pc_eigenvalues_prtn$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_prtn$pct[3],digits = 0),"%)",sep=""))

q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")
leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,1.2),nrow=1)
Sys.time()
```
#### 14.4.4. PC1~PC3 vs. Growth rate
```{r Prtn,warning=FALSE, message=FALSE,fig.height=3,fig.width=12,dpi=300}
df_prtn_growth<-as.data.frame(read_csv(paste(getwd(),"PRTN_DataSet","Table_S23_growth.csv",sep="/")))
figtemp<-left_join(pc_scores_prtn,Sch_prtn_meta[!is.na(Sch_prtn_meta$New_Col_name),which(colnames(Sch_prtn_meta) %in% qw(New_Col_name,Media))],by=c("Condition"="New_Col_name"))
figtemp<-left_join(figtemp,df_prtn_growth)
figtemp$Media<-factor(figtemp$Media,levels = unique(figtemp$Media))
p1<-ggscatter(data=figtemp,x="PC1",y="Growth_Rate",alpha=0.5,legend="",color = "Media",
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,xlab = "PC1",ylab = "Growth rate [1/h]")

p2<-ggscatter(data=figtemp,x="PC2",y="Growth_Rate",alpha=0.5,legend="",color = "Media",
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=10,label.y=1.5))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,xlab = "PC2",ylab = "Growth rate [1/h]")

p3<-ggscatter(data=figtemp,x="PC3",y="Growth_Rate",alpha=0.5,legend="",color = "Media",
  cor.coef = T,
  cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=-10,label.y=1.5))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,xlab = "PC3",ylab = "Growth rate [1/h]")

q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")
leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,1.2),nrow=1)
Sys.time()
```
#### 14.4.5. Correlation between rotation (eigenvector) for PC1~PC3
```{r Prtn,warning=FALSE, message=FALSE,fig.height=4.5,fig.width=11,dpi=300}
my_lim<-c(0,0.65)

# Env vs. Prtn
temp5<-data.frame(Env_PC=rep(seq(1,3),each=3),Prtn_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp5)){
  temp1<-as.data.frame(res.pca.global_env$rotation[,temp5$Env_PC[i]])
  colnames(temp1)<-"rot.env"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_prtn$rotation[,temp5$Prtn_PC[i]])
  colnames(temp2)<-"rot.prtn"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp5$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.prtn, method = "spearman")$estimate
  temp5$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.prtn, method = "spearman")$p.value
}

temp5.1<-matrix(rep(0,9),nrow=3)
rownames(temp5.1)<-mypcs
colnames(temp5.1)<-mypcs
temp5.2<-temp5.1
for(i in 1:nrow(temp5.1)){
  for(j in 1:ncol(temp5.1)){
    temp5.1[i,j]<-temp5[temp5$Env_PC==i&temp5$Prtn_PC==j,]$Cor
    temp5.2[i,j]<-temp5[temp5$Env_PC==i&temp5$Prtn_PC==j,]$p.value
  }
}
temp5.3<-abs(temp5.1)

# Evo vs. Prtn
temp6<-data.frame(Evo_PC=rep(seq(1,3),each=3),Prtn_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp6)){
  temp1<-as.data.frame(res.pca.global_evo$rotation[,temp6$Evo_PC[i]])
  colnames(temp1)<-"rot.evo"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_prtn$rotation[,temp6$Prtn_PC[i]])
  colnames(temp2)<-"rot.prtn"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp6$Cor[i]<-cor.test(temp4$rot.evo,temp4$rot.prtn, method = "spearman")$estimate
  temp6$p.value[i]<-cor.test(temp4$rot.evo,temp4$rot.prtn, method = "spearman")$p.value
}

temp6.1<-matrix(rep(0,9),nrow=3)
rownames(temp6.1)<-mypcs
colnames(temp6.1)<-mypcs
temp6.2<-temp6.1
for(i in 1:nrow(temp6.1)){
  for(j in 1:ncol(temp6.1)){
    temp6.1[i,j]<-temp6[temp6$Evo_PC==i&temp6$Prtn_PC==j,]$Cor
    temp6.2[i,j]<-temp6[temp6$Evo_PC==i&temp6$Prtn_PC==j,]$p.value
  }
}
temp6.3<-abs(temp6.1)

# Mut vs. Prtn
temp7<-data.frame(Mut_PC=rep(seq(1,3),each=3),Prtn_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp7)){
  temp1<-as.data.frame(res.pca.global_mut$rotation[,temp7$Mut_PC[i]])
  colnames(temp1)<-"rot.mut"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_prtn$rotation[,temp7$Prtn_PC[i]])
  colnames(temp2)<-"rot.prtn"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp7$Cor[i]<-cor.test(temp4$rot.mut,temp4$rot.prtn, method = "spearman")$estimate
  temp7$p.value[i]<-cor.test(temp4$rot.mut,temp4$rot.prtn, method = "spearman")$p.value
}

temp7.1<-matrix(rep(0,9),nrow=3)
rownames(temp7.1)<-mypcs
colnames(temp7.1)<-mypcs
temp7.2<-temp7.1
for(i in 1:nrow(temp7.1)){
  for(j in 1:ncol(temp7.1)){
    temp7.1[i,j]<-temp7[temp7$Mut_PC==i&temp7$Prtn_PC==j,]$Cor
    temp7.2[i,j]<-temp7[temp7$Mut_PC==i&temp7$Prtn_PC==j,]$p.value
  }
}
temp7.3<-abs(temp7.1)

my_lab_fontsize<-6

p4<-ggcorrplot::ggcorrplot(temp5.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Env", y = "Prtn")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p4<-ggpar(p4,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p5<-ggcorrplot::ggcorrplot(temp6.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Evo", y = "Prtn")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p5<-ggpar(p5,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p6<-ggcorrplot::ggcorrplot(temp7.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Prtn")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p6<-ggpar(p6,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

my_fontsize<-16
q2<-ggpar(p6,legend = "right",legend.title="",font.legend = my_fontsize)+
  scale_fill_continuous(limits = my_lim, breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "Abs. of Spearman's R",title.position = "left",ticks.linewidth = 1))+
  theme(legend.title = element_text(angle = 90, hjust = 10, vjust = 0.5,size=18),legend.title.align=10)
leg2 <- ggpubr::get_legend(q2)

my_tmp<-data.frame(x=seq(min(my_lim),max(my_lim),by=0.01),y=seq(min(my_lim),max(my_lim),by=0.01),z=seq(min(my_lim),max(my_lim),by=0.01))
p_led<-ggscatter(my_tmp,x="x",y="y",fill="z",legend.title = "|R|",legend="right")+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
q2<-ggpar(p_led,font.legend = my_fontsize)
leg2 <- ggpubr::get_legend(q2)

grid.arrange(p4,p5,p6,as_ggplot(leg2),widths=c(1,1,1,0.4),nrow=1)
Sys.time()
```
#
#
#
# 15. The Strs dataset
Maeda & Iwasawa 2020 Nature communications
Transcriptome data of mutants under a non-stressful condition.
https://www.nature.com/articles/s41467-020-19713-w
## 15.1. Load data
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
genename.mds42<-as.data.frame(read_csv(dgenename.mds42))
genename.mds42<-genename.mds42[,c(2:ncol(genename.mds42))]
colnames(genename.mds42)[colnames(genename.mds42)=="locus_tag"]<-"Bnb"
colnames(genename.mds42)[colnames(genename.mds42)=="GeneID"]<-"EntrezID"

MaeIwadat<-read_csv(paste(getwd(),"STRS_DataSet","MIdat2020supl4.csv",sep="/"))
my_col<-colnames(MaeIwadat)
my_col<-str_replace_all(string = my_col,pattern = "-",replacement = "_")
colnames(MaeIwadat)<-my_col
# MaeIwadat<-as.data.frame(read.csv(paste(getwd(),"STRS_DataSet","MIdat2020supl4.csv",sep="/")))
# temp<-read_csv(paste(getwd(),"STRS_DataSet","MIdat2020supl4_colnames.csv",sep="/"))
# temp<-as.data.frame(read.csv(paste(getwd(),"STRS_DataSet","MIdat2020supl4_colnames.csv",sep="/")))
# colnames(MaeIwadat)<-temp$colnames
colnames(MaeIwadat)[colnames(MaeIwadat)=="gene"]<-"ProbeName"
df_maeiwa_condition<-data.frame(Condition=colnames(MaeIwadat)[colnames(MaeIwadat) %nin% qw(ProbeName)])
df_maeiwa_condition$STRS<-"No"
for(i in 1:nrow(df_maeiwa_condition)){
  if(df_maeiwa_condition$Condition[i] %nin% qw(MDS42_n1,MDS42_n2,MDS42_n3,MDS42_n4)){
    df_maeiwa_condition$STRS[i]<-str_sub(df_maeiwa_condition$Condition[i], start=1, end=-3)
  }
  if(str_count(df_maeiwa_condition$STRS[i],"^X")==1){
    df_maeiwa_condition$STRS[i]<-str_sub(df_maeiwa_condition$STRS[i], start=2, end=-1)
  }
}


# Remove the same conditions used in the Env dataset. 
my_abx<-qw(MER,H2O2,TP,CPFX)
MaeIwadat<-MaeIwadat[,which(colnames(MaeIwadat) %nin% df_maeiwa_condition[df_maeiwa_condition$STRS %in% my_abx,]$Condition)]
df_maeiwa_condition<-df_maeiwa_condition[df_maeiwa_condition$STRS %nin% my_abx,]

# log2-transform
for(i in 1:nrow(df_maeiwa_condition)){
  MaeIwadat[,which(colnames(MaeIwadat)==df_maeiwa_condition$Condition[i])]<-log(10,2)*MaeIwadat[,which(colnames(MaeIwadat)==df_maeiwa_condition$Condition[i])]
}

temp<-MaeIwadat
temp<-left_join(temp,probe2id[(probe2id$IStag==F)&(probe2id$Count==1)&(!grepl("\\+",probe2id$ECKnb)),which(colnames(probe2id) %in% qw(ProbeName,ECKnb))],by="ProbeName")
temp<-left_join(temp,genename.mds42[,which(colnames(genename.mds42) %in% qw(Bnb,ECKnb))],by="ECKnb")
# omit #Bnb=NA
temp<-temp[!is.na(temp$Bnb),]

# remove duplicated Bnbs
temp1<-data.frame(Bnb=unique(temp$Bnb),Count=1)
for(i in 1:nrow(temp1)){
  temp1$Count[i]<-nrow(temp[temp$Bnb==temp1$Bnb[i],])
}
temp1<-temp[temp$Bnb %in% temp1[temp1$Count>1,]$Bnb,]
temp<-temp[temp$ProbeName %nin% temp1[!grepl("[A-Z]$",temp1$ProbeName),]$ProbeName,]
MaeIwadat<-temp[,which(colnames(temp) %in% c("Bnb",df_maeiwa_condition$Condition))]
MaeIwadat<-MaeIwadat %>% select(Bnb,everything())

# ggscatter(MaeIwadat,x="B_Cl_AlaE1",y="B_Cl_AlaE4",size=0.5,alpha=0.2)
ggscatter(MaeIwadat,x="MDS42_n1",y="MDS42_n2",size=0.5,alpha=0.2)
Sys.time()
```
## 15.2. Basic statistics of the dataset
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
my_venn<-list(
  Three=dBnb.comp$Bnb,
  Strs=MaeIwadat$Bnb
)

# ggVennDiagram(figtemp)
venn <- process_data(Venn(my_venn))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Three=length(venn$regionData$item[[which(venn$regionData$name=="Three")]]),
               Strs=length(venn$regionData$item[[which(venn$regionData$name=="Strs")]]),
               "Three&Strs"=length(venn$regionData$item[[which(venn$regionData$name=="Three/Strs")]])))
my_tmp<-v$colors
my_tmp[names(my_tmp)=="Three"][[1]]<-"gray50"
my_tmp[names(my_tmp)!="Three"][[1]]<-"yellow"
v$colors<-my_tmp
v$labels<-c("","")
par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = v$colors, quantities = TRUE) # Three(gray), 

Sys.time()
```
## 15.3. DM 
### 15.3.1. Calc DM
```{r Strs,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=3,dpi=300}
temp3<-MaeIwadat[,which(colnames(MaeIwadat) %nin% "Bnb")]
temp5<-createEmptyDf(nrow=nrow(temp3),ncol = 3,colnames = qw(mean,var,sd))
temp5$mean<-rowMeans(as.matrix(temp3),na.rm=T)
temp5$var<-rowVars(as.matrix(temp3),na.rm=T)
temp5$sd<-rowSds(as.matrix(temp3),na.rm=T)
MaeIwadat<-cbind(MaeIwadat,temp5)

MaeIwadat<-MaeIwadat[order(MaeIwadat$mean),] # sorting 
MaeIwadat$RunmedVar<-runmed(MaeIwadat$sd,41)
MaeIwadat$lmg=MaeIwadat$mean
# MaeIwadat<-transform(MaeIwadat,RunmedVar=runmed(MaeIwadat$sd,41),lmg=MaeIwadat$mean)
lmgn<-data.frame(x=MaeIwadat$lmg,y=MaeIwadat$RunmedVar)
sp<-smooth.spline(lmgn,df=12)
x<-seq(min(lmgn$x),max(lmgn$x),length=1000)
pred<-as.data.frame(x)
pred<-transform(pred,smsp=predict(sp,x)[2])
colnames(pred)<-c("x","smoothsp")
pred2<-MaeIwadat$mean
pred2<-transform(pred2,y=predict(sp,pred2)[2])
colnames(pred2)<-c("x","smoothsp")

MaeIwadat$SmoothSpline<-pred2$smoothsp
MaeIwadat$DMstrs<-MaeIwadat$sd-MaeIwadat$SmoothSpline
# MaeIwadat<-transform(MaeIwadat,SmoothSpline=pred2$smoothsp)
# MaeIwadat<-transform(MaeIwadat,DMstrs=MaeIwadat$sd-MaeIwadat$SmoothSpline) # calculate distance from smooth spline to SD 

df_strs_maeiwa<-left_join(df_DMall,MaeIwadat,by="Bnb")

## Running median
# mycolpal<-get_palette("lancet",2)
ggscatter(df_strs_maeiwa,x="mean",y="sd",
          cor.coef = TRUE,cor.coeff.args = list(method = "spearman", label.x = 12, label.y=1.7, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression("Standard deviation [a.u.]"))+
  geom_line(data=df_strs_maeiwa,aes(x=mean,y=RunmedVar), colour="darkorange", size=0.5)+
  geom_line(data=pred,aes(x=x,y=smoothsp), colour="deepskyblue", size=0.5) # "deepskyblue"

ggscatter(df_strs_maeiwa,x="mean",y="DMstrs",
          cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
          cor.coeff.args = list(method = "spearman", label.x = 12, label.y=1.3, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression(DM[strs]~"[a.u.]"))


Sys.time()
```
### 15.3.2. Comprison of DMs
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
# DMenv vs. DMstrs
p1<-ggscatter(df_strs_maeiwa,x="DMenv",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=2,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.7, 3.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.2, 2.1))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[strs]~"[a.u.]"))
p1<-ggpar(p1,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMevo vs. DMstrs
p2<-ggscatter(df_strs_maeiwa,x="DMevo",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.2, 2.1))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[strs]~"[a.u.]"))
p2<-ggpar(p2,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMmut vs. DMstrs
p3<-ggscatter(df_strs_maeiwa,x="DMmut",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.2, 2.1))+
  # scale_x_continuous(breaks = get_breaks(by = 0.5, from = 0),limits =  c(-0.5, 2))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[strs]~"[a.u.]"))
p3<-ggpar(p3,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)

p1+p2+p3

# DMenv_wo_abx vs. DMstrs
p4<-ggscatter(df_strs_maeiwa,x="DMenv_wo_abx",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=2,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.7, 3.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.2, 2.1))+
  xlab(expression(DM[env_wo_abx]~"[a.u.]"))+
  ylab(expression(DM[strs]~"[a.u.]"))
p4<-ggpar(p4,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)
p4+blank+blank
Sys.time()
```
## 15.4. PCA
### 15.4.1. Detect outliers in PC1~PC3
```{r Strs,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-MaeIwadat
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% df_maeiwa_condition$Condition)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
mi_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    mi_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(mi_outlier[!is.na(mi_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
### 15.4.2. Remove outliers if it is necessary and examine PCA
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-MaeIwadat
if(!is.na(mi_outlier)){
  temp0<-temp0[,colnames(temp0) %nin% mi_outlier]
}
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% df_maeiwa_condition$Condition)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
res.pca.global_strs_mi <- prcomp(figtemp, scale. = T)
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_strs_mi <- res.pca.global_strs_mi$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_strs_mi$label<-rownames(pc_scores_strs_mi)

pc_eigenvalues_strs_mi <- res.pca.global_strs_mi$sdev^2
pc_eigenvalues_strs_mi <- tibble(PC = factor(1:length(pc_eigenvalues_strs_mi)), 
                         variance = pc_eigenvalues_strs_mi) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained

# Scree plot
figtemp1<-pc_eigenvalues_strs_mi
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree_strs_mi_maeiwa<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,plot_type = "l",
        color="gray20",
       xlab="Components",ylab="Fraction of variance explained (%)")+
  # scale_x_continuous(breaks = get_breaks(by = 4, from = 0),limits = c(0,16))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_strs_mi$pct_cum[k] ,"% variance",sep=""))

Sys.time()
```
### 15.4.3. 2D plot (PC1~PC3)
```{r Strs,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
figtemp<-left_join(pc_scores_strs_mi,df_maeiwa_condition)
figtemp[figtemp$STRS=="No",]$STRS<-"Ancestor"
figtemp$STRS<-factor(figtemp$STRS,levels = unique(figtemp$STRS))
p1<-ggscatter(data=figtemp,x="PC1",y="PC2",alpha=0.5,legend="",color = "STRS")+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,
      xlab = paste("PC1 (", round(pc_eigenvalues_strs_mi$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_strs_mi$pct[2],digits = 0),"%)",sep=""))

p2<-ggscatter(data=figtemp,x="PC2",y="PC3",alpha=0.5,legend="",color = "STRS")+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,
      xlab = paste("PC2 (", round(pc_eigenvalues_strs_mi$pct[2],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_strs_mi$pct[3],digits = 0),"%)",sep=""))

p3<-ggscatter(data=figtemp,x="PC1",y="PC3",alpha=0.5,legend="",color = "STRS")+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,
      xlab = paste("PC1 (", round(pc_eigenvalues_strs_mi$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_strs_mi$pct[3],digits = 0),"%)",sep=""))

q<-ggpar(p1,legend = "right",font.ytickslab= c(12),legend.title = "")
leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,1),nrow=1)
Sys.time()
```
### 15.4.3.2. 3D plot (PC1~PC3)
```{r Strs,warning=FALSE, message=FALSE,fig.height=4,fig.width=4,dpi=300}
my_temp<-pc_scores_strs_mi %>% select(Condition,PC1,PC2,PC3)
figtemp<-left_join(my_temp,df_maeiwa_condition)
figtemp[figtemp$STRS=="No",]$STRS<-"Ancestor"
figtemp$STRS<-factor(figtemp$STRS,levels = unique(figtemp$STRS))
my_temp<-data.frame(STRS=levels(figtemp$STRS),label=seq(1,length(unique(figtemp$STRS))))
figtemp<-left_join(figtemp,my_temp)


par(mar=c(2, 0.1, 0.1, 0.1))
# scatter3D(figtemp$PC1, figtemp$PC2, figtemp$PC3,colkey = FALSE,col = gg_color_hue(length(unique(figtemp$STRS))),phi = 0,alpha=0.3,
#           xlab="PC1 (17%)", ylab="PC2 (6%)", zlab="PC3 (5%)",pch = 20, cex = 2,colvar = figtemp$label,cex.lab = 1)
Sys.time()
```
### 15.4.4. Correlation between rotation (eigenvector) for PC1~PC3
```{r Strs,warning=FALSE, message=FALSE,fig.height=4.5,fig.width=11,dpi=300}
my_lim<-c(0,0.65)

# Env vs. STRS
temp5<-data.frame(Env_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp5)){
  temp1<-as.data.frame(res.pca.global_env$rotation[,temp5$Env_PC[i]])
  colnames(temp1)<-"rot.env"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_mi$rotation[,temp5$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp5$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.strs, method = "spearman")$estimate
  temp5$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.strs, method = "spearman")$p.value
}

temp5.1<-matrix(rep(0,9),nrow=3)
rownames(temp5.1)<-mypcs
colnames(temp5.1)<-mypcs
temp5.2<-temp5.1
for(i in 1:nrow(temp5.1)){
  for(j in 1:ncol(temp5.1)){
    temp5.1[i,j]<-temp5[temp5$Env_PC==i&temp5$STRS_PC==j,]$Cor
    temp5.2[i,j]<-temp5[temp5$Env_PC==i&temp5$STRS_PC==j,]$p.value
  }
}
temp5.3<-abs(temp5.1)


# Evo vs. STRS
temp6<-data.frame(Evo_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp6)){
  temp1<-as.data.frame(res.pca.global_evo$rotation[,temp6$Evo_PC[i]])
  colnames(temp1)<-"rot.evo"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_mi$rotation[,temp6$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp6$Cor[i]<-cor.test(temp4$rot.evo,temp4$rot.strs, method = "spearman")$estimate
  temp6$p.value[i]<-cor.test(temp4$rot.evo,temp4$rot.strs, method = "spearman")$p.value
}

temp6.1<-matrix(rep(0,9),nrow=3)
rownames(temp6.1)<-mypcs
colnames(temp6.1)<-mypcs
temp6.2<-temp6.1
for(i in 1:nrow(temp6.1)){
  for(j in 1:ncol(temp6.1)){
    temp6.1[i,j]<-temp6[temp6$Evo_PC==i&temp6$STRS_PC==j,]$Cor
    temp6.2[i,j]<-temp6[temp6$Evo_PC==i&temp6$STRS_PC==j,]$p.value
  }
}
temp6.3<-abs(temp6.1)

# Mut vs. STRS
temp7<-data.frame(Mut_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp7)){
  temp1<-as.data.frame(res.pca.global_mut$rotation[,temp7$Mut_PC[i]])
  colnames(temp1)<-"rot.mut"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_mi$rotation[,temp7$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp7$Cor[i]<-cor.test(temp4$rot.mut,temp4$rot.strs, method = "spearman")$estimate
  temp7$p.value[i]<-cor.test(temp4$rot.mut,temp4$rot.strs, method = "spearman")$p.value
}

temp7.1<-matrix(rep(0,9),nrow=3)
rownames(temp7.1)<-mypcs
colnames(temp7.1)<-mypcs
temp7.2<-temp7.1
for(i in 1:nrow(temp7.1)){
  for(j in 1:ncol(temp7.1)){
    temp7.1[i,j]<-temp7[temp7$Mut_PC==i&temp7$STRS_PC==j,]$Cor
    temp7.2[i,j]<-temp7[temp7$Mut_PC==i&temp7$STRS_PC==j,]$p.value
  }
}
temp7.3<-abs(temp7.1)

my_lab_fontsize<-6

p4<-ggcorrplot::ggcorrplot(temp5.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Env", y = "Strs")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p4<-ggpar(p4,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p5<-ggcorrplot::ggcorrplot(temp6.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Evo", y = "Strs")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p5<-ggpar(p5,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p6<-ggcorrplot::ggcorrplot(temp7.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Strs")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p6<-ggpar(p6,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

my_fontsize<-16
q2<-ggpar(p6,legend = "right",legend.title="",font.legend = my_fontsize)+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.1),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "Abs. of Spearman's R",title.position = "left",ticks.linewidth = 1))+
  theme(legend.title = element_text(angle = 90, hjust = 10, vjust = 0.5,size=18),legend.title.align=10)
leg2 <- ggpubr::get_legend(q2)

my_tmp<-data.frame(x=seq(min(my_lim),max(my_lim),by=0.01),y=seq(min(my_lim),max(my_lim),by=0.01),z=seq(min(my_lim),max(my_lim),by=0.01))
p_led<-ggscatter(my_tmp,x="x",y="y",fill="z",legend.title = "|R|",legend="right")+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
q2<-ggpar(p_led,font.legend = my_fontsize)
leg2 <- ggpubr::get_legend(q2)

grid.arrange(p4,p5,p6,as_ggplot(leg2),widths=c(1,1,1,0.4),nrow=1)
Sys.time()
```
## 15.5. Analysis of commonly muated transcriptional regulators
### 15.5.1. define mutagenic conditions
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=12,dpi=300}
temp1<-as.data.frame(read_xlsx(paste(getwd(),"STRS_DataSet","41467_2020_19713_MOESM6_ESM.xlsx",sep="/"),sheet = "All mutations",skip = 1,range = "A2:D1450"))
colnames(temp1)[1]<-"Chemicals"
for(i in 1:nrow(temp1)){
  if(is.na(temp1$Chemicals[i])){
    temp1$Chemicals[i]<-temp1$Chemicals[i-1]
  }
  if(is.na(temp1$strain_name[i])){
    temp1$strain_name[i]<-temp1$strain_name[i-1]
  }
}
temp1$Stressor<-str_remove_all(string = temp1$strain_name,pattern = "E[1-9]$")
my_df<-tibble(strain_name=unique(temp1$strain_name),NbMut=0)
my_df$Stressor<-str_remove_all(string = my_df$strain_name,pattern = "E[1-9]$")
for(i in 1:nrow(my_df)){
  my_tmp<-temp1[temp1$strain_name==my_df$strain_name[i],]
  my_df$NbMut[i]<-nrow(my_tmp)
}
ggdotplot(my_df,x="Stressor",y="NbMut",color = "strain_name",legend="")+
  yscale("log10", .format = TRUE)+
  rotate_x_text()+
  geom_hline(yintercept = 10,color="red")

mutagenic_strs<-c("GAH","MMC","NQO","PLM")

```
### 15.5.2. Mutated regulatory genes identified
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=12,dpi=300}
temp1<-as.data.frame(read_xlsx(paste(getwd(),"STRS_DataSet","41467_2020_19713_MOESM6_ESM.xlsx",sep="/"),sheet = "Common mutations and primers",skip = 1,range = "A2:E135"))
colnames(temp1)[1]<-"Stressor"
for(i in 1:nrow(temp1)){
  if(is.na(temp1$Stressor[i])){
    temp1$Stressor[i]<-temp1$Stressor[i-1]
    temp1$Chemicals[i]<-temp1$Chemicals[i-1]
  }
  # if(is.na(temp1$strain_name[i])){
  #   temp1$strain_name[i]<-temp1$strain_name[i-1]
  # }
}

temp1<-temp1[temp1$Stressor %in% df_maeiwa_condition$STRS,]
temp1<-temp1[!is.na(temp1$`Evolved strain and mutation`),]
temp1<-left_join(temp1,probe2id[,which(colnames(probe2id) %in% qw(ProbeName,Bnb))],by=c("gene"="ProbeName"))
temp1<-temp1[!grepl(pattern = "intergenic",x = temp1$gene),]
temp1[(temp1$Stressor=="FTD")&(temp1$gene=="ompR/envZ"),]$gene<-"ompR"
temp1[(temp1$Stressor=="FTD")&(temp1$gene=="ompR"),]$Bnb<-"b3405"
temp1<-temp1[!is.na(temp1$Bnb),]
temp1[(temp1$gene=="ompR"),]$`Evolved strain and mutation`<-"FTDE1 (ompR, L228M (CTG→ATG)), FTDE3 (ompR, E193K (GAA→AAA))"
my_tr<-unique(TRN$regulator_id)
temp1$TR_flag<-"Non-TR"
for(i in 1:nrow(temp1)){
  if(length(grep(temp1$Bnb[i],my_tr))>0){
    temp1$TR_flag[i]<-"TR"
  }
}
temp1<-temp1[temp1$Stressor %nin% mutagenic_strs,]
temp1[temp1$gene=="ybjK",]$gene<-"rcdA"

my_DMs<-df_DMall[,which(colnames(df_DMall) %in% qw(gene,Bnb,DMenv,DMevo,DMmut,DMenv_wo_abx))]
my_DMs<-full_join(my_DMs,MaeIwadat[,which(colnames(MaeIwadat) %in% qw(Bnb,DMstrs))])

df_mt<-temp1[temp1$TR_flag=="TR",]
df_mt$regulator<-NA
df_mt$targets<-NA
df_mt$DMenv<-NA
df_mt$DMevo<-NA
df_mt$DMmut<-NA
df_mt$DMstrs<-NA
df_mt$DMenv_wo_abx<-NA

for(i in 1:nrow(df_mt)){
  temp3<-TRN[str_detect(string = TRN$regulator_id,pattern = df_mt$Bnb[i]),]
  df_mt$regulator[i]<-paste(unique(temp3$regulator),collapse = ",")
  df_mt$targets[i]<-paste(unique(temp3$gene_name),collapse = ",")
  df_mt$DMenv[i]<-mean(my_DMs[my_DMs$Bnb %in% temp3$gene_id,]$DMenv)
  df_mt$DMevo[i]<-mean(my_DMs[my_DMs$Bnb %in% temp3$gene_id,]$DMevo)
  df_mt$DMmut[i]<-mean(my_DMs[my_DMs$Bnb %in% temp3$gene_id,]$DMmut)
  df_mt$DMstrs[i]<-mean(my_DMs[my_DMs$Bnb %in% temp3$gene_id,]$DMstrs)
  df_mt$DMenv_wo_abx[i]<-mean(my_DMs[my_DMs$Bnb %in% temp3$gene_id,]$DMenv_wo_abx)
}

df_mt_unique<-df_mt %>% distinct(Bnb,.keep_all = TRUE)
df_mt_unique <-df_mt_unique %>% select(gene,Bnb,regulator,targets,everything())
df_mt_unique <-df_mt_unique[,which(colnames(df_mt_unique) %in% qw(gene,Bnb,regulator,targets))]
df_mt_unique$Stressors<-NA
for(i in 1:nrow(df_mt_unique)){
  df_mt_unique$Stressors[i]<-paste(unique(df_mt[df_mt$Bnb == df_mt_unique$Bnb[i],]$Stressor),collapse = ", ")
}

my_df<-dBnb %>% select(Bnb,gene)
my_tmp<-MaeIwadat %>% select(Bnb,DMstrs)
my_df<-left_join(my_df,my_tmp)
colnames(my_df)[colnames(my_df)=="DMstrs"]<-"DMstrs_me"

my_df$TR_flag<-"Non-TR"
my_tr<-unique(unlist(str_split(TRN$regulator_id,",")))
my_tr<-my_tr[order(my_tr)]
my_df[my_df$Bnb %in% my_tr,]$TR_flag<-"TR"

my_df$TG_flag<-"Unregulated"
my_tg<-unique(TRN$gene_id)
my_tg<-my_tg[order(my_tg)]
my_df[my_df$Bnb %in% my_tg,]$TG_flag<-"Regulated"

my_df$MutTR_flag<-"Others"
my_df[my_df$Bnb %in% df_mt_unique$Bnb,]$MutTR_flag<-"Mutated TRs"

my_df$TGofMutTR_flag<-"Others"
my_tr<-unique(unlist(str_split(df_mt_unique$regulator,",")))
my_tg<-unique(TRN[TRN$regulator %in% my_tr,]$gene_id)
my_df[my_df$Bnb %in% my_tg,]$TGofMutTR_flag<-"TGs of mutTRs"
my_df$TGofMutTR_flag<-factor(my_df$TGofMutTR_flag,levels=c("Others","TGs of mutTRs"))

my_tr<-df_mt_unique$Bnb
my_tg<-unique(TRN[TRN$regulator_id %in% my_tr,]$gene_id)
# length(my_tg)
my_df$TGofCommonMutTR_flag<-"Others"
my_df[my_df$Bnb %in% my_tg,]$TGofCommonMutTR_flag<-"TGs of common mutTRs"
my_df$TGofCommonMutTR_flag<-factor(my_df$TGofCommonMutTR_flag,levels=c("Others","TGs of common mutTRs"))
# nrow(my_df[my_df$TGofCommonMutTR_flag=="TGs of common mutTRs",])

my_tmp<-df_DMall %>% select(Bnb,gene,DMenv,DMevo,DMmut,DMenv_wo_abx)
df_maeiwa_tg<-left_join(my_df,my_tmp)
df_maeiwa_tg<-df_maeiwa_tg %>% select(Bnb,gene,DMenv,DMevo,DMmut,DMenv_wo_abx,DMstrs_me,everything())
# nrow(df_maeiwa_tg[df_maeiwa_tg$TGofCommonMutTR_flag=="TGs of common mutTRs",])

Sys.time()
```
### 15.5.3. Check IC50s
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=12,dpi=300}
df_maeiwa_chem<-as.data.frame(read_xlsx(paste(getwd(),"STRS_DataSet","41467_2020_19713_MOESM4_ESM.xlsx",sep="/"),skip = 1))
my_tmp<-unique(unlist(str_split(paste(df_mt_unique$Stressors,collapse = ", "),pattern = ", ")))
my_df<-data.frame(Stressors=my_tmp,Stressor=NA)
for(i in 1:nrow(my_df)){
  my_df$Stressor[i]<-df_maeiwa_chem[grepl(pattern = my_df$Stressors[i],x = df_maeiwa_chem$Abbreviation,ignore.case = T),]$Abbreviation
}

my_tmp<-as.data.frame(read_xlsx(paste(getwd(),"STRS_DataSet","41467_2020_19713_MOESM5_ESM.xlsx",sep="/"),skip = 1,sheet = "IC50s of reconstructed mutants"))
my_tmp[my_tmp$Strains=="ybjK",]$Strains<-"rcdA"
my_tmp[my_tmp$Strains=="ycfQ",]$Strains<-"comR"
my_tmp<-my_tmp[my_tmp$Strains %in% df_mt_unique$gene,]
my_mut<-my_tmp$Strains

my_tmp2<-as.data.frame(read_xlsx(paste(getwd(),"STRS_DataSet","41467_2020_19713_MOESM5_ESM.xlsx",sep="/"),skip = 1,sheet = "IC50s of evolved strains"))
colnames(my_tmp2)[1]<-"Strains"
my_co<-c("M9E1","M9E2","M9E3","M9E4")
my_tmp2<-my_tmp2[my_tmp2$Strains %in% my_co,]
my_co<-my_tmp2[1,]
my_co$Strains<-"control"
for(i in 2:ncol(my_tmp2)){
  my_co[,i]<-mean(my_tmp2[,i])
}

my_tmp<-bind_rows(my_tmp,my_co)

i<-1
for(i in 1:nrow(df_mt_unique)){
  my_list<-unlist(str_split(df_mt_unique$Stressors[i],pattern = ", "))
  my_list<-my_df[my_df$Stressors %in% my_list,]$Stressor
  my_tmp3<-my_tmp %>%select(Strains,my_list)
  my_tmp3<-my_tmp3[my_tmp3$Strains %in% c(df_mt_unique$gene[i],"control"),]
  my_tmp3$Strains<-factor(my_tmp3$Strains,levels = c(df_mt_unique$gene[i],"control"))
  my_tmp3<-my_tmp3 %>% pivot_longer(cols = colnames(my_tmp3)[2:ncol(my_tmp3)],names_to = "Stressor",values_to = "log2 IC50")
  p<-ggbarplot(my_tmp3,x = "Stressor",y="log2 IC50",fill="Strains",position = position_dodge(0.5),width = 0.5)
  print(p)
}

Sys.time()
```
### 15.5.4. Simmilarity between genes with high DMs and the target genes of commonly mutated regulators in adaptive evolution toward resistance to stressors
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=15,dpi=300}
figtemp<-df_maeiwa_tg[df_maeiwa_tg$TG_flag=="Regulated",]

figtemp1<-figtemp[!is.na(figtemp$DMenv),]

p5<-ggviolin(figtemp1,x="TGofCommonMutTR_flag",y="DMenv",fill="TGofCommonMutTR_flag",legend="",xlab="",add="boxplot",palette = c("#00AFBB", "#FC4E07"), add.params = list(fill = "white"))+
  scale_y_continuous(limits = c(-1,4))+
  ylab(expression(DM[env]~"[a.u.]"))+
  annotate("text",x=1,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMenv))&(figtemp1$TGofCommonMutTR_flag=="Others"),]),hjust = 0.5,vjust = 4)+
  annotate("text",x=2,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMenv))&(figtemp1$TGofCommonMutTR_flag=="TGs of common mutTRs"),]),hjust = 0.5,vjust = 4)+
  scale_x_discrete(labels = c("Targets of\nother TRs","Targets of\ncommonly\nmuated TRs"))+
  stat_compare_means(size=5,label.y = 3.7)
p5<-ggpar(p5,font.y=c(14))

p6<-ggviolin(figtemp1,x="TGofCommonMutTR_flag",y="DMevo",fill="TGofCommonMutTR_flag",legend="",xlab="",add="boxplot",palette = c("#00AFBB", "#FC4E07"), add.params = list(fill = "white"))+
  scale_y_continuous(limits = c(-1,4))+
  ylab(expression(DM[evo]~"[a.u.]"))+
  annotate("text",x=1,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMevo))&(figtemp1$TGofCommonMutTR_flag=="Others"),]),hjust = 0.5,vjust = 4)+
  annotate("text",x=2,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMevo))&(figtemp1$TGofCommonMutTR_flag=="TGs of common mutTRs"),]),hjust = 0.5,vjust = 4)+
  scale_x_discrete(labels = c("Targets of\nother TRs","Targets of\ncommonly\nmuated TRs"))+
  stat_compare_means(size=5,label.y = 3.7)
p6<-ggpar(p6,font.y=c(14))

p7<-ggviolin(figtemp1,x="TGofCommonMutTR_flag",y="DMmut",fill="TGofCommonMutTR_flag",legend="",xlab="",add="boxplot",palette = c("#00AFBB", "#FC4E07"), add.params = list(fill = "white"))+
  scale_y_continuous(limits = c(-1,5))+
  ylab(expression(DM[mut]~"[a.u.]"))+
  annotate("text",x=1,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMmut))&(figtemp1$TGofCommonMutTR_flag=="Others"),]),hjust = 0.5,vjust = 4)+
  annotate("text",x=2,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMmut))&(figtemp1$TGofCommonMutTR_flag=="TGs of common mutTRs"),]),hjust = 0.5,vjust = 4)+
  scale_x_discrete(labels = c("Targets of\nother TRs","Targets of\ncommonly\nmuated TRs"))+
  stat_compare_means(size=5,label.y = 4.7)
p7<-ggpar(p7,font.y=c(14))

p9<-ggviolin(figtemp1,x="TGofCommonMutTR_flag",y="DMenv_wo_abx",fill="TGofCommonMutTR_flag",legend="",xlab="",add="boxplot",palette = c("#00AFBB", "#FC4E07"), add.params = list(fill = "white"))+
  scale_y_continuous(limits = c(-1,4.5))+
  ylab(expression(DM[env_wo_abx]~"[a.u.]"))+
  annotate("text",x=1,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMenv))&(figtemp1$TGofCommonMutTR_flag=="Others"),]),hjust = 0.5,vjust = 4)+
  annotate("text",x=2,y=Inf,label= nrow(figtemp1[(!is.na(figtemp1$DMenv))&(figtemp1$TGofCommonMutTR_flag=="TGs of common mutTRs"),]),hjust = 0.5,vjust = 4)+
  scale_x_discrete(labels = c("Targets of\nother TRs","Targets of\ncommonly\nmuated TRs"))+
  stat_compare_means(size=5,label.y = 4)
p9<-ggpar(p9,font.y=c(14))

figtemp2<-figtemp[!is.na(figtemp$DMstrs_me),]

p8<-ggviolin(figtemp2,x="TGofCommonMutTR_flag",y="DMstrs_me",fill="TGofCommonMutTR_flag",legend="",xlab="",add="boxplot",palette = c("#00AFBB", "#FC4E07"), add.params = list(fill = "white"))+
  scale_y_continuous(limits = c(-0.5,2.7))+
  ylab(expression(DM[strs]~"[a.u.]"))+
  annotate("text",x=1,y=Inf,label= nrow(figtemp2[figtemp2$TGofCommonMutTR_flag=="Others",]),hjust = 0.5,vjust = 4)+
  annotate("text",x=2,y=Inf,label= nrow(figtemp2[figtemp2$TGofCommonMutTR_flag=="TGs of common mutTRs",]),hjust = 0.5,vjust = 4)+
  scale_x_discrete(labels = c("Targets of\nother TRs","Targets of\ncommonly\nmuated TRs"))+
  stat_compare_means(size=5,label.y = 2.5)
p8<-ggpar(p8,font.y=c(14))

p5+p9+p6+p7+p8+
  plot_layout(nrow = 1)

Sys.time()
```
### 15.5.5. Expression levels of target genes of the commonly mutated regulators
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=6,dpi=300}
my_smpl<-list()
for (i in 1:nrow(df_mt)){
  # df_mt$`Evolved strain and mutation`[i]
  my_tmp<-str_split(string = df_mt$`Evolved strain and mutation`[i],pattern = "\\), ")
  my_tmp<-unlist(my_tmp)
  my_tmp<-str_extract(string = my_tmp,pattern = "^[[:alnum:]]+ \\(")
  my_tmp<-str_remove_all(string = my_tmp,pattern = " \\(")
  my_smpl[[i]]<-paste(my_tmp,collapse=",")
}
my_smpl<-unlist(my_smpl)
my_smpl<-str_split(string = paste(my_smpl,collapse = ","),pattern = ",")
my_smpl<-unique(unlist(my_smpl))
my_smpl<-my_smpl[my_smpl!="NA"]
my_co<-c("M9E1","M9E2","M9E3","M9E4")
my_oth<-df_maeiwa_condition[df_maeiwa_condition$STRS %nin% qw(M9,No),]$Condition
my_oth<-my_oth[my_oth %nin% my_smpl]
print(paste("Control strains (",length(my_co),"): ",paste(my_co,collapse = ","),sep=""))
print(paste("MT strains (",length(my_smpl),"): ",paste(my_smpl,collapse = ","),sep=""))
print(paste("WT strains (",length(my_oth),"): ",paste(my_oth,collapse = ","),sep=""))
my_tg<-str_split(string = paste(df_mt$targets,collapse=","),pattern = ",")
my_tg<-unique(unlist(my_tg))
my_tg<-dBnb[dBnb$gene %in% my_tg, ]$Bnb
print(paste("# of commonly mutated TRs (",nrow(df_mt_unique),"): ",paste(df_mt_unique$regulator,collapse = ","),sep=""))
print(paste("# of target genes (",length(my_tg),"): ",paste(my_tg,collapse = ","),sep=""))

my_df<-MaeIwadat[MaeIwadat$Bnb %in% my_tg,]
my_df<-my_df %>% select(c("Bnb",my_co,my_smpl,my_oth))
my_df2<-data.frame(Bnb=my_df$Bnb,Co_exp=NA,MT_exp=NA,WT_exp=NA)
i<-1
for(i in 1:nrow(my_df)){
  my_df2$Co_exp[i]<-rowMeans(as.matrix(my_df[i,colnames(my_df) %in% my_co]))
  my_df2$MT_exp[i]<-rowMeans(as.matrix(my_df[i,colnames(my_df) %in% my_smpl]))
  my_df2$WT_exp[i]<-rowMeans(as.matrix(my_df[i,colnames(my_df) %in% my_oth]))
}
my_df2$MT<-my_df2$MT_exp-my_df2$Co_exp
my_df2$WT<-my_df2$WT_exp-my_df2$Co_exp
my_df3<-my_df2 %>% select(Bnb,MT,WT)
my_df3<-my_df3 %>% pivot_longer(cols = c(MT,WT),names_to = "Type",values_to = "log2_fold_change")
my_df3$Type<-factor(my_df3$Type,levels = qw(WT,MT))

my_df4<-my_df2 %>% select(Bnb,Co_exp,MT_exp,WT_exp)
colnames(my_df4)<-qw(Bnb,Co,MT,WT)
my_df4<-my_df4 %>% pivot_longer(cols = qw(MT,WT),names_to = "Type",values_to = "Expression")
my_df4$Type<-factor(my_df4$Type,levels = qw(WT,MT))

p1<-ggscatter(my_df4,x="Co",y="Expression",alpha=0.5,color = "Type",
              xlim=c(3.9,16),ylim=c(3.9,16),
              legend=c(0.8,0.3),legend.title="",font.legend = c(12),
              palette = c("gray50","#F8766D"),size=1)+
  xlab(expression(atop(log[2]~"mRNA expression level","of control strains [a.u.]")))+
  ylab(expression(atop(log[2]~"mRNA expression level","of resistant mutants [a.u.]")))+
  geom_abline(slope = 1,intercept = 0,color="black",linewidth=0.3)
p2<-ggboxplot(my_df3,x="Type",y="log2_fold_change",color = "Type",size=1,
              add = "jitter",add.params = list(alpha=0.5,size=0.7),
              xlab="Type of commonly mutated TR",
              palette = c("gray50","#F8766D"),legend="")+
  ylab(expression(log[2]~"fold change [a.u.]"))+
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1)))+
  stat_compare_means(size=5,label.y = 1.9)
p1+p2
Sys.time()
```
#
#
#
# 16. The Strs_hp and Strs_he dataset
Horinouchi 2017
https://www.nature.com/articles/s41598-017-14335-7
## 16.1. The Strs_hp dataset 
Analysis for transcriptional plasticity
Transcriptome data of a parental strain under different stressful conditions
### 16.1.1. Load the Strs_hp dataset
```{r Strs_hp,warning=FALSE, message=FALSE,fig.height=2,fig.width=2,dpi=300}
Horidat<-as.data.frame(read.csv(paste(getwd(),"STRS_HPHE_DataSet","Horinouchi2017S2_parent.csv",sep="/")))
temp<-data.frame(colnames=colnames(Horidat))
colnames(Horidat)[colnames(Horidat)=="name"]<-"ProbeName"
colnames(Horidat)[colnames(Horidat)=="NoStress"]<-"No"
df_hori_condition<-data.frame(Condition=colnames(Horidat)[colnames(Horidat) %nin% qw(ProbeName)])
df_hori_condition$STRS<-df_hori_condition$Condition
df_hori_condition[df_hori_condition$Condition=="CoCl",]$STRS<-"CoCl2"
df_hori_condition[df_hori_condition$Condition=="SoC",]$STRS<-"Na2CO3"
df_hori_condition<-df_hori_condition[order(df_hori_condition$STRS),]
# Remove the same conditions used in PRECISE
my_strs<-qw(NaCl,KCl,Lac)
Horidat<-Horidat[,which(colnames(Horidat) %nin% df_hori_condition[df_hori_condition$STRS %in% my_strs,]$Condition)]
df_hori_condition<-df_hori_condition[df_hori_condition$STRS %nin% my_strs,]
df_hori_condition$STRS<-factor(df_hori_condition$STRS,levels = qw(CoCl2,Na2CO3,Mal,MCL,Cro,MG,BuOH,CPC,No))
# Remove the genes whose variance are zero
temp<-rowVars(as.matrix(Horidat[,which(colnames(Horidat) %nin% "ProbeName")]))
names(temp)<-seq(length(temp))
Horidat<-Horidat[-as.numeric(names(temp[temp==0])),]

# log2-transform
for(i in 1:nrow(df_hori_condition)){
  Horidat[,which(colnames(Horidat)==df_hori_condition$Condition[i])]<-log(10,2)*Horidat[,which(colnames(Horidat)==df_hori_condition$Condition[i])]
}

temp<-Horidat
temp<-left_join(temp,probe2id[(probe2id$IStag==F)&(probe2id$Count==1)&(!grepl("\\+",probe2id$ECKnb)),which(colnames(probe2id) %in% qw(ProbeName,ECKnb))],by="ProbeName")
temp<-left_join(temp,genename.mds42[,which(colnames(genename.mds42) %in% qw(Bnb,ECKnb))],by="ECKnb")
# omit #Bnb=NA
temp<-temp[!is.na(temp$Bnb),]

# remove duplicated Bnbs
temp1<-data.frame(Bnb=unique(temp$Bnb),Count=1)
for(i in 1:nrow(temp1)){
  temp1$Count[i]<-nrow(temp[temp$Bnb==temp1$Bnb[i],])
}
temp1<-temp[temp$Bnb %in% temp1[temp1$Count>1,]$Bnb,]
temp<-temp[temp$ProbeName %nin% temp1[!grepl("[A-Z]$",temp1$ProbeName),]$ProbeName,]
Horidat<-temp[,which(colnames(temp) %in% c("Bnb",df_hori_condition$Condition))]

# ggscatter(Horidat,x="B_Cl_AlaE1",y="B_Cl_AlaE4",size=0.5,alpha=0.2)
ggscatter(Horidat,x="No",y="CoCl",size=0.5,alpha=0.2)
Sys.time()

```
### 16.1.2. Number of genes in the dataset
```{r Strs,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
# venn
my_venn<-list(
  Three=dBnb.comp$Bnb,
  Strs_h=Horidat$Bnb
)

# ggVennDiagram(figtemp)
venn <- process_data(Venn(my_venn))
venn$regionData$name
venn$regionData$count
# venn$regionData$item
v<-venneuler(c(Three=length(venn$regionData$item[[which(venn$regionData$name=="Three")]]),
               Strs_h=length(venn$regionData$item[[which(venn$regionData$name=="Strs_h")]]),
               "Three&Strs_h"=length(venn$regionData$item[[which(venn$regionData$name=="Three/Strs_h")]])))
my_tmp<-v$colors
my_tmp[names(my_tmp)=="Three"][[1]]<-"gray50"
my_tmp[names(my_tmp)!="Three"][[1]]<-"yellow"
v$colors<-my_tmp
v$labels<-c("","")
par(mar=c(0, 0, 0, 0)) # Make large margins
plot(v,col = v$colors, quantities = TRUE) # Three(gray), 

Sys.time()
```
### 16.1.2. DM
#### 16.1.2.1. Calc DM
```{r Strs_hp,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=3,dpi=300}
temp3<-Horidat[,which(colnames(Horidat) %nin% "Bnb")]
temp5<-createEmptyDf(nrow=nrow(temp3),ncol = 3,colnames = qw(mean,var,sd))
temp5$mean<-rowMeans(as.matrix(temp3),na.rm=T)
temp5$var<-rowVars(as.matrix(temp3),na.rm=T)
temp5$sd<-rowSds(as.matrix(temp3),na.rm=T)
Horidat<-cbind(Horidat,temp5)

Horidat<-Horidat[order(Horidat$mean),] # sorting 
Horidat<-transform(Horidat,RunmedVar=runmed(Horidat$sd,41),lmg=Horidat$mean)
lmgn<-data.frame(x=Horidat$lmg,y=Horidat$RunmedVar)
sp<-smooth.spline(lmgn,df=12)
x<-seq(min(lmgn$x),max(lmgn$x),length=1000)
pred<-as.data.frame(x)
pred<-transform(pred,smsp=predict(sp,x)[2])
colnames(pred)<-c("x","smoothsp")
pred2<-Horidat$mean
pred2<-transform(pred2,y=predict(sp,pred2)[2])
colnames(pred2)<-c("x","smoothsp")
Horidat<-transform(Horidat,SmoothSpline=pred2$smoothsp)
Horidat<-transform(Horidat,DMstrs=Horidat$sd-Horidat$SmoothSpline) # calculate distance from smooth spline to SD 

df_strs_hori<-left_join(df_DMall,Horidat,by="Bnb")

## Running median
# mycolpal<-get_palette("lancet",2)
ggscatter(df_strs_hori,x="mean",y="sd",
          cor.coef = TRUE,cor.coeff.args = list(method = "spearman", label.x = 12, label.y=2.7, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression("Standard deviation [a.u.]"))+
  geom_line(data=df_strs_hori,aes(x=mean,y=RunmedVar), colour="darkorange", size=0.5)+
  geom_line(data=pred,aes(x=x,y=smoothsp), colour="deepskyblue", size=0.5) # "deepskyblue"

ggscatter(df_strs_hori,x="mean",y="DMstrs",
          cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
          cor.coeff.args = list(method = "spearman", label.x = 12, label.y=2.3, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression(DM[strs_hp]~"[a.u.]"))


Sys.time()
```
#### 16.1.2.2. Comprison of DMs
```{r Strs_hp,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
# DMenv vs. DMstrs
p1<-ggscatter(df_strs_hori,x="DMenv",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=2,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.7, 3.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 2.1))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[strs_hp]~"[a.u.]"))
p1<-ggpar(p1,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMevo vs. DMstrs
p2<-ggscatter(df_strs_hori,x="DMevo",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 2.1))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[strs_hp]~"[a.u.]"))
p2<-ggpar(p2,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMmut vs. DMstrs
p3<-ggscatter(df_strs_hori,x="DMmut",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 2.1))+
  # scale_x_continuous(breaks = get_breaks(by = 0.5, from = 0),limits =  c(-0.5, 2))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[strs_hp]~"[a.u.]"))
p3<-ggpar(p3,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)

p1+p2+p3

Sys.time()
```
### 16.1.3. PCA
#### 16.1.3.1. Detect outliers in PC1~PC3
```{r Strs,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-Horidat
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% df_hori_condition$Condition)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
hp_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    hp_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(hp_outlier[!is.na(hp_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
#### 16.1.3.2. Remove outliers if it is necessary and examine PCA
```{r Strs_hp,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-Horidat
if(!is.na(hp_outlier)){
  temp0<-temp0[,colnames(temp0) %nin% hp_outlier]
}
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% df_hori_condition$Condition)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
res.pca.global_strs_hp <- prcomp(figtemp, scale. = T)
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_strs <- res.pca.global_strs_hp$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_strs$label<-rownames(pc_scores_strs)

pc_eigenvalues_strs <- res.pca.global_strs_hp$sdev^2
pc_eigenvalues_strs <- tibble(PC = factor(1:length(pc_eigenvalues_strs)), 
                         variance = pc_eigenvalues_strs) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained

# Scree plot
figtemp1<-pc_eigenvalues_strs
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree_strs_hori<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,plot_type = "l",
        color="gray20",
       xlab="Components",ylab="Fraction of variance explained (%)")+
  # scale_x_continuous(breaks = get_breaks(by = 4, from = 0),limits = c(0,16))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_strs$pct_cum[k] ,"% variance",sep=""))

Sys.time()
```
#### 16.1.3.3. 2D plot (PC1~PC3)
```{r Strs_hp,warning=FALSE, message=FALSE,fig.height=3,fig.width=11,dpi=300}
figtemp<-left_join(pc_scores_strs,df_hori_condition)
figtemp$STRS<-factor(figtemp$STRS,levels = levels(df_hori_condition$STRS))

p1<-ggscatter(data=figtemp,x="PC1",y="PC2",alpha=0.5,legend="",color = "STRS",palette = c(gg_color_hue(8),"gray40"))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,
      xlab = paste("PC1 (", round(pc_eigenvalues_strs$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_strs$pct[2],digits = 0),"%)",sep=""))

p2<-ggscatter(data=figtemp,x="PC2",y="PC3",alpha=0.5,legend="",color = "STRS",palette = c(gg_color_hue(8),"gray40"))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,
      xlab = paste("PC2 (", round(pc_eigenvalues_strs$pct[2],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_strs$pct[3],digits = 0),"%)",sep=""))

p3<-ggscatter(data=figtemp,x="PC1",y="PC3",alpha=0.5,legend="",color = "STRS",palette = c(gg_color_hue(8),"gray40"))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,
      xlab = paste("PC1 (", round(pc_eigenvalues_strs$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_strs$pct[3],digits = 0),"%)",sep=""))

q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")
leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,0.3),nrow=1)
Sys.time()
```
#### 16.1.3.4. Correlation between rotation (eigenvector) for PC1~PC3
```{r Strs_hp,warning=FALSE, message=FALSE,fig.height=4.5,fig.width=11,dpi=300}
my_lim<-c(0,0.65)
# Env vs. STRS
temp5<-data.frame(Env_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp5)){
  temp1<-as.data.frame(res.pca.global_env$rotation[,temp5$Env_PC[i]])
  colnames(temp1)<-"rot.env"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_hp$rotation[,temp5$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp5$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.strs, method = "spearman")$estimate
  temp5$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.strs, method = "spearman")$p.value
}

temp5.1<-matrix(rep(0,9),nrow=3)
rownames(temp5.1)<-mypcs
colnames(temp5.1)<-mypcs
temp5.2<-temp5.1
for(i in 1:nrow(temp5.1)){
  for(j in 1:ncol(temp5.1)){
    temp5.1[i,j]<-temp5[temp5$Env_PC==i&temp5$STRS_PC==j,]$Cor
    temp5.2[i,j]<-temp5[temp5$Env_PC==i&temp5$STRS_PC==j,]$p.value
  }
}
temp5.3<-abs(temp5.1)


# Evo vs. STRS
temp6<-data.frame(Evo_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp6)){
  temp1<-as.data.frame(res.pca.global_evo$rotation[,temp6$Evo_PC[i]])
  colnames(temp1)<-"rot.evo"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_hp$rotation[,temp6$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp6$Cor[i]<-cor.test(temp4$rot.evo,temp4$rot.strs, method = "spearman")$estimate
  temp6$p.value[i]<-cor.test(temp4$rot.evo,temp4$rot.strs, method = "spearman")$p.value
}

temp6.1<-matrix(rep(0,9),nrow=3)
rownames(temp6.1)<-mypcs
colnames(temp6.1)<-mypcs
temp6.2<-temp6.1
for(i in 1:nrow(temp6.1)){
  for(j in 1:ncol(temp6.1)){
    temp6.1[i,j]<-temp6[temp6$Evo_PC==i&temp6$STRS_PC==j,]$Cor
    temp6.2[i,j]<-temp6[temp6$Evo_PC==i&temp6$STRS_PC==j,]$p.value
  }
}
temp6.3<-abs(temp6.1)

# Mut vs. STRS
temp7<-data.frame(Mut_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp7)){
  temp1<-as.data.frame(res.pca.global_mut$rotation[,temp7$Mut_PC[i]])
  colnames(temp1)<-"rot.mut"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_hp$rotation[,temp7$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp7$Cor[i]<-cor.test(temp4$rot.mut,temp4$rot.strs, method = "spearman")$estimate
  temp7$p.value[i]<-cor.test(temp4$rot.mut,temp4$rot.strs, method = "spearman")$p.value
}

temp7.1<-matrix(rep(0,9),nrow=3)
rownames(temp7.1)<-mypcs
colnames(temp7.1)<-mypcs
temp7.2<-temp7.1
for(i in 1:nrow(temp7.1)){
  for(j in 1:ncol(temp7.1)){
    temp7.1[i,j]<-temp7[temp7$Mut_PC==i&temp7$STRS_PC==j,]$Cor
    temp7.2[i,j]<-temp7[temp7$Mut_PC==i&temp7$STRS_PC==j,]$p.value
  }
}
temp7.3<-abs(temp7.1)

my_lab_fontsize<-6

p4<-ggcorrplot::ggcorrplot(temp5.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Env", y = "Strs_hp")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p4<-ggpar(p4,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p5<-ggcorrplot::ggcorrplot(temp6.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Evo", y = "Strs_hp")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p5<-ggpar(p5,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p6<-ggcorrplot::ggcorrplot(temp7.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Strs_hp")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p6<-ggpar(p6,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

my_fontsize<-16
q2<-ggpar(p6,legend = "right",legend.title="",font.legend = my_fontsize)+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.1),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "Abs. of Spearman's R",title.position = "left",ticks.linewidth = 1))+
  theme(legend.title = element_text(angle = 90, hjust = 10, vjust = 0.5,size=18),legend.title.align=10)
leg2 <- ggpubr::get_legend(q2)

my_tmp<-data.frame(x=seq(min(my_lim),max(my_lim),by=0.01),y=seq(min(my_lim),max(my_lim),by=0.01),z=seq(min(my_lim),max(my_lim),by=0.01))
p_led<-ggscatter(my_tmp,x="x",y="y",fill="z",legend.title = "|R|",legend="right")+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
q2<-ggpar(p_led,font.legend = my_fontsize)
leg2 <- ggpubr::get_legend(q2)


grid.arrange(p4,p5,p6,as_ggplot(leg2),widths=c(1,1,1,0.4),nrow=1)
Sys.time()
```
## 16.2. The Strs_he dataset
Analysis for evolvability
Transcriptome data of mutants under a non-stressful condition
### 16.2.1. Load the Strs_he dataset
```{r Strs_he,warning=FALSE, message=FALSE,fig.height=2,fig.width=2,dpi=300}
Hori_evo_dat<-as.data.frame(read.csv(paste(getwd(),"STRS_HPHE_DataSet","Horinouchi2017S2_evo_nostress.csv",sep="/")))
temp<-data.frame(colnames=colnames(Hori_evo_dat))
colnames(Hori_evo_dat)[colnames(Hori_evo_dat)=="name"]<-"ProbeName"
df_hori_evo_condition<-data.frame(Condition=colnames(Hori_evo_dat)[colnames(Hori_evo_dat) %nin% qw(ProbeName)])
df_hori_evo_condition$STRS<-str_sub(df_hori_evo_condition$Condition,start = 1,end=-3)

# Remove the same conditions used in PRECISE
my_strs<-qw(NaCl,KCl,Lac)
Hori_evo_dat<-Hori_evo_dat[,which(colnames(Hori_evo_dat) %nin% df_hori_evo_condition[df_hori_evo_condition$STRS %in% my_strs,]$Condition)]
df_hori_evo_condition<-df_hori_evo_condition[df_hori_evo_condition$STRS %nin% my_strs,]
df_hori_evo_condition$STRS<-factor(df_hori_evo_condition$STRS,levels=qw(CoCl2,Na2CO3,Mal,MCL,Cro,MG,BuOH,CPC))
# # Remove the genes whose variance are zero
# temp<-rowVars(as.matrix(Hori_evo_dat[,which(colnames(Hori_evo_dat) %nin% "ProbeName")]))
# names(temp)<-seq(length(temp))
# Hori_evo_dat<-Hori_evo_dat[-as.numeric(names(temp[temp==0])),]

# log2-transform
for(i in 1:nrow(df_hori_evo_condition)){
  Hori_evo_dat[,which(colnames(Hori_evo_dat)==df_hori_evo_condition$Condition[i])]<-log(10,2)*Hori_evo_dat[,which(colnames(Hori_evo_dat)==df_hori_evo_condition$Condition[i])]
}

temp<-Hori_evo_dat
temp<-left_join(temp,probe2id[(probe2id$IStag==F)&(probe2id$Count==1)&(!grepl("\\+",probe2id$ECKnb)),which(colnames(probe2id) %in% qw(ProbeName,ECKnb))],by="ProbeName")
temp<-left_join(temp,genename.mds42[,which(colnames(genename.mds42) %in% qw(Bnb,ECKnb))],by="ECKnb")
# omit #Bnb=NA
temp<-temp[!is.na(temp$Bnb),]

# remove duplicated Bnbs
temp1<-data.frame(Bnb=unique(temp$Bnb),Count=1)
for(i in 1:nrow(temp1)){
  temp1$Count[i]<-nrow(temp[temp$Bnb==temp1$Bnb[i],])
}
temp1<-temp[temp$Bnb %in% temp1[temp1$Count>1,]$Bnb,]
temp<-temp[temp$ProbeName %nin% temp1[!grepl("[A-Z]$",temp1$ProbeName),]$ProbeName,]
Hori_evo_dat<-temp[,which(colnames(temp) %in% c("Bnb",df_hori_evo_condition$Condition))]

ggscatter(Hori_evo_dat,x="Na2CO3_1",y="CoCl2_1",size=0.5,alpha=0.2)
Sys.time()
```
### 16.2.2. DM
#### 16.2.2.1. Calc DM
```{r Strs_he,warning=FALSE, message=FALSE,fig.height=2.5,fig.width=3,dpi=300}
temp3<-Hori_evo_dat[,which(colnames(Hori_evo_dat) %nin% "Bnb")]
temp5<-createEmptyDf(nrow=nrow(temp3),ncol = 3,colnames = qw(mean,var,sd))
temp5$mean<-rowMeans(as.matrix(temp3),na.rm=T)
temp5$var<-rowVars(as.matrix(temp3),na.rm=T)
temp5$sd<-rowSds(as.matrix(temp3),na.rm=T)
Hori_evo_dat<-cbind(Hori_evo_dat,temp5)

Hori_evo_dat<-Hori_evo_dat[order(Hori_evo_dat$mean),] # sorting 
Hori_evo_dat<-transform(Hori_evo_dat,RunmedVar=runmed(Hori_evo_dat$sd,41),lmg=Hori_evo_dat$mean)
lmgn<-data.frame(x=Hori_evo_dat$lmg,y=Hori_evo_dat$RunmedVar)
sp<-smooth.spline(lmgn,df=12)
x<-seq(min(lmgn$x),max(lmgn$x),length=1000)
pred<-as.data.frame(x)
pred<-transform(pred,smsp=predict(sp,x)[2])
colnames(pred)<-c("x","smoothsp")
pred2<-Hori_evo_dat$mean
pred2<-transform(pred2,y=predict(sp,pred2)[2])
colnames(pred2)<-c("x","smoothsp")
Hori_evo_dat<-transform(Hori_evo_dat,SmoothSpline=pred2$smoothsp)
Hori_evo_dat<-transform(Hori_evo_dat,DMstrs=Hori_evo_dat$sd-Hori_evo_dat$SmoothSpline) # calculate distance from smooth spline to SD 

df_strs_hori_evo<-left_join(df_DMall,Hori_evo_dat,by="Bnb")

## Running median
# mycolpal<-get_palette("lancet",2)
ggscatter(df_strs_hori_evo,x="mean",y="sd",
          cor.coef = TRUE,cor.coeff.args = list(method = "spearman", label.x = 12, label.y=1.6, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression("Standard deviation [a.u.]"))+
  geom_line(data=df_strs_hori_evo,aes(x=mean,y=RunmedVar), colour="darkorange", size=0.5)+
  geom_line(data=pred,aes(x=x,y=smoothsp), colour="deepskyblue", size=0.5) # "deepskyblue"

ggscatter(df_strs_hori_evo,x="mean",y="DMstrs",
          cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
          cor.coeff.args = list(method = "spearman", label.x = 12, label.y=1, label.sep = "\n",size=4),
          alpha = 0.2, size=0.3,legend="")+
  # scale_x_continuous(breaks = get_breaks(by = 5, from = 0),limits =  c(0, 12))+
  xlab(expression(log[2]~"mRNA expression level [a.u.]"))+
  ylab(expression(DM[strs_he]~"[a.u.]"))


Sys.time()
```
#### 16.2.2.2. Comprison of DMs
```{r Strs_he,warning=FALSE, message=FALSE,fig.height=3,fig.width=9,dpi=300}
# DMenv vs. DMstrs
p1<-ggscatter(df_strs_hori_evo,x="DMenv",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=2,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.7, 3.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 2.1))+
  xlab(expression(DM[env]~"[a.u.]"))+
  ylab(expression(DM[strs_he]~"[a.u.]"))
p1<-ggpar(p1,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMevo vs. DMstrs
p2<-ggscatter(df_strs_hori_evo,x="DMevo",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_x_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.5, 2.5))+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 2.1))+
  xlab(expression(DM[evo]~"[a.u.]"))+
  ylab(expression(DM[strs_he]~"[a.u.]"))
p2<-ggpar(p2,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)


# DMmut vs. DMstrs
p3<-ggscatter(df_strs_hori_evo,x="DMmut",y="DMstrs",
          cor.coef = T,cor.coeff.args = list(method = "spearman",label.sep = "\n",size=4,label.x=1.5,label.y=1.5),
          alpha = 0.2, size=0.3,legend="")+
  scale_y_continuous(breaks = get_breaks(by = 1, from = 0),limits =  c(-0.3, 2.1))+
  # scale_x_continuous(breaks = get_breaks(by = 0.5, from = 0),limits =  c(-0.5, 2))+
  xlab(expression(DM[mut]~"[a.u.]"))+
  ylab(expression(DM[strs_he]~"[a.u.]"))
p3<-ggpar(p3,font.y=my_fontsize,font.x=my_fontsize,font.xtickslab = my_fontsize,font.ytickslab = my_fontsize)

p1+p2+p3

Sys.time()
```
### 16.2.3. Examine PCA
#### 16.2.3.1. Detect outliers in PC1~PC3
```{r Strs,warning=FALSE, message=FALSE,fig.height=4,fig.width=12,dpi=300}
temp0<-Hori_evo_dat
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% df_hori_evo_condition$Condition)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
my_res <- prcomp(figtemp, scale. = T)

my_tmp<-apply(my_res$x , 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))
my_tmp<-unlist(my_tmp)
he_outlier<-NA
if(length(my_tmp)>0){
  my_tmp<-names(my_tmp)[grepl(pattern = "^PC[1-3]\\.",x = names(my_tmp))]
  if(length(my_tmp)>0){
    my_tmp<-str_remove_all(string = my_tmp,pattern = "^PC[1-3]\\.")
    he_outlier<-unique(my_tmp)
  }
}

print(paste("Outliers: ",length(he_outlier[!is.na(he_outlier)]),"out of",nrow(my_res$x)))
Sys.time()
```
#### 16.2.3.2. Remove outliers if it is necessary and examine PCA
```{r Strs_he,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
temp0<-Hori_evo_dat
if(!is.na(he_outlier)){
  temp0<-temp0[,colnames(temp0) %nin% he_outlier]
}
temp0<-temp0[order(temp0$Bnb,decreasing = F),]
rownames(temp0)<-temp0$Bnb
temp0<-na.omit(temp0[,which(colnames(temp0) %in% df_hori_evo_condition$Condition)])
temp0<-t(temp0) # transpose the matrix so genes are as columns
figtemp<-as.data.frame(temp0)
figtemp<-figtemp %>% select_if(negate(anyNA))
res.pca.global_strs_he <- prcomp(figtemp, scale. = T)
# The PC scores are stored in the "x" value of the prcomp object
pc_scores_strs <- res.pca.global_strs_he$x %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "Condition")
pc_scores_strs$label<-rownames(pc_scores_strs)

pc_eigenvalues_strs <- res.pca.global_strs_he$sdev^2
pc_eigenvalues_strs <- tibble(PC = factor(1:length(pc_eigenvalues_strs)), 
                         variance = pc_eigenvalues_strs) %>%   
  mutate(pct = variance/sum(variance)*100) %>% # add a new column with the percent variance
  mutate(pct_cum = cumsum(pct)) # add another column with the cumulative variance explained

# Scree plot
figtemp1<-pc_eigenvalues_strs
figtemp1$PC<-as.numeric(figtemp1$PC)
dscree_strs_hori_evo<-figtemp1
ggline(figtemp1,x="PC",y="pct_cum",numeric.x.axis = T,plot_type = "l",
        color="gray20",
       xlab="Components",ylab="Fraction of variance explained (%)")+
  # scale_x_continuous(breaks = get_breaks(by = 4, from = 0),limits = c(0,16))+
  scale_y_continuous(breaks = get_breaks(by = 20, from = 0),limits = c(0,100))

# Variance explatind by PC1~PCk
k<-3
print(paste("PC1~PC",k, " explains ", pc_eigenvalues_strs$pct_cum[k] ,"% variance",sep=""))

Sys.time()
```
### 16.2.3.3. 2D plot (PC1~PC3)
```{r Strs_he,warning=FALSE, message=FALSE,fig.height=3,fig.width=11,dpi=300}
figtemp<-left_join(pc_scores_strs,df_hori_evo_condition)
figtemp$STRS<-factor(figtemp$STRS,levels = levels(figtemp$STRS))
p1<-ggscatter(data=figtemp,x="PC1",y="PC2",alpha=0.5,legend="",color = "STRS",palette = gg_color_hue(8))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p1<-ggpar(p1,
      xlab = paste("PC1 (", round(pc_eigenvalues_strs$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC2 (", round(pc_eigenvalues_strs$pct[2],digits = 0),"%)",sep=""))

p2<-ggscatter(data=figtemp,x="PC2",y="PC3",alpha=0.5,legend="",color = "STRS",palette = gg_color_hue(8))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p2<-ggpar(p2,
      xlab = paste("PC2 (", round(pc_eigenvalues_strs$pct[2],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_strs$pct[3],digits = 0),"%)",sep=""))

p3<-ggscatter(data=figtemp,x="PC1",y="PC3",alpha=0.5,legend="",color = "STRS",palette = gg_color_hue(8))+
  scale_y_continuous(expand = expansion(mult = c(0.2, 0.1)))
p3<-ggpar(p3,
      xlab = paste("PC1 (", round(pc_eigenvalues_strs$pct[1],digits = 0),"%)",sep=""),
      ylab = paste("PC3 (", round(pc_eigenvalues_strs$pct[3],digits = 0),"%)",sep=""))

q<-ggpar(p1,legend = "right",font.ytickslab= c(10),legend.title = "")
leg <- ggpubr::get_legend(q)

grid.arrange(p1,p2,p3,as_ggplot(leg),widths=c(1,1,1,0.3),nrow=1)
Sys.time()
```
### 16.2.3.4. Correlation between rotation (eigenvector) for PC1~PC3
```{r Strs_he,warning=FALSE, message=FALSE,fig.height=4.5,fig.width=11,dpi=300}
my_lim<-c(0,0.65)

# Env vs. STRS
temp5<-data.frame(Env_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp5)){
  temp1<-as.data.frame(res.pca.global_env$rotation[,temp5$Env_PC[i]])
  colnames(temp1)<-"rot.env"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_he$rotation[,temp5$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp5$Cor[i]<-cor.test(temp4$rot.env,temp4$rot.strs, method = "spearman")$estimate
  temp5$p.value[i]<-cor.test(temp4$rot.env,temp4$rot.strs, method = "spearman")$p.value
}

temp5.1<-matrix(rep(0,9),nrow=3)
rownames(temp5.1)<-mypcs
colnames(temp5.1)<-mypcs
temp5.2<-temp5.1
for(i in 1:nrow(temp5.1)){
  for(j in 1:ncol(temp5.1)){
    temp5.1[i,j]<-temp5[temp5$Env_PC==i&temp5$STRS_PC==j,]$Cor
    temp5.2[i,j]<-temp5[temp5$Env_PC==i&temp5$STRS_PC==j,]$p.value
  }
}
temp5.3<-abs(temp5.1)


# Evo vs. STRS
temp6<-data.frame(Evo_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp6)){
  temp1<-as.data.frame(res.pca.global_evo$rotation[,temp6$Evo_PC[i]])
  colnames(temp1)<-"rot.evo"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_he$rotation[,temp6$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp6$Cor[i]<-cor.test(temp4$rot.evo,temp4$rot.strs, method = "spearman")$estimate
  temp6$p.value[i]<-cor.test(temp4$rot.evo,temp4$rot.strs, method = "spearman")$p.value
}

temp6.1<-matrix(rep(0,9),nrow=3)
rownames(temp6.1)<-mypcs
colnames(temp6.1)<-mypcs
temp6.2<-temp6.1
for(i in 1:nrow(temp6.1)){
  for(j in 1:ncol(temp6.1)){
    temp6.1[i,j]<-temp6[temp6$Evo_PC==i&temp6$STRS_PC==j,]$Cor
    temp6.2[i,j]<-temp6[temp6$Evo_PC==i&temp6$STRS_PC==j,]$p.value
  }
}
temp6.3<-abs(temp6.1)

# Mut vs. STRS
temp7<-data.frame(Mut_PC=rep(seq(1,3),each=3),STRS_PC=rep(seq(1,3),3),Cor=NA,p.value=NA)
for(i in 1:nrow(temp7)){
  temp1<-as.data.frame(res.pca.global_mut$rotation[,temp7$Mut_PC[i]])
  colnames(temp1)<-"rot.mut"
  temp1$Bnb<-rownames(temp1)
  temp2<-as.data.frame(res.pca.global_strs_he$rotation[,temp7$STRS_PC[i]])
  colnames(temp2)<-"rot.strs"
  temp2$Bnb<-rownames(temp2)
  temp4<-left_join(temp1,temp2)
  temp7$Cor[i]<-cor.test(temp4$rot.mut,temp4$rot.strs, method = "spearman")$estimate
  temp7$p.value[i]<-cor.test(temp4$rot.mut,temp4$rot.strs, method = "spearman")$p.value
}

temp7.1<-matrix(rep(0,9),nrow=3)
rownames(temp7.1)<-mypcs
colnames(temp7.1)<-mypcs
temp7.2<-temp7.1
for(i in 1:nrow(temp7.1)){
  for(j in 1:ncol(temp7.1)){
    temp7.1[i,j]<-temp7[temp7$Mut_PC==i&temp7$STRS_PC==j,]$Cor
    temp7.2[i,j]<-temp7[temp7$Mut_PC==i&temp7$STRS_PC==j,]$p.value
  }
}
temp7.3<-abs(temp7.1)

my_lab_fontsize<-6

p4<-ggcorrplot::ggcorrplot(temp5.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Env", y = "Strs_he")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p4<-ggpar(p4,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p5<-ggcorrplot::ggcorrplot(temp6.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Evo", y = "Strs_he")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p5<-ggpar(p5,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

p6<-ggcorrplot::ggcorrplot(temp7.3,
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize,tl.cex = 16,outline.col = "gray40")+
  scale_fill_gradient(limit = my_lim, low = "white", high =  "#E46726")+
  labs(x = "Mut", y = "Strs_he")+
  theme(
    axis.title.x = element_text(angle = 0, color = 'grey20',size=24,vjust = -1),
    axis.title.y = element_text(angle = 90, color = 'grey20',size=24,vjust = 2)
  )
p6<-ggpar(p6,legend.title = "Spearman's R",legend = "")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())

my_fontsize<-16
q2<-ggpar(p6,legend = "right",legend.title="",font.legend = my_fontsize)+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.1),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "Abs. of Spearman's R",title.position = "left",ticks.linewidth = 1))+
  theme(legend.title = element_text(angle = 90, hjust = 10, vjust = 0.5,size=18),legend.title.align=10)
leg2 <- ggpubr::get_legend(q2)

my_tmp<-data.frame(x=seq(min(my_lim),max(my_lim),by=0.01),y=seq(min(my_lim),max(my_lim),by=0.01),z=seq(min(my_lim),max(my_lim),by=0.01))
p_led<-ggscatter(my_tmp,x="x",y="y",fill="z",legend.title = "|R|",legend="right")+
  scale_fill_continuous(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                        guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
q2<-ggpar(p_led,font.legend = my_fontsize)
leg2 <- ggpubr::get_legend(q2)

grid.arrange(p4,p5,p6,as_ggplot(leg2),widths=c(1,1,1,0.4),nrow=1)
Sys.time()
```
#
#
#
# 17. Bow-tie decomposition
The formal bow-tie deconvolution was defined in https://ieeexplore.ieee.org/document/5977625
The code is based on https://royalsocietypublishing.org/doi/10.1098/rsif.2021.0069
## 17.1. Load tables of RegulonDB Cell Sensing
http://regulondb.ccg.unam.mx/menu/about_regulondb/cellsensing/index.jsp
https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=62adcb9d37fa4257d6d5e74cdade4870ff9faf45
```{r Bowtie,warning=FALSE, message=FALSE,fig.height=2,fig.width=2,dpi=300}
# Rename
my_cure<-read.xlsx(paste(getwd(),"RegulonDB","Cell_Sensing","Cell_Sensing.xlsx",sep="/"),sheet = "Cure")
my_cure<-my_cure[!is.na(my_cure$Ecocyc),]

# E-TC
#  Include TFs forming part of two-component systems, the signals are sensed by the sensor component (most of them located in the bacterial periplasm)
df_cs_e_tc<-read.xlsx(paste(getwd(),"RegulonDB","Cell_Sensing","Cell_Sensing.xlsx",sep="/"),sheet = "E-TC")
colnames(df_cs_e_tc)[1]<-"regulator"
df_cs_e_tc$regulator<-str_replace(df_cs_e_tc$regulator, pattern="/", replacement="")
df_cs_e_tc$regulator<-str_replace(df_cs_e_tc$regulator, pattern="\\?", replacement="")
for(i in 1:nrow(df_cs_e_tc)){
  temp<-my_cure[my_cure$RegulonDB==df_cs_e_tc$regulator[i],]
  if(nrow(temp)==1){
    df_cs_e_tc$regulator[i]<-temp$Ecocyc
  }
}
df_cs_e_tc<-df_cs_e_tc[df_cs_e_tc$regulator %in% unique(TRN$regulator),]

# E-TM
#  Include TFs sensing transportable metabolites; these TFs work in close connection with transport systems that sense/introduce specific metabolites from the milieu. The metabolites bound/sensed by this kind of TF are not anymore modified by the cell metabolism except those modifications inherent to the transport events (e.g. phosphorylation).
df_cs_e_tm<-read.xlsx(paste(getwd(),"RegulonDB","Cell_Sensing","Cell_Sensing.xlsx",sep="/"),sheet = "E-TM")
colnames(df_cs_e_tm)[1]<-"regulator"
i<-17
for(i in 1:nrow(df_cs_e_tm)){
  df_cs_e_tm$regulator[i]<-str_split(df_cs_e_tm$regulator[i],pattern = " ")[[1]][1]
}
for(i in 1:nrow(df_cs_e_tm)){
  temp<-my_cure[my_cure$RegulonDB==df_cs_e_tm$regulator[i],]
  if(nrow(temp)==1){
    df_cs_e_tm$regulator[i]<-temp$Ecocyc
  }
}
df_cs_e_tm<-df_cs_e_tm[df_cs_e_tm$regulator %in% unique(TRN$regulator),]

# H
# Includes hybrid TFs, thus named because both bind/sense metabolites produced by the cell and transported from the milieu (mostly for sensing amino acids that in addition to be synthesized by the cell can be transported into the cell from the milieu).
df_cs_h<-read.xlsx(paste(getwd(),"RegulonDB","Cell_Sensing","Cell_Sensing.xlsx",sep="/"),sheet = "H")
colnames(df_cs_h)[1]<-"regulator"
df_cs_h$regulator<-str_replace(df_cs_h$regulator, pattern="\\*", replacement="")
for(i in 1:nrow(df_cs_h)){
  df_cs_h$regulator[i]<-str_split(df_cs_h$regulator[i],pattern = " ")[[1]][1]
}
for(i in 1:nrow(df_cs_h)){
  temp<-my_cure[my_cure$RegulonDB==df_cs_h$regulator[i],]
  if(nrow(temp)==1){
    df_cs_h$regulator[i]<-temp$Ecocyc
  }
}
df_cs_h<-df_cs_h[df_cs_h$regulator %in% unique(TRN$regulator),]

# I-SM
# Include TFs binding/sensing metabolites generated by the cellular metabolism; by enzymatic reactions (sugar, nucleotides, cofactors, etc) or generated as byproducts of biochemical reactions (e.g. redox potential).
df_cs_i_sm<-read.xlsx(paste(getwd(),"RegulonDB","Cell_Sensing","Cell_Sensing.xlsx",sep="/"),sheet = "I-SM")
colnames(df_cs_i_sm)[1]<-"regulator"
for(i in 1:nrow(df_cs_i_sm)){
  df_cs_i_sm$regulator[i]<-str_split(df_cs_i_sm$regulator[i],pattern = " ")[[1]][1]
}
for(i in 1:nrow(df_cs_i_sm)){
  temp<-my_cure[my_cure$RegulonDB==df_cs_i_sm$regulator[i],]
  if(nrow(temp)==1){
    df_cs_i_sm$regulator[i]<-temp$Ecocyc
  }
}
df_cs_i_sm<-df_cs_i_sm[df_cs_i_sm$regulator %in% unique(TRN$regulator),]


# I-DB
#  Include DNA-bending TFs for nucleoid or chromosome remodeling and compaction, the activity of this kind of TF is not directly affected by signal effectors but possibly by DNA supercoiling or macromolecular crowding.
df_cs_i_db<-read.xlsx(paste(getwd(),"RegulonDB","Cell_Sensing","Cell_Sensing.xlsx",sep="/"),sheet = "I-DB")
colnames(df_cs_i_db)[1]<-"regulator"
df_cs_i_db<-df_cs_i_db[df_cs_i_db$regulator %in% unique(TRN$regulator),]



Sys.time()
```
## 17.2. Load tables of EcoCyc Cell Sensing
```{r Bowtie,warning=FALSE, message=FALSE,fig.height=2,fig.width=2,dpi=300}
df_ec_eff<-as.data.frame(read_csv(paste(getwd(),"ECOCYC","2022_10_09_Ecocyc","transcription_factors_effectors.csv",sep="/")))
df_ec_eff<-df_ec_eff[!is.na(df_ec_eff$Effectors),]
df_ec_eff<-df_ec_eff[df_ec_eff$regulator %in% unique(TRN$regulator),]

sensory_tfs<-unique(c(df_cs_e_tc$regulator,
                    df_cs_e_tm$regulator,
                    df_cs_h$regulator,
                    df_cs_i_sm$regulator,
                    df_cs_i_db$regulator,
                    df_ec_eff$regulator))
#print(sensory_tfs)

Sys.time()
```
## 17.3. Load output of bow-tie decomposition using the networkx package in Python
```{r Bowtie,warning=FALSE, message=FALSE,fig.height=2,fig.width=2,dpi=300}
df_bt_core<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","core.csv",sep="/"),col_names = "node"))
df_bt_core$bt_role<-"core"
df_bt_inp<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","inp.csv",sep="/"),col_names = "node"))
df_bt_inp$bt_role<-"inp"
df_bt_intendrils<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","intendrils.csv",sep="/"),col_names = "node"))
df_bt_intendrils$bt_role<-"intendrils"
df_bt_others<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","others.csv",sep="/"),col_names = "node"))
df_bt_others$bt_role<-"others"
df_bt_out<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","out.csv",sep="/"),col_names = "node"))
df_bt_out$bt_role<-"out"
df_bt_outtendrils<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","outtendrils.csv",sep="/"),col_names = "node"))
df_bt_outtendrils$bt_role<-"outtendrils"
df_bt_tubes<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","tubes.csv",sep="/"),col_names = "node"))
df_bt_tubes$bt_role<-"tubes"
df_bt<-bind_rows(df_bt_core,df_bt_inp,df_bt_intendrils,df_bt_others,df_bt_out,df_bt_outtendrils,df_bt_tubes)
df_bt_regulators<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","regulators.csv",sep="/")))
df_bt$regulator_flag<-"Non-TR"
df_bt[df_bt$node %in% unique(df_bt_regulators$regulator),]$regulator_flag<-"TR"
df_bt$sensor_flag<-"Non-SR" # non sensory regulator
df_bt[df_bt$node %in% sensory_tfs,]$sensor_flag<-"SR"

TRhigh<-unique(c(dgsea_trn_evo[(dgsea_trn_evo$NES>0)&(dgsea_trn_evo$qvalue<0.05),]$ID,
  dgsea_trn_env[(dgsea_trn_env$NES>0)&(dgsea_trn_env$qvalue<0.05),]$ID,
  dgsea_trn_mut[(dgsea_trn_mut$NES>0)&(dgsea_trn_mut$qvalue<0.05),]$ID))

df_bt$impact_flag<-"TRlow" # low impact
df_bt[df_bt$node %in% TRhigh,]$impact_flag<-"TRhigh"
df_bt[df_bt$node %in% TRcom,]$impact_flag<-"TRcom"

df_ecoli_net<-as.data.frame(read_csv(paste(getwd(),"Bow_tie_GRN","output","nx_network_all.csv",sep="/")))
Sys.time()
```
## 17.4. Visualization of bow-tie structure of E. coli gene regulatory network
```{r Bowtie,warning=FALSE, message=FALSE,fig.height=20,fig.width=23,dpi=300}
df_bt_vis<-df_bt
my_palette_info<-data.frame(bt_role=qw(inp,intendrils,core,out,outtendrils,tubes,others),
                            # node_color=c("#42B540FF","#0099B4FF","#ED0000FF","#00468BFF","#925E9FFF","#FDAF91FF","#ADB6B6FF"))
                      node_color=c("#fde725","white","#6ece58","#1f9e89","white","white","white"))
df_bt_vis<-left_join(df_bt_vis,my_palette_info)
my_palette_info<-data.frame(impact_flag=qw(TRlow,TRhigh,TRcom),
                            node_color_impact=get_palette("Reds",3))
df_bt_vis<-left_join(df_bt_vis,my_palette_info)
my_range_info<-data.frame(bt_role=qw(inp,intendrils,core,out,outtendrils,tubes,others),size=0)
for(i in 1:nrow(my_range_info)){
  my_range_info$size[i]<-nrow(df_bt_vis[df_bt_vis$bt_role==my_range_info$bt_role[i],])
}
my_range_info$xmean<-c(0,0,2,5,5,2,2)
my_range_info$ymean<-c(1,3,0,1,3,1.8,4)
my_range_info$sd<-c(0.1,0.1,0.1,0.3,0.1,0.1,0.1)
df_bt_vis$layout_x<-NA
df_bt_vis$layout_y<-NA
set.seed(123)
for(i in 1:nrow(my_range_info)){
  df_bt_vis[df_bt_vis$bt_role==my_range_info$bt_role[i],]$layout_x<-rnorm(n = my_range_info$size[i],
                                                                          mean = my_range_info$xmean[i],
                                                                          sd =my_range_info$sd[i])
  df_bt_vis[df_bt_vis$bt_role==my_range_info$bt_role[i],]$layout_y<-rnorm(n = my_range_info$size[i],
                                                                          mean = my_range_info$ymean[i],
                                                                          sd =my_range_info$sd[i])
}
set.seed(NULL)
g<-as_tbl_graph(df_ecoli_net,directed = T)
g<-g %>%
    activate(nodes) %>%
    left_join(df_bt_vis, by = c("name" = "node"))

temp<-data.frame(node=V(g)$name)
temp$order<-seq(1:nrow(temp))
df_bt_vis<-left_join(df_bt_vis,temp)
df_bt_vis<-df_bt_vis[order(df_bt_vis$order),]

LO<-as.matrix(df_bt_vis[,which(colnames(df_bt_vis) %in% qw(layout_x,layout_y))])
plot(g, vertex.color=V(g)$node_color, layout=LO,vertex.size=3,vertex.label="")


Sys.time()
```
## 17.5. Visualization of core network
```{r Bowtie,warning=FALSE, message=FALSE,fig.height=9,fig.width=10,dpi=300}
# Node info
df_bt_vis_core<-df_bt_vis
df_bt_vis_core<-df_bt_vis_core[df_bt_vis_core$bt_role=="core",]
df_bt_vis_core$impact_flag<-factor(df_bt_vis_core$impact_flag,levels = qw(TRlow,TRhigh,TRcom))
df_bt_vis_core<-df_bt_vis_core[order(df_bt_vis_core$impact_flag),]
df_bt_vis_core$id<-seq(1:nrow(df_bt_vis_core))-1
df_bt_vis_core$angle <- 0 + 360 * df_bt_vis_core$id / nrow(df_bt_vis_core)
df_bt_vis_core$hjust <- ifelse((df_bt_vis_core$angle > 90)&(df_bt_vis_core$angle<270), 1, 0)
df_bt_vis_core$angle <- ifelse((df_bt_vis_core$angle > 90)&(df_bt_vis_core$angle<270), df_bt_vis_core$angle+180, df_bt_vis_core$angle)
df_bt_vis_core$labcol<-mycolpal[2]
df_bt_vis_core$lab<-df_bt_vis_core$node
df_bt_vis_core[df_bt_vis_core$impact_flag=="TRcom",]$lab<-paste(df_bt_vis_core[df_bt_vis_core$impact_flag=="TRcom",]$lab,"*",sep = "")
df_bt_vis_core$node_color_sensor<-ifelse(df_bt_vis_core$sensor_flag=="SR","#ED0000FF","#1B1919EF")
df_bt_vis_core$sensor_flag<-factor(df_bt_vis_core$sensor_flag,levels=c("SR","Non-SR"))

# Edge info
df_bt_vis_core_edgeinfo<-df_ecoli_net[(df_ecoli_net$source %in% df_bt_vis_core$node)&(df_ecoli_net$target %in% df_bt_vis_core$node),]
df_bt_vis_core_edgeinfo$edgecol<-"gray40"
df_bt_vis_core_edgeinfo$edgewidth<-0.8

g_core<-as_tbl_graph(df_bt_vis_core_edgeinfo,directed = T)
g_core<-g_core %>%
    activate(nodes) %>%
    left_join(df_bt_vis_core, by = c("name" = "node"))

# order the nodes
g_core <- permute(g_core, match(V(g_core)$name, df_bt_vis_core$node))

g_core %>% 
  ggraph::ggraph(layout="circle")+
  ggraph::geom_edge_fan(alpha=0.5,arrow = arrow(length = unit(3, 'mm')),aes(color = as.factor(edgecol),width=edgewidth))+
  # ggraph::geom_edge_loop(alpha=0.5,arrow = arrow(length = unit(3, 'mm')),aes(color = as.factor(edgecol),width=edgewidth),angle_calc = "along")+
  ggraph::geom_node_point(aes(x=x*1.05,y=y*1.05), size=5,shape=21, fill="#6ece58")+
  ggraph::geom_node_text(aes(x = x*1.2, y=y*1.2,angle=angle, label=lab, hjust=hjust,colour=sensor_flag), size=7) +
  scale_color_manual(values = c("#ED0000FF","#1B1919EF"))+
  ggraph::theme_graph(background="white",)+
  theme(plot.margin=unit(c(0,0,0,0),"cm"))+
  expand_limits(x = c(-2, 2), y = c(-2, 2))+
  scale_size(range = c(5, 10))

Sys.time()
```
## 17.6. Statistics of Bow-tie structure
```{r Bowtie,warning=FALSE, message=FALSE,fig.height=3,fig.width=3,dpi=300}
nrow(df_bt_vis)
nrow(df_bt_vis[df_bt_vis$sensor_flag=="SR",]) # Nb of SR
nrow(df_bt_vis_core[df_bt_vis_core$sensor_flag=="SR",]) # Nb of SR in core
nrow(df_bt_vis[(df_bt_vis$bt_role=="inp")&(df_bt_vis$sensor_flag=="SR"),]) # Nb of SR in inp
nrow(df_bt_vis[(df_bt_vis$bt_role=="intendrils")&(df_bt_vis$sensor_flag=="SR"),]) # Nb of SR in intendrils
nrow(df_bt_vis[(df_bt_vis$bt_role=="out")&(df_bt_vis$sensor_flag=="SR"),]) # Nb of SR in out
nrow(df_bt_vis[(df_bt_vis$bt_role=="outtendrils")&(df_bt_vis$sensor_flag=="SR"),]) # Nb of SR in outtendrils
nrow(df_bt_vis[(df_bt_vis$bt_role=="tubes")&(df_bt_vis$sensor_flag=="SR"),]) # Nb of SR in tubes
nrow(df_bt_vis[(df_bt_vis$bt_role=="others")&(df_bt_vis$sensor_flag=="SR"),]) # Nb of SR in others

df_bt_stat=data.frame(bt_role=qw(inp,intendrils,core,out,outtendrils,tubes,others),
                      label=qw(IN,INTENDRILS, SCC,OUT,OUTTENDRILS,TUBES,OTHERS),
                      size=0,TR_size=0,Non_TR_size=0,SR_size=0,Non_SR_size=0)

df_bt_stat$label<-factor(df_bt_stat$label,levels = qw(IN,INTENDRILS, SCC,OUT,OUTTENDRILS,TUBES,OTHERS))

for(i in 1:nrow(df_bt_stat)){
  df_bt_stat$size[i]<-nrow(df_bt[df_bt$bt_role==df_bt_stat$bt_role[i],])
  df_bt_stat$TR_size[i]<-nrow(df_bt[(df_bt$regulator_flag=="TR")&(df_bt$bt_role==df_bt_stat$bt_role[i]),])
  df_bt_stat$Non_TR_size[i]<-nrow(df_bt[(df_bt$regulator_flag=="Non-TR")&(df_bt$bt_role==df_bt_stat$bt_role[i]),])
  df_bt_stat$SR_size[i]<-nrow(df_bt[(df_bt$sensor_flag=="SR")&(df_bt$bt_role==df_bt_stat$bt_role[i]),])
  df_bt_stat$Non_SR_size[i]<-nrow(df_bt[(df_bt$regulator_flag=="TR")&(df_bt$sensor_flag=="Non-SR")&(df_bt$bt_role==df_bt_stat$bt_role[i]),])
}

df_bt_stat_long<-data.frame(bt_role=rep(qw(inp,intendrils,core,out,outtendrils,tubes,others),2),
                            label=rep(qw(IN,INTENDRILS, SCC,OUT,OUTTENDRILS,TUBES,OTHERS),2),
                            regulator_flag=rep(c("TR","Non-TR"),each=7),
                            sensor_flag=rep(c("SR","Non-SR"),each=7),
                            size=0,size_sr=0)
for(i in 1:nrow(df_bt_stat_long)){
  if(df_bt_stat_long$regulator_flag[i]=="TR"){
    df_bt_stat_long$size[i]<-df_bt_stat[(df_bt_stat$bt_role==df_bt_stat_long$bt_role[i]),]$TR_size
  }else{
    df_bt_stat_long$size[i]<-df_bt_stat[(df_bt_stat$bt_role==df_bt_stat_long$bt_role[i]),]$Non_TR_size
  }
  if(df_bt_stat_long$sensor_flag[i]=="SR"){
    df_bt_stat_long$size_sr[i]<-df_bt_stat[(df_bt_stat$bt_role==df_bt_stat_long$bt_role[i]),]$SR_size
  }else{
    df_bt_stat_long$size_sr[i]<-df_bt_stat[(df_bt_stat$bt_role==df_bt_stat_long$bt_role[i]),]$Non_SR_size
  }
}
df_bt_stat_long$regulator_flag<-factor(df_bt_stat_long$regulator_flag,levels=c("TR","Non-TR"))
df_bt_stat_long$sensor_flag<-factor(df_bt_stat_long$sensor_flag,levels=c("SR","Non-SR"))

ggbarplot(df_bt_stat_long,x="label",y="size",fill="regulator_flag",
          legend.title="",legend="right",xlab="",ylab="Number of nodes",ylim=c(0,200),
  position = position_dodge(0.9),size=0)+rotate_x_text()

ggbarplot(df_bt_stat_long,x="label",y="size_sr",fill="sensor_flag",
          legend.title="",legend="right",xlab="",ylab="Number of TR",palette = c("#ED0000FF", "#1B1919EF"),
  position = position_dodge(0.9),size=0)+rotate_x_text()

ggbarplot(df_bt_stat,x="label",y="TR_size",xlab="",ylab="Number of reuglators",fill="label", color="label",palette =c("#42B540FF","#0099B4FF","#ED0000FF","#00468BFF","#925E9FFF","#FDAF91FF","#ADB6B6FF"),legend="",size=0)+
  rotate_x_text()

ggbarplot(df_bt_stat,x="label",y="SR_size",xlab="",ylab="Number of reuglators",fill="label", color="label",palette =c("#42B540FF","#0099B4FF","#ED0000FF","#00468BFF","#925E9FFF","#FDAF91FF","#ADB6B6FF"),legend="",size=0)+
  rotate_x_text()

Sys.time()
```
## 17.7. Add Stimuli sector to the netwrok
```{r Bowtie,warning=FALSE, message=FALSE,fig.height=20,fig.width=23,dpi=300}
df_bt_stml<-data.frame(source=rep("Stimuli",nrow(df_bt_vis[df_bt_vis$sensor_flag=="SR",])),
                       target=df_bt_vis[df_bt_vis$sensor_flag=="SR",]$node)
g_stml<-as_tbl_graph(df_bt_stml,directed = T)
g_stml<-igraph::union(g_stml,g)
g_stml<- set.vertex.attribute(g_stml, "node_color", "Stimuli", "#1B1919EF")
g_stml<- set.vertex.attribute(g_stml, "layout_x", "Stimuli", 3)
g_stml<- set.vertex.attribute(g_stml, "layout_y", "Stimuli", 2.7)
g_stml<- set.vertex.attribute(g_stml, name = "node_size", value = 3)
g_stml<- set.vertex.attribute(g_stml, "node_size", "Stimuli", 10)
g_stml<-set_edge_attr(g_stml, name = "edge_col",value="gray70")
g_stml<-set_edge_attr(g_stml, name = "edge_lty",value="solid")
for(i in 1:ecount(g_stml)){
  if(V(g_stml)$name[get.edges(g_stml,i)[1]]=="Stimuli"){
    g_stml<- set_edge_attr(g_stml, name = "edge_col",index = i,"#ED0000FF")
    g_stml<- set_edge_attr(g_stml, name = "edge_lty",index = i,"dashed")
  }
}
temp<-as_data_frame(g_stml,what = "vertices")
LO<-as.matrix(temp[,which(colnames(temp) %in% qw(layout_x,layout_y))])
plot(g_stml, vertex.color=V(g_stml)$node_color, layout=LO,vertex.size=V(g_stml)$node_size,vertex.label="",
     edge.color=E(g_stml)$edge_col,edge.lty=E(g_stml)$edge_lty)

Sys.time()
```
## 17.8 Edge density within SCC
12 regulators are sampled.
Self loops are omitted.
```{r TRN,warning=FALSE, message=FALSE,fig.height=3.1,fig.width=3,dpi=300}
temp0<-trn.stat.table[trn.stat.table$all_target>4,] %>% distinct(regulator,.keep_all = T)
# temp0<-trn.stat.table %>% distinct(regulator,.keep_all = T)
temp0<-temp0[temp0$regulator %in% df_bt_vis_core$node,]
trial<-10000
TRcom_scc<-TRcom
TRcom_scc<-TRcom_scc[TRcom_scc %in% df_bt_vis_core$node]
temp<-createEmptyDf(nrow=trial,ncol = 4,colnames = qw(test,regulators,ratio.TRcom,test.btw))
for(i in 1:nrow(temp)){
  temp1<-temp0[sample(seq(1,nrow(temp0)),size = length(TRcom_scc),replace = F),]$regulator
  temp$regulators[i]<-paste(temp1,collapse = ",")
  temp$ratio.TRcom[i]<-length(temp1[temp1 %in% TRcom_scc])/length(TRcom_scc)
  temp2<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$who %in% temp1 & dnet_edgeinfo_ts.global$whom %in% temp1,]
  temp2<-temp2[temp2$who!=temp2$whom,] # remove self-loop
  g<-as_tbl_graph(temp2,directed = T)
  # temp$test[i]<-ecount(g)/((length(TRcom)*(length(TRcom)-1))+length(TRcom)) # consider self-loop
  temp$test[i]<-ecount(g)/((length(TRcom_scc)*(length(TRcom_scc)-1))) # not consider self-loop
  temp1.2<-temp0[temp0$regulator %nin% temp1,] # the rest TFs
  temp2.2<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$who %in% temp1 & dnet_edgeinfo_ts.global$whom %in% temp1.2$regulator,]
  temp2.3<-dnet_edgeinfo_ts.global[dnet_edgeinfo_ts.global$whom %in% temp1 & dnet_edgeinfo_ts.global$who %in% temp1.2$regulator,]
  temp2.4<-bind_rows(temp2.2,temp2.3)
  temp$test.btw[i]<-nrow(temp2.4)/(2*length(TRcom_scc)*nrow(temp1.2))
}
temp3<-dnet_edgeinfo_ts.global.sbst
temp3<-temp3[temp3$who!=temp3$whom,]
g<-as_tbl_graph(temp3,directed = T)
# temp3<-ecount(g)/((length(TRcom)*(length(TRcom)-1))+length(TRcom)) # consider self-loop
temp3<-ecount(g)/((length(TRcom_scc)*(length(TRcom_scc)-1))) # not consider self-loop
print(paste("Edge density of TRcom within SCC is",temp3))
temp4<-nrow(as.data.frame(temp[temp$test>temp3,]))/trial

# Edge density within randomly sampled 13 regulators
df.rndmpick.edgedens_scc<-temp
p<-gghistogram(df.rndmpick.edgedens_scc,x="test",y="..density..",bins=30,
            fill="black",
            color="darkgray",alpha=1,size=0,
            # add="mean",add.params = list(color="black",linetype=2),
            xlab="Edge density",ylab="Relative frequency [a.u.]")+
  geom_vline(xintercept = temp3,linetype=1,color=mycolpal[2])+
  scale_x_continuous(breaks = get_breaks(by = 0.1, from = 0),limits =  c(0, 0.3))
p<-ggpar(p,font.x=c(17),font.y = c(17),font.xtickslab = c(17),font.ytickslab = c(17))
if(nrow(as.data.frame(df.rndmpick.edgedens_scc[df.rndmpick.edgedens_scc$test>temp3,]))==0){
  p<-p+annotate("text",x=temp3,y=10,label=paste("p< ",1/trial ,sep=""),hjust = 0.3,vjust = 0, parse=F,size=6)
}else{
  p<-p+annotate("text",x=temp3,y=10,label=paste("p= ",temp4 ,sep=""),hjust = 0.3,vjust = 0, parse=F,size=6)
}
p

Sys.time()
```
## 17.9 Impact of Sensory Regulators on DM values
```{r TRN,warning=FALSE, message=FALSE,fig.height=3.1,fig.width=6,dpi=300}
my_df<-df.regcount.subcomp
my_df$SRs<-NA
my_df$SRcount<-0
my_df$NonTRcomSRs<-NA
my_df$NonTRcomSRcount<-0
i<-1
for (i in 1:nrow(my_df)){
  if(!is.na(my_df$regulators[i])){
    my_regs<-str_split(my_df$regulators[i],",")[[1]]
    my_regs<-my_regs[my_regs %in% df_bt_stml$target]
    if(length(my_regs)>0){
      my_df$SRs[i]<-paste(my_regs,collapse = ",")
      my_df$SRcount[i]<-length(my_regs)
    }
    my_regs<-my_regs[my_regs %nin% TRcom]
    if(length(my_regs)>0){
      my_df$NonTRcomSRs[i]<-paste(my_regs,collapse = ",")
      my_df$NonTRcomSRcount[i]<-length(my_regs)
    }
  }
}

my_df2<-my_df
my_df2$No

ggscatter(my_df,x="SRcount",y="DMenv",alpha=0.1,color="TRcomflag2",legend="right",legend.title="")+
  stat_cor(method = "spearman",label.x = 4,label.y =2)

ggscatter(my_df[my_df$Count.allregs==5,],x="NonTRcomSRcount",y="DMenv",alpha=0.1,color="TRcomflag2",legend="right",legend.title="")+
  stat_cor(method = "spearman",label.x = 0,label.y = 2)

ggscatter(my_df,y="SRcount",x="Count.allregs",xlab="Number of unique regulators", ylab="Number of unique SRs",alpha=0.1)+
  stat_cor(method = "spearman",label.x = 0,label.y=9)

my_y<-"DMenv"
my_n<-7
ggscatter(my_df[my_df$Count.allregs==my_n,],y=my_y,x="SRcount",xlab="Number of unique SRs", ylab=my_y,alpha=0.3,color="TRcomflag2",legend="right",legend.title="")+
  stat_cor(method = "spearman",label.x = 0,label.y = 2)
ggscatter(my_df[my_df$Count.allregs==my_n,],y=my_y,x="NonTRcomSRcount",xlab="Number of unique Non-TRcom SRs", ylab=my_y,alpha=0.3,color="TRcomflag2",legend="right",legend.title="")+
  stat_cor(method = "spearman",label.x = 0,label.y = 2)


Sys.time()
```
## 17.10 Enrichment analysis (Hypergeometric test) for TRcom and TRhigh
Testing whether TRcoms and TRhghs are enriched in the SCC sector
```{r TRN,warning=FALSE, message=FALSE,fig.height=3.1,fig.width=6,dpi=300}
## TERM2GENE
my_t2g<-df_bt[df_bt$regulator_flag=="TR",] %>% select(bt_role,node)
## TRcom
my_gene<-TRcom
my_res<-enricher(gene = my_gene,pAdjustMethod = "BH",pvalueCutoff = 1,TERM2GENE = my_t2g)
my_df<-my_res@result
my_df$TR<-"TRcom"

## TRhigh
my_gene<-TRhigh
my_res<-enricher(gene = my_gene,pAdjustMethod = "BH",pvalueCutoff = 1,TERM2GENE = my_t2g)
my_tmp<-my_res@result
my_tmp$TR<-"TRhigh"

my_df<-bind_rows(my_df,my_tmp)
rownames(my_df)<-seq(1,nrow(my_df))
my_df$qval.fin<-qvalue(my_df$pvalue)$qvalue
my_df[my_df$qval.fin<0.05,]
Sys.time()
```
## 17.11 Enrichment analysis (Hypergeometric test) for SR
Testing whether TRs in SCC are enriched in SRs
```{r TRN,warning=FALSE, message=FALSE,fig.height=3.1,fig.width=6,dpi=300}
## TERM2GENE
my_t2g<-df_bt[df_bt$regulator_flag=="TR",] %>% select(sensor_flag,node)
## Gene
my_gene<-df_bt[df_bt$bt_role=="core",]$node
my_res<-enricher(gene = my_gene,pAdjustMethod = "BH",pvalueCutoff = 1,TERM2GENE = my_t2g)
my_df<-my_res@result
my_df
Sys.time()
```
## 17.12 Number of SR in TRhigh or in TRcom
```{r TRN,warning=FALSE, message=FALSE,fig.height=3.1,fig.width=6,dpi=300}
# Number of SR in TRhigh
my_df<-df_bt[df_bt$impact_flag %in% qw(TRhigh,TRcom),]
print (paste("Number of SR & TRhigh: ",nrow(my_df[my_df$sensor_flag=="SR",]),sep=""))

my_df<-df_bt[(df_bt$bt_role %in% qw(core,inp))&(df_bt$sensor_flag=="SR"),]
print (paste("Number of SR in SCC and IN: ",nrow(my_df),sep=""))
Sys.time()
```

#
#
#
# 18. Summary
## 18.1. correlation between DM values
```{r Summary,warning=FALSE, message=FALSE,fig.height=3,fig.width=4,dpi=300}
# DM values
my_df<-df_DMall[,which(colnames(df_DMall) %in% qw(gene,Bnb,DMenv,DMevo,DMmut))]
my_tmp<-df_strs_maeiwa[,which(colnames(df_strs_maeiwa) %in% qw(Bnb,DMstrs))]
my_df<-left_join(my_df,my_tmp)
my_tmp<-df_strs_hori[,which(colnames(df_strs_hori) %in% qw(Bnb,DMstrs))]
colnames(my_tmp)[colnames(my_tmp)=="DMstrs"]<-"DMstrs_hp"
my_df<-left_join(my_df,my_tmp)
my_tmp<-df_strs_hori_evo[,which(colnames(df_strs_hori_evo) %in% qw(Bnb,DMstrs))]
colnames(my_tmp)[colnames(my_tmp)=="DMstrs"]<-"DMstrs_he"
my_df<-left_join(my_df,my_tmp)
my_tmp<-df_prtn_sch[,which(colnames(df_prtn_sch) %in% qw(Bnb,DMprtn))]
my_df<-left_join(my_df,my_tmp)
my_tmp<-df_tani[df_tani$HL_label=="High",]
my_tmp<-my_tmp %>% select(Bnb,TNDM)
colnames(my_tmp)[colnames(my_tmp)=="TNDM"]<-"DMnoise_ch"
my_df<-left_join(my_df,my_tmp)
my_df<-my_df %>% select(gene,Bnb,everything())
my_df<-my_df[order(my_df$Bnb),]

my_mat<-matrix(data=NA,nrow=5,ncol=3)
rownames(my_mat)<-qw(DMstrs_hp,DMstrs_he,DMstrs,DMprtn,DMnoise_ch)
colnames(my_mat)<-qw(DMenv,DMevo,DMmut)
for(i in 1:nrow(my_mat)){
  for(j in 1:ncol(my_mat)){
    my_tmp<-my_df %>% select(rownames(my_mat)[i],colnames(my_mat)[j])
    my_mat[i,j]<-cor.test(my_tmp[,1],my_tmp[,2], method = "spearman")$estimate
  }
}

my_fontsize<-16
p<-ggcorrplot::ggcorrplot(t(my_mat),
                           legend.title = "Spearman's R",
  lab = TRUE,lab_size = my_lab_fontsize*0.8,tl.cex = my_fontsize,outline.col = "gray40")+
  scale_fill_gradient(limit = c(0,1.0),breaks = seq(0,1,0.2), low = "white", high =  "#2171B5",
                       guide=guide_colourbar(title= "R",title.position = "top",ticks.linewidth = 1))
p<-ggpar(p,legend.title = "R",font.legend = my_fontsize)+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())
p

Sys.time()
```
## 18.2. correlation between loadings of PC1 in Env and those of PC1-3 in other datasets
```{r Summary,warning=FALSE, message=FALSE,fig.height=6,fig.width=4,dpi=300}
# DM values
# Env
my_df<-as.data.frame(res.pca.global_env$rotation)
my_df<-my_df %>% select(PC1,PC2,PC3)
colnames(my_df)<-paste("Env",seq(1,3),sep="_PC")
my_df$Bnb<-rownames(my_df)
my_df<-my_df %>% select(Bnb,everything())

# Strs
my_tmp<-as.data.frame(res.pca.global_strs_mi$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Strs_hp
my_tmp<-as.data.frame(res.pca.global_strs_hp$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs_hp",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Strs_he
my_tmp<-as.data.frame(res.pca.global_strs_he$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs_he",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Prtn
my_tmp<-as.data.frame(res.pca.global_prtn$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Prtn",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)




my_mat<-matrix(data=NA,nrow=15,ncol=3)
rownames(my_mat)<-paste(rep(qw(Strs,Strs_hp,Strs_he,Prtn),each=3),rep(qw(PC1,PC2,PC3),5),sep="_")
colnames(my_mat)<-paste("Env",seq(1,3),sep="_PC")
for(i in 1:nrow(my_mat)){
  for(j in 1:ncol(my_mat)){
    my_tmp<-my_df %>% select(rownames(my_mat)[i],colnames(my_mat)[j])
    my_mat[i,j]<-cor.test(my_tmp[,1],my_tmp[,2], method = "spearman")$estimate
  }
}
my_mat<-abs(my_mat)

my_fontsize<-16
p<-ggcorrplot::ggcorrplot(t(my_mat),
  lab = TRUE,lab_size = my_lab_fontsize*0.8,tl.cex = my_fontsize,outline.col = "gray40")+
  scale_fill_gradient(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                       guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
p<-ggpar(p,legend.title = "|R|",font.legend = my_fontsize)+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())
p

Sys.time()
```
## 18.3. correlation between loadings of PC1 in Evo and those of PC1-3 in other datasets
```{r Summary,warning=FALSE, message=FALSE,fig.height=6,fig.width=4,dpi=300}
# DM values
# Evo
my_df<-as.data.frame(res.pca.global_evo$rotation)
my_df<-my_df %>% select(PC1,PC2,PC3)
colnames(my_df)<-paste("Evo",seq(1,3),sep="_PC")
my_df$Bnb<-rownames(my_df)
my_df<-my_df %>% select(Bnb,everything())

# Strs
my_tmp<-as.data.frame(res.pca.global_strs_mi$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Strs_hp
my_tmp<-as.data.frame(res.pca.global_strs_hp$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs_hp",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Strs_he
my_tmp<-as.data.frame(res.pca.global_strs_he$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs_he",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Prtn
my_tmp<-as.data.frame(res.pca.global_prtn$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Prtn",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)




my_mat<-matrix(data=NA,nrow=15,ncol=3)
rownames(my_mat)<-paste(rep(qw(Strs,Strs_hp,Strs_he,Prtn),each=3),rep(qw(PC1,PC2,PC3),5),sep="_")
colnames(my_mat)<-paste("Evo",seq(1,3),sep="_PC")
for(i in 1:nrow(my_mat)){
  for(j in 1:ncol(my_mat)){
    my_tmp<-my_df %>% select(rownames(my_mat)[i],colnames(my_mat)[j])
    my_mat[i,j]<-cor.test(my_tmp[,1],my_tmp[,2], method = "spearman")$estimate
  }
}
my_mat<-abs(my_mat)

my_fontsize<-16
p<-ggcorrplot::ggcorrplot(t(my_mat),
  lab = TRUE,lab_size = my_lab_fontsize*0.8,tl.cex = my_fontsize,outline.col = "gray40")+
  scale_fill_gradient(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                       guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
p<-ggpar(p,legend.title = "|R|",font.legend = my_fontsize)+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())
p

Sys.time()
```
## 18.4. correlation between loadings of PC1 in Mut and those of PC1-3 in other datasets
```{r Summary,warning=FALSE, message=FALSE,fig.height=6,fig.width=4,dpi=300}
# DM values
# Mut
my_df<-as.data.frame(res.pca.global_mut$rotation)
my_df<-my_df %>% select(PC1,PC2,PC3)
colnames(my_df)<-paste("Mut",seq(1,3),sep="_PC")
my_df$Bnb<-rownames(my_df)
my_df<-my_df %>% select(Bnb,everything())

# Strs
my_tmp<-as.data.frame(res.pca.global_strs_mi$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Strs_hp
my_tmp<-as.data.frame(res.pca.global_strs_hp$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs_hp",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Strs_he
my_tmp<-as.data.frame(res.pca.global_strs_he$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Strs_he",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)

# Prtn
my_tmp<-as.data.frame(res.pca.global_prtn$rotation)
my_tmp<-my_tmp %>% select(PC1,PC2,PC3)
colnames(my_tmp)<-paste("Prtn",seq(1,3),sep="_PC")
my_tmp$Bnb<-rownames(my_tmp)
my_df<-left_join(my_df,my_tmp)




my_mat<-matrix(data=NA,nrow=15,ncol=3)
rownames(my_mat)<-paste(rep(qw(Strs,Strs_hp,Strs_he,Prtn),each=3),rep(qw(PC1,PC2,PC3),5),sep="_")
colnames(my_mat)<-paste("Mut",seq(1,3),sep="_PC")
for(i in 1:nrow(my_mat)){
  for(j in 1:ncol(my_mat)){
    my_tmp<-my_df %>% select(rownames(my_mat)[i],colnames(my_mat)[j])
    my_mat[i,j]<-cor.test(my_tmp[,1],my_tmp[,2], method = "spearman")$estimate
  }
}
my_mat<-abs(my_mat)

my_fontsize<-16
p<-ggcorrplot::ggcorrplot(t(my_mat),
  lab = TRUE,lab_size = my_lab_fontsize*0.8,tl.cex = my_fontsize,outline.col = "gray40")+
  scale_fill_gradient(limits = c(0,0.65), breaks = seq(0,0.6,0.2),low = "white", high =  "#E46726",
                       guide=guide_colourbar(title= "|R|",title.position = "top",ticks.linewidth = 1))
p<-ggpar(p,legend.title = "|R|",font.legend = my_fontsize)+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank())
p

Sys.time()
```


#
#
#
# 20. Output for supplementary tables
## 20.1. Output tables
```{r Out,warning=FALSE, message=FALSE,fig.height=6,fig.width=9,dpi=300}
# log2-transformed normalized transcriptome data for the Mut dataset
outtemp<-datlog.norm[,which(colnames(datlog.norm) %in% c(qw(gene,Bnb),qw(AncHa_rep1_0,AncHa_rep2_0,AncHb_rep1_0,AncHb_rep2_0),MAlin.sel.R30,MAlin.sel.R10))]
outtemp<-outtemp %>% select(gene,Bnb,everything())
outtemp<-outtemp[order(outtemp$Bnb),]
# write.csv(outtemp,paste(getwd(),"Output","MA_transcriptome.csv",sep="/"),row.names=F)

# DM values
outtemp<-df_DMall[,which(colnames(df_DMall) %in% qw(gene,Bnb,DMenv,DMevo,DMmut))]
outtemp1<-df_strs_maeiwa[,which(colnames(df_strs_maeiwa) %in% qw(Bnb,DMstrs))]
outtemp<-left_join(outtemp,outtemp1)
outtemp1<-df_strs_hori[,which(colnames(df_strs_hori) %in% qw(Bnb,DMstrs))]
colnames(outtemp1)[colnames(outtemp1)=="DMstrs"]<-"DMstrs_hp"
outtemp<-left_join(outtemp,outtemp1)
outtemp1<-df_strs_hori_evo[,which(colnames(df_strs_hori_evo) %in% qw(Bnb,DMstrs))]
colnames(outtemp1)[colnames(outtemp1)=="DMstrs"]<-"DMstrs_he"
outtemp<-left_join(outtemp,outtemp1)
outtemp1<-df_prtn_sch[,which(colnames(df_prtn_sch) %in% qw(Bnb,DMprtn))]
outtemp<-left_join(outtemp,outtemp1)
outtemp1<-df_tani[,which(colnames(df_tani) %in% qw(Bnb,TNDM))]
colnames(outtemp1)[colnames(outtemp1)=="TNDM"]<-"DMnoise_ch"
outtemp<-left_join(outtemp,outtemp1)
outtemp<-outtemp %>% select(gene,Bnb,everything())
outtemp<-outtemp[order(outtemp$Bnb),]
# write.csv(outtemp,paste(getwd(),"Output","DM_values.csv",sep="/"),row.names=F)

# TRN
outtemp<-TRN
outtemp<-outtemp %>% select(regulator,regulator_id,gene_name,gene_id,effect,source,category,everything())
outtemp<-outtemp[,which(colnames(outtemp) %in% qw(regulator,regulator_id,gene_name,gene_id,effect,source,category))]
colnames(outtemp)<-qw(Regulator,Regulator_Bnb,Target,Target_Bnb,Effect,Source,Category)
# write.csv(outtemp,paste(getwd(),"Output","TRN.csv",sep="/"),row.names=F)

# KEGG pathway
outtemp<-df_DMall_cat[,which(colnames(df_DMall_cat) %nin% "EntrezID")]
outtemp<-outtemp %>% select(label,everything())
# write.csv(outtemp,paste(getwd(),"Output","KEGG_pathway.csv",sep="/"),row.names=F)

# GO
outtemp<-df_DMall_gocat
outtemp<-outtemp %>% select(ID,"Term Name",parent,parentTerm,Category,Count,GeneRatio,Bnb,DMenv,DMevo,DMmut,everything())
outtemp<-outtemp[,which(colnames(outtemp) %in% c("ID","Term Name",qw(parent,parentTerm,Category,Count,GeneRatio,Bnb,DMenv,DMevo,DMmut)))]
outtemp$Bnb<-str_replace_all(outtemp$Bnb,pattern = "_",replacement = ",")
# write.csv(outtemp,paste(getwd(),"Output","GO_slim.csv",sep="/"),row.names=F)

# PRECISE dataset
outtemp<-fileinfo_plasticity.pr2
outtemp<-outtemp %>% select(sample_ID,sample,rep_id,Strain.Description,Strain,Base.Media,Carbon.Source..g.L.,Nitrogen.Source..g.L.,Electron.Acceptor,Trace.Element.Mixture,Supplement,Temperature..C.,pH,Culture.Type,Growth.Rate..1.hr.,everything())
outtemp<-outtemp[,which(colnames(outtemp) %in%  qw(sample_ID,sample,rep_id,Strain.Description,Strain,Base.Media,Carbon.Source..g.L.,Nitrogen.Source..g.L.,Electron.Acceptor,Trace.Element.Mixture,Supplement,Temperature..C.,pH,Culture.Type,Growth.Rate..1.hr.))]
# write.csv(outtemp,paste(getwd(),"Output","PRECISE_used.csv",sep="/"),row.names=F)

# Common targets
outtemp<-df_mt_unique %>% select(gene,Bnb,Stressors,targets)
# write.csv(outtemp,paste(getwd(),"Output","Common_targets.csv",sep="/"),row.names=F)

# Bow-tie decomposition
outtemp<-df_bt[,which(colnames(df_bt) %nin% "impact_flag")]
outtemp<-outtemp %>% select(node,bt_role,regulator_flag,sensor_flag)
colnames(outtemp)<-qw(Node,Sector,TRorNon_TR,SRorNon_SR)
# write.csv(outtemp,paste(getwd(),"Output","Bow_tie_deconv.csv",sep="/"),row.names=F)

Sys.time()
```



